{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def MST_input_maker(vcf_path, sex_info_file, offspring_presence_threshold, mendelian_threshold):\n",
    "    \n",
    "    \"\"\"\n",
    "    Usage: MST_input_maker  <vcf_path>  <sex_info_file>  <offspring_presence_threshold>  <mendelian_threshold>\n",
    "    \n",
    "    ### Informative locus filters:\n",
    "    -Loci are used only if the number of offspring present is equal to or above the percentage specified by the user\n",
    "    -Loci are used only if they are heterozygous in one sex and homozygous in another.\n",
    "    \n",
    "    ###Null allele filters:\n",
    "\n",
    "    Null allele in Parents:\n",
    "    -If an offspring is found to have an allele that isn't present in the parent, this locus is discarded\n",
    "\n",
    "    Null allele in offspring:\n",
    "    -If only 1 offspring is homozygous for the parental minor allele then that sample is assumed to possess a null allele and the data for that individual is coded as missing. **Note, the sample presence/absence filter works after this step, so some loci may be pushed over the \"missing data\" limit by this null allele filter. \n",
    "    -If two or more samples are found to be homozygous for the minor parental allele, it is more likely that this is due to the homozygous parent having a null allele. In this case, the locus is again discarded.\n",
    "    -Also, loci with non-mendelian properties indicative of allele dropout or similar are also filtered\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    import sys\n",
    "    import vcf\n",
    "    from collections import Counter\n",
    "    import pprint as pp\n",
    "    from pydoc import help\n",
    "    \n",
    "\n",
    "    alteredvcfpath = \"%s%s\" % (vcf_path, \".altered\")\n",
    "\n",
    "    oldvcf = open(vcf_path, 'r').readlines()\n",
    "    alteredvcf = open(alteredvcfpath, 'w')    \n",
    "    \n",
    "    for line in oldvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "    \n",
    "    altered_vcf = open(alteredvcfpath, 'r')\n",
    "    \n",
    "    myvcf = vcf.Reader(altered_vcf)\n",
    "        \n",
    "    MST_path = \"%s/\" % (vcf_path.rpartition(\"/\")[0])\n",
    "    \n",
    "    loci_used_for_male_map = 0\n",
    "    loci_used_for_female_map = 0\n",
    "    \n",
    "    Loc_w_null_alleles = 0\n",
    "    loc_w_excess_hom = 0\n",
    "    \n",
    "    offspring = []\n",
    "    \n",
    "    femMSTfile = open(\"%sMST_female_map_input.txt\" % (MST_path), 'w')\n",
    "    malMSTfile = open(\"%sMST_male_map_input.txt\" % (MST_path), 'w')\n",
    "    log_file = open(\"%sMST_input_maker.log\" % (MST_path), 'w')\n",
    "\n",
    "\n",
    "    sex_file = open(sex_info_file, 'r').readlines()\n",
    "\n",
    "    for line in sex_file: ## Note, only samples in the sex_info file are used\n",
    "        \n",
    "        sample_name = line.split()[0]\n",
    "        sex = line.split()[1]                    \n",
    "        \n",
    "        if sex == \"F\" or sex == \"f\":\n",
    "            Mother = sample_name\n",
    "            log_file.write(\"Mother = %s\" % (sample_name))\n",
    "        \n",
    "        elif sex == \"M\" or sex == \"m\":\n",
    "            Father = sample_name\n",
    "            log_file.write(\"\\nFather = %s\\n\" % (sample_name))        \n",
    "    \n",
    "        elif sex == \"O\" or sex == \"o\":\n",
    "            offspring.append(sample_name)        \n",
    "        \n",
    "    offspring = sorted(offspring) ## sort the list of offspring - used to retain correct order throughout script\n",
    "    \n",
    "   \n",
    "    \n",
    "    femmap_file_lines = []\n",
    "    malmap_file_lines = []\n",
    "    \n",
    "    \n",
    "    for record in myvcf:\n",
    "        \n",
    "        Mat_ref = None\n",
    "        Mat_alt = None\n",
    "        Pat_ref = None\n",
    "        Pat_alt = None\n",
    "        mat_het = None\n",
    "        pat_het = None\n",
    "        Mat_genotype = None\n",
    "        Pat_genotype = None\n",
    "        Mat_genotype_bases = None\n",
    "        Pat_genotype_bases = None\n",
    "        Non_par_allele = False\n",
    "        N_offspring = 0\n",
    "        offspr_w_allele_dropout = 0\n",
    "        N_offsp_hom_for_minor_par_allele = 0\n",
    "        \n",
    "        Sample_headers = []\n",
    "        sample_names = []\n",
    "        Parental_alleles = []\n",
    "        offspring_gt_types = []\n",
    "        \n",
    "        offspring_gt_types_dict = {}\n",
    "        offspring_gt_bases_dict = {}\n",
    "        allele_dropout_scores = {}\n",
    "        \n",
    "        ## lists for constructing the lines of the MST output file\n",
    "        malemap_offspring_haploid_genotypes = []\n",
    "        malemap_offspring_haploid_genotypes_compli = []\n",
    "        femalemap_offspring_haploid_genotypes = []\n",
    "        femalemap_offspring_haploid_genotypes_compli = []\n",
    "    \n",
    "    \n",
    "        Loc_Id = \"%s_%s\" % (record.ID, record.POS)\n",
    "        \n",
    "        log_file.write(\"\\n\\nLocus: %s -------------\\n\" % (Loc_Id))\n",
    "        \n",
    "        for sample in record.samples:\n",
    "            \n",
    "            name = sample.sample\n",
    "            genotype = sample['GT']\n",
    "            allele_dropout_scores[record] = {}\n",
    "            \n",
    "            sample_names.append(name)\n",
    "            \n",
    "            ## Get Mother and Father information\n",
    "            \n",
    "            if name == Mother:\n",
    "                if not sample.called == False:\n",
    "                    Mat_ref = sample.gt_bases.split(\"/\")[0]\n",
    "                    Mat_alt = sample.gt_bases.split(\"/\")[1]\n",
    "                    Mat_genotype = genotype\n",
    "                    Mat_genotype_bases = sample.gt_bases\n",
    "                    \n",
    "                    Parental_alleles.append(Mat_ref)\n",
    "                    Parental_alleles.append(Mat_alt)\n",
    "                    \n",
    "                    if sample.gt_type == 0 or sample.gt_type == 2:\n",
    "                        mat_het = False\n",
    "                        log_file.write(\"\\nMother is homozygous (%s)\" % (sample.gt_bases))\n",
    "                    elif sample.gt_type == 1:\n",
    "                        mat_het = True   \n",
    "                        log_file.write(\"\\nMother is heterozygous (%s)\" % (sample.gt_bases))\n",
    "                elif sample.called == False:\n",
    "                    log_file.write(\"\\nLocus not called in mother\")\n",
    "                    \n",
    "            elif name == Father:\n",
    "                if not sample.called == False:\n",
    "                    Pat_ref = sample.gt_bases.split(\"/\")[0]\n",
    "                    Pat_alt = sample.gt_bases.split(\"/\")[1]\n",
    "                    Pat_genotype = genotype\n",
    "                    Pat_genotype_bases = sample.gt_bases\n",
    "                    \n",
    "                    Parental_alleles.append(Pat_ref)\n",
    "                    Parental_alleles.append(Pat_alt)\n",
    "                    \n",
    "                    if sample.gt_type == 0 or sample.gt_type == 2:\n",
    "                        pat_het = False\n",
    "                        log_file.write(\"\\nFather is homozygous (%s)\" % (sample.gt_bases))\n",
    "                    elif sample.gt_type == 1:\n",
    "                        pat_het = True\n",
    "                        log_file.write(\"\\nFather is heterozygous (%s)\" % (sample.gt_bases))\n",
    "                elif sample.called == False:\n",
    "                    log_file.write(\"\\nLocus not called in father\")\n",
    "            \n",
    "        ## Get parental allele information and filter for unacceptable loci\n",
    "        \n",
    "        \n",
    "        Par_allele_counts = Counter(Parental_alleles)\n",
    "        \n",
    "        \n",
    "        if None in Par_allele_counts or Mat_genotype == None or Pat_genotype == None: ## If there is a missing call in the parents, discard locus\n",
    "            log_file.write(\"\\nMissing data in parents, locus discarded\")\n",
    "            \n",
    "        elif len(Par_allele_counts) > 2: ## If there are more than two alleles in the parents, discard locus\n",
    "            log_file.write(\"\\nparents contain more than 2 alleles, locus not used\")\n",
    "        \n",
    "        elif pat_het == True and mat_het == True: ## If both parents are heterozygous, discard locus\n",
    "            log_file.write(\"\\nBoth parents heterozygous, locus not used\")\n",
    "            \n",
    "        elif pat_het == False and mat_het == False: ## If both parents are homozygous, discard locus\n",
    "            log_file.write(\"\\nBoth parents homozygous, locus not used\")\n",
    "            \n",
    "        elif len(Par_allele_counts) == 2 and not 2 in Par_allele_counts.values():\n",
    "            for allele in Par_allele_counts:\n",
    "                if Par_allele_counts[allele] == 3:\n",
    "                    Par_major_allele = allele\n",
    "                elif Par_allele_counts[allele] == 1:\n",
    "                    Par_minor_allele = allele\n",
    "        \n",
    "        ## Get offspring information -----------------------------------------------------------------------\n",
    "               \n",
    "            for sample in record.samples:            \n",
    "                name = sample.sample\n",
    "                genotype = sample['GT']            \n",
    "        \n",
    "                if name in offspring:  ## Only samples included in the pop_map file are used\n",
    "                    N_offspring += 1\n",
    "                        \n",
    "                    ### Filter sample genotypes for signs of allele droupout \n",
    "                                    \n",
    "                    if genotype == None:                    \n",
    "                            \n",
    "                        offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                        offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "                                            \n",
    "                    elif genotype == Mat_genotype or genotype == Pat_genotype or genotype == Mat_genotype[::-1] or genotype == Pat_genotype[::-1]:\n",
    "                            \n",
    "                        offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                        offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "                                \n",
    "                    elif sample.gt_bases == \"%s/%s\" % (Par_minor_allele,Par_minor_allele):\n",
    "                        log_file.write(\"\\n%s is homozygous for minor parental allele (%s), likely allele dropout in offspring (if rare) or parent (if common)\" % (name, sample.gt_bases))\n",
    "                        offspr_w_allele_dropout += 1\n",
    "                            \n",
    "                        offspring_gt_types.append(None) # Count as missing data\n",
    "                        offspring_gt_bases_dict[name] = (None, None)\n",
    "                        N_offsp_hom_for_minor_par_allele += 1\n",
    "                            \n",
    "                    else:\n",
    "                        off_ref = sample.gt_bases.split(\"/\")[0]\n",
    "                        off_alt = sample.gt_bases.split(\"/\")[1]\n",
    "                                \n",
    "                        if off_ref not in Parental_alleles or off_alt not in Parental_alleles:\n",
    "                            offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                            offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "                                \n",
    "                            log_file.write(\"\\n%s contains non-parental allele in genotype (%s), likely allele dropout in a Parent\" % (name, sample.gt_bases))\n",
    "                            Non_par_allele = True\n",
    "                \n",
    "            counted = Counter(offspring_gt_types) ## Count the numbers of each genotype at this locus\n",
    "            \n",
    "            ## Caluculate the percentage of offspring missing at the locus\n",
    "            \n",
    "            if None in counted:\n",
    "                perc_offspring_missing = counted[None]/N_offspring\n",
    "            else:\n",
    "                perc_offspring_missing = 0\n",
    "            \n",
    "            ## calculate the percentages of homozygotes/heterozygotes for mendelian frequency filters below\n",
    "            \n",
    "        \n",
    "            \n",
    "            perc_hom_0 = counted[0]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_hom_2 = counted[2]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_het = counted[1]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_off_w_all_dropout = offspr_w_allele_dropout/N_offspring\n",
    "                  \n",
    "            \n",
    "            ## Filter loci with non-mendelian genotype proportions and write data to repsective map files\n",
    "            if perc_offspring_missing > 1-offspring_presence_threshold:\n",
    "                log_file.write(\"\\nToo many missing genotypes, locus not used\")\n",
    "            \n",
    "            elif Non_par_allele == True:\n",
    "                log_file.write(\"\\nNon parental allele found, locus not used\")\n",
    "            \n",
    "            elif N_offsp_hom_for_minor_par_allele > 1: ## if there is more than one offspring homozygous for the minor parental allele, then it is likely to be allele dropout in the parent, locus is discarded\n",
    "                log_file.write(\"\\nMore than one offspring homozygous for minor allele, locus not used\")\n",
    "            \n",
    "            elif perc_hom_0 > mendelian_threshold or perc_hom_2 > mendelian_threshold:\n",
    "                log_file.write(\"\\nHomozygosity excess (>%s)\" % (mendelian_threshold))\n",
    "                log_file.write(\"\\nLocus discarded, Perc_hom_0= %s, Perc_hom_2= %s\" % (perc_hom_0, perc_hom_2))\n",
    "                loc_w_excess_hom += 1\n",
    "            \n",
    "            #elif perc_off_w_all_dropout > \n",
    "        \n",
    "            # Get data for Female map--------------------------------------------------------------------------------------------------------               \n",
    "                \n",
    "            elif mat_het == True and pat_het == False:\n",
    "                log_file.write(\"\\n\\n###Used in Female map###\\n\")\n",
    "                loci_used_for_female_map += 1\n",
    "                femalemap_offspring_haploid_genotypes.append(\"\\n%s\" % (Loc_Id)) ## make line for the record (not used unless criteria below are met)\n",
    "                femalemap_offspring_haploid_genotypes_compli.append(\"\\ncompli_%s\" % (Loc_Id)) ## make line for the complimentary record (not usd unless criteria below are met)\n",
    "                \n",
    "                for i in offspring: ## retain order throughout\n",
    "                    if i in sample_names: ## make sure all samples in the pop_codes file are in the vcf\n",
    "                        if offspring_gt_bases_dict[i][1] == Mat_genotype_bases:\n",
    "                            off_MST = \"B\"\n",
    "                            off_compli = \"a\"\n",
    "                            femalemap_offspring_haploid_genotypes.append(\"B\")\n",
    "                            femalemap_offspring_haploid_genotypes_compli.append(\"a\")\n",
    "                        elif offspring_gt_bases_dict[i][1] == Pat_genotype_bases:\n",
    "                            off_MST = \"A\"\n",
    "                            off_compli = \"b\"\n",
    "                            femalemap_offspring_haploid_genotypes.append(\"A\")\n",
    "                            femalemap_offspring_haploid_genotypes_compli.append(\"b\")\n",
    "                        elif offspring_gt_bases_dict[i][1] == None:\n",
    "                            off_MST = \"U\"\n",
    "                            off_compli = \"U\"\n",
    "                            femalemap_offspring_haploid_genotypes.append(\"U\")\n",
    "                            femalemap_offspring_haploid_genotypes_compli.append(\"U\")\n",
    "                        else:\n",
    "                            log_file.write(\"\\nNo genotype match, Mat_gt = %s, Pat_gt = %s, off_gt = %s\" %(Mat_genotype_bases, Pat_genotype_bases, offspring_gt_bases_dict[i][1]))\n",
    "                            femalemap_offspring_haploid_genotypes.append(\"U\")\n",
    "                            femalemap_offspring_haploid_genotypes_compli.append(\"U\")\n",
    "                            \n",
    "                femmap_file_lines.append(\"\\t\".join(femalemap_offspring_haploid_genotypes))\n",
    "                femmap_file_lines.append(\"\\t\".join(femalemap_offspring_haploid_genotypes_compli))\n",
    "                \n",
    "                log_file.write(\"\\n\\nPercentage homozygous offspring for reference allele: %s\" % (perc_hom_0))\n",
    "                log_file.write(\"\\nPercentage homozygous offspring for alternative allele: %s\" % (perc_hom_2))\n",
    "                log_file.write(\"\\nPercentage heterozygous offspring for reference allele: %s\" % (perc_het))\n",
    "                log_file.write(\"\\nPerc offspring with probable allele dropout: %s\" % (offspr_w_allele_dropout/N_offspring))\n",
    "                            \n",
    "            # Do the same to get data forMale map----------------------------------------------------------------------------------------------                \n",
    "            \n",
    "            elif pat_het == True and mat_het == False: \n",
    "                log_file.write(\"\\n\\n###Used in Male map###\\n\")\n",
    "                loci_used_for_male_map += 1\n",
    "                malemap_offspring_haploid_genotypes.append(\"\\n%s\" % (Loc_Id)) ## make line for the record (not used unless criteria below are met)\n",
    "                malemap_offspring_haploid_genotypes_compli.append(\"\\ncompli_%s\" % (Loc_Id)) ## make line for the complimentary record (not used unless criteria below are met)            \n",
    "                    \n",
    "                for i in offspring:\n",
    "                    if i in sample_names:\n",
    "                        if offspring_gt_bases_dict[i][1] == Mat_genotype_bases:\n",
    "                            off_MST = \"A\"\n",
    "                            off_compli = \"b\"\n",
    "                            malemap_offspring_haploid_genotypes.append(\"A\")\n",
    "                            malemap_offspring_haploid_genotypes_compli.append(\"b\")\n",
    "                        elif offspring_gt_bases_dict[i][1] == Pat_genotype_bases:\n",
    "                            off_MST = \"B\"\n",
    "                            off_compli = \"a\"\n",
    "                            malemap_offspring_haploid_genotypes.append(\"B\")\n",
    "                            malemap_offspring_haploid_genotypes_compli.append(\"a\")\n",
    "                        elif offspring_gt_bases_dict[i][1] == None:\n",
    "                            off_MST = \"U\"\n",
    "                            off_compli = \"U\"\n",
    "                            malemap_offspring_haploid_genotypes.append(\"U\")\n",
    "                            malemap_offspring_haploid_genotypes_compli.append(\"U\")\n",
    "                        else:\n",
    "                            log_file.write(\"\\nNo genotype match, Mat_gt = %s, Pat_gt = %s, off_gt = %s\" %(Mat_genotype_bases, Pat_genotype_bases, offspring_gt_bases_dict[i][1]))\n",
    "                            malemap_offspring_haploid_genotypes.append(\"\\tU\")\n",
    "                            malemap_offspring_haploid_genotypes_compli.append(\"\\tU\")\n",
    "                \n",
    "                malmap_file_lines.append(\"\\t\".join(malemap_offspring_haploid_genotypes))\n",
    "                malmap_file_lines.append(\"\\t\".join(malemap_offspring_haploid_genotypes_compli))\n",
    "                \n",
    "                \n",
    "                log_file.write(\"\\n\\nPercentage homozygous offspring for reference allele: %s\" % (perc_hom_0))\n",
    "                log_file.write(\"\\nPercentage homozygous offspring for alternative allele: %s\" % (perc_hom_2))\n",
    "                log_file.write(\"\\nPercentage heterozygous offspring for reference allele: %s\" % (perc_het))\n",
    "                log_file.write(\"\\nPerc offspring with probable allele dropout= %s\" % (offspr_w_allele_dropout/N_offspring))\n",
    "    \n",
    "    log_file.write(\"\\n\\nSUMMARY:\\nN loci used in female map: %s\" % (loci_used_for_female_map))\n",
    "    log_file.write(\"\\nN loci used in male map: %s\" % (loci_used_for_male_map))\n",
    "    \n",
    "    \n",
    "    ## Make file headers\n",
    "    N_offspring = 0\n",
    "    Sample_headers = []        \n",
    "    Sample_headers.append(\"locus_name\")\n",
    "    for i in offspring:\n",
    "        if i in sample_names:\n",
    "            N_offspring += 1\n",
    "            Sample_headers.append(\"\\t%s\" % (i))\n",
    "    Sample_header_line = ''.join(Sample_headers)\n",
    "    \n",
    "    \n",
    "    fem_header = \"\\\n",
    "    population_type DH\\n\\\n",
    "    population_name Female_map\\n\\\n",
    "    distance_function kosambi\\n\\\n",
    "    cut_off_p_value 0.000005\\n\\\n",
    "    no_map_dist 30\\n\\\n",
    "    no_map_size 1\\n\\\n",
    "    missing_threshold 0.25\\n\\\n",
    "    estimation_before_clustering no\\n\\\n",
    "    detect_bad_data yes\\n\\\n",
    "    objective_function ML\\n\\\n",
    "    number_of_loci %s\\n\\\n",
    "    number_of_individual %s\\n\\n\" % (loci_used_for_female_map*2, N_offspring)\n",
    "    \n",
    "    male_header = \"\\\n",
    "    population_type DH\\n\\\n",
    "    population_name Female_map\\n\\\n",
    "    distance_function kosambi\\n\\\n",
    "    cut_off_p_value 0.000005\\n\\\n",
    "    no_map_dist 30\\n\\\n",
    "    no_map_size 1\\n\\\n",
    "    missing_threshold 0.25\\n\\\n",
    "    estimation_before_clustering no\\n\\\n",
    "    detect_bad_data yes\\n\\\n",
    "    objective_function ML\\n\\\n",
    "    number_of_loci %s\\n\\\n",
    "    number_of_individual %s\\n\\n\" % (loci_used_for_male_map*2, N_offspring)\n",
    "    \n",
    "    femMSTfile.write(fem_header)\n",
    "    femMSTfile.write(Sample_header_line)\n",
    "    \n",
    "    for line in femmap_file_lines:\n",
    "        femMSTfile.write(line)\n",
    "    \n",
    "    malMSTfile.write(male_header)\n",
    "    malMSTfile.write(Sample_header_line)\n",
    "    \n",
    "    for line in malmap_file_lines:\n",
    "        malMSTfile.write(line)\n",
    "        \n",
    "        \n",
    "    femMSTfile.close()\n",
    "    malMSTfile.close()\n",
    "    log_file.close()\n",
    "                   \n",
    "    print \"N loci used in female map: %s\" % (loci_used_for_female_map)\n",
    "    print \"N loci used in male map: %s\" % (loci_used_for_male_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Cline\n",
    "\n",
    "if len(sys.argv) == 1:\n",
    "    # sys.exit(help(VCF_MSTmap_converter))\n",
    "    print VCF_MSTmap_converter.__doc__\n",
    "elif len(sys.argv) < 4: ## If not enough args are supplied print error message\n",
    "    sys.exit(\"\\n##Error, not enough arguments, run script with no arguments to see help message\\n\")\n",
    "\n",
    "elif len(sys.argv) == 4:\n",
    "    specified_vcfpath = sys.argv[1]\n",
    "    sex_info_file = sys.argv[2]\n",
    "    sample_presence_threshold = float(sys.argv[3])\n",
    "\n",
    "    VCF_MSTmap_converter(specified_vcfpath, sex_info_file, sample_presence_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Args\n",
    "\n",
    "vcf_file = \"/home/djeffrie/Data/Pperezi/Stacks_outs/batch_1.haplotypes1000.vcf\"\n",
    "mend_thresh = 0.75\n",
    "offsp_pres = 0.75\n",
    "sex_info = \"/home/djeffrie/Data/Pperezi/Stacks_outs/pop_map_link.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 513,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N loci used in female map: 10\n",
      "N loci used in male map: 11\n"
     ]
    }
   ],
   "source": [
    "MST_input_maker(vcf_file, sex_info, offsp_pres, mend_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Informative locus filters:\n",
    "    \n",
    "    Loci are used only if the number of offspring present is equal to or above the percentage specified by the user\n",
    "    \n",
    "    Loci are used only if they are heterozygous in one sex and homozygous in another.\n",
    "    \n",
    "\n",
    "###Null allele filters:\n",
    "Null allele in Parents:\n",
    "\n",
    "    If an offspring is found to have an allele that isn't present in the parent, this locus is discarded\n",
    "\n",
    "Null allele in offspring:\n",
    "\n",
    "    If only 1 offspring is homozygous for the parental minor allele then that sample is assumed to possess a null allele and the data for that individual is coded as missing. **Note, the sample presence/absence filter works after this step, so some loci may be pushed over the \"missing data\" limit by this null allele filter. \n",
    "    \n",
    "    If two or more samples are found to be homozygous for the minor parental allele, it is more likely that this is due to the homozygous parent having a null allele. In this case, the locus is again discarded.\n",
    "    \n",
    "    Also, loci with non-mendelian properties indicative of allele dropout or similar are also filtered"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
