{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bootstrapping sex linked marker sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import MISC_RAD_tools as MISC\n",
    "import SLMF_lightweight as SLMF_L"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook I will first compile all of the final stacks outputs for each species dataset (or subset) and then, for each one, I will randomise the male and female assignments accross the samples. For 100 random male female assignments I will then run the sex linked marker finding analyses, using exactly the parameters used to identify the final set of sex linked markers in the paper. The only difference is the male and female assignments. \n",
    "\n",
    "This randomisation will give an idea of what the false positive rate in the dataset is. For example, a skew towards more males or females in the dataset may make false positives of one type or another more likely. Also, if there are several populations in the data and males and females are not distributed evenly among them then population structure could look like sex linkage. However, randomising male and female assignments across all samples will allow us to account for this. \n",
    "\n",
    "Due to the prohibitively long amount of time that it would take, I will not do 1000 randomisations for each species, although this is what I would prefer. Instead I will just do 100, this should still give a reasonably good estimate for the false positive rate. \n",
    "\n",
    "With regards to how these randomisations will be used to judge the validity of the dataset, I will look for sample sets where the number of sex-linked markers found using the correct male/female assignments is above the 95th percentile of the distribution of sex-linked markers found in the randomisations. \n",
    "\n",
    "In collaboration with the genome mapping, this should help validate the sex-linked markers sets found.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### First, make a list of all of the parameter disctionaries, which contain paths and parameters used for finding sex linked markers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "Parameter_dictionaries = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. chiricahuensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Lchiri = {}\n",
    "Parameter_dict_Lchiri[\"Name\"] = \"Lchiri\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Lchiri['Catalog'] =  \"/home/djeffrie/Data/RADseq/Lchricahuensis/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Lchiri['VCF'] =  \"/home/djeffrie/Data/RADseq/Lchricahuensis/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Lchiri['Pop_map'] = \"/home/djeffrie/Data/RADseq/Lchricahuensis/Sex_ID_info_reassigned.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Lchiri['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Lchiri['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Lchiri['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Lchiri['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Lchiri['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Lchiri['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Lchiri['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Lchiri['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Lchiri['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Lchiri['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Lchiri['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Lchiri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. montezumae\n",
    "\n",
    "No sex linked markers found"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. pipiens (HOR / SWE populations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_LpipHORSWE = {}\n",
    "Parameter_dict_LpipHORSWE[\"Name\"] = \"LpipHORSWE\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_LpipHORSWE['Catalog'] =  \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_LpipHORSWE['VCF'] =  \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/HOR_SWE_populations/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_LpipHORSWE['Pop_map'] = \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/HOR_SWE_populations/Sex_ID_info_kept_certain.txt\"\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_LpipHORSWE['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_LpipHORSWE['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_LpipHORSWE['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_LpipHORSWE['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_LpipHORSWE['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_LpipHORSWE['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_LpipHORSWE['heterogamtic_heterozygosity_threshold'] = 0.8 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_LpipHORSWE['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_LpipHORSWE['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_LpipHORSWE['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_LpipHORSWE['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_LpipHORSWE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. tarahumarae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ltarah = {}\n",
    "Parameter_dict_Ltarah[\"Name\"] = \"Ltarah\"\n",
    "\n",
    "working_dir = \"/home/djeffrie/Data/RADseq/Ltarahumarae/Sex_linked_markers/\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ltarah['Catalog'] =  \"%s/batch_1.catalog.tags.tsv.gz\" % working_dir ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ltarah['VCF'] =  \"%s/batch_1.vcf\"  % working_dir ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ltarah['Pop_map'] = \"%s/Sex_ID_info_kept_2.txt\"  % working_dir ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ltarah['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ltarah['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ltarah['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ltarah['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ltarah['homogametic_REF_allele_freq'] = 0.9 ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ltarah['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ltarah['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ltarah['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ltarah['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ltarah['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ltarah['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ltarah)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. berlandieri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_R_ber = {}\n",
    "Parameter_dict_R_ber[\"Name\"] = \"R_ber\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_R_ber['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_R_ber['VCF'] =  \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_R_ber['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/Sex_ID_info_kept_2.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_R_ber['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_R_ber['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_R_ber['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_R_ber['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_R_ber['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_R_ber['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_R_ber['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_R_ber['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_R_ber['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_R_ber['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_R_ber['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_R_ber)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### P. perezi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Pper = {}\n",
    "Parameter_dict_Pper[\"Name\"] = \"Pper\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Pper['Catalog'] =  \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Pper['VCF'] =  \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Pper['Pop_map'] = \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/Sex_info_ID.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Pper['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Pper['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Pper['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Pper['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Pper['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Pper['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Pper['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Pper['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Pper['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Pper['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Pper['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Pper)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. arvalis (all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rarv = {}\n",
    "Parameter_dict_Rarv[\"Name\"] = \"Rarv\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rarv['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rarvalis_NEW/Stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rarv['VCF'] =  \"/home/djeffrie/Data/RADseq/Rarvalis_NEW/Stacks/Populations_kept_2/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rarv['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rarvalis_NEW//Stacks/Sex_ID_info_kept_2.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rarv['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rarv['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rarv['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rarv['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rarv['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rarv['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rarv['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rarv['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rarv['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rarv['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rarv['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rarv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. chensinensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rchen = {}\n",
    "Parameter_dict_Rchen[\"Name\"] = \"Rchen\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rchen['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rchen['VCF'] =  \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/batch_1_strict_kept_3.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rchen['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/Sex_ID_info_kept_3.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rchen['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rchen['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rchen['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rchen['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rchen['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rchen['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rchen['heterogamtic_heterozygosity_threshold'] = 0.7 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rchen['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rchen['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rchen['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rchen['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rchen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. dalmatina "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rdal = {}\n",
    "Parameter_dict_Rdal[\"Name\"] = \"Rdal\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rdal['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rdal['VCF'] =  \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rdal['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rdal['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rdal['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rdal['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rdal['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rdal['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rdal['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rdal['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rdal['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rdal['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rdal['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rdal['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rdal)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. iberica all northern Spain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ribe_Sp = {}\n",
    "Parameter_dict_Ribe_Sp[\"Name\"] = \"Ribe_Sp\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ribe_Sp['Catalog'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ribe_Sp['VCF'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ribe_Sp['Pop_map'] = \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/Sex_ID_info_N_spain.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ribe_Sp['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ribe_Sp['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_Sp['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_Sp['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ribe_Sp['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ribe_Sp['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ribe_Sp['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ribe_Sp['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_Sp['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_Sp['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ribe_Sp['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ribe_Sp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. iberica \"family\" (which isn't actually a family)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ribe_fam = {}\n",
    "Parameter_dict_Ribe_fam[\"Name\"] = \"Ribe_fam\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ribe_fam['Catalog'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ribe_fam['VCF'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ribe_fam['Pop_map'] = \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ribe_fam['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ribe_fam['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_fam['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_fam['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ribe_fam['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ribe_fam['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ribe_fam['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ribe_fam['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_fam['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_fam['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ribe_fam['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ribe_fam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. italica (CM pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rita = {}\n",
    "Parameter_dict_Rita[\"Name\"] = \"Rita\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rita['Catalog'] =  \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rita['VCF'] =  \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rita['Pop_map'] = \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/CM_sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rita['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rita['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rita['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rita['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rita['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rita['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rita['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rita)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. italica (Fam)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rita_fam = {}\n",
    "Parameter_dict_Rita_fam[\"Name\"] = \"Rita_fam\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rita_fam['Catalog'] =  \"/home/djeffrie/Data/RADseq/Ritalica3/Stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rita_fam['VCF'] =  \"/home/djeffrie/Data/RADseq/Ritalica3/Stacks/Populations_all/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rita_fam['Pop_map'] = \"/home/djeffrie/Data/RADseq/Ritalica3/Stacks/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rita_fam['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rita_fam['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita_fam['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita_fam['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rita_fam['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rita_fam['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rita_fam['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rita_fam['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita_fam['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita_fam['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rita_fam['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rita_fam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. kukinoris (Nanping only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rkuk = {}\n",
    "Parameter_dict_Rkuk[\"Name\"] = \"Rkuk\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rkuk['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rkuk['VCF'] =  \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rkuk['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rkuk['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rkuk['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rkuk['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rkuk['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rkuk['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rkuk['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rkuk['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rkuk['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rkuk['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rkuk['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rkuk['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rkuk)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. yavapaiensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ryav = {}\n",
    "Parameter_dict_Ryav[\"Name\"] = \"Ryav\"\n",
    "\n",
    "## NOTE. DICT KEY NAMES MUST NOT BE CHANGED!\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ryav['Catalog'] =  \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ryav['VCF'] =  \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/batch_1.vcf.altered\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ryav['Pop_map'] = \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ryav['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ryav['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ryav['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ryav['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ryav['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ryav['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ryav['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ryav['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ryav['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ryav['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ryav['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ryav)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6 species for which sex linked markers were found that will be randomised\n"
     ]
    }
   ],
   "source": [
    "print \"There are %s species for which sex linked markers were found that will be randomised\" % len(Parameter_dictionaries)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The program will work as follows:\n",
    "\n",
    "For every species, it will\n",
    "    1. make a new directory\n",
    "    2. Move the sex information file and vcf file into that directory\n",
    "    3. Make the randomised sex information files and renamed vcf files\n",
    "    4. Run the randomisation prgroam for that species and collect results\n",
    "    5. Write results to an output file\n",
    "    6. Tidy up loose files\n",
    "    7. Plot the results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Define a program to help parallelise the analyses\n",
    "\n",
    "\n",
    "def Super_SLM_finder_parallel(popmap, Parameter_dict):\n",
    "    import os\n",
    "    \n",
    "    Parameter_dict['Pop_map'] = popmap\n",
    "    Parameter_dict['VCF'] = \"%s.vcf\" % popmap.rpartition(\".\")[0]\n",
    "    results_dict = {}\n",
    "    results_dict[\"XYset\"], results_dict[\"ZWset\"], results_dict[\"Detailed\"] = SLMF_L.Super_SLM_finder(Parameter_dict, \"111\", verbose = False, write_files=False, plot=False)\n",
    "    \n",
    "    os.remove(Parameter_dict['VCF']) ## remove VCFs as they are used\n",
    "    os.remove(\"%s.altered\" % Parameter_dict['VCF'])\n",
    "    os.remove(\"%s.all_frequencies.tsv\" % Parameter_dict['VCF'])\n",
    "    \n",
    "\n",
    "    return results_dict\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing dataset in /home/djeffrie/Data/RADseq/Ritalica3/Stacks/Populations_all/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Ritalica3/Stacks/Populations_all/Randomisations\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:  5.2min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 12.9min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Ritalica3/Stacks/Populations_all/Randomisations/Randomisations_Rita_fam.txt\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 1\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "X_het markers -> X_het markers -> X_het markers -> X_het markers ->  []\n",
      " []\n",
      " []\n",
      " []\n",
      "Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "Final number of ZW tags = 1\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n"
     ]
    }
   ],
   "source": [
    "from joblib import Parallel, delayed\n",
    "import multiprocessing\n",
    "import os\n",
    "import shutil\n",
    "from random import shuffle\n",
    "\n",
    "results_dict = {}\n",
    "\n",
    "for dataset in Parameter_dictionaries:\n",
    "    print \"processing dataset in\", dataset[\"VCF\"]\n",
    "    # 1. Make a new directory next in the VCF path. \n",
    "    Randomisation_dir = \"%s/Randomisations\" % dataset[\"VCF\"].rpartition(\"/\")[0]\n",
    "    if not os.path.exists(Randomisation_dir):\n",
    "        os.makedirs(Randomisation_dir)\n",
    "    \n",
    "    # 2. Copy sex info and VCF into that folder\n",
    "    \n",
    "    shutil.copyfile(dataset[\"Pop_map\"], \"%s/Sex_ID_info.txt\" % Randomisation_dir)  ## sex info\n",
    "    shutil.copyfile(dataset[\"VCF\"], \"%s/batch_1.vcf\" % Randomisation_dir)  ## sex info\n",
    "    \n",
    "    orig_vcf = \"%s/batch_1.vcf\" % Randomisation_dir\n",
    "    \n",
    "    print \"\\nRandomisations happening in %s\" % Randomisation_dir\n",
    "    \n",
    "    ## 3. Make the randomised sex info files\n",
    "\n",
    "    orig_popmap_path = \"%s/Sex_ID_info.txt\" % Randomisation_dir\n",
    "    orig_popmap = open(orig_popmap_path, 'r').readlines()\n",
    "\n",
    "    sexes = []\n",
    "    samples = []\n",
    "    IDs = []\n",
    "    randomisations = []\n",
    "\n",
    "    for line in orig_popmap:\n",
    "        sexes.append(line.strip().split()[1])\n",
    "        samples.append(line.strip().split()[0])\n",
    "        IDs.append(line.strip().split()[2])\n",
    "\n",
    "    popmaps = []\n",
    "    for i in range(100):\n",
    "        popmap_path = \"%s/rand_popmap_%s.txt\" % (Randomisation_dir, i)\n",
    "        popmaps.append(popmap_path)\n",
    "        rand_popmap = open(popmap_path, 'w')\n",
    "        shuffle(sexes)\n",
    "        randomisations.append(sexes)\n",
    "        \n",
    "        for i in range(len(samples)):\n",
    "            rand_popmap.write(\"%s\\t%s\\t%s\\n\" % (samples[i], sexes[i], IDs[i]))\n",
    "\n",
    "        rand_popmap.close()\n",
    "    \n",
    "    print \"\\nRandom sex info files made\"\n",
    "    \n",
    "    ## make new VCFs for parallelised analyses\n",
    "    \n",
    "    for i in popmaps:\n",
    "        new_vcf = \"%s.vcf\" % i.rpartition(\".\")[0]\n",
    "        shutil.copyfile(orig_vcf, new_vcf)\n",
    "        \n",
    "    ## 4. Run the randomisations\n",
    "    \n",
    "    print \"\\nRunning randomisations\\n\"\n",
    "    \n",
    "    results_dict[dataset[\"Name\"]] = Parallel(n_jobs=4, verbose = 1)(delayed(Super_SLM_finder_parallel)(i, dataset) for i in popmaps)\n",
    "    \n",
    "    ## 5. Output the results for each species after the species is complete (i.e. checkpoints)\n",
    "    \n",
    "    outfile = open(\"%s/Randomisations_%s.txt\" % (Randomisation_dir, dataset[\"Name\"]), 'w')\n",
    "    \n",
    "    for Randomisation in results_dict[dataset[\"Name\"]]:\n",
    "\n",
    "        XYfreq = len(Randomisation[\"Detailed\"][\"XY\"][\"freq\"])\n",
    "        XYhet = len(Randomisation[\"Detailed\"][\"XY\"][\"het\"])\n",
    "        Ytags = len(Randomisation[\"Detailed\"][\"XY\"][\"Ytags\"])\n",
    "    \n",
    "        ZWfreq = len(Randomisation[\"Detailed\"][\"ZW\"][\"freq\"])\n",
    "        ZWhet = len(Randomisation[\"Detailed\"][\"ZW\"][\"het\"])\n",
    "        Wtags = len(Randomisation[\"Detailed\"][\"ZW\"][\"Wtags\"])\n",
    "                \n",
    "        line = \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (dataset[\"Name\"],XYfreq,XYhet,Ytags,ZWfreq,ZWhet,Wtags)\n",
    "    \n",
    "        outfile.write(line)\n",
    "        \n",
    "    outfile.close()\n",
    "    \n",
    "    print \"Results outputted to %s/Randomisations_%s.txt\" % (Randomisation_dir, dataset[\"Name\"])\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## So all went well, randomisations completed in less than 24 hours with 100 randomisations per dataset. \n",
    "\n",
    "### Now to plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def removeOutliers(x, outlierConstant):\n",
    "    \"\"\"\n",
    "    An outlier here is defined as a value more than 1.5* the IQ range\n",
    "    but constants higher than 1.5 can be used to be more stringent\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    a = np.array(x)\n",
    "    upper_quartile = np.percentile(a, 75)\n",
    "    lower_quartile = np.percentile(a, 25)\n",
    "    IQR = (upper_quartile - lower_quartile) * outlierConstant\n",
    "    quartileSet = (lower_quartile - IQR, upper_quartile + IQR)\n",
    "    resultList = []\n",
    "    for y in a.tolist():\n",
    "        if y > quartileSet[0] and y < quartileSet[1]:\n",
    "            resultList.append(y)\n",
    "    return resultList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def randomisation_plotter(Randomisation_directory, realdatapath, Outs_constant = 1.5, plot = True):\n",
    "    \n",
    "    import os\n",
    "    from matplotlib import pyplot as plt\n",
    "    import numpy as np\n",
    "    \n",
    "    randomisation_files = []\n",
    "    \n",
    "    for root, dirs, files in os.walk(Randomisation_directory):\n",
    "        for fil in files:\n",
    "            if fil.startswith(\"Randomisation\") and fil.endswith(\"txt\") and \"sigs\" not in fil:\n",
    "                randomisation_files.append(\"%s/%s\" % (root, fil))\n",
    "\n",
    "    rando_sig_file = open(\"%s/Randomisation_sigs.txt\" % randomisation_directory, 'w')\n",
    "            \n",
    "    rando_sig_file.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (\"species\", \\\n",
    "                                 \"realXYfreq_sig\", \\\n",
    "                                 \"realXYhet_sig\", \\\n",
    "                                 \"Ytags_sig\", \\\n",
    "                                 \"ZW_freqs_sig\", \\\n",
    "                                 \"ZW_hets_sig\", \\\n",
    "                                 \"W_tagss_sig\", \\\n",
    "                                 \"XYvsZW_freqs_sig\",\\\n",
    "                                 \"XYvsZW_tagss_sig\"))\n",
    "    \n",
    "    for randomisation_filepath in randomisation_files:\n",
    "        \n",
    "        print \"\\nPlotting %s\\n\" % randomisation_filepath.rpartition(\"/\")[2].rpartition(\".\")[0]\n",
    "\n",
    "        randomisations = open(randomisation_filepath, 'r').readlines()\n",
    "\n",
    "        XYfreqs = []\n",
    "        XYhets = []\n",
    "        Ytags = []\n",
    "\n",
    "        ZW_freqs = []\n",
    "        ZW_hets = []\n",
    "        W_tagss = []\n",
    "\n",
    "        XYvsZW_freqs = []\n",
    "        XYvsZW_hets = []\n",
    "        XYvsZW_tagss = []\n",
    "\n",
    "        ## Get randomisation data\n",
    "        \n",
    "        for line in randomisations:\n",
    "            #print line.split()\n",
    "            species = line.split()[0]\n",
    "            XYfreq = line.split()[1]\n",
    "            XYhet = line.split()[2]\n",
    "            Y_tags = line.split()[3]\n",
    "\n",
    "            ZW_freq = line.split()[4]\n",
    "            ZW_het = line.split()[5]\n",
    "            W_tags = line.split()[6]\n",
    "\n",
    "            XYvsZW_freq = int(XYfreq) - int(ZW_freq)\n",
    "            XYvsZW_het = int(XYhet) - int(ZW_het)\n",
    "            XYvsZW_tags = int(Y_tags) - int(W_tags)\n",
    "\n",
    "            XYfreqs.append(int(XYfreq))\n",
    "            XYhets.append(int(XYhet))\n",
    "            Ytags.append(int(Y_tags))\n",
    "            ZW_freqs.append(int(ZW_freq))\n",
    "            ZW_hets.append(int(ZW_het))\n",
    "            W_tagss.append(int(W_tags))\n",
    "            XYvsZW_freqs.append(int(XYvsZW_freq))\n",
    "            XYvsZW_hets.append(int(XYvsZW_het))\n",
    "            XYvsZW_tagss.append(int(XYvsZW_tags))\n",
    "    \n",
    "\n",
    "        ## Get the real data from the file\n",
    "        realdata = open(realdatapath, 'r').readlines()\n",
    "\n",
    "        for line in realdata:\n",
    "            if species == line.split()[0]:\n",
    "                realXYfreq = int(line.split()[1])\n",
    "                realXYhet = int(line.split()[2])\n",
    "                realYtags = int(line.split()[3])\n",
    "                realZWfreq = int(line.split()[4])\n",
    "                realZWhet = int(line.split()[5])\n",
    "                realWtags = int(line.split()[6])\n",
    "        \n",
    "\n",
    "        realXYvsZWfreq = realXYfreq - realZWfreq\n",
    "        realXYvsZWhet = realXYhet - realZWhet\n",
    "        realYvsWtags = realYtags - realWtags\n",
    "\n",
    "        \n",
    "        ## Test to see if the real data is significantly higher than the randomisations\n",
    "        \n",
    "        ## First check that the IQR is greater than zero, otherwise there the downstream stats won't work\n",
    "            ## if the IQR is greater than zero, remove any outliers in the datasets\n",
    "        \n",
    "        if not (np.percentile(XYfreqs, 75) - np.percentile(XYfreqs, 25)) * Outs_constant == 0:\n",
    "            XYfreqs_no_outs = removeOutliers(XYfreqs, Outs_constant)\n",
    "        else:\n",
    "            XYfreqs_no_outs = XYfreqs\n",
    "            \n",
    "        if not (np.percentile(XYhets, 75) - np.percentile(XYhets, 25)) * Outs_constant == 0:\n",
    "            XYhets_no_outs = removeOutliers(XYhets, Outs_constant)\n",
    "        else:\n",
    "            XYhets_no_outs = XYhets\n",
    "            \n",
    "        if not (np.percentile(Ytags, 75) - np.percentile(Ytags, 25)) * Outs_constant == 0:\n",
    "            Ytags_no_outs = removeOutliers(Ytags, Outs_constant)\n",
    "        else:\n",
    "            Ytags_no_outs = Ytags\n",
    "            \n",
    "        if not (np.percentile(ZW_freqs, 75) - np.percentile(ZW_freqs, 25)) * Outs_constant == 0:\n",
    "            ZW_freqs_no_outs = removeOutliers(ZW_freqs, Outs_constant)\n",
    "        else:\n",
    "            ZW_freqs_no_outs = ZW_freqs\n",
    "            \n",
    "        if not (np.percentile(ZW_hets, 75) - np.percentile(ZW_hets, 25)) * Outs_constant == 0:\n",
    "            ZW_hets_no_outs = removeOutliers(ZW_hets, Outs_constant)\n",
    "        else:\n",
    "            ZW_hets_no_outs = ZW_hets\n",
    "            \n",
    "        if not (np.percentile(W_tagss, 75) - np.percentile(W_tagss, 25)) * Outs_constant == 0:\n",
    "            W_tagss_no_outs = removeOutliers(W_tagss, Outs_constant)\n",
    "        else:\n",
    "            W_tagss_no_outs = W_tagss\n",
    "        \n",
    "        \n",
    "        if not (np.percentile(XYvsZW_freqs, 75) - np.percentile(XYvsZW_freqs, 25)) * Outs_constant == 0:\n",
    "            XYvsZW_freqs_no_outs = removeOutliers(XYvsZW_freqs, Outs_constant)\n",
    "        else:\n",
    "            XYvsZW_freqs_no_outs = XYvsZW_freqs\n",
    "        \n",
    "        if not (np.percentile(XYvsZW_hets, 75) - np.percentile(XYvsZW_hets, 25)) * Outs_constant == 0:\n",
    "            XYvsZW_hets_no_outs = removeOutliers(XYvsZW_hets, Outs_constant)\n",
    "        else:\n",
    "            XYvsZW_hets_no_outs = XYvsZW_hets\n",
    "            \n",
    "        if not (np.percentile(XYvsZW_tagss, 75) - np.percentile(XYvsZW_tagss, 25)) * Outs_constant == 0:\n",
    "            XYvsZW_tagss_no_outs = removeOutliers(XYvsZW_tagss, Outs_constant)\n",
    "        else:\n",
    "            XYvsZW_tagss_no_outs = XYvsZW_tagss\n",
    "\n",
    "        ## set \"Non-sig\" as default\n",
    "        \n",
    "        realXYfreq_sig = \"Non_sig\"\n",
    "        realXYhet_sig = \"Non_sig\"\n",
    "        Ytags_sig = \"Non_sig\"\n",
    "        ZW_freqs_sig = \"Non_sig\"\n",
    "        ZW_hets_sig = \"Non_sig\"\n",
    "        W_tagss_sig = \"Non_sig\"\n",
    "        XYvsZW_freqs_sig = \"Non_sig\"\n",
    "        XYvsZW_hets_sig = \"Non_sig\"\n",
    "        XYvsZW_tagss_sig = \"Non_sig\"\n",
    "        \n",
    "        \n",
    "        ## if the real value is higher than the 95th or 99th percentile, then record sig level\n",
    "        \n",
    "        if realXYfreq > np.round(np.percentile(XYfreqs_no_outs, 95),3):\n",
    "            realXYfreq_sig = 95\n",
    "        if realXYfreq > np.round(np.percentile(XYfreqs_no_outs, 99),3):\n",
    "            realXYfreq_sig = 99\n",
    "\n",
    "        if realXYhet > np.round(np.percentile(XYhets_no_outs, 95),3):\n",
    "            realXYhet_sig = 95\n",
    "        if realXYhet > np.round(np.percentile(XYhets_no_outs, 99),3):\n",
    "            realXYhet_sig = 99\n",
    "            \n",
    "        if realYtags > np.round(np.percentile(Ytags_no_outs, 95),3):\n",
    "            Ytags_sig = 95\n",
    "        if realYtags > np.round(np.percentile(Ytags_no_outs, 99),3):\n",
    "            Ytags_sig = 99\n",
    "            \n",
    "        if realZWfreq > np.round(np.percentile(ZW_freqs_no_outs, 95),3):\n",
    "            ZW_freqs_sig = 95\n",
    "        if realZWfreq > np.round(np.percentile(ZW_freqs_no_outs, 99),3):\n",
    "            ZW_freqs_sig = 99\n",
    "            \n",
    "        if realZWhet > np.round(np.percentile(ZW_hets_no_outs, 95),3):\n",
    "            ZW_hets_sig = 95\n",
    "        if realZWhet > np.round(np.percentile(ZW_hets_no_outs, 99),3):\n",
    "            ZW_hets_sig = 99\n",
    "        \n",
    "        if realWtags > np.round(np.percentile(W_tagss_no_outs, 95),3):\n",
    "            W_tagss_sig = 95\n",
    "        if realWtags > np.round(np.percentile(W_tagss_no_outs, 99),3):\n",
    "            W_tagss_sig = 99    \n",
    "        \n",
    "        if realXYvsZWfreq > np.round(np.percentile(XYvsZW_freqs_no_outs, 95),3):\n",
    "            XYvsZW_freqs_sig = 95\n",
    "        if realXYvsZWfreq > np.round(np.percentile(XYvsZW_freqs_no_outs, 99),3):\n",
    "            XYvsZW_freqs_sig = 99\n",
    "            \n",
    "        if realXYvsZWhet > np.round(np.percentile(XYvsZW_hets_no_outs, 95),3):\n",
    "            XYvsZW_hets_sig = 95\n",
    "        if realXYvsZWhet > np.round(np.percentile(XYvsZW_hets_no_outs, 99),3):\n",
    "            XYvsZW_hets_sig = 99\n",
    "        \n",
    "        if realYvsWtags > np.round(np.percentile(XYvsZW_tagss_no_outs, 95),3):\n",
    "            XYvsZW_tagss_sig = 95\n",
    "        if realYvsWtags > np.round(np.percentile(XYvsZW_tagss_no_outs, 99),3):\n",
    "            XYvsZW_tagss_sig = 99    \n",
    "\n",
    "        print \"XYfreq_sig\", realXYfreq_sig\n",
    "        print \"XYhet_sig\", realXYhet_sig\n",
    "        print \"Ytags_sig\", Ytags_sig\n",
    "        print \"ZW_freqs_sig\", ZW_freqs_sig\n",
    "        print \"ZW_hets_sig\", ZW_hets_sig\n",
    "        print \"Wtags_sig\", W_tagss_sig\n",
    "        print \"XYvsZW_freqs_sig\", XYvsZW_freqs_sig\n",
    "        print \"XYvsZW_hets_sig\", XYvsZW_hets_sig\n",
    "        print \"XYvsZW_tagss_sig\", XYvsZW_tagss_sig\n",
    "        \n",
    "        \n",
    "        rando_sig_file.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (species, \\\n",
    "                                 realXYfreq_sig, \\\n",
    "                                 realXYhet_sig, \\\n",
    "                                 Ytags_sig, \\\n",
    "                                 ZW_freqs_sig, \\\n",
    "                                 ZW_hets_sig, \\\n",
    "                                 W_tagss_sig, \\\n",
    "                                 XYvsZW_freqs_sig,\\\n",
    "                                 XYvsZW_tagss_sig))\n",
    "        \n",
    "        if plot == True:\n",
    "\n",
    "            fig = plt.figure(figsize = (20,10))\n",
    "\n",
    "            fig.add_subplot(3,3,1)\n",
    "\n",
    "            counts, bins, bars = plt.hist(XYfreqs, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(XYfreqs) >= realXYfreq:\n",
    "                plt.xlim((0,max(XYfreqs)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realXYfreq+10))\n",
    "            plt.vlines(realXYfreq, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"XY frequency method\")\n",
    "            if realXYfreq_sig == 99 or realXYfreq_sig == 95:\n",
    "                plt.text(realXYfreq+1, max(counts)*0.6, \"Sig: %s\" % realXYfreq_sig)\n",
    "\n",
    "\n",
    "            fig.add_subplot(3,3,2)\n",
    "            counts, bins, bars = plt.hist(XYhets, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(XYhets) >= realXYhet:\n",
    "                plt.xlim((0,max(XYhets)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realXYhet+10))\n",
    "            plt.vlines(realXYhet, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"XY heterozygosity method\")\n",
    "            if realXYhet_sig == 99 or realXYhet_sig == 95:\n",
    "                plt.text(realXYhet+1, max(counts)*0.6, \"Sig: %s\" % realXYhet_sig)\n",
    "\n",
    "                \n",
    "            fig.add_subplot(3,3,3)\n",
    "            counts, bins, bars = plt.hist(Ytags, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(Ytags) >= realYtags:\n",
    "                plt.xlim((0,max(Ytags)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realYtags+10))\n",
    "            plt.vlines(realYtags, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"Y specific tags method\")\n",
    "            if Ytags_sig == 99 or Ytags_sig == 95:\n",
    "                plt.text(realYtags+1, max(counts)*0.6, \"Sig: %s\" % Ytags_sig)\n",
    "                \n",
    "                \n",
    "            fig.add_subplot(3,3,4)\n",
    "            counts, bins, bars = plt.hist(ZW_freqs, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(ZW_freqs) >= realZWfreq:\n",
    "                plt.xlim((0,max(ZW_freqs)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realZWfreq+10))\n",
    "            plt.vlines(realZWfreq, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"ZW frequency method\")\n",
    "            plt.ylabel(\"Number of Randomisations\")\n",
    "            if ZW_freqs_sig == 99 or ZW_freqs_sig == 95:\n",
    "                plt.text(realZWfreq+1, max(counts)*0.6, \"Sig: %s\" % ZW_freqs_sig)\n",
    "                \n",
    "                \n",
    "            fig.add_subplot(3,3,5)\n",
    "            counts, bins, bars = plt.hist(ZW_hets, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(ZW_hets) >= realZWhet:\n",
    "                plt.xlim((0,max(ZW_hets)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realZWhet+10))\n",
    "            plt.vlines(realZWhet, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"ZW heterozygosity method\")\n",
    "            if ZW_hets_sig == 99 or ZW_hets_sig == 95:\n",
    "                plt.text(realZWhet+1, max(counts)*0.6, \"Sig: %s\" % ZW_hets_sig)\n",
    "                \n",
    "\n",
    "            fig.add_subplot(3,3,6)\n",
    "            counts, bins, bars = plt.hist(W_tagss, 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(W_tagss) >= realWtags:\n",
    "                plt.xlim((0,max(W_tagss)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realWtags +10))\n",
    "            plt.vlines(realWtags, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"W specific tags method\")\n",
    "            if W_tagss_sig == 99 or W_tagss_sig == 95:\n",
    "                plt.text(realWtags+1, max(counts)*0.6, \"Sig: %s\" % W_tagss_sig)\n",
    "                \n",
    "                \n",
    "            fig.add_subplot(3,3,7)\n",
    "            counts, bins, bars = plt.hist(XYvsZW_freqs, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(XYvsZW_freqs) >= realXYvsZWfreq:\n",
    "                plt.xlim((0,max(XYvsZW_freqs)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realXYvsZWfreq+10))\n",
    "            plt.vlines(realXYvsZWfreq, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"XY - ZW frequency method\")\n",
    "            if XYvsZW_freqs_sig == 99 or XYvsZW_freqs_sig == 95:\n",
    "                plt.text(realXYvsZWfreq+1, max(counts)*0.6, \"Sig: %s\" % XYvsZW_freqs_sig)\n",
    "                \n",
    "                \n",
    "            fig.add_subplot(3,3,8)\n",
    "            counts, bins, bars = plt.hist(XYvsZW_hets, bins= 50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(XYvsZW_hets) >= realXYvsZWhet:\n",
    "                plt.xlim((0,max(XYvsZW_hets)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realXYvsZWhet+10))\n",
    "            plt.vlines(realXYvsZWhet, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"XY - ZW heterozygosity method\")\n",
    "            plt.xlabel(\"Number of loci\")\n",
    "            if XYvsZW_hets_sig == 99 or XYvsZW_hets_sig == 95:\n",
    "                plt.text(realXYvsZWhet+1, max(counts)*0.6, \"Sig: %s\" % XYvsZW_hets_sig)\n",
    "                \n",
    "                \n",
    "            fig.add_subplot(3,3,9)\n",
    "            counts, bins, bars = plt.hist(XYvsZW_tagss, bins=50, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "            if max(XYvsZW_tagss) >= realYvsWtags:\n",
    "                plt.xlim((0,max(XYvsZW_tagss)+10))\n",
    "            else:\n",
    "                plt.xlim((0,realYvsWtags+10))\n",
    "            plt.vlines(realYvsWtags, 0, max(counts)*0.75, color = \"red\")\n",
    "            plt.title(\"Y specific tags - W specific tags\")\n",
    "            if XYvsZW_tagss_sig == 99 or XYvsZW_tagss_sig == 95:\n",
    "                plt.text(realYvsWtags+1, max(counts)*0.6, \"Sig: %s\" % XYvsZW_tagss_sig)\n",
    "                \n",
    "\n",
    "            fig.savefig(\"%s.pdf\" % randomisation_filepath.rpartition(\".\")[0])\n",
    "            fig.show()\n",
    "\n",
    "    rando_sig_file.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Plotting Randomisations_LpipHORSWE\n",
      "\n",
      "XYfreq_sig 95\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig 95\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Rkuk\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig 99\n",
      "Wtags_sig 99\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Rchen\n",
      "\n",
      "XYfreq_sig Non_sig\n",
      "XYhet_sig Non_sig\n",
      "Ytags_sig Non_sig\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig Non_sig\n",
      "XYvsZW_hets_sig Non_sig\n",
      "XYvsZW_tagss_sig Non_sig\n",
      "\n",
      "Plotting Randomisations_Pper\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Ryav\n",
      "\n",
      "XYfreq_sig Non_sig\n",
      "XYhet_sig Non_sig\n",
      "Ytags_sig 95\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig Non_sig\n",
      "XYvsZW_hets_sig Non_sig\n",
      "XYvsZW_tagss_sig 95\n",
      "\n",
      "Plotting Randomisations_Ribe_fam\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig 99\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Ltarah\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig 95\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Rber\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig Non_sig\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig Non_sig\n",
      "\n",
      "Plotting Randomisations_Rdal\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig Non_sig\n",
      "ZW_freqs_sig 99\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig Non_sig\n",
      "\n",
      "Plotting Randomisations_Lchiri\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig 99\n",
      "ZW_hets_sig 99\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 95\n",
      "\n",
      "Plotting Randomisations_Rita\n",
      "\n",
      "XYfreq_sig Non_sig\n",
      "XYhet_sig Non_sig\n",
      "Ytags_sig Non_sig\n",
      "ZW_freqs_sig Non_sig\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 95\n",
      "XYvsZW_hets_sig Non_sig\n",
      "XYvsZW_tagss_sig Non_sig\n",
      "\n",
      "Plotting Randomisations_Rarv\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig 99\n",
      "ZW_hets_sig Non_sig\n",
      "Wtags_sig Non_sig\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n",
      "\n",
      "Plotting Randomisations_Ribe_Sp\n",
      "\n",
      "XYfreq_sig 99\n",
      "XYhet_sig 99\n",
      "Ytags_sig 99\n",
      "ZW_freqs_sig 99\n",
      "ZW_hets_sig 99\n",
      "Wtags_sig 99\n",
      "XYvsZW_freqs_sig 99\n",
      "XYvsZW_hets_sig 99\n",
      "XYvsZW_tagss_sig 99\n"
     ]
    }
   ],
   "source": [
    "myrealdatapath = \"/home/djeffrie/Data/RADseq/Randomisations/Real_data.txt\"\n",
    "myrandomisation_directory = \"/home/djeffrie/Data/RADseq/Randomisations/\"\n",
    "\n",
    "randomisation_plotter(myrandomisation_directory, myrealdatapath, plot=False, Outs_constant=2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now I want to test statistically whether the real number of observed sex linked markers is higher than the simulated ones. . . \n",
    "\n",
    "The proceedure will be:\n",
    "\n",
    "For every species:\n",
    "    1. remove outliers from the distributions\n",
    "    2. check to see if the real data is above the 95th and 99th percentile\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "170.2\n",
      "77.2\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "XYfreqs_removed_outliers = removeOutliers(XYfreqs, 2)\n",
    "\n",
    "\n",
    "print np.round(np.percentile(XYfreqs, 95),3)\n",
    "print np.round(np.percentile(XYfreqs_removed_outliers, 95),3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3,\n",
       " 8,\n",
       " 26,\n",
       " 3,\n",
       " 9,\n",
       " 13,\n",
       " 7,\n",
       " 7,\n",
       " 26,\n",
       " 24,\n",
       " 20,\n",
       " 37,\n",
       " 147,\n",
       " 8,\n",
       " 216,\n",
       " 586,\n",
       " 16,\n",
       " 11,\n",
       " 16,\n",
       " 147,\n",
       " 96,\n",
       " 25,\n",
       " 11,\n",
       " 174,\n",
       " 184,\n",
       " 17,\n",
       " 26,\n",
       " 143,\n",
       " 56,\n",
       " 20,\n",
       " 36,\n",
       " 4,\n",
       " 12,\n",
       " 21,\n",
       " 76,\n",
       " 13,\n",
       " 25,\n",
       " 9,\n",
       " 80,\n",
       " 24,\n",
       " 54,\n",
       " 69,\n",
       " 123,\n",
       " 5,\n",
       " 63,\n",
       " 15,\n",
       " 17,\n",
       " 170,\n",
       " 55,\n",
       " 7,\n",
       " 2,\n",
       " 28,\n",
       " 11,\n",
       " 17,\n",
       " 24,\n",
       " 2,\n",
       " 15,\n",
       " 109,\n",
       " 4,\n",
       " 91,\n",
       " 20,\n",
       " 14,\n",
       " 17,\n",
       " 33,\n",
       " 33,\n",
       " 4,\n",
       " 3,\n",
       " 7,\n",
       " 10,\n",
       " 7,\n",
       " 19,\n",
       " 22,\n",
       " 10,\n",
       " 10,\n",
       " 36,\n",
       " 18,\n",
       " 57,\n",
       " 3,\n",
       " 26,\n",
       " 53,\n",
       " 6,\n",
       " 27,\n",
       " 6,\n",
       " 17,\n",
       " 27,\n",
       " 1,\n",
       " 22,\n",
       " 51,\n",
       " 11,\n",
       " 5,\n",
       " 89,\n",
       " 12,\n",
       " 27,\n",
       " 14,\n",
       " 22,\n",
       " 78,\n",
       " 33,\n",
       " 486,\n",
       " 27,\n",
       " 15]"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "XYfreqs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
