{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bootstrapping sex linked marker sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import MISC_RAD_tools as MISC\n",
    "import SLMF_lightweight as SLMF_L"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook I will first compile all of the final stacks outputs for each species dataset (or subset) and then, for each one, I will randomise the male and female assignments accross the samples. For 100 random male female assignments I will then run the sex linked marker finding analyses, using exactly the parameters used to identify the final set of sex linked markers in the paper. The only difference is the male and female assignments. \n",
    "\n",
    "This randomisation will give an idea of what the false positive rate in the dataset is. For example, a skew towards more males or females in the dataset may make false positives of one type or another more likely. Also, if there are several populations in the data and males and females are not distributed evenly among them then population structure could look like sex linkage. However, randomising male and female assignments across all samples will allow us to account for this. \n",
    "\n",
    "Due to the prohibitively long amount of time that it would take, I will not do 1000 randomisations for each species, although this is what I would prefer. Instead I will just do 100, this should still give a reasonably good estimate for the false positive rate. \n",
    "\n",
    "With regards to how these randomisations will be used to judge the validity of the dataset, I will look for sample sets where the number of sex-linked markers found using the correct male/female assignments is above the 95th percentile of the distribution of sex-linked markers found in the randomisations. \n",
    "\n",
    "In collaboration with the genome mapping, this should help validate the sex-linked markers sets found.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### First, make a list of all of the parameter disctionaries, which contain paths and parameters used for finding sex linked markers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "Parameter_dictionaries = []"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. chiricahuensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Lchiri = {}\n",
    "Parameter_dict_Lchiri[\"Name\"] = \"Lchiri\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Lchiri['Catalog'] =  \"/home/djeffrie/Data/RADseq/Lchricahuensis/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Lchiri['VCF'] =  \"/home/djeffrie/Data/RADseq/Lchricahuensis/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Lchiri['Pop_map'] = \"/home/djeffrie/Data/RADseq/Lchricahuensis/Sex_ID_info_reassigned.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Lchiri['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Lchiri['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Lchiri['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Lchiri['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Lchiri['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Lchiri['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Lchiri['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Lchiri['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Lchiri['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Lchiri['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Lchiri['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Lchiri)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. montezumae\n",
    "\n",
    "No sex linked markers found"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. pipiens (HOR / SWE populations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_LpipHORSWE = {}\n",
    "Parameter_dict_LpipHORSWE[\"Name\"] = \"LpipHORSWE\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_LpipHORSWE['Catalog'] =  \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_LpipHORSWE['VCF'] =  \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/HOR_SWE_populations/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_LpipHORSWE['Pop_map'] = \"/home/djeffrie/Data/RADseq/Lpipiens/Lpip_all_stacks/HOR_SWE_populations/Sex_ID_info_kept_certain.txt\"\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_LpipHORSWE['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_LpipHORSWE['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_LpipHORSWE['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_LpipHORSWE['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_LpipHORSWE['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_LpipHORSWE['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_LpipHORSWE['heterogamtic_heterozygosity_threshold'] = 0.8 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_LpipHORSWE['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_LpipHORSWE['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_LpipHORSWE['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_LpipHORSWE['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_LpipHORSWE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### L. tarahumarae"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ltarah = {}\n",
    "Parameter_dict_Ltarah[\"Name\"] = \"Ltarah\"\n",
    "\n",
    "working_dir = \"/home/djeffrie/Data/RADseq/Ltarahumarae/Sex_linked_markers/\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ltarah['Catalog'] =  \"%s/batch_1.catalog.tags.tsv.gz\" % working_dir ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ltarah['VCF'] =  \"%s/batch_1.vcf\"  % working_dir ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ltarah['Pop_map'] = \"%s/Sex_ID_info_kept_2.txt\"  % working_dir ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ltarah['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ltarah['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ltarah['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ltarah['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ltarah['homogametic_REF_allele_freq'] = 0.9 ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ltarah['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ltarah['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ltarah['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ltarah['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ltarah['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ltarah['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ltarah)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. berlandieri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_R_ber = {}\n",
    "Parameter_dict_R_ber[\"Name\"] = \"R_ber\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_R_ber['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_R_ber['VCF'] =  \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_R_ber['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rberlandieri/Sex_linked_markers/Sex_ID_info_kept_2.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_R_ber['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_R_ber['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_R_ber['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_R_ber['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_R_ber['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_R_ber['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_R_ber['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_R_ber['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_R_ber['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_R_ber['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_R_ber['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_R_ber)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### P. perezi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Pper = {}\n",
    "Parameter_dict_Pper[\"Name\"] = \"Pper\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Pper['Catalog'] =  \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Pper['VCF'] =  \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Pper['Pop_map'] = \"/home/djeffrie/Data/RADseq/Pperezi/Sex_linked_markers/Sex_info_ID.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Pper['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Pper['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Pper['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Pper['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Pper['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Pper['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Pper['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Pper['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Pper['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Pper['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Pper['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Pper)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. arvalis (all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rarv = {}\n",
    "Parameter_dict_Rarv[\"Name\"] = \"Rarv\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rarv['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rarvalis_NEW/Stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rarv['VCF'] =  \"/home/djeffrie/Data/RADseq/Rarvalis_NEW/Stacks/Populations_kept_2/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rarv['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rarvalis_NEW//Stacks/Sex_ID_info_kept_2.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rarv['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rarv['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rarv['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rarv['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rarv['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rarv['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rarv['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rarv['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rarv['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rarv['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rarv['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rarv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. chensinensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rchen = {}\n",
    "Parameter_dict_Rchen[\"Name\"] = \"Rchen\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rchen['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rchen['VCF'] =  \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/batch_1_strict_kept_3.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rchen['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rchensinensis/Sex_linked_markers/Sex_ID_info_kept_3.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rchen['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rchen['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rchen['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rchen['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rchen['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rchen['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rchen['heterogamtic_heterozygosity_threshold'] = 0.7 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rchen['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rchen['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rchen['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rchen['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rchen)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. dalmatina "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rdal = {}\n",
    "Parameter_dict_Rdal[\"Name\"] = \"Rdal\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rdal['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rdal['VCF'] =  \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rdal['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rdal['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rdal['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rdal['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rdal['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rdal['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rdal['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rdal['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rdal['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rdal['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rdal['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rdal['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rdal)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. iberica all northern Spain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ribe_Sp = {}\n",
    "Parameter_dict_Ribe_Sp[\"Name\"] = \"Ribe_Sp\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ribe_Sp['Catalog'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ribe_Sp['VCF'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ribe_Sp['Pop_map'] = \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/Sex_ID_info_N_spain.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ribe_Sp['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ribe_Sp['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_Sp['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_Sp['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ribe_Sp['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ribe_Sp['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ribe_Sp['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ribe_Sp['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_Sp['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_Sp['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ribe_Sp['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ribe_Sp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. iberica \"family\" (which isn't actually a family)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ribe_fam = {}\n",
    "Parameter_dict_Ribe_fam[\"Name\"] = \"Ribe_fam\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ribe_fam['Catalog'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ribe_fam['VCF'] =  \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ribe_fam['Pop_map'] = \"/home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ribe_fam['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ribe_fam['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_fam['coverage_threshold1'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_fam['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ribe_fam['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ribe_fam['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ribe_fam['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ribe_fam['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ribe_fam['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ribe_fam['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ribe_fam['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ribe_fam)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### R. italica (CM pop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rita = {}\n",
    "Parameter_dict_Rita[\"Name\"] = \"Rita\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rita['Catalog'] =  \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rita['VCF'] =  \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rita['Pop_map'] = \"/home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/CM_sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rita['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rita['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rita['homogametic_REF_allele_freq'] = 0.9 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rita['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rita['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rita['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rita['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rita['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rita['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rita)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. kukinoris (Nanping only)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Rkuk = {}\n",
    "Parameter_dict_Rkuk[\"Name\"] = \"Rkuk\"\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Rkuk['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Rkuk['VCF'] =  \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Rkuk['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Rkuk['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Rkuk['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rkuk['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rkuk['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Rkuk['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Rkuk['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Rkuk['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Rkuk['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Rkuk['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Rkuk['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Rkuk['sex_presence_threshold'] =  0.7 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Rkuk)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "### R. yavapaiensis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict_Ryav = {}\n",
    "Parameter_dict_Ryav[\"Name\"] = \"Ryav\"\n",
    "\n",
    "## NOTE. DICT KEY NAMES MUST NOT BE CHANGED!\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict_Ryav['Catalog'] =  \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict_Ryav['VCF'] =  \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/batch_1.vcf.altered\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict_Ryav['Pop_map'] = \"/home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/Sex_ID_info.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict_Ryav['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict_Ryav['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ryav['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ryav['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "Parameter_dict_Ryav['homogametic_REF_allele_freq'] = 1 ## (Default = 0.95) The sex linked SNP will be the minor allele, so a check is done to make sure that the homogametic sex is above the threshold specified for the major allele. In theory this should be 1. But allowing for some error 0.95 is used as a default. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict_Ryav['homogamtic_homozygosity_threshold'] = 1 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict_Ryav['heterogamtic_heterozygosity_threshold'] = 0.6 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict_Ryav['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict_Ryav['coverage_threshold2'] = 7 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict_Ryav['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict_Ryav['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "Parameter_dictionaries.append(Parameter_dict_Ryav)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6 species for which sex linked markers were found that will be randomised\n"
     ]
    }
   ],
   "source": [
    "print \"There are %s species for which sex linked markers were found that will be randomised\" % len(Parameter_dictionaries)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "The program will work as follows:\n",
    "\n",
    "For every species, it will\n",
    "    1. make a new directory\n",
    "    2. Move the sex information file and vcf file into that directory\n",
    "    3. Make the randomised sex information files and renamed vcf files\n",
    "    4. Run the randomisation prgroam for that species and collect results\n",
    "    5. Write results to an output file\n",
    "    6. Tidy up loose files\n",
    "    7. Plot the results\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Define a program to help parallelise the analyses\n",
    "\n",
    "\n",
    "def Super_SLM_finder_parallel(popmap, Parameter_dict):\n",
    "    import os\n",
    "    \n",
    "    Parameter_dict['Pop_map'] = popmap\n",
    "    Parameter_dict['VCF'] = \"%s.vcf\" % popmap.rpartition(\".\")[0]\n",
    "    results_dict = {}\n",
    "    results_dict[\"XYset\"], results_dict[\"ZWset\"], results_dict[\"Detailed\"] = SLMF_L.Super_SLM_finder(Parameter_dict, \"111\", verbose = False, write_files=False, plot=False)\n",
    "    \n",
    "    os.remove(Parameter_dict['VCF']) ## remove VCFs as they are used\n",
    "    os.remove(\"%s.altered\" % Parameter_dict['VCF'])\n",
    "    os.remove(\"%s.all_frequencies.tsv\" % Parameter_dict['VCF'])\n",
    "    \n",
    "\n",
    "    return results_dict\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processing dataset in /home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/Randomisations\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:  3.5min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed:  8.4min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Rdalmatina/Populations_final/Randomisations/Randomisations_Rdal.txt\n",
      "processing dataset in /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/Randomisations\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative Z linked markers = 10\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked tags = 10\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 2\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 29\n",
      "Number of putative X linked snps = 43\n",
      "Number of putative X linked snps = 14\n",
      "Number of putative X linked snps = 31\n",
      "Number of putative X linked tags = 29\n",
      "Number of putative X linked tags = 43\n",
      "Number of putative X linked tags = 14\n",
      "Number of putative X linked tags = 31\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked markers = 24\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked markers = 37\n",
      "Number of putative Z linked tags = 33\n",
      "Number of putative Z linked tags = 24\n",
      "Number of putative Z linked tags = 33\n",
      "Number of putative Z linked tags = 37\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['14257_1311610', '16303_1499796', '17290_1590672', '20102_1849328', '20689_1903311', '20689_1903340', '20689_1903358', '20689_1903382', '24221_2228330', '30262_2784028', '38621_3553111', '40213_3699569', '43891_4037893', '47601_4379264', '50863_4679355', '54225_4988670', '62217_5723955', '62217_5723964', '66692_6135590', '66692_6135644', '69609_6404013', '72539_6673539', '73987_6806766', '73987_6806768', '76315_7020958', '79508_7314684', '84623_7785310', '101343_9323541', '106486_9796704']\n",
      "X_het markers ->  ['3154_290090', '19096_1756756', '19096_1756775', '22043_2027895', '23483_2160366', '24151_2221885', '25549_2350471', '26418_2430387', '26423_2430861', '26818_2467177', '27347_2515849', '29632_2726121', '40855_3758654', '43058_3961281', '46301_4259660', '47042_4327799', '47042_4327861', '47073_4330661', '52677_4846243', '52985_4874551', '52985_4874569', '53994_4967397', '61788_5684497', '65718_6046000', '65718_6046051', '67722_6230373', '72678_6686317', '74509_6854792', '77234_7105472', '77316_7113037', '80890_7441810', '81472_7495370', '83824_7711767', '86094_7920622', '88276_8121343', '88450_8137358', '89170_8203557', '90893_8362108', '91195_8389914', '93570_8608440', '95054_8744941', '99468_9151057', '101929_9377439']\n",
      "X_het markers ->  ['4192_385584', '6661_612783', '6661_612790', '9414_866019', '25739_2367924', '26541_2441690', '41740_3840048', '43640_4014874', '60673_5581847', '60673_5581859', '64992_5979242', '72576_6676925', '77260_7107879', '88444_8136833']\n",
      "X_het markers ->  ['8827_812082', '9843_905527', '11856_1090722', '11856_1090750', '14257_1311610', '22661_2084805', '22661_2084813', '25001_2300037', '25001_2300060', '27453_2525667', '32347_2975851', '35016_3221402', '35016_3221409', '38824_3571733', '38824_3571761', '41867_3851762', '48354_4448508', '53700_4940397', '56062_5157704', '75231_6921202', '76354_7024506', '82938_7630284', '89114_8198476', '90012_8281044', '98250_9038986', '101310_9320484', '103468_9518982', '103652_9535969', '104516_9615395', '104582_9621478', '106710_9817320']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 56\n",
      "Number of Putative Y linked tags: 9\n",
      "Number of Putative Y linked tags: 29\n",
      "Number of Putative Y linked tags: 36\n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 7\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 1\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 91\n",
      "Final number of XY tags = 70\n",
      "Final number of XY tags = 47\n",
      "Final number of XY tags = 68\n",
      "Final number of ZW tags = 39\n",
      "Final number of ZW tags = 33\n",
      "Final number of ZW tags = 38\n",
      "Final number of ZW tags = 39\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 14\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 23\n",
      "Number of putative X linked tags = 14\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 23\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 5\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 22\n",
      "Number of putative X linked snps = 25\n",
      "Number of putative X linked snps = 21\n",
      "Number of putative X linked snps = 35\n",
      "Number of putative X linked tags = 22\n",
      "Number of putative X linked tags = 25\n",
      "Number of putative X linked tags = 21\n",
      "Number of putative X linked tags = 35\n",
      "Number of putative Z linked markers = 49\n",
      "Number of putative Z linked markers = 18\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked markers = 18\n",
      "Number of putative Z linked tags = 49\n",
      "Number of putative Z linked tags = 18\n",
      "Number of putative Z linked tags = 33\n",
      "Number of putative Z linked tags = 18\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['6716_617841', '6910_635712', '6967_640927', '11939_1098348', '19096_1756756', '19096_1756770', '19096_1756775', '25887_2381534', '25887_2381603', '41858_3850936', '42508_3910734', '42508_3910737', '43620_4013041', '45883_4221237', '54777_5039420', '68030_6258685', '85604_7875529', '88444_8136811', '93962_8644495', '95127_8751676', '95127_8751684', '108674_9997972']\n",
      "X_het markers ->  ['13645_1255324', '20143_1853083', '20143_1853084', '20143_1853109', '20143_1853115', '20143_1853117', '20143_1853122', '20143_1853125', '22080_2031322', '24903_2291037', '26752_2461109', '32989_3034977', '39660_3648677', '45883_4221237', '54184_4984909', '69782_6419896', '70670_6501557', '70670_6501588', '81103_7461404', '81887_7533566', '90893_8362108', '93348_8587939', '94457_8689962', '94457_8690016', '102589_9438188']\n",
      "X_het markers ->  ['13645_1255324', '14257_1311610', '22039_2027589', '30262_2784028', '35016_3221402', '40213_3699569', '66692_6135590', '66692_6135644', '75243_6922306', '75243_6922307', '76315_7020958', '76708_7057055', '79508_7314684', '80672_7421797', '97187_8941175', '98090_9024231', '101899_9374631', '101899_9374670', '101899_9374680', '101899_9374683', '104582_9621478']\n",
      "X_het markers ->  ['6910_635712', '6967_640927', '11939_1098348', '14905_1371201', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '24464_2250629', '25212_2319421', '29466_2710792', '29466_2710802', '42508_3910734', '42508_3910737', '42877_3944656', '43000_3955963', '50334_4630726', '54936_5054048', '55747_5128688', '75243_6922306', '75243_6922307', '76788_7064478', '83028_7638524', '85345_7851731', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '91459_8414153', '91459_8414207', '97005_8924390', '97005_8924403', '101268_9316611', '106093_9760510']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 5\n",
      "Number of Putative Y linked tags: 21\n",
      "Number of Putative Y linked tags: 56\n",
      "Number of Putative Y linked tags: 10\n",
      "Number of Putative W linked tags: 12\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 7\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 40\n",
      "Final number of XY tags = 48\n",
      "Final number of XY tags = 81\n",
      "Final number of XY tags = 66\n",
      "Final number of ZW tags = 64\n",
      "Final number of ZW tags = 27\n",
      "Final number of ZW tags = 37\n",
      "Final number of ZW tags = 25\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 7\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative X linked snps = 8\n",
      "Number of putative X linked tags = 7\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 20\n",
      "Number of putative X linked tags = 8\n",
      "Number of putative Z linked markers = 12\n",
      "Number of putative Z linked markers = 10\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked tags = 12\n",
      "Number of putative Z linked tags = 10\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 5\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 46\n",
      "Number of putative X linked snps = 22\n",
      "Number of putative X linked snps = 59\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked tags = 46\n",
      "Number of putative X linked tags = 22\n",
      "Number of putative X linked tags = 59\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative Z linked markers = 36\n",
      "Number of putative Z linked markers = 35\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 29\n",
      "Number of putative Z linked tags = 36\n",
      "Number of putative Z linked tags = 35\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 29\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['1154_106099', '1154_106100', '4300_395564', '4739_435953', '6716_617841', '9007_828587', '11637_1070535', '12045_1108074', '12097_1112916', '12097_1112924', '12116_1114624', '12200_1122329', '15305_1407987', '15305_1408032', '15407_1417370', '22446_2064953', '23264_2140257', '25404_2337093', '25887_2381534', '25887_2381603', '28256_2599550', '41858_3850936', '43982_4046279', '44059_4053383', '44059_4053393', '45883_4221237', '49335_4538793', '49645_4567260', '51391_4727923', '52670_4845557', '67546_6214172', '70720_6506162', '70720_6506163', '71305_6560029', '71305_6560045', '72145_6637278', '76074_6998777', '85340_7851263', '85604_7875529', '86274_7937129', '88552_8146701', '90891_8361911', '92927_8549217', '97221_8944287', '101700_9356358', '108674_9997972']\n",
      "X_het markers ->  ['1479_136012', '9843_905527', '10724_986575', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '23592_2170409', '32347_2975851', '34042_3131799', '53958_4964063', '55684_5122895', '56932_5237736', '60108_5529903', '74766_6878428', '74766_6878429', '80317_7389134', '85419_7858491', '90012_8281044', '93348_8587939', '99186_9125076']\n",
      "X_het markers ->  ['134_12255', '1479_136012', '3426_315112', '4710_433241', '6716_617841', '8815_810900', '10724_986575', '10867_999683', '14940_1374448', '18115_1666509', '19096_1756756', '19096_1756775', '19770_1818825', '21118_1942836', '21118_1942846', '25723_2366461', '26818_2467177', '27347_2515849', '30277_2785417', '30277_2785472', '32224_2964542', '33893_3118090', '34178_3144296', '35906_3303279', '38614_3552421', '42003_3864211', '44597_4102878', '49322_4537577', '52440_4824455', '54434_5007868', '55684_5122895', '56367_5185745', '56776_5223359', '57754_5313290', '57754_5313354', '58005_5336448', '59322_5457571', '60108_5529903', '64344_5919572', '64870_5967986', '71041_6535722', '78058_7181328', '79039_7271565', '79866_7347626', '83324_7665769', '85419_7858491', '86373_7946291', '88009_8096826', '88009_8096829', '88957_8184033', '90786_8352270', '90871_8360078', '91324_8401797', '91471_8415326', '92095_8472683', '93739_8623977', '95081_8747445', '96284_8858129', '108674_9997972']\n",
      "X_het markers ->  ['624_57387', '1479_136012', '21118_1942836', '21118_1942846', '24769_2278666', '35472_3263359', '36331_3342447', '43341_3987291', '48443_4456680', '52440_4824455', '76074_6998777', '81887_7533566', '83253_7659234', '85451_7861470', '91459_8414153', '91459_8414207', '95897_8822520', '98832_9092509', '102554_9434950']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 98\n",
      "Number of Putative Y linked tags: 91\n",
      "Number of Putative Y linked tags: 47\n",
      "Number of Putative Y linked tags: 29\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 10\n",
      "Number of Putative W linked tags: 5\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 150\n",
      "Final number of XY tags = 114\n",
      "Final number of XY tags = 120\n",
      "Final number of XY tags = 55\n",
      "Final number of ZW tags = 50\n",
      "Final number of ZW tags = 39\n",
      "Final number of ZW tags = 39\n",
      "Final number of ZW tags = 39\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 9\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 22\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked tags = 9\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 22\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative Z linked markers = 9\n",
      "Number of putative Z linked markers = 9\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 12\n",
      "Number of putative Z linked tags = 9\n",
      "Number of putative Z linked tags = 9\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 12\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 38\n",
      "Number of putative X linked snps = 16\n",
      "Number of putative X linked snps = 60\n",
      "Number of putative X linked snps = 37\n",
      "Number of putative X linked tags = 38\n",
      "Number of putative X linked tags = 16\n",
      "Number of putative X linked tags = 60\n",
      "Number of putative X linked tags = 37\n",
      "Number of putative Z linked markers = 13\n",
      "Number of putative Z linked markers = 43\n",
      "Number of putative Z linked markers = 46\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked tags = 13\n",
      "Number of putative Z linked tags = 43\n",
      "Number of putative Z linked tags = 46\n",
      "Number of putative Z linked tags = 21\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['6910_635712', '6967_640927', '8852_814315', '11939_1098348', '17528_1612536', '24464_2250629', '26945_2478938', '28376_2610531', '28743_2644274', '28743_2644288', '28743_2644299', '28743_2644302', '28743_2644357', '34748_3196808', '36490_3357006', '36490_3357072', '36490_3357074', '36584_3365652', '38296_3523202', '42508_3910734', '42508_3910737', '58063_5341733', '69806_6422146', '72110_6634062', '72888_6705631', '72888_6705696', '75078_6907111', '76299_7019463', '80874_7440395', '84941_7814524', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '88183_8112836', '93413_8593973', '93413_8593978', '95388_8775639']\n",
      "X_het markers ->  ['1479_136012', '21118_1942836', '21118_1942846', '24769_2278666', '35016_3221409', '35472_3263359', '52440_4824455', '55747_5128688', '76074_6998777', '76708_7057055', '81887_7533566', '83253_7659234', '95471_8783248', '95471_8783295', '95897_8822520', '98832_9092509']\n",
      "X_het markers ->  ['2446_224991', '3426_315112', '6716_617841', '8600_791143', '8815_810900', '10724_986575', '14940_1374448', '19770_1818825', '21118_1942836', '21118_1942846', '23264_2140257', '25274_2325163', '25649_2359648', '26580_2445293', '28354_2608521', '30202_2778530', '34780_3199680', '34780_3199682', '35906_3303279', '38614_3552421', '42003_3864211', '44597_4102878', '49322_4537577', '52440_4824455', '52481_4828241', '52919_4868547', '55684_5122895', '56089_5160163', '56089_5160169', '56207_5171032', '56367_5185745', '59322_5457571', '60108_5529903', '71041_6535722', '72145_6637278', '74766_6878428', '74766_6878429', '79866_7347626', '82380_7578885', '82380_7578888', '82592_7598465', '82593_7598551', '85419_7858491', '86483_7956370', '86966_8000816', '88957_8184033', '89114_8198476', '90871_8360078', '91324_8401797', '91471_8415326', '91834_8448657', '91834_8448689', '91834_8448694', '95081_8747445', '95852_8818362', '96284_8858129', '99186_9125076', '100846_9277792', '104608_9623870', '108674_9997972']\n",
      "X_het markers ->  ['12849_1182050', '12849_1182052', '13100_1205184', '20689_1903311', '20689_1903340', '20689_1903358', '20689_1903382', '25739_2367924', '32797_3017309', '32797_3017318', '32989_3034977', '48295_4443120', '49322_4537577', '57840_5321199', '57840_5321240', '65459_6022144', '69782_6419896', '76299_7019427', '79596_7322786', '82411_7581753', '82411_7581783', '82592_7598465', '86039_7915530', '86829_7988248', '88542_8145861', '90797_8353264', '91471_8415326', '91834_8448654', '91834_8448657', '91834_8448689', '91834_8448694', '93358_8588856', '93358_8588857', '97187_8941175', '100590_9254252', '105510_9706909', '105510_9706910']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 15\n",
      "Number of Putative Y linked tags: 20\n",
      "Number of Putative Y linked tags: 262\n",
      "Number of Putative Y linked tags: 10\n",
      "Number of Putative W linked tags: 12\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 2\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 59\n",
      "Final number of XY tags = 40\n",
      "Final number of XY tags = 338\n",
      "Final number of XY tags = 50\n",
      "Final number of ZW tags = 34\n",
      "Final number of ZW tags = 50\n",
      "Final number of ZW tags = 51\n",
      "Final number of ZW tags = 35\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 10\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 15\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 10\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 15\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 2\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 26\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 57\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked tags = 26\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 57\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative Z linked markers = 14\n",
      "Number of putative Z linked markers = 38\n",
      "Number of putative Z linked markers = 20\n",
      "Number of putative Z linked markers = 38\n",
      "Number of putative Z linked tags = 14\n",
      "Number of putative Z linked tags = 38\n",
      "Number of putative Z linked tags = 20\n",
      "Number of putative Z linked tags = 38\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['15897_1462447', '20562_1891698', '25163_2314957', '25535_2349194', '25535_2349195', '27347_2515849', '32294_2971046', '33893_3118090', '38236_3517669', '40213_3699569', '47617_4380764', '48718_4481977', '54777_5039420', '57207_5263036', '70406_6477313', '70559_6491368', '70606_6495723', '70606_6495725', '70670_6501557', '70670_6501588', '71197_6550083', '84370_7762036', '85340_7851263', '88444_8136811', '88444_8136833', '89981_8278246']\n",
      "X_het markers ->  ['10724_986575', '20102_1849328', '25104_2309532', '30767_2830486', '33043_3039949', '39119_3598888', '50822_4675564', '52919_4868547', '55747_5128688', '60108_5529903', '75243_6922306', '75243_6922307', '76708_7057055', '82380_7578885', '82380_7578888', '85419_7858491', '86966_8000816', '99186_9125076', '100846_9277792']\n",
      "X_het markers ->  ['8815_810900', '10603_975468', '10707_984978', '14940_1374448', '15305_1407987', '15305_1408032', '17517_1611504', '18189_1673314', '18189_1673347', '18189_1673364', '18189_1673367', '19096_1756756', '19096_1756770', '19096_1756775', '22336_2054857', '22336_2054858', '22336_2054882', '22336_2054899', '25319_2329300', '25887_2381534', '25887_2381603', '26190_2409449', '26580_2445293', '26727_2458879', '26945_2478938', '27478_2527917', '28354_2608521', '33137_3048527', '33137_3048531', '33137_3048557', '33137_3048574', '33137_3048578', '33137_3048582', '33137_3048592', '34780_3199680', '34780_3199682', '35025_3222236', '35025_3222298', '39079_3595233', '46441_4272550', '51189_4709381', '51391_4727923', '54222_4988345', '54222_4988346', '56207_5171032', '59692_5491658', '59692_5491661', '72145_6637278', '72336_6654864', '72336_6654910', '72899_6706709', '74926_6893172', '76299_7019427', '82592_7598465', '82593_7598551', '88174_8111999', '101700_9356358']\n",
      "X_het markers ->  ['2457_226010', '2457_226039', '2736_251645', '2736_251666', '9928_913366', '11582_1065536', '22929_2109403', '23177_2132255', '23905_2199203', '24675_2270041', '28402_2612917', '30585_2813788', '31585_2905737', '32294_2971046', '33341_3067342', '33504_3082332', '33766_3106426', '33951_3123475', '34505_3174409', '34573_3180687', '37848_3481982', '40416_3718273', '55242_5082189', '55242_5082264', '57207_5263036', '58928_5421368', '59866_5507617', '60634_5578244', '68207_6275008', '68452_6297544', '69324_6377769', '69324_6377781', '70406_6477313', '76467_7034924', '77251_7107035', '77251_7107039', '77495_7129483', '86039_7915530', '89301_8215642', '92356_8496696', '93600_8611182', '100311_9228557']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 9\n",
      "Number of Putative Y linked tags: 27\n",
      "Number of Putative Y linked tags: 14\n",
      "Number of Putative Y linked tags: 7\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 5\n",
      "Number of Putative W linked tags: 5\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 38\n",
      "Final number of XY tags = 53\n",
      "Final number of XY tags = 75\n",
      "Final number of XY tags = 64\n",
      "Final number of ZW tags = 17\n",
      "Final number of ZW tags = 44\n",
      "Final number of ZW tags = 25\n",
      "Final number of ZW tags = 45\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 1\n",
      "Number of putative X linked snps = 15\n",
      "Number of putative X linked snps = 15\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 1\n",
      "Number of putative X linked tags = 15\n",
      "Number of putative X linked tags = 15\n",
      "Number of putative Z linked markers = 9\n",
      "Number of putative Z linked markers = 11\n",
      "Number of putative Z linked markers = 8\n",
      "Number of putative Z linked markers = 22\n",
      "Number of putative Z linked tags = 9\n",
      "Number of putative Z linked tags = 11\n",
      "Number of putative Z linked tags = 8\n",
      "Number of putative Z linked tags = 22\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked snps = 14\n",
      "Number of putative X linked snps = 51\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative X linked tags = 14\n",
      "Number of putative X linked tags = 51\n",
      "Number of putative X linked tags = 36\n",
      "Number of putative Z linked markers = 26\n",
      "Number of putative Z linked markers = 36\n",
      "Number of putative Z linked markers = 57\n",
      "Number of putative Z linked markers = 54\n",
      "Number of putative Z linked tags = 26\n",
      "Number of putative Z linked tags = 36\n",
      "Number of putative Z linked tags = 57\n",
      "Number of putative Z linked tags = 54\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['10424_958977', '10424_958983', '11119_1022901', '18115_1666509', '23015_2117371', '25507_2346615', '30149_2773635', '32056_2949081', '32056_2949119', '32056_2949135', '32056_2949139', '32056_2949144', '33691_3099558', '34708_3193083', '37113_3414374', '41178_3788327', '41674_3833953', '47617_4380764', '50334_4630726', '53384_4911251', '53384_4911328', '56062_5157704', '56519_5199736', '66676_6134143', '67827_6240054', '70606_6495723', '70606_6495725', '71197_6550083', '73588_6770081', '75689_6963315', '76708_7057055', '84941_7814524', '87359_8036993', '87650_8063775', '87650_8063776', '90952_8367562', '91816_8447010', '93413_8593973', '93413_8593978', '93945_8642858', '102777_9455473', '107181_9860609']\n",
      "X_het markers ->  ['11119_1022901', '11508_1058728', '15327_1410046', '17355_1596646', '25794_2372967', '26818_2467177', '33893_3118090', '34042_3131799', '41621_3829094', '43640_4014831', '43640_4014874', '81829_7528228', '97422_8962805', '107181_9860609']\n",
      "X_het markers ->  ['3014_277282', '4307_396200', '10659_980615', '10748_988804', '10748_988805', '11119_1022901', '15984_1470478', '18099_1665049', '21927_2017201', '21927_2017250', '23015_2117371', '32056_2949081', '32056_2949119', '32056_2949135', '32056_2949139', '32056_2949144', '34178_3144296', '34485_3172572', '34708_3193083', '37113_3414374', '40213_3699569', '41674_3833953', '42859_3943005', '42859_3943018', '47617_4380764', '49322_4537577', '50334_4630726', '59450_5469374', '59475_5471621', '59475_5471662', '61613_5668353', '66676_6134143', '70606_6495723', '70606_6495725', '71197_6550083', '73588_6770081', '76074_6998777', '76635_7050362', '85604_7875529', '87359_8036993', '87650_8063775', '87650_8063776', '88444_8136833', '90952_8367562', '91471_8415326', '91816_8447010', '98250_9038986', '98250_9038996', '102777_9455473', '105263_9684181', '107181_9860609']\n",
      "X_het markers ->  ['6408_589504', '11119_1022901', '11125_1023431', '11125_1023457', '21927_2017201', '21927_2017250', '23733_2183428', '25163_2314957', '25649_2359648', '26727_2458800', '26727_2458837', '26727_2458874', '26727_2458875', '27453_2525667', '31478_2895910', '34178_3144296', '49322_4537577', '51823_4767691', '52919_4868547', '59475_5471621', '59475_5471662', '77260_7107879', '77480_7128088', '77924_7168958', '77924_7168976', '80317_7389134', '81829_7528228', '82592_7598465', '84884_7809274', '91471_8415326', '91834_8448657', '91834_8448689', '91834_8448694', '95651_8799811', '95651_8799860', '107181_9860609']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 51\n",
      "Number of Putative Y linked tags: 7\n",
      "Number of Putative Y linked tags: 36\n",
      "Number of Putative Y linked tags: 71\n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 8\n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 95\n",
      "Final number of XY tags = 22\n",
      "Final number of XY tags = 100\n",
      "Final number of XY tags = 115\n",
      "Final number of ZW tags = 36\n",
      "Final number of ZW tags = 55\n",
      "Final number of ZW tags = 66\n",
      "Final number of ZW tags = 75\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 16\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked tags = 16\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative Z linked markers = 7\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked tags = 7\n",
      "Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 4\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 55\n",
      "Number of putative X linked snps = 37\n",
      "Number of putative X linked snps = 40\n",
      "Number of putative X linked snps = 31\n",
      "Number of putative X linked tags = 55\n",
      "Number of putative X linked tags = 37\n",
      "Number of putative X linked tags = 40\n",
      "Number of putative X linked tags = 31\n",
      "Number of putative Z linked markers = 19\n",
      "Number of putative Z linked markers = 26\n",
      "Number of putative Z linked markers = 27\n",
      "Number of putative Z linked markers = 29\n",
      "Number of putative Z linked tags = 19\n",
      "Number of putative Z linked tags = 26\n",
      "Number of putative Z linked tags = 27\n",
      "Number of putative Z linked tags = 29\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['3426_315112', '10567_972080', '10567_972103', '10567_972104', '10567_972129', '11681_1074643', '13529_1244628', '14257_1311610', '18189_1673314', '18189_1673347', '18189_1673364', '18189_1673367', '19096_1756756', '19096_1756770', '19096_1756775', '19770_1818825', '21118_1942836', '21118_1942846', '22161_2038733', '22161_2038749', '22161_2038765', '24221_2228330', '27347_2515849', '34178_3144296', '35906_3303279', '38614_3552421', '38866_3575641', '42003_3864211', '44597_4102878', '47601_4379264', '49322_4537577', '52440_4824455', '56367_5185745', '56776_5223359', '58005_5336448', '59322_5457571', '66692_6135590', '66692_6135644', '70720_6506162', '70720_6506163', '71041_6535722', '72336_6654864', '72336_6654910', '76315_7020958', '76349_7024024', '79866_7347626', '82592_7598465', '85340_7851263', '88957_8184033', '90871_8360078', '91324_8401797', '91471_8415326', '93739_8623977', '95081_8747445', '96284_8858129']\n",
      "X_het markers ->  ['6716_617841', '10707_984978', '15305_1407987', '15305_1408032', '15942_1466592', '15942_1466593', '15942_1466596', '15942_1466623', '17517_1611504', '19096_1756756', '19096_1756770', '19096_1756775', '25319_2329300', '25887_2381534', '25887_2381603', '26190_2409449', '26580_2445293', '26727_2458879', '34780_3199680', '34780_3199682', '41858_3850936', '43728_4022913', '45883_4221237', '51391_4727923', '52670_4845557', '67546_6214172', '70797_6513254', '70797_6513258', '72145_6637278', '72336_6654864', '72336_6654910', '72899_6706709', '85604_7875529', '89784_8260069', '96399_8868673', '101700_9356358', '108674_9997972']\n",
      "X_het markers ->  ['5714_525657', '5919_544501', '8706_800947', '9458_870120', '22661_2084736', '22661_2084781', '22661_2084805', '22661_2084808', '22661_2084813', '23986_2206677', '26727_2458800', '26727_2458837', '26727_2458874', '26727_2458875', '26727_2458879', '28924_2660955', '34042_3131799', '46947_4319073', '52670_4845557', '56932_5237736', '62275_5729266', '62275_5729285', '67516_6211427', '67546_6214172', '73987_6806766', '73987_6806768', '77589_7138181', '78720_7242169', '78720_7242178', '78765_7246369', '80971_7449283', '83226_7656755', '84916_7812230', '89807_8262164', '89807_8262165', '90286_8306231', '90286_8306270', '96399_8868673', '103585_9529778', '103655_9536257']\n",
      "X_het markers ->  ['13100_1205184', '13645_1255324', '15887_1461572', '15942_1466592', '15942_1466593', '15942_1466596', '15942_1466623', '20918_1924443', '21172_1947744', '21172_1947783', '22661_2084736', '22661_2084781', '22661_2084805', '22661_2084808', '22661_2084813', '25649_2359648', '26727_2458879', '26818_2467177', '34642_3187031', '38727_3562850', '53183_4892813', '61888_5693641', '70797_6513254', '70797_6513258', '80971_7449283', '91834_8448654', '91834_8448657', '91834_8448689', '91834_8448694', '94457_8689962', '94457_8690016']\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of Putative Y linked tags: 31\n",
      "Number of Putative Y linked tags: 9\n",
      "Number of Putative Y linked tags: 51\n",
      "Number of Putative Y linked tags: 22\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 7\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 3\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 96\n",
      "Final number of XY tags = 48\n",
      "Final number of XY tags = 94\n",
      "Final number of XY tags = 57\n",
      "Final number of ZW tags = 27\n",
      "Final number of ZW tags = 36\n",
      "Final number of ZW tags = 29\n",
      "Final number of ZW tags = 36\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 12\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 11\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 12\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 11\n",
      "Number of putative Z linked markers = 19\n",
      "Number of putative Z linked markers = 10\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:  9.5min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 21.7min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/Populations_N_spain/Randomisations/Randomisations_Ribe_Sp.txt\n",
      "processing dataset in /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/Randomisations\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "Number of putative Z linked tags = 19\n",
      "Number of putative Z linked tags = 10\n",
      "Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 1\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative X linked snps = 38\n",
      "Number of putative X linked snps = 33\n",
      "Number of putative X linked snps = 24\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked tags = 38\n",
      "Number of putative X linked tags = 33\n",
      "Number of putative X linked tags = 24\n",
      "Number of putative X linked tags = 20\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative Z linked markers = 46\n",
      "Number of putative Z linked markers = 16\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked markers = 24\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 46\n",
      "Number of putative Z linked tags = 16\n",
      "Number of putative Z linked tags = 21\n",
      "Number of putative Z linked tags = 24\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122X_het markers ->  ['3044_279993', '11856_1090722', '11856_1090750', '15984_1470478', '25549_2350471', '33865_3115550', '33865_3115571', '34178_3144296', '50334_4630726', '50497_4645652', '51823_4767691', '52913_4867975', '54936_5054048', '55747_5128688', '59450_5469374', '59475_5471621', '59475_5471662', '60214_5539684', '60595_5574671', '64675_5950079', '66676_6134143', '77260_7107879', '79021_7269898', '86531_7960780', '87359_8036993', '87650_8063775', '87650_8063776', '88450_8137358', '90306_8308072', '90952_8367562', '91816_8447010', '93600_8611182', '96278_8857572', '97005_8924390', '97005_8924403', '98999_9107907', '102777_9455473', '107181_9860609']\n",
      "X_het markers ->  ['15887_1461572', '15999_1471854', '15999_1471907', '17290_1590672', '17517_1611504', '20562_1891698', '23409_2153626', '24036_2211286', '24464_2250629', '27478_2527917', '29402_2704918', '29835_2744746', '29835_2744761', '29835_2744763', '29835_2744793', '31140_2864809', '34126_3139586', '34748_3196808', '38296_3523202', '39079_3595233', '52670_4845557', '54222_4988345', '54222_4988346', '54777_5039420', '58063_5341733', '63746_5864574', '67546_6214172', '72333_6654589', '73987_6806766', '73987_6806768', '76299_7019463', '82863_7623346', '96399_8868673']\n",
      "X_het markers ->  ['9843_905527', '18115_1666509', '23241_2138089', '25507_2346590', '25507_2346615', '29827_2744041', '33691_3099558', '33893_3118090', '35016_3221402', '35016_3221409', '41178_3788327', '44407_4085439', '44582_4101495', '47617_4380764', '55684_5122895', '59866_5507617', '70606_6495723', '70606_6495725', '71197_6550083', '75231_6921202', '90012_8281044', '93600_8611182', '98250_9038986', '99276_9133328']\n",
      "X_het markers ->  ['22336_2054857', '22336_2054858', '22336_2054882', '22336_2054899', '25549_2350471', '26746_2460580', '53700_4940397', '66692_6135590', '66692_6135644', '67829_6240198', '77765_7154337', '77765_7154359', '85556_7871132', '86234_7933513', '96357_8864787', '96357_8864801', '101343_9323541', '102713_9449538', '104400_9604736', '104400_9604737']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of Putative Y linked tags: 22\n",
      "Number of Putative Y linked tags: 18\n",
      "Number of Putative Y linked tags: 28\n",
      "Number of Putative Y linked tags: 42\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of Putative W linked tags: 4\n",
      "Number of Putative W linked tags: 6\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 1\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 63\n",
      "Final number of XY tags = 59\n",
      "Final number of XY tags = 55\n",
      "Final number of XY tags = 70\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Final number of ZW tags = 63\n",
      "Final number of ZW tags = 30\n",
      "Final number of ZW tags = 25\n",
      "Final number of ZW tags = 26\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative X linked snps = 7\n",
      "Number of putative X linked snps = 14\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 16\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 7\n",
      "Number of putative X linked tags = 14\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 16\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative Z linked markers = 7\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 2\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative Z linked tags = 7\n",
      "Number of putative Z linked tags = 5\n",
      "Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 2\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative X linked snps = 26\n",
      "Number of putative X linked snps = 28\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked tags = 26\n",
      "Number of putative X linked tags = 28\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative X linked tags = 36\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative Z linked markers = 31\n",
      "Number of putative Z linked markers = 30\n",
      "Number of putative Z linked markers = 28\n",
      "Number of putative Z linked markers = 18\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 31\n",
      "Number of putative Z linked tags = 30\n",
      "Number of putative Z linked tags = 28\n",
      "Number of putative Z linked tags = 18\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122X_het markers ->  ['9843_905527', '14257_1311610', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '22039_2027589', '24769_2278666', '30262_2784028', '32347_2975851', '35016_3221402', '35016_3221409', '41867_3851762', '66552_6122766', '67179_6180403', '73406_6753314', '75231_6921202', '76078_6999128', '76708_7057055', '83709_7701146', '85340_7851263', '90012_8281044', '98570_9068373', '101535_9341208', '102554_9434950']\n",
      "X_het markers ->  ['17517_1611504', '22929_2109403', '24221_2228330', '24464_2250629', '27478_2527917', '28354_2608521', '34748_3196808', '36490_3357006', '36490_3357072', '36490_3357074', '36584_3365652', '38296_3523202', '39079_3595233', '44831_4124390', '44831_4124413', '46441_4272550', '54222_4988345', '54222_4988346', '58063_5341733', '62317_5733086', '66692_6135590', '66692_6135644', '67829_6240198', '76299_7019427', '76299_7019463', '86234_7933513', '89981_8278246', '106486_9796704']\n",
      "X_het markers ->  ['2457_226010', '2457_226039', '8600_791143', '8827_812082', '9928_913366', '10231_941208', '10529_968590', '11055_1017043', '23177_2132255', '23905_2199203', '26148_2405538', '30202_2778530', '30585_2813788', '31585_2905737', '32045_2948091', '33341_3067342', '33504_3082332', '33766_3106426', '33951_3123475', '34126_3139586', '34505_3174409', '34573_3180687', '40416_3718273', '51064_4697882', '55242_5082189', '55242_5082264', '58928_5421368', '58950_5423373', '60634_5578244', '68207_6275008', '68452_6297544', '69324_6377769', '69324_6377781', '71064_6537859', '75243_6922306', '75243_6922307', '76955_7079816', '77495_7129483', '89301_8215642', '92356_8496696', '94263_8672142', '94263_8672194']\n",
      "X_het markers ->  ['1479_136012', '2457_226010', '2457_226039', '6661_612783', '6661_612790', '8600_791143', '9309_856412', '9414_866019', '10386_955468', '18153_1670030', '20800_1913535', '21537_1981336', '23241_2138089', '23241_2138090', '23241_2138091', '26541_2441690', '30202_2778530', '37517_3451508', '37517_3451522', '49031_4510807', '49336_4538883', '51064_4697882', '51331_4722440', '59866_5507617', '61468_5654988', '61513_5659168', '71064_6537859', '72576_6676925', '75243_6922306', '75243_6922307', '76955_7079816', '78461_7218412', '80219_7380092', '83226_7656755', '88252_8119134', '93600_8611182']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of Putative Y linked tags: 41\n",
      "Number of Putative Y linked tags: 11\n",
      "Number of Putative Y linked tags: 5\n",
      "Number of Putative Y linked tags: 37\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 6\n",
      "Number of Putative W linked tags: 2\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 73\n",
      "Final number of XY tags = 49\n",
      "Final number of XY tags = 66\n",
      "Final number of XY tags = 87\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Final number of ZW tags = 39\n",
      "Final number of ZW tags = 34\n",
      "Final number of ZW tags = 37\n",
      "Final number of ZW tags = 22\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 8\n",
      "Number of putative X linked snps = 14\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 8\n",
      "Number of putative X linked tags = 14\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked markers = 3\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 5\n",
      "Number of putative Z linked tags = 3\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative X linked snps = 21\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative X linked snps = 24\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked tags = 21\n",
      "Number of putative X linked tags = 20\n",
      "Number of putative X linked tags = 36\n",
      "Number of putative X linked tags = 24\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative Z linked markers = 34\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked markers = 22\n",
      "Number of putative Z linked markers = 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative Z linked tags = 21\n",
      "Number of putative Z linked tags = 22\n",
      "Number of putative Z linked tags = 19\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122X_het markers ->  ['134_12255', '10748_988804', '10748_988805', '10867_999683', '13529_1244628', '18115_1666509', '21912_2015843', '26423_2430861', '27330_2514314', '31225_2872695', '32224_2964542', '34178_3144296', '35016_3221402', '38866_3575641', '52985_4874551', '52985_4874569', '58005_5336448', '58063_5341733', '59475_5471621', '59475_5471662', '81829_7528228']\n",
      "X_het markers ->  ['13820_1271423', '22161_2038733', '22161_2038749', '22161_2038765', '23383_2151160', '34178_3144296', '36584_3365652', '40416_3718273', '47547_4374255', '60403_5557076', '64992_5979242', '72336_6654864', '72336_6654910', '73263_6740196', '77495_7129527', '83028_7638524', '86039_7915530', '89981_8278246', '91737_8439724', '98027_9018410']\n",
      "X_het markers ->  ['8010_736856', '9618_884832', '14257_1311610', '21166_1947192', '22161_2038733', '22161_2038749', '22161_2038765', '22929_2109403', '25284_2326063', '27350_2516146', '32083_2951629', '32669_3005524', '32938_3030268', '34748_3196808', '38621_3553111', '39586_3641903', '40156_3694308', '47601_4379264', '58063_5341733', '66692_6135590', '66692_6135644', '67215_6183726', '68796_6329184', '69634_6406329', '76004_6992329', '76315_7020958', '76470_7035180', '77789_7156579', '79981_7358189', '84623_7785310', '87162_8018899', '89114_8198476', '92668_8525455', '94701_8712428', '99572_9160574', '99572_9160575']\n",
      "X_het markers ->  ['6369_585949', '11582_1065536', '21912_2015843', '22336_2054857', '22336_2054858', '22336_2054882', '22336_2054899', '22929_2109403', '26423_2430861', '27753_2553196', '27753_2553217', '27753_2553248', '27753_2553256', '27753_2553273', '32695_3007856', '32695_3007917', '46441_4272550', '49655_4568211', '53994_4967397', '57303_5271828', '64675_5950079', '86039_7915530', '86531_7960780', '89170_8203557']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of Putative Y linked tags: 28\n",
      "Number of Putative Y linked tags: 8\n",
      "Number of Putative Y linked tags: 10\n",
      "Number of Putative Y linked tags: 7\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of Putative W linked tags: 4\n",
      "Number of Putative W linked tags: 6\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 3\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 52\n",
      "Final number of XY tags = 31\n",
      "Final number of XY tags = 53\n",
      "Final number of XY tags = 42\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Final number of ZW tags = 43\n",
      "Final number of ZW tags = 31\n",
      "Final number of ZW tags = 28\n",
      "Final number of ZW tags = 25\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative X linked snps = 14\n",
      "Number of putative X linked snps = 12\n",
      "Number of putative X linked snps = 7\n",
      "Number of putative X linked snps = 4\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 14\n",
      "Number of putative X linked tags = 12\n",
      "Number of putative X linked tags = 7\n",
      "Number of putative X linked tags = 4\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 7\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 7\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative X linked snps = 32\n",
      "Number of putative X linked snps = 35\n",
      "Number of putative X linked snps = 30\n",
      "Number of putative X linked snps = 14\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked tags = 32\n",
      "Number of putative X linked tags = 35\n",
      "Number of putative X linked tags = 30\n",
      "Number of putative X linked tags = 14\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative Z linked markers = 28\n",
      "Number of putative Z linked markers = 32\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 43\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 28\n",
      "Number of putative Z linked tags = 32\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 43\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122X_het markers ->  ['6910_635712', '8815_810900', '14940_1374448', '20562_1891674', '21172_1947744', '21172_1947783', '25507_2346590', '25507_2346615', '25887_2381534', '25887_2381603', '27607_2539845', '30767_2830486', '32882_3025065', '37113_3414374', '41858_3850936', '42859_3943005', '42859_3943018', '45883_4221237', '46441_4272550', '51189_4709381', '53876_4956569', '73588_6770081', '74926_6893172', '86531_7960780', '89170_8203557', '90384_8315317', '95127_8751676', '95127_8751684', '98832_9092509', '103421_9514703', '103899_9558665', '108674_9997972']\n",
      "X_het markers ->  ['9149_841643', '11119_1022901', '21172_1947744', '21172_1947783', '25805_2373985', '26752_2461109', '26818_2467177', '27588_2538086', '27588_2538094', '27588_2538096', '27607_2539845', '29466_2710792', '29466_2710802', '29827_2744041', '34485_3172572', '34708_3193083', '46345_4263683', '50497_4645652', '54936_5054048', '66676_6134143', '67516_6211427', '69782_6419896', '87359_8036993', '87650_8063775', '87650_8063776', '90952_8367562', '91816_8447010', '94457_8689962', '94457_8690016', '95471_8783248', '95471_8783295', '97005_8924390', '97005_8924403', '105263_9684181', '107181_9860609']\n",
      "X_het markers ->  ['1546_142218', '3420_314578', '9127_839684', '12922_1188812', '20562_1891698', '23015_2117371', '24903_2291037', '33893_3118090', '41307_3800217', '41867_3851762', '47617_4380764', '52670_4845557', '54538_5017439', '67546_6214172', '70606_6495723', '70606_6495725', '71197_6550083', '73588_6770081', '73987_6806766', '73987_6806768', '75689_6963315', '81887_7533566', '83459_7678175', '86533_7960962', '86533_7960969', '88444_8136811', '91435_8411992', '98832_9092509', '102589_9438188', '103585_9529778']\n",
      "X_het markers ->  ['5714_525657', '20562_1891698', '22661_2084736', '22661_2084781', '22661_2084805', '22661_2084808', '22661_2084813', '25649_2359648', '37113_3414374', '41674_3833953', '91834_8448657', '91834_8448689', '91834_8448694', '103655_9536257']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of Putative Y linked tags: 16\n",
      "Number of Putative Y linked tags: 11\n",
      "Number of Putative Y linked tags: 35\n",
      "Number of Putative Y linked tags: 56\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of Putative W linked tags: 7\n",
      "Number of Putative W linked tags: 4\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:  8.5min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 19.5min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Riberica/Stacks_trimmed/IDd/FAMILY/New_family_assignments/Randomisations/Randomisations_Ribe_fam.txt\n",
      "processing dataset in /home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/Randomisations\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 61\n",
      "Final number of XY tags = 54\n",
      "Final number of XY tags = 72\n",
      "Final number of XY tags = 73\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 39\n",
      "Final number of ZW tags = 36\n",
      "Final number of ZW tags = 30\n",
      "Final number of ZW tags = 50\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 17\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 9\n",
      "Number of putative X linked snps = 6\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative X linked tags = 17\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 9\n",
      "Number of putative X linked tags = 6\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 16\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 16\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 47\n",
      "Number of putative X linked snps = 26\n",
      "Number of putative X linked snps = 31\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 47\n",
      "Number of putative X linked tags = 26\n",
      "Number of putative X linked tags = 31\n",
      "Number of putative X linked tags = 20\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 24\n",
      "Number of putative Z linked markers = 16\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 28\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 24\n",
      "Number of putative Z linked tags = 16\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 28\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29X_het markers ->  ['8600_791143', '8815_810900', '10707_984978', '12643_1163106', '12643_1163107', '14940_1374448', '22161_2038733', '22161_2038749', '22161_2038765', '22929_2109403', '25274_2325163', '25319_2329300', '25887_2381534', '25887_2381603', '30202_2778530', '32224_2964542', '32695_3007856', '32695_3007917', '35025_3222236', '35025_3222298', '39052_3592711', '39443_3628708', '51189_4709381', '56207_5171032', '59692_5491658', '59692_5491661', '64992_5979242', '65106_5989748', '72145_6637278', '72336_6654864', '72336_6654910', '73263_6740196', '74926_6893172', '77495_7129483', '77765_7154337', '77765_7154359', '79754_7337326', '79754_7337346', '82593_7598551', '86483_7956370', '88174_8111999', '88753_8165205', '88753_8165210', '89114_8198476', '91737_8439724', '95852_8818362', '104608_9623870']\n",
      "X_het markers ->  ['6910_635712', '18189_1673314', '18189_1673347', '18189_1673364', '18189_1673367', '18918_1740372', '19096_1756756', '19096_1756770', '19096_1756775', '26945_2478938', '33137_3048527', '33137_3048531', '33137_3048557', '33137_3048574', '33137_3048578', '33137_3048582', '33137_3048592', '46441_4272550', '56932_5237736', '75550_6950543', '75550_6950547', '76299_7019427', '91834_8448654', '91834_8448657', '91834_8448689', '91834_8448694']\n",
      "X_het markers ->  ['1479_136012', '10867_999683', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '19096_1756756', '19096_1756775', '21118_1942836', '21118_1942846', '21927_2017201', '21927_2017250', '25212_2319421', '25887_2381534', '25887_2381603', '32224_2964542', '35065_3225930', '43000_3955963', '52440_4824455', '64344_5919572', '66128_6083768', '68030_6258685', '75243_6922306', '75243_6922307', '83226_7656755', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '88444_8136833']\n",
      "X_het markers ->  ['13100_1205184', '13873_1276266', '14257_1311610', '27350_2516146', '29279_2693640', '29835_2744763', '29835_2744793', '34642_3187031', '34701_3192417', '34701_3192450', '34748_3196808', '38655_3556223', '38727_3562850', '47601_4379264', '71500_6577934', '86925_7997022', '91969_8461077', '93690_8619437', '95471_8783248', '95471_8783295']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative Y linked tags: 68\n",
      "Number of Putative Y linked tags: 6\n",
      "Number of Putative Y linked tags: 17\n",
      "Number of Putative Y linked tags: 18\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of Putative W linked tags: 1\n",
      "Number of Putative W linked tags: 8\n",
      "Number of Putative W linked tags: 7\n",
      "Number of Putative W linked tags: 2\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 129\n",
      "Final number of XY tags = 32\n",
      "Final number of XY tags = 57\n",
      "Final number of XY tags = 39\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 27\n",
      "Final number of ZW tags = 27\n",
      "Final number of ZW tags = 30\n",
      "Final number of ZW tags = 43\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 3\n",
      "Number of putative X linked snps = 7\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 4\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative X linked tags = 3\n",
      "Number of putative X linked tags = 7\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 4\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 3\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of putative Z linked tags = 5\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 3\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 27\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 23\n",
      "Number of putative X linked snps = 21\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 27\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 23\n",
      "Number of putative X linked tags = 21\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 34\n",
      "Number of putative Z linked markers = 34\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 11\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 11\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29X_het markers ->  ['6967_640927', '8827_812082', '14257_1311610', '39916_3672203', '39916_3672208', '43785_4028203', '48311_4444532', '50334_4630726', '53700_4940397', '54225_4988670', '54677_5030267', '60941_5606492', '62217_5723955', '62217_5723964', '67179_6180403', '69609_6404013', '76354_7024506', '76536_7041240', '82303_7571870', '86483_7956370', '89114_8198476', '92927_8549217', '93413_8593973', '93413_8593978', '95852_8818362', '105486_9704659', '106093_9760510']\n",
      "X_het markers ->  ['134_12255', '10386_955468', '22661_2084736', '22661_2084781', '22661_2084808', '22929_2109403', '28354_2608521', '32558_2995270', '53994_4967397', '65106_5989748', '80672_7421797', '86039_7915530', '98027_9018410', '101899_9374631', '101899_9374670', '101899_9374680', '101899_9374683', '102589_9438188', '108674_9997972']\n",
      "X_het markers ->  ['5848_537959', '5848_537963', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '25212_2319421', '30784_2832071', '32347_2975851', '43000_3955963', '54766_5038420', '59667_5489336', '75143_6913085', '75243_6922306', '75243_6922307', '76078_6999128', '83028_7638524', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '89337_8218997']\n",
      "X_het markers ->  ['13873_1276266', '20562_1891698', '24036_2211286', '24464_2250629', '26752_2461109', '29835_2744746', '29835_2744761', '29835_2744763', '32989_3034977', '52670_4845557', '54777_5039420', '63746_5864574', '67546_6214172', '69782_6419896', '73987_6806766', '73987_6806768', '76708_7057055', '76708_7057132', '79508_7314684', '82863_7623346', '86818_7987183']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative Y linked tags: 19\n",
      "Number of Putative Y linked tags: 24\n",
      "Number of Putative Y linked tags: 21\n",
      "Number of Putative Y linked tags: 16\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 2\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 47\n",
      "Final number of XY tags = 50\n",
      "Final number of XY tags = 62\n",
      "Final number of XY tags = 41\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 40\n",
      "Final number of ZW tags = 38\n",
      "Final number of ZW tags = 26\n",
      "Final number of ZW tags = 16\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 8\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 44\n",
      "Number of putative X linked snps = 13\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative X linked tags = 8\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 44\n",
      "Number of putative X linked tags = 13\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 3\n",
      "Number of putative Z linked markers = 14\n",
      "Number of putative Z linked markers = 31\n",
      "Number of putative Z linked markers = 8\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of putative Z linked tags = 3\n",
      "Number of putative Z linked tags = 14\n",
      "Number of putative Z linked tags = 31\n",
      "Number of putative Z linked tags = 8\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:  9.9min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 22.6min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Ritalica/Sex_linked_markers/Pop_CM/Randomisations/Randomisations_Rita.txt\n",
      "processing dataset in /home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/batch_1.vcf\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/Randomisations\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 25\n",
      "Number of putative X linked snps = 17\n",
      "Number of putative X linked snps = 27\n",
      "Number of putative X linked snps = 16\n",
      "Number of putative X linked snps = 45\n",
      "Number of putative X linked snps = 60\n",
      "Number of putative X linked snps = 66\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 25\n",
      "Number of putative X linked tags = 17\n",
      "Number of putative X linked tags = 27\n",
      "Number of putative X linked tags = 16\n",
      "Number of putative X linked tags = 45\n",
      "Number of putative X linked tags = 60\n",
      "Number of putative X linked tags = 66\n",
      "Number of putative X linked tags = 42\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 25\n",
      "Number of putative Z linked markers = 25\n",
      "Number of putative Z linked markers = 25\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 19\n",
      "Number of putative Z linked markers = 70\n",
      "Number of putative Z linked markers = 43\n",
      "Number of putative Z linked markers = 23\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 25\n",
      "Number of putative Z linked tags = 25\n",
      "Number of putative Z linked tags = 25\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 19\n",
      "Number of putative Z linked tags = 70\n",
      "Number of putative Z linked tags = 43\n",
      "Number of putative Z linked tags = 23\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " X_het markers ->  ['1479_136012', '2457_226010', '2457_226039', '6661_612783', '6661_612790', '9414_866019', '9928_913366', '20562_1891698', '23177_2132255', '23905_2199203', '24436_2248034', '26541_2441690', '28402_2612917', '29402_2704918', '30585_2813788', '31585_2905737', '32045_2948091', '33341_3067342', '33504_3082332', '33766_3106426', '33893_3118090', '33951_3123475', '34505_3174409', '34573_3180687', '40416_3718273', '54777_5039420', '55242_5082189', '55242_5082264', '58928_5421368', '60634_5578244', '68207_6275008', '68452_6297544', '69324_6377769', '69324_6377781', '69588_6402053', '69782_6419896', '72576_6676925', '77495_7129483', '79596_7322786', '88542_8145861', '89301_8215642', '90893_8362108', '91901_8454847', '92356_8496696', '102589_9438188']\n",
      "X_het markers ->  ['3044_279993', '10748_988804', '10748_988805', '11119_1022901', '11856_1090722', '11856_1090750', '15984_1470478', '18099_1665049', '23015_2117371', '25507_2346590', '25507_2346615', '29827_2744041', '30767_2830486', '32056_2949081', '32056_2949119', '32056_2949135', '32056_2949139', '32056_2949144', '34178_3144296', '34708_3193083', '37113_3414374', '41674_3833953', '42859_3943005', '42859_3943018', '46345_4263683', '47617_4380764', '50334_4630726', '50497_4645652', '52913_4867975', '54936_5054048', '55747_5128688', '59450_5469374', '59475_5471621', '59475_5471662', '60214_5539684', '60595_5574671', '66676_6134143', '70606_6495723', '70606_6495725', '71197_6550083', '73588_6770081', '75231_6921202', '79021_7269898', '86531_7960780', '87359_8036993', '87650_8063775', '87650_8063776', '88450_8137358', '90952_8367562', '91816_8447010', '93600_8611182', '95471_8783248', '95471_8783295', '97005_8924390', '97005_8924403', '98250_9038986', '98250_9038996', '98999_9107907', '102777_9455473', '107181_9860609']\n",
      "X_het markers ->  ['8600_791143', '8827_812082', '9040_831598', '9149_841643', '10675_982060', '12249_1126866', '12249_1126868', '12998_1195738', '22336_2054857', '22336_2054858', '22336_2054882', '22336_2054899', '23805_2190034', '25805_2373985', '26234_2413484', '27688_2547235', '28402_2612917', '29614_2724477', '29827_2744041', '29846_2745776', '29846_2745781', '30202_2778530', '30690_2823415', '31640_2910875', '33941_3122527', '35914_3304010', '36955_3399811', '38824_3571733', '38824_3571761', '45883_4221237', '48627_4473643', '49818_4583201', '51064_4697882', '52532_4832944', '52618_4840777', '53183_4892813', '55676_5122177', '55684_5122895', '58815_5410912', '58941_5422546', '59878_5508720', '59878_5508758', '61378_5646720', '68861_6335155', '69324_6377769', '69324_6377781', '69490_6393007', '71064_6537859', '73014_6717281', '76557_7043208', '76955_7079816', '78254_7199288', '78687_7239181', '81433_7491822', '82588_7598058', '82588_7598097', '84776_7799346', '88018_8097572', '88018_8097593', '89301_8215642', '91899_8454695', '92668_8525455', '94701_8712428', '94716_8713813', '99491_9153129', '105841_9737342']\n",
      "X_het markers ->  ['624_57387', '1479_136012', '1728_158953', '4739_435953', '5848_537959', '5848_537963', '5919_544501', '7083_651634', '7083_651635', '8706_800947', '10754_989369', '12097_1112916', '12097_1112924', '24903_2291037', '26727_2458800', '36331_3342447', '46947_4319073', '50334_4630726', '62275_5729266', '62275_5729285', '67516_6211427', '73987_6806766', '73987_6806768', '75243_6922306', '75243_6922307', '76078_6999128', '77589_7138181', '78720_7242169', '78720_7242178', '78765_7246369', '83226_7656755', '84916_7812230', '90286_8306231', '90286_8306270', '90891_8361911', '91435_8411992', '91459_8414153', '91459_8414207', '97005_8924390', '97005_8924403', '101700_9356358', '103585_9529778']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 5452\n",
      "Number of loci with too few samples = 5446\n",
      "Number of loci with too few samples = 5432\n",
      "Number of loci with too few samples = 5442\n",
      "Number of Putative Y linked tags: 16\n",
      "Number of Putative Y linked tags: 26\n",
      "Number of Putative Y linked tags: 13\n",
      "Number of Putative Y linked tags: 60\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of Putative W linked tags: 5\n",
      "Number of Putative W linked tags: 4\n",
      "Number of Putative W linked tags: 15\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative X linked snps = 24\n",
      "Number of putative X linked snps = 16\n",
      "Number of putative X linked snps = 8\n",
      "Number of putative X linked snps = 23\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative X linked tags = 24\n",
      "Number of putative X linked tags = 16\n",
      "Number of putative X linked tags = 8\n",
      "Number of putative X linked tags = 23\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 18\n",
      "Number of putative Z linked markers = 14\n",
      "Number of putative Z linked markers = 12\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 18\n",
      "Number of putative Z linked tags = 14\n",
      "Number of putative Z linked tags = 12\n",
      "Final number of XY tags = 69\n",
      "Final number of XY tags = 88\n",
      "Final number of XY tags = 80\n",
      "Final number of XY tags = 111\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Final number of ZW tags = 27\n",
      "Final number of ZW tags = 88\n",
      "Final number of ZW tags = 67\n",
      "Final number of ZW tags = 30\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =X_het markers ->  ['2418_222379', '5775_531294', '10719_986109', '12635_1162347', '17029_1566651', '22551_2074666', '25874_2380367', '28616_2632672', '29620_2725003', '33240_3058001', '36083_3319568', '38482_3540338', '39640_3646872', '40288_3706464', '68688_6319257', '69440_6388397', '69440_6388450', '69882_6429141', '71660_6592648', '72326_6653971', '73060_6721486', '74301_6835688', '81452_7493522', '117304_10791934']\n",
      "X_het markers ->  ['4545_418126', '5821_535453', '6985_642563', '14873_1368315', '15818_1455204', '18291_1682756', '22286_2050295', '31508_2898695', '33334_3066644', '37179_3420466', '41451_3813416', '51966_4780842', '52496_4829588', '71373_6566292', '72725_6690651', '86175_7928057']\n",
      "X_het markers ->  ['6590_606258', '20646_1899350', '33725_3102620', '40270_3704760', '60526_5568311', '70171_6455649', '74093_6816542', '74455_6849841']\n",
      "X_het markers ->  ['2248_206788', '2492_229201', '13964_1284611', '13964_1284612', '16368_1505776', '16368_1505788', '22058_2029272', '22720_2090199', '25827_2376016', '25827_2376047', '26047_2396320', '38750_3564920', '50325_4629855', '54688_5031217', '56113_5162378', '62342_5735461', '70659_6500571', '70659_6500578', '72654_6684132', '72654_6684169', '76333_7022630', '77268_7108578', '84751_7797073']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of Putative Y linked tags: 32\n",
      "Number of Putative Y linked tags: 49\n",
      "Number of Putative Y linked tags: 56\n",
      "Number of Putative Y linked tags: 30\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of Putative W linked tags: 19\n",
      "Number of Putative W linked tags: 19\n",
      "Number of Putative W linked tags: 9\n",
      "Number of Putative W linked tags: 36\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked snps = 8\n",
      "Number of putative X linked snps = 10\n",
      "Number of putative X linked snps = 13\n",
      "Number of putative X linked snps = 7\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 8\n",
      "Number of putative X linked tags = 10\n",
      "Number of putative X linked tags = 13\n",
      "Number of putative X linked tags = 7\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked markers = 11\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 5\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 11\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 5\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 79\n",
      "Final number of XY tags = 82\n",
      "Final number of XY tags = 91\n",
      "Final number of XY tags = 67\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 49\n",
      "Final number of ZW tags = 61\n",
      "Final number of ZW tags = 48\n",
      "Final number of ZW tags = 71\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 27\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative X linked snps = 31\n",
      "Number of putative X linked snps = 33\n",
      "Number of putative X linked snps = 13\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative X linked snps = 30\n",
      "Number of putative X linked snps = 33\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 27\n",
      "Number of putative X linked tags = 20\n",
      "Number of putative X linked tags = 31\n",
      "Number of putative X linked tags = 33\n",
      "Number of putative X linked tags = 13\n",
      "Number of putative X linked tags = 36\n",
      "Number of putative X linked tags = 30\n",
      "Number of putative X linked tags = 33\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 36\n",
      "Number of putative Z linked markers = 34\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked markers = 27\n",
      "Number of putative Z linked markers = 32\n",
      "Number of putative Z linked markers = 35\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 36\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative Z linked tags = 33\n",
      "Number of putative Z linked tags = 21\n",
      "Number of putative Z linked tags = 27\n",
      "Number of putative Z linked tags = 32\n",
      "Number of putative Z linked tags = 35\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " X_het markers ->  ['134_12255', '10724_986575', '30767_2830486', '35472_3263359', '41178_3788327', '41858_3850936', '56615_5208540', '60108_5529903', '61513_5659168', '75243_6922306', '75243_6922307', '85419_7858491', '89170_8203557']\n",
      "X_het markers ->  ['8305_764029', '8305_764036', '17512_1611053', '17512_1611061', '17512_1611062', '17512_1611068', '17512_1611070', '21583_1985603', '21583_1985628', '22039_2027589', '23986_2206677', '26190_2409449', '27607_2539819', '28924_2660955', '30767_2830486', '34042_3131799', '39916_3672203', '39916_3672208', '56932_5237736', '62275_5729266', '62275_5729285', '80672_7421797', '82045_7548068', '83709_7701146', '90286_8306231', '90286_8306270', '90367_8313697', '90384_8315317', '96646_8891424', '101535_9341208', '101899_9374631', '101899_9374670', '101899_9374680', '101899_9374683', '102554_9434950', '103655_9536257']\n",
      "X_het markers ->  ['8600_791143', '18115_1666509', '18305_1684008', '21557_1983165', '21557_1983174', '21557_1983177', '21557_1983185', '21557_1983239', '25657_2360375', '27330_2514314', '27588_2538086', '27588_2538094', '27588_2538096', '30202_2778530', '30398_2796603', '36955_3399811', '43000_3955963', '43640_4014831', '43640_4014874', '51064_4697882', '69324_6377769', '69324_6377781', '71064_6537859', '76955_7079816', '77260_7107879', '87880_8084957', '89301_8215642', '94716_8713813', '97422_8962805', '107181_9860609']\n",
      "X_het markers ->  ['134_12255', '11508_1058728', '15942_1466592', '15942_1466593', '15942_1466596', '15942_1466623', '18918_1740372', '22878_2104747', '25649_2359648', '26727_2458800', '26727_2458837', '26727_2458874', '26727_2458875', '28636_2634498', '30539_2809535', '32224_2964542', '34042_3131799', '41621_3829094', '56932_5237736', '65106_5989748', '70797_6513254', '70797_6513258', '77480_7128088', '80831_7436421', '82592_7598465', '91834_8448654', '91834_8448657', '91834_8448689', '91834_8448694', '95651_8799811', '95651_8799860', '97422_8962805', '104608_9623870']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 5446\n",
      "Number of loci with too few samples = 5455\n",
      "Number of loci with too few samples = 5455\n",
      "Number of loci with too few samples = 5452\n",
      "Number of Putative Y linked tags: 12\n",
      "Number of Putative Y linked tags: 6\n",
      "Number of Putative Y linked tags: 10\n",
      "Number of Putative Y linked tags: 37\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of Putative W linked tags: 6\n",
      "Number of Putative W linked tags: 5\n",
      "Number of Putative W linked tags: 3\n",
      "Number of Putative W linked tags: 4\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative X linked snps = 16\n",
      "Number of putative X linked snps = 15\n",
      "Number of putative X linked snps = 24\n",
      "Number of putative X linked snps = 25\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative X linked tags = 16\n",
      "Number of putative X linked tags = 15\n",
      "Number of putative X linked tags = 24\n",
      "Number of putative X linked tags = 25\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked markers = 10\n",
      "Number of putative Z linked markers = 22\n",
      "Number of putative Z linked markers = 9\n",
      "Number of putative Z linked markers = 14\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 10\n",
      "Number of putative Z linked tags = 22\n",
      "Number of putative Z linked tags = 9\n",
      "Number of putative Z linked tags = 14\n",
      "Final number of XY tags = 32\n",
      "Final number of XY tags = 52\n",
      "Final number of XY tags = 51\n",
      "Final number of XY tags = 76\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Final number of ZW tags = 37\n",
      "Final number of ZW tags = 35\n",
      "Final number of ZW tags = 35\n",
      "Final number of ZW tags = 44\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =X_het markers ->  ['11468_1055032', '12687_1167177', '17529_1612657', '26312_2420629', '30005_2760389', '52962_4872494', '54688_5031217', '56749_5220862', '62342_5735461', '65598_6034936', '65823_6055697', '66383_6107231', '69123_6359255', '74093_6816542', '75446_6941005', '76333_7022630']\n",
      "X_het markers ->  ['1378_126763', '19507_1794570', '22551_2074666', '26392_2427998', '30507_2806627', '33093_3044508', '48181_4432607', '54442_5008606', '54442_5008621', '59034_5431067', '61910_5695699', '63302_5823763', '70593_6494502', '75264_6924268', '85928_7905362']\n",
      "X_het markers ->  ['1958_180101', '6742_620240', '13095_1204716', '14408_1325519', '17578_1617143', '32048_2948379', '34344_3159591', '34344_3159613', '34909_3211607', '38665_3557122', '40386_3715477', '49610_4564117', '50626_4657511', '50739_4667925', '51567_4744108', '54723_5034452', '56520_5199759', '57000_5243977', '63789_5868562', '70189_6457309', '75904_6983091', '78280_7201679', '79826_7343938', '111504_10258347']\n",
      "X_het markers ->  ['872_80216', '872_80219', '1058_97319', '4564_419818', '5775_531294', '12239_1125916', '12936_1190087', '15621_1437073', '17450_1605356', '23477_2159876', '24811_2282612', '24811_2282613', '25572_2352556', '27688_2547234', '31951_2939412', '32031_2946790', '32031_2946843', '35845_3297669', '38537_3545379', '40179_3696388', '61934_5697863', '68157_6270407', '70910_6523693', '81452_7493522', '85928_7905362']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of Putative Y linked tags: 49\n",
      "Number of Putative Y linked tags: 59\n",
      "Number of Putative Y linked tags: 54\n",
      "Number of Putative Y linked tags: 29\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of Putative W linked tags: 11\n",
      "Number of Putative W linked tags: 12\n",
      "Number of Putative W linked tags: 20\n",
      "Number of Putative W linked tags: 22\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 12\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 12\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 1\n",
      "Number of putative Z linked markers = 4\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 1\n",
      "Number of putative Z linked tags = 4\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 89\n",
      "Final number of XY tags = 94\n",
      "Final number of XY tags = 105\n",
      "Final number of XY tags = 82\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 44\n",
      "Final number of ZW tags = 70\n",
      "Final number of ZW tags = 63\n",
      "Final number of ZW tags = 68\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed: 30.7min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 69.9min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Rkukinoris/Stacks/Populations_Nanping_kept_altered/Randomisations/Randomisations_Rkuk.txt\n",
      "processing dataset in /home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/batch_1.vcf.altered\n",
      "\n",
      "Randomisations happening in /home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/RandomisationsNumber of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative X linked snps = 33\n",
      "Number of putative X linked snps = 24\n",
      "Number of putative X linked snps = 38\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked snps = 37\n",
      "Number of putative X linked snps = 53\n",
      "Number of putative X linked snps = 34\n",
      "Number of putative X linked snps = 26\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 36\n",
      "Number of putative X linked tags = 33\n",
      "Number of putative X linked tags = 24\n",
      "Number of putative X linked tags = 38\n",
      "Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of putative X linked tags = 37\n",
      "Number of putative X linked tags = 53\n",
      "Number of putative X linked tags = 34\n",
      "Number of putative X linked tags = 26\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 43\n",
      "Number of putative Z linked markers = 20\n",
      "Number of putative Z linked markers = 27\n",
      "Number of putative Z linked markers = 23\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 12\n",
      "Number of putative Z linked markers = 27\n",
      "Number of putative Z linked markers = 32\n",
      "Number of putative Z linked markers = 21\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 43\n",
      "Number of putative Z linked tags = 20\n",
      "Number of putative Z linked tags = 27\n",
      "Number of putative Z linked tags = 23\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of putative Z linked tags = 12\n",
      "Number of putative Z linked tags = 27\n",
      "Number of putative Z linked tags = 32\n",
      "Number of putative Z linked tags = 21\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      " 38\n",
      " 38\n",
      " 38\n",
      " 38\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "X_het markers ->  ['3014_277282', '8852_814315', '10659_980615', '11508_1058728', '17528_1612536', '18115_1666509', '19030_1750725', '24464_2250629', '27753_2553196', '27753_2553217', '27753_2553248', '27753_2553256', '27753_2553273', '28743_2644274', '28743_2644288', '28743_2644299', '28743_2644302', '28743_2644357', '29466_2710792', '29466_2710802', '41621_3829094', '58063_5341733', '62491_5749125', '69806_6422146', '72110_6634062', '72888_6705631', '72888_6705696', '75243_6922306', '75243_6922307', '77316_7113037', '80874_7440395', '83226_7656755', '88444_8136833', '90306_8308072', '92005_8464393', '94701_8712428', '95388_8775639']\n",
      "X_het markers ->  ['3154_290090', '15327_1410046', '20143_1853083', '20143_1853084', '20143_1853109', '20143_1853115', '20143_1853117', '20143_1853122', '20143_1853125', '22043_2027895', '22161_2038733', '22161_2038749', '22161_2038765', '22661_2084736', '22661_2084781', '22661_2084808', '23483_2160366', '24151_2221885', '24436_2248034', '25549_2350471', '26418_2430387', '26423_2430861', '33436_3076039', '40855_3758654', '46301_4259660', '47042_4327799', '47042_4327861', '47073_4330661', '52677_4846243', '52985_4874551', '52985_4874569', '53994_4967397', '61788_5684497', '65718_6046000', '65718_6046051', '67722_6230373', '72678_6686317', '74509_6854792', '77234_7105472', '79776_7339337', '80890_7441810', '81472_7495370', '83824_7711767', '86039_7915530', '86094_7920622', '88276_8121343', '91195_8389914', '93348_8587939', '93570_8608440', '95054_8744941', '98027_9018410', '99468_9151057', '108674_9997972']\n",
      "X_het markers ->  ['6910_635712', '17512_1611053', '17512_1611061', '17512_1611062', '17512_1611068', '17512_1611070', '17518_1611577', '20481_1884174', '20481_1884185', '20481_1884190', '20481_1884205', '20481_1884235', '20562_1891687', '25319_2329300', '26190_2409449', '28376_2610531', '31225_2872695', '35472_3263359', '37255_3427387', '37255_3427406', '39916_3672203', '39916_3672208', '41858_3850936', '53700_4940397', '54160_4982637', '75550_6950543', '75550_6950547', '79561_7319550', '82045_7548068', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '96646_8891424']\n",
      "X_het markers ->  ['134_12255', '7677_706252', '11508_1058728', '11661_1072760', '12762_1174037', '18115_1666509', '21927_2017201', '21927_2017250', '25055_2304991', '27347_2515849', '29632_2726121', '29827_2744041', '35016_3221402', '41621_3829094', '43058_3961281', '51235_4713607', '64344_5919572', '67502_6210109', '67502_6210175', '69782_6419896', '75231_6921202', '90893_8362108', '93600_8611182', '98250_9038986', '105281_9685852', '105281_9685853']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of putative X linked snps = 1\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 1\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 1\n",
      "Number of putative Z linked markers = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 1\n",
      "Number of putative Z linked tags = 0\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "\n",
      "\n",
      "Random sex info files made\n",
      "\n",
      "Running randomisations\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 5461\n",
      "Number of loci with too few samples = 5447\n",
      "Number of loci with too few samples = 5443\n",
      "Number of loci with too few samples = 5457\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of Putative Y linked tags: 16\n",
      "Number of Putative Y linked tags: 13\n",
      "Number of Putative Y linked tags: 5\n",
      "Number of Putative Y linked tags: 23\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative W linked tags: 8\n",
      "Number of Putative W linked tags: 2\n",
      "Number of Putative W linked tags: 5\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 22\n",
      "Number of putative X linked snps = 27\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 22\n",
      "Number of putative X linked tags = 27\n",
      "Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of loci = 48475\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked markers = 10\n",
      "Number of putative Z linked markers = 12\n",
      "Number of putative Z linked markers = 11\n",
      "Number of putative Z linked markers = 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 10\n",
      "Number of putative Z linked tags = 12\n",
      "Number of putative Z linked tags = 11\n",
      "Number of putative Z linked tags = 19\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Final number of XY tags = 57\n",
      "Final number of XY tags = 70\n",
      "Final number of XY tags = 40\n",
      "Final number of XY tags = 56\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      " 38\n",
      " 38\n",
      " 38\n",
      " 38\n",
      "Final number of ZW tags = 26\n",
      "Final number of ZW tags = 31\n",
      "Final number of ZW tags = 38\n",
      "Final number of ZW tags = 25\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =X_het markers ->  ['3337_306966', '4555_418977', '11266_1036415', '11711_1077388', '16167_1487311', '17529_1612657', '18919_1740521', '20986_1930710', '33306_3064098', '34063_3133795', '35498_3265756', '51157_4706442', '57065_5249961', '62349_5736037', '66936_6158112', '68504_6302310', '68504_6302322', '70135_6452364', '71420_6570560']\n",
      "X_het markers ->  ['664_61004', '1631_150012', '15778_1451576', '24474_2251556', '35139_3232787', '37466_3446792', '41322_3801568', '56476_5195769', '56476_5195781', '56589_5206172', '58389_5371772', '58658_5396469', '61786_5684271', '62769_5774704', '62769_5774716', '70203_6458626', '71660_6592648', '75013_6901136', '77094_7092602']\n",
      "X_het markers ->  ['877_80656', '2413_221952', '11436_1052041', '13735_1263570', '13735_1263615', '16305_1499987', '16610_1528078', '16610_1528086', '18592_1710390', '18699_1720228', '28270_2600795', '28894_2658192', '32346_2975822', '51233_4713380', '60238_5541876', '67364_6197410', '67364_6197435', '67863_6243364', '68404_6293092', '70438_6480232', '71957_6619989', '76653_7052034']\n",
      "X_het markers ->  ['211_19339', '6674_613935', '11991_1103164', '20407_1877401', '20885_1921368', '22150_2037722', '22551_2074630', '23578_2169166', '28515_2623342', '32387_2979589', '35620_3276984', '35620_3277039', '36808_3386322', '37907_3487363', '51515_4739362', '53909_4959611', '58611_5392198', '61559_5663349', '61559_5663357', '65841_6057313', '70905_6523224', '72729_6691053', '72941_6710546', '73663_6776988', '76180_7008508', '76180_7008559', '81411_7489760']\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked snps = 1\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9Number of putative X linked tags = 1\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 1\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 1\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 1\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of Putative Y linked tags: 41\n",
      "Number of Putative Y linked tags: 57\n",
      "Number of Putative Y linked tags: 52\n",
      "Number of Putative Y linked tags: 55\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of Putative W linked tags: 19\n",
      "Number of Putative W linked tags: 13\n",
      "Number of Putative W linked tags: 15\n",
      "Number of Putative W linked tags: 8\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "X_het markers ->  ['11005_1023421']\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  ['35221_3275472']\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked snps = 12\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked snps = 1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of putative X linked tags = 6\n",
      "Number of putative X linked tags = 12\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative X linked tags = 1\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 11\n",
      "Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 14\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25Number of putative Z linked tags = 11\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 14\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 95\n",
      "Final number of XY tags = 106\n",
      "Final number of XY tags = 97\n",
      "Final number of XY tags = 120\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 71\n",
      "Final number of ZW tags = 45\n",
      "Final number of ZW tags = 53\n",
      "Final number of ZW tags = 50\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Final number of XY tags = 2\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 1\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Final number of ZW tags = 2\n",
      "Final number of ZW tags = 3\n",
      "Final number of ZW tags = 1\n",
      "Final number of ZW tags = 0\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 26\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked snps = 36\n",
      "Number of putative X linked snps = 34\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked snps = 27\n",
      "Number of putative X linked snps = 23\n",
      "Number of putative X linked snps = 15\n",
      "Number of putative X linked snps = 33\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 26\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative X linked tags = 36\n",
      "Number of putative X linked tags = 34\n",
      "Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of putative X linked tags = 27\n",
      "Number of putative X linked tags = 23\n",
      "Number of putative X linked tags = 15\n",
      "Number of putative X linked tags = 33\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 23\n",
      "Number of putative Z linked markers = 34\n",
      "Number of putative Z linked markers = 42\n",
      "Number of putative Z linked markers = 52\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked markers = 26\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked markers = 30\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 23\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative Z linked tags = 42\n",
      "Number of putative Z linked tags = 52\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of putative Z linked tags = 33\n",
      "Number of putative Z linked tags = 26\n",
      "Number of putative Z linked tags = 21\n",
      "Number of putative Z linked tags = 30\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      " 38\n",
      " 38\n",
      " 38\n",
      " 38\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "X_het markers ->  ['1479_136012', '10386_955468', '14257_1311610', '16303_1499796', '20102_1849328', '20689_1903311', '20689_1903340', '20689_1903358', '20689_1903382', '24221_2228330', '24903_2291037', '28354_2608521', '30262_2784028', '38621_3553111', '47601_4379264', '66692_6135590', '66692_6135644', '72539_6673539', '76315_7020958', '79508_7314684', '84623_7785310', '85340_7851263', '89114_8198476', '99582_9161471', '99582_9161539', '104918_9652375', '106486_9796704']\n",
      "X_het markers ->  ['5714_525657', '8745_804469', '10529_968590', '10707_984978', '25104_2309532', '26190_2409449', '33504_3082332', '39443_3628708', '53700_4940397', '54225_4988670', '69367_6381709', '72539_6673539', '78120_7187005', '79754_7337326', '79754_7337346', '86483_7956370', '86741_7980135', '95852_8818362', '101899_9374631', '101899_9374670', '101899_9374680', '101899_9374683', '106710_9817320']\n",
      "X_het markers ->  ['13100_1205184', '21172_1947744', '21172_1947783', '30767_2830486', '32294_2971046', '32882_3025065', '50334_4630726', '57207_5263036', '70406_6477313', '90384_8315317', '97005_8924390', '97005_8924403', '98250_9038986', '98250_9038996', '100311_9228557']\n",
      "X_het markers ->  ['4068_374219', '11119_1022901', '14352_1320380', '15327_1410046', '17512_1611053', '17512_1611061', '17512_1611062', '17512_1611068', '17512_1611070', '17528_1612536', '25507_2346615', '33893_3118090', '36438_3352216', '37113_3414374', '41674_3833953', '41828_3848140', '43640_4014831', '43640_4014874', '47617_4380764', '48465_4458735', '70606_6495723', '70606_6495725', '71197_6550083', '81829_7528228', '82045_7548068', '84941_7814524', '86269_7936666', '86269_7936703', '86269_7936704', '86269_7936706', '96278_8857572', '97422_8962805', '107181_9860609']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6\n",
      "SUMMARY:\n",
      "Number of males: 6Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of putative Z linked markers = 2\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked markers = 5\n",
      "Number of putative Z linked markers = 7\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "Number of putative Z linked tags = 2\n",
      "Number of putative Z linked tags = 4\n",
      "Number of putative Z linked tags = 5\n",
      "Number of putative Z linked tags = 7\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of males: 12\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 5446\n",
      "Number of loci with too few samples = 5455\n",
      "Number of loci with too few samples = 5449\n",
      "Number of loci with too few samples = 5442\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of Putative Y linked tags: 47\n",
      "Number of Putative Y linked tags: 16\n",
      "Number of Putative Y linked tags: 9\n",
      "Number of Putative Y linked tags: 5\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 4\n",
      "Number of Putative W linked tags: 10\n",
      "Number of Putative W linked tags: 9\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "Number of loci with enough data = 15678\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 27\n",
      "Number of putative X linked snps = 21\n",
      "Number of putative X linked snps = 11\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 27\n",
      "Number of putative X linked tags = 21\n",
      "Number of putative X linked tags = 11\n",
      "Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of loci = 48475\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked markers = 12\n",
      "Number of putative Z linked markers = 21\n",
      "Number of putative Z linked markers = 15\n",
      "Number of putative Z linked markers = 12\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 12\n",
      "Number of putative Z linked tags = 21\n",
      "Number of putative Z linked tags = 15\n",
      "Number of putative Z linked tags = 12\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Final number of XY tags = 80\n",
      "Final number of XY tags = 46\n",
      "Final number of XY tags = 30\n",
      "Final number of XY tags = 38\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      " 38\n",
      " 38\n",
      " 38\n",
      " 38\n",
      "Final number of ZW tags = 38\n",
      "Final number of ZW tags = 32\n",
      "Final number of ZW tags = 35\n",
      "Final number of ZW tags = 53\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =X_het markers ->  ['4351_400212', '4351_400241', '6166_567192', '6166_567193', '14304_1315932', '20822_1915625', '24567_2260124', '24567_2260126', '29442_2708586', '32563_2995726', '35614_3276471', '36083_3319568', '56559_5203399', '63085_5803774', '63414_5834045', '67386_6199508', '69206_6366906', '71756_6601529', '90192_8297597']\n",
      "X_het markers ->  ['2836_260869', '11905_1095260', '11905_1095261', '14254_1311350', '14783_1360007', '18919_1740521', '19295_1775075', '19295_1775095', '21901_2014842', '23188_2133226', '23188_2133227', '26952_2479511', '27235_2505588', '32190_2961425', '32190_2961452', '32535_2993168', '32535_2993195', '35498_3265756', '37179_3420466', '50477_4643827', '50477_4643879', '50477_4643881', '50477_4643885', '55940_5146415', '56315_5180915', '62914_5788027', '95231_8761198']\n",
      "X_het markers ->  ['3306_304111', '12390_1139880', '20743_1908312', '22259_2047744', '22259_2047747', '29793_2740887', '30200_2778398', '30615_2816559', '35248_3242787', '46539_4281558', '47607_4379841', '49564_4559844', '52434_4823849', '64540_5937640', '72268_6648582', '74798_6881342', '76022_6993974', '78673_7237903', '78995_7267499', '79308_7296335', '93370_8590016']\n",
      "X_het markers ->  ['1389_127712', '22793_2096917', '26812_2466664', '26812_2466704', '36255_3335411', '36255_3335444', '47153_4338068', '49853_4586440', '57613_5300396', '73221_6736251', '73581_6769427']\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30\n",
      "SUMMARY:\n",
      "Number of males: 30 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of males: 36\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9\n",
      "SUMMARY:\n",
      "Number of males: 9Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 1\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of males: 10\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 1\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "Number of loci with too few samples = 2903\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of Putative Y linked tags: 40\n",
      "Number of Putative Y linked tags: 42\n",
      "Number of Putative Y linked tags: 70\n",
      "Number of Putative Y linked tags: 47\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "Number of loci with low MAF = 3061\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of Putative W linked tags: 16\n",
      "Number of Putative W linked tags: 35\n",
      "Number of Putative W linked tags: 17\n",
      "Number of Putative W linked tags: 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative X linked snps = 4\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 2\n",
      "Number of putative X linked snps = 2\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of putative X linked tags = 4\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 2\n",
      "Number of putative X linked tags = 2\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 8\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 6\n",
      "Number of putative Z linked markers = 0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed: 14.1min\n",
      "[Parallel(n_jobs=4)]: Done 100 out of 100 | elapsed: 32.1min finished\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Results outputted to /home/djeffrie/Data/RADseq/Ryavapaiensis/Sex_linked_markers/Randomisations/Randomisations_Ryav.txt\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "Final number of ZW tags = 0\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "SUMMARY:\n",
      "Number of males: 25\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked tags = 8\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 6\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Final number of XY tags = 85\n",
      "Final number of XY tags = 110\n",
      "Final number of XY tags = 126\n",
      "Final number of XY tags = 90\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 3792Number of loci = 3792Number of loci = 3792Number of loci = 3792Final number of ZW tags = 50\n",
      "Final number of ZW tags = 90\n",
      "Final number of ZW tags = 72\n",
      "Final number of ZW tags = 82\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "Number of males: 13\n",
      "Number of loci = 42934Number of loci = 42934Number of loci = 42934Number of loci = 42934\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of Putative W linked tags: 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "\n",
      "\n",
      "\n",
      "Number of samples =Number of samples =Number of samples =Number of samples = 64\n",
      " 64\n",
      " 64\n",
      " 64\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      "\n",
      " ### DONE! ###\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "Number of loci = 12198Number of loci = 12198Number of loci = 12198Number of loci = 12198 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "Number of loci with too few samples = 642\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with too few samples = 555\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "\n",
      " ## After merging tags accross methods ## \n",
      "Number of loci with low MAF = 13329\n",
      "Number of loci with low MAF = 13329\n",
      "Number of loci with low MAF = 13329\n",
      "Number of loci with low MAF = 13329\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      "Number of loci with enough data = 3149\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 29604\n",
      "Number of loci with enough data = 29604\n",
      "Number of loci with enough data = 29604\n",
      "Number of loci with enough data = 29604\n",
      " 18\n",
      " 18\n",
      " 18\n",
      " 18\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of loci with enough data = 3566\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 2\n",
      "Final number of XY tags = 0\n",
      "Final number of XY tags = 0\n",
      "Number of putative X linked snps = 38\n",
      "Number of putative X linked snps = 138\n",
      "Number of putative X linked snps = 46\n",
      "Number of putative X linked snps = 37\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of loci with too few samples = 3975\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Number of loci with low MAF = 7120\n",
      "Final number of ZW tags = 2\n",
      "Final number of ZW tags = 5\n",
      "Final number of ZW tags = 5\n",
      "Final number of ZW tags = 7\n",
      "Number of putative X linked tags = 38\n",
      "Number of putative X linked tags = 138\n",
      "Number of putative X linked tags = 46\n",
      "Number of putative X linked tags = 37\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of loci with low MAF = 1989\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "Number of loci with enough data = 21163\n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "\n",
      "##### Using SNP frequency approach #### \n",
      "Number of putative Z linked markers = 22\n",
      "Number of putative Z linked markers = 92\n",
      "Number of putative Z linked markers = 47\n",
      "Number of putative Z linked markers = 34\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of loci with enough data = 6233\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative X linked snps = 22\n",
      "Number of putative X linked snps = 19\n",
      "Number of putative X linked snps = 25\n",
      "Number of putative X linked snps = 37\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked tags = 22\n",
      "Number of putative Z linked tags = 92\n",
      "Number of putative Z linked tags = 47\n",
      "Number of putative Z linked tags = 34\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked snps = 31\n",
      "Number of putative X linked snps = 42\n",
      "Number of putative X linked snps = 20\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "Number of putative X linked tags = 22\n",
      "Number of putative X linked tags = 19\n",
      "Number of putative X linked tags = 25\n",
      "Number of putative X linked tags = 37\n",
      "Number of loci = 48475Number of loci = 48475Number of loci = 48475Number of loci = 48475\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative X linked tags = 31\n",
      "Number of putative X linked tags = 42\n",
      "Number of putative X linked tags = 20\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 28\n",
      "Number of putative Z linked markers = 17\n",
      "Number of putative Z linked markers = 28\n",
      "Number of putative Z linked markers = 34\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of putative Z linked markers = 48\n",
      "Number of putative Z linked markers = 28\n",
      "Number of putative Z linked markers = 43\n",
      "Number of putative Z linked markers = 16\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "X_het markers ->  []\n",
      "Number of putative Z linked tags = 28\n",
      "Number of putative Z linked tags = 17\n",
      "Number of putative Z linked tags = 28\n",
      "Number of putative Z linked tags = 34\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of putative Z linked tags = 48\n",
      "Number of putative Z linked tags = 28\n",
      "Number of putative Z linked tags = 43\n",
      "Number of putative Z linked tags = 16\n",
      "\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      "\n",
      "***DONE!***\n",
      " 38\n",
      " 38\n",
      " 38\n",
      " 38\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      " ### DONE! ### \n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci with too few samples = 0\n",
      "Number of loci = 42934Number of loci = 42934Number of loci = 42934Number of loci = 42934\n",
      "\n",
      "\n",
      "\n",
      "Number of loci = 4122Number of loci = 4122Number of loci = 4122Number of loci = 4122\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "SUMMARY:\n",
      "Number of males: 29\n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " \n",
      "##### Using SNP heterozygosity approach #####\n",
      " Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "Number of loci with low MAF = 17877\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "X_het markers ->  ['3044_279993', '8697_800072', '8815_810900', '10675_982060', '11856_1090722', '11856_1090750', '13529_1244628', '13603_1251450', '14940_1374448', '18115_1666509', '23241_2138089', '25242_2322199', '34178_3144296', '34618_3184788', '34618_3184851', '34751_3197057', '38866_3575641', '42047_3868284', '50497_4645652', '52913_4867975', '54936_5054048', '58005_5336448', '59475_5471621', '59475_5471662', '60595_5574671', '66676_6134143', '70311_6468598', '85078_7827107', '85345_7851731', '87359_8036993', '87650_8063775', '87650_8063776', '88450_8137358', '90952_8367562', '91816_8447010', '92927_8549217', '93600_8611182', '96267_8856546', '97005_8924390', '97005_8924403', '98999_9107907', '107181_9860609']\n",
      "X_het markers ->  ['1479_136012', '5714_525657', '13873_1276266', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '25121_2311052', '25121_2311053', '25739_2367924', '30398_2796603', '32347_2975851', '42635_3922388', '43640_4014831', '43640_4014874', '44803_4121842', '54766_5038420', '62899_5786630', '62899_5786639', '69324_6377769', '69324_6377781', '85674_7882004', '89259_8211795', '89301_8215642', '90893_8362108', '93690_8619437', '99582_9161471', '99582_9161539', '102451_9425424', '104918_9652375']\n",
      "X_het markers ->  ['1479_136012', '2457_226010', '2457_226039', '6661_612783', '6661_612790', '9414_866019', '10386_955468', '10707_984978', '17518_1611577', '18153_1670030', '22039_2027589', '22661_2084736', '22661_2084781', '22661_2084808', '23241_2138089', '23241_2138090', '23241_2138091', '26541_2441690', '32850_3022148', '32850_3022161', '32850_3022170', '32850_3022171', '41178_3788327', '50402_4636938', '55684_5122895', '58112_5346243', '65804_6053910', '65804_6053920', '72576_6676925', '76078_6999128', '80672_7421797', '80674_7421945', '89114_8198476', '93348_8587939', '98541_9065761', '99276_9133328', '99582_9161471', '99582_9161539', '100846_9277792', '101899_9374631', '101899_9374683', '104918_9652375']\n",
      "X_het markers ->  ['14257_1311610', '17905_1647180', '17905_1647192', '17905_1647203', '17905_1647217', '17905_1647241', '25347_2331858', '27347_2515849', '29632_2726121', '30262_2784028', '32347_2975851', '54766_5038420', '59866_5507617', '66692_6135590', '66692_6135644', '76315_7020958', '79508_7314684', '85340_7851263', '90893_8362108', '98027_9018410']\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of loci with enough data = 30597\n",
      "Number of samples =Number of samples =Number of samples =Number of samples =\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "\n",
      "##### Using Sex specific tag approach ##### \n",
      "Number of samples =Number of samples =Number of samples =Number of samples =Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of males: 35\n",
      "Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of loci = 28284Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 5\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked snps = 4\n",
      " 19\n",
      " 19\n",
      " 19\n",
      " 19\n",
      "\n",
      "\n",
      "\n",
      "\n",
      " 66\n",
      " 66\n",
      " 66\n",
      " 66\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n",
      "Number of Putative Y linked tags: 0\n"
     ]
    }
   ],
   "source": [
    "from joblib import Parallel, delayed\n",
    "import multiprocessing\n",
    "import os\n",
    "import shutil\n",
    "from random import shuffle\n",
    "\n",
    "results_dict = {}\n",
    "\n",
    "for dataset in Parameter_dictionaries:\n",
    "    print \"processing dataset in\", dataset[\"VCF\"]\n",
    "    # 1. Make a new directory next in the VCF path. \n",
    "    Randomisation_dir = \"%s/Randomisations\" % dataset[\"VCF\"].rpartition(\"/\")[0]\n",
    "    if not os.path.exists(Randomisation_dir):\n",
    "        os.makedirs(Randomisation_dir)\n",
    "    \n",
    "    # 2. Copy sex info and VCF into that folder\n",
    "    \n",
    "    shutil.copyfile(dataset[\"Pop_map\"], \"%s/Sex_ID_info.txt\" % Randomisation_dir)  ## sex info\n",
    "    shutil.copyfile(dataset[\"VCF\"], \"%s/batch_1.vcf\" % Randomisation_dir)  ## sex info\n",
    "    \n",
    "    orig_vcf = \"%s/batch_1.vcf\" % Randomisation_dir\n",
    "    \n",
    "    print \"\\nRandomisations happening in %s\" % Randomisation_dir\n",
    "    \n",
    "    ## 3. Make the randomised sex info files\n",
    "\n",
    "    orig_popmap_path = \"%s/Sex_ID_info.txt\" % Randomisation_dir\n",
    "    orig_popmap = open(orig_popmap_path, 'r').readlines()\n",
    "\n",
    "    sexes = []\n",
    "    samples = []\n",
    "    IDs = []\n",
    "    randomisations = []\n",
    "\n",
    "    for line in orig_popmap:\n",
    "        sexes.append(line.strip().split()[1])\n",
    "        samples.append(line.strip().split()[0])\n",
    "        IDs.append(line.strip().split()[2])\n",
    "\n",
    "    popmaps = []\n",
    "    for i in range(100):\n",
    "        popmap_path = \"%s/rand_popmap_%s.txt\" % (Randomisation_dir, i)\n",
    "        popmaps.append(popmap_path)\n",
    "        rand_popmap = open(popmap_path, 'w')\n",
    "        shuffle(sexes)\n",
    "        randomisations.append(sexes)\n",
    "        \n",
    "        for i in range(len(samples)):\n",
    "            rand_popmap.write(\"%s\\t%s\\t%s\\n\" % (samples[i], sexes[i], IDs[i]))\n",
    "\n",
    "        rand_popmap.close()\n",
    "    \n",
    "    print \"\\nRandom sex info files made\"\n",
    "    \n",
    "    ## make new VCFs for parallelised analyses\n",
    "    \n",
    "    for i in popmaps:\n",
    "        new_vcf = \"%s.vcf\" % i.rpartition(\".\")[0]\n",
    "        shutil.copyfile(orig_vcf, new_vcf)\n",
    "        \n",
    "    ## 4. Run the randomisations\n",
    "    \n",
    "    print \"\\nRunning randomisations\\n\"\n",
    "    \n",
    "    results_dict[dataset[\"Name\"]] = Parallel(n_jobs=4, verbose = 1)(delayed(Super_SLM_finder_parallel)(i, dataset) for i in popmaps)\n",
    "    \n",
    "    ## 5. Output the results for each species after the species is complete (i.e. checkpoints)\n",
    "    \n",
    "    outfile = open(\"%s/Randomisations_%s.txt\" % (Randomisation_dir, dataset[\"Name\"]), 'w')\n",
    "    \n",
    "    for Randomisation in results_dict[dataset[\"Name\"]]:\n",
    "\n",
    "        XYfreq = len(Randomisation[\"Detailed\"][\"XY\"][\"freq\"])\n",
    "        XYhet = len(Randomisation[\"Detailed\"][\"XY\"][\"het\"])\n",
    "        Ytags = len(Randomisation[\"Detailed\"][\"XY\"][\"Ytags\"])\n",
    "    \n",
    "        ZWfreq = len(Randomisation[\"Detailed\"][\"ZW\"][\"freq\"])\n",
    "        ZWhet = len(Randomisation[\"Detailed\"][\"ZW\"][\"het\"])\n",
    "        Wtags = len(Randomisation[\"Detailed\"][\"ZW\"][\"Wtags\"])\n",
    "                \n",
    "        line = \"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (dataset[\"Name\"],XYfreq,XYhet,Ytags,ZWfreq,ZWhet,Wtags)\n",
    "    \n",
    "        outfile.write(line)\n",
    "        \n",
    "    outfile.close()\n",
    "    \n",
    "    print \"Results outputted to %s/Randomisations_%s.txt\" % (Randomisation_dir, dataset[\"Name\"])\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## So all went well, randomisations completed in less than 24 hours with 100 randomisations per dataset. \n",
    "\n",
    "### Now to plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "results_dict1 ## contains first half\n",
    "results_dict ## contains 2nd half."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "randomisation_files = []\n",
    "for root, dirs, files in os.walk(\"/home/djeffrie/Data/RADseq/Randomisations\"):\n",
    "    for fil in files:\n",
    "        if fil.startswith(\"Randomisation\"):\n",
    "            randomisation_files.append(\"%s/%s\" % (root, fil))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "randomisation_filepath = \"/home/djeffrie/Data/RADseq/Randomisations/Randomisations_Ryav.txt\"\n",
    "\n",
    "randomisations = open(randomisation_filepath, 'r').readlines()\n",
    "\n",
    "XYfreqs = []\n",
    "XYhets = []\n",
    "Ytags = []\n",
    "\n",
    "ZW_freqs = []\n",
    "ZW_hets = []\n",
    "W_tagss = []\n",
    "\n",
    "XYvsZW_freqs = []\n",
    "XYvsZW_hets = []\n",
    "XYvsZW_tagss = []\n",
    "\n",
    "for line in randomisations:\n",
    "    #print line.split()\n",
    "    species = line.split()[0]\n",
    "    XYfreq = line.split()[1]\n",
    "    XYhet = line.split()[2]\n",
    "    Y_tags = line.split()[3]\n",
    "    \n",
    "    ZW_freq = line.split()[4]\n",
    "    ZW_het = line.split()[5]\n",
    "    W_tags = line.split()[6]\n",
    "    \n",
    "    XYvsZW_freq = int(XYfreq) - int(ZW_freq)\n",
    "    XYvsZW_het = int(XYhet) - int(ZW_het)\n",
    "    XYvsZW_tags = int(Y_tags) - int(W_tags)\n",
    "\n",
    "    XYfreqs.append(int(XYfreq))\n",
    "    XYhets.append(int(XYhet))\n",
    "    Ytags.append(int(Y_tags))\n",
    "    ZW_freqs.append(int(ZW_freq))\n",
    "    ZW_hets.append(int(ZW_het))\n",
    "    W_tagss.append(int(W_tags))\n",
    "    XYvsZW_freqs.append(int(XYvsZW_freq))\n",
    "    XYvsZW_hets.append(int(XYvsZW_het))\n",
    "    XYvsZW_tagss.append(int(XYvsZW_tags))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function vlines in module matplotlib.pyplot:\n",
      "\n",
      "vlines(x, ymin, ymax, colors=u'k', linestyles=u'solid', label=u'', hold=None, **kwargs)\n",
      "    Plot vertical lines.\n",
      "    \n",
      "    Plot vertical lines at each `x` from `ymin` to `ymax`.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    x : scalar or 1D array_like\n",
      "        x-indexes where to plot the lines.\n",
      "    \n",
      "    ymin, ymax : scalar or 1D array_like\n",
      "        Respective beginning and end of each line. If scalars are\n",
      "        provided, all lines will have same length.\n",
      "    \n",
      "    colors : array_like of colors, optional, default: 'k'\n",
      "    \n",
      "    linestyles : ['solid' | 'dashed' | 'dashdot' | 'dotted'], optional\n",
      "    \n",
      "    label : string, optional, default: ''\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    lines : `~matplotlib.collections.LineCollection`\n",
      "    \n",
      "    Other parameters\n",
      "    ----------------\n",
      "    kwargs : `~matplotlib.collections.LineCollection` properties.\n",
      "    \n",
      "    See also\n",
      "    --------\n",
      "    hlines : horizontal lines\n",
      "    \n",
      "    Examples\n",
      "    ---------\n",
      "    .. plot:: mpl_examples/pylab_examples/vline_hline_demo.py\n",
      "    \n",
      "    \n",
      "    \n",
      "    Additional kwargs: hold = [True|False] overrides default hold state\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(plt.vlines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "num must be 0 <= num <= 1, not 3",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-207-3b91c87b7363>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[0mfig\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfigsize\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m(\u001b[0m\u001b[1;36m20\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m10\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mfig\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd_subplot\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[0mcounts\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbars\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mhist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mXYfreqs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbins\u001b[0m\u001b[1;33m=\u001b[0m \u001b[1;36m20\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0medgecolor\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"lightblue\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcolor\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"royalblue\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/djeffrie/anaconda/lib/python2.7/site-packages/matplotlib/figure.pyc\u001b[0m in \u001b[0;36madd_subplot\u001b[1;34m(self, *args, **kwargs)\u001b[0m\n\u001b[0;32m    962\u001b[0m                     \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_axstack\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0max\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    963\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 964\u001b[1;33m             \u001b[0ma\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msubplot_class_factory\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mprojection_class\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    965\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    966\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_axstack\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0ma\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/home/djeffrie/anaconda/lib/python2.7/site-packages/matplotlib/axes/_subplots.pyc\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, fig, *args, **kwargs)\u001b[0m\n\u001b[0;32m     62\u001b[0m                     raise ValueError(\n\u001b[0;32m     63\u001b[0m                         \"num must be 0 <= num <= {maxn}, not {num}\".format(\n\u001b[1;32m---> 64\u001b[1;33m                             maxn=rows*cols, num=num))\n\u001b[0m\u001b[0;32m     65\u001b[0m                 \u001b[1;32mif\u001b[0m \u001b[0mnum\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     66\u001b[0m                     warnings.warn(\"The use of 0 (which ends up being the \"\n",
      "\u001b[1;31mValueError\u001b[0m: num must be 0 <= num <= 1, not 3"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fd52fa10890>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "\n",
    "fig = plt.figure(figsize = (20,10))\n",
    "\n",
    "fig.add_subplot(1,1,3)\n",
    "\n",
    "counts, bins, bars = plt.hist(XYfreqs, bins= 20, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "plt.xlim((0,200))\n",
    "plt.vlines(100, 0, max(counts)*0.75, color = \"red\")\n",
    "plt.show()\n",
    "\n",
    "fig.add_subplot(2,1,3)\n",
    "counts, bins, bars = plt.hist(XYhets, bins= 20, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "plt.xlim((0,200))\n",
    "plt.vlines(100, 0, max(counts)*0.75, color = \"red\")\n",
    "plt.show()\n",
    "\n",
    "fig.add_subplot(3,1,3)\n",
    "counts, bins, bars = plt.hist(Ytags, bins= 20, edgecolor = \"lightblue\", color = \"royalblue\")\n",
    "plt.xlim((0,200))\n",
    "plt.vlines(100, 0, max(counts)*0.75, color = \"red\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[  8.   7.  18.  14.  20.  12.   5.   4.   4.   3.   1.   1.   0.   0.   0.\n",
      "   0.   1.   1.   0.   1.]\n"
     ]
    }
   ],
   "source": [
    "print counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
