{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 607,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import vcf\n",
    "from collections import Counter\n",
    "import pprint as pp\n",
    "from pydoc import help\n",
    "\n",
    "\n",
    "\n",
    "def VCF_MSTmap_converter(vcfpath, sex_info, sample_threshold):\n",
    "\n",
    "    \"\"\"\n",
    "    USAGE:  VCF_to_MSTmap.py  <vcfpath>  <sex_info file>  <sample_presence_threshold>\n",
    "         \n",
    "        From VCF format, filter for linkage informative loci, duplicate each locus\n",
    "        in the complimentary phase and output to MSTmap input format.\n",
    "        \n",
    "        Two MSTmap input files are made, one containing loci that are heterozygous in females and homozygous in males\n",
    "        and a second containing loci heterozygous in males and homozygous in females. \n",
    "        \n",
    "        In both files, the (uninformative) allele from the homozygous parent is removed from all genotypes to conform to the double haploid ('DH') data format in MSTmap.\n",
    "        eg. \n",
    "        \n",
    "        if Maternal genotpe = AB and Paternal genotype = B\n",
    "            if offspring genotype == AB\n",
    "                new offspring genpytpe = A\n",
    "                \n",
    "            else if offspring genotype == A (dipliod homozygous A)\n",
    "                new offspring genpytpe = A (haploid homozygous A) \n",
    "                \n",
    "        # In the latter example, the codes for diploid homozygous (AA) and haploid (A) are the same, but using population_type = DH\n",
    "        # in MSTmap will ensure that the genoytpe is correctly interpreted as haploid.\n",
    "           \n",
    "        \n",
    "        Lastly, each locus is replicated with the complimentary phase in the MSTmap input files produced. This allows for the construction of\n",
    "        a linkage map without a priori knowledge of marker phase. Marker names for complimentaty phase loci are changed\n",
    "        to \"Compli_Loc_id\" and genotype states are lower case instead of upper case.\n",
    "        \n",
    "        MSTmap input files are written to the same path as the VCF file called \"MSTmap_input_male_het.txt\" and \"MSTmap_input_female_het.txt\"\n",
    "        \n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    \n",
    "    ## open vcf and alter it so vcf can parse it properly (annoying) =============\n",
    "    \n",
    "    alteredvcfpath = \"%s%s\" %(vcfpath, \".altered\")\n",
    "    \n",
    "    myvcf = open(vcfpath, 'r').readlines()\n",
    "    alteredvcf = open(alteredvcfpath, 'w')\n",
    "    \n",
    "    for line in myvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "    \n",
    "    \n",
    "    ## codings for MSTmap file ===================================================\n",
    "    \n",
    "    codes = {\"0/0\":\"A\", \"1/1\":\"B\", \"0/1\":\"AB\", \"1/0\":\"AB\", None:\"U\" }\n",
    "    Complicodes = {\"A\":\"B\", \"B\":\"A\", \"AB\":\"AB\", \"U\":\"U\" }\n",
    "    \n",
    "    \n",
    "    ## Get the sex info for samples =============================================\n",
    "    \n",
    "    sex_file = open(sex_info, 'r').readlines()\n",
    "    offspring = []\n",
    "    sample_counter = 0\n",
    "    \n",
    "    for line in sex_file: ## Note, only samples in the sex_info file are used\n",
    "        \n",
    "        sample_counter += 1\n",
    "        sample_name = line.split()[0]\n",
    "        sex = line.split()[1]\n",
    "        \n",
    "            \n",
    "        if sex == \"F\" or sex == \"f\":\n",
    "            Mother = sample_name\n",
    "            print \"Mother =\", sample_name\n",
    "            \n",
    "        elif sex == \"M\" or sex == \"m\":\n",
    "            Father = sample_name\n",
    "            print \"Father =\", sample_name\n",
    "        \n",
    "        elif sex == \"O\" or sex == \"o\":\n",
    "            offspring.append(sample_name)\n",
    "        \n",
    "    offspring = sorted(offspring)\n",
    "    \n",
    "    \n",
    "    print \"Number of offspring =\", len(offspring)\n",
    "    \n",
    "            \n",
    "    \n",
    "    ## Get and convert sample data ===============================================\n",
    "    \n",
    "    Loc_Ids = []\n",
    "    Mat_Pat_het_discards = 0\n",
    "    Mat_pat_hom_discards = 0\n",
    "    Mat_pat_missing_discards = 0\n",
    "    missing_data_discards = 0\n",
    "    total_loci = 0\n",
    "    \n",
    "\n",
    "    \n",
    "    \n",
    "    ## Lists (of info for MSTmap file lines) for the loci that pass filters\n",
    "       \n",
    "    male_het_loci_kept = []\n",
    "    female_het_loci_kept = []\n",
    "    \n",
    "    male_het_loci_kept_compli = []\n",
    "    female_het_loci_kept_compli = []\n",
    "    \n",
    "    Male_ids = []\n",
    "    Female_ids = []\n",
    "            \n",
    "    \n",
    "    myvcf = vcf.Reader(open(alteredvcfpath, 'r')) ## read in vcf\n",
    "\n",
    "    info = \"no\"\n",
    "    \n",
    "    for record in myvcf:\n",
    "        \n",
    "        total_loci += 1\n",
    "        \n",
    "        if record.num_called < (sample_counter)*sample_threshold: ## discard if too much missing data\n",
    "            missing_data_discards += 1\n",
    "        \n",
    "        elif record.num_called >= (sample_counter)*sample_threshold: ## if locus has data for more than 0.75 samples\n",
    "\n",
    "            Loc_Id = \"%s_%s\" % (record.ID, record.POS)\n",
    "            \n",
    "            \n",
    "            offspring_genotypes = {}\n",
    "            \n",
    "            ## get genotypes for each sample at this locus and convert to MSTmap format\n",
    "            \n",
    "            \n",
    "            \n",
    "            for sample in record.samples: \n",
    "                \n",
    "                name = sample.sample\n",
    "                genotype = sample['GT']\n",
    "                \n",
    "                \n",
    "                ## Get the MST codes and complimentary codes for the informative loci\n",
    "                if name == Mother:\n",
    "                    Maternal_genotype = codes[genotype]\n",
    "                    \n",
    "                elif name == Father:\n",
    "                    Paternal_genotype = codes[genotype]\n",
    "                    \n",
    "                elif name in offspring:\n",
    "                    offspring_genotypes[name] = codes[genotype]\n",
    "            \n",
    "            \n",
    "                \n",
    "            \n",
    "            ### Filter for informative loci, remove uninformative allele and produce complimentary phase loci\n",
    "                \n",
    "            \n",
    "            femmap_offspring_haploid_genotypes = []\n",
    "            femmap_offspring_haploid_genotypes_compli = []\n",
    "            \n",
    "            malemap_offspring_haploid_genotypes = []\n",
    "            malemap_offspring_haploid_genotypes_compli = []\n",
    "            \n",
    "            ## Female mapping data    \n",
    "            if Maternal_genotype == \"AB\" and Paternal_genotype == \"A\" or Paternal_genotype == \"B\":\n",
    "\n",
    "                female_het_loci_kept.append(\"\\n%s\" % (Loc_Id)) ## make line for the record (not used unless criteria below are met)\n",
    "                female_het_loci_kept_compli.append(\"\\ncompli_%s\" % (Loc_Id)) ## make line for the complimentary record (not used unless criteria below are met)\n",
    "                \n",
    "                for sample in offspring: ## in the order of the samples in the ordered offspring name list\n",
    "                    \n",
    "                    if offspring_genotypes[sample] == \"A\":\n",
    "                        femmap_offspring_haploid_genotypes.append(\"\\tA\")\n",
    "                        femmap_offspring_haploid_genotypes_compli.append(\"\\tb\")\n",
    "                    elif offspring_genotypes[sample] == \"B\":\n",
    "                        femmap_offspring_haploid_genotypes.append(\"\\tB\")\n",
    "                        femmap_offspring_haploid_genotypes_compli.append(\"\\ta\")\n",
    "                    elif offspring_genotypes[sample] == \"AB\" and Paternal_genotype == \"A\":\n",
    "                        femmap_offspring_haploid_genotypes.append(\"\\tB\")\n",
    "                        femmap_offspring_haploid_genotypes_compli.append(\"\\ta\")\n",
    "                    elif offspring_genotypes[sample] == \"AB\" and Paternal_genotype == \"B\":\n",
    "                        femmap_offspring_haploid_genotypes.append(\"\\tA\")\n",
    "                        femmap_offspring_haploid_genotypes_compli.append(\"\\tb\")\n",
    "                    elif offspring_genotypes[sample] == \"U\":\n",
    "                        femmap_offspring_haploid_genotypes.append(\"\\tU\")\n",
    "                        femmap_offspring_haploid_genotypes_compli.append(\"\\tU\")\n",
    "\n",
    "                            \n",
    "                female_het_loci_kept.append(\"\".join(femmap_offspring_haploid_genotypes))\n",
    "                female_het_loci_kept_compli.append(\"\".join(femmap_offspring_haploid_genotypes_compli)) ## complimentary phase ???? \n",
    "                \n",
    "                \n",
    "            \n",
    "            ## Male mapping data            \n",
    "            elif Paternal_genotype == \"AB\" and Maternal_genotype == \"A\" or Maternal_genotype == \"B\":\n",
    "                     \n",
    "                male_het_loci_kept.append(\"\\n%s\" % (Loc_Id)) ## make line for the record (not used unless criteria below are met)\n",
    "                male_het_loci_kept_compli.append(\"\\ncompli_%s\" % (Loc_Id)) ## make line for the complimentary record (not used unless criteria below are met)\n",
    "            \n",
    "                for sample in offspring:\n",
    "                    \n",
    "                    if offspring_genotypes[sample] == \"A\":\n",
    "                        malemap_offspring_haploid_genotypes.append(\"\\tA\")\n",
    "                        malemap_offspring_haploid_genotypes_compli.append(\"\\tb\")\n",
    "                    elif offspring_genotypes[sample] == \"B\":\n",
    "                        malemap_offspring_haploid_genotypes.append(\"\\tB\")\n",
    "                        malemap_offspring_haploid_genotypes_compli.append(\"\\ta\")\n",
    "                    elif offspring_genotypes[sample] == \"AB\" and Maternal_genotype == \"A\":\n",
    "                        malemap_offspring_haploid_genotypes.append(\"\\tB\")\n",
    "                        malemap_offspring_haploid_genotypes_compli.append(\"\\ta\")\n",
    "                    elif offspring_genotypes[sample] == \"AB\" and Maternal_genotype == \"B\":\n",
    "                        malemap_offspring_haploid_genotypes.append(\"\\tA\")\n",
    "                        malemap_offspring_haploid_genotypes_compli.append(\"\\tb\")\n",
    "                    elif offspring_genotypes[sample] == \"U\":\n",
    "                        malemap_offspring_haploid_genotypes.append(\"\\tU\")\n",
    "                        malemap_offspring_haploid_genotypes_compli.append(\"\\tU\")                            \n",
    "\n",
    "                male_het_loci_kept.append(\"\".join(malemap_offspring_haploid_genotypes))\n",
    "                male_het_loci_kept_compli.append(\"\".join(malemap_offspring_haploid_genotypes_compli))\n",
    "                \n",
    "\n",
    "            ## Discard loci that are heterozygous in both\n",
    "            elif Paternal_genotype == \"AB\" and Maternal_genotype == \"AB\":\n",
    "                \n",
    "                Mat_Pat_het_discards += 1\n",
    "                   \n",
    "            elif Paternal_genotype == \"U\" or Maternal_genotype == \"U\":\n",
    "                \n",
    "                Mat_pat_missing_discards += 1\n",
    "                \n",
    "            elif Paternal_genotype == \"A\" or Paternal_genotype == \"B\" and Maternal_genotype == \"A\" or Maternal_genotype == \"B\":\n",
    "                \n",
    "                Mat_pat_hom_discards += 1\n",
    "                \n",
    "               \n",
    "                                    ## WRITING FILE ##\n",
    "    ## Set up the new file ==============================================================\n",
    "    \n",
    "    fem_het_MST_file = open(\"%s/MSTmap_input_female_het.txt\" % (vcfpath.rpartition(\"/\")[0]), 'w') ## in same path as VCF\n",
    "    male_het_MST_file = open(\"%s/MSTmap_input_male_het.txt\" % (vcfpath.rpartition(\"/\")[0]), 'w') ## in same path as VCF\n",
    "                                          \n",
    "    ### Headers ==========================================================================\n",
    "    \n",
    "    ## Fem_het\n",
    "    file_headers = \"\\\n",
    "population_type DH\\n\\\n",
    "population_name Female_map\\n\\\n",
    "distance_function kosambi\\n\\\n",
    "cut_off_p_value 0.000005\\n\\\n",
    "no_map_dist 30\\n\\\n",
    "no_map_size 1\\n\\\n",
    "missing_threshold 0.25\\n\\\n",
    "estimation_before_clustering no\\n\\\n",
    "detect_bad_data yes\\n\\\n",
    "objective_function ML\\n\\\n",
    "number_of_loci %s\\n\\\n",
    "number_of_individual %s\\n\\n\" % (len(female_het_loci_kept), len(offspring))\n",
    "            \n",
    "    \n",
    "    fem_het_MST_file.write(file_headers)\n",
    "        \n",
    "    sample_headers = \"locus_name\\t%s\" % (\"\\t\".join(offspring))\n",
    "    fem_het_MST_file.write(sample_headers)\n",
    "                                          \n",
    "        \n",
    "    for line in female_het_loci_kept:\n",
    "        fem_het_MST_file.write(line) \n",
    "    for line in female_het_loci_kept_compli:\n",
    "        fem_het_MST_file.write(line)\n",
    "        \n",
    "    fem_het_MST_file.close()\n",
    "\n",
    "                                          \n",
    "    ## Male_het\n",
    "    file_headers = \"\\\n",
    "population_type DH\\n\\\n",
    "population_name Male_map\\n\\\n",
    "distance_function kosambi\\n\\\n",
    "cut_off_p_value 0.000005\\n\\\n",
    "no_map_dist 30\\n\\\n",
    "no_map_size 1\\n\\\n",
    "missing_threshold 0.25\\n\\\n",
    "estimation_before_clustering no\\n\\\n",
    "detect_bad_data yes\\n\\\n",
    "objective_function ML\\n\\\n",
    "number_of_loci %s\\n\\\n",
    "number_of_individual %s\\n\\n\" % (len(male_het_loci_kept), len(offspring))\n",
    "            \n",
    "    \n",
    "    male_het_MST_file.write(file_headers)\n",
    "        \n",
    "    sample_headers = \"locus_name\\t%s\" % (\"\\t\".join(offspring))\n",
    "    male_het_MST_file.write(sample_headers)\n",
    "                                          \n",
    "    \n",
    "    for line in male_het_loci_kept:\n",
    "        male_het_MST_file.write(line)\n",
    "    for line in male_het_loci_kept_compli:\n",
    "        male_het_MST_file.write(line)\n",
    "        \n",
    "    male_het_MST_file.close()\n",
    "                                          \n",
    "                                          \n",
    "    print \"Total number of loci in VCF =\", total_loci\n",
    "    print \"Number of loci thrown out due to low sample presence =\", missing_data_discards\n",
    "    print \"Number of loci thrown out becuase heterozygous in mother and father =\", Mat_Pat_het_discards\n",
    "    print \"Number of loci thrown out because homozygous in mother and father =\", Mat_pat_hom_discards\n",
    "    print \"Number of loci thrown out because missing in mother or father =\", Mat_pat_missing_discards\n",
    "    print \"loci in MSTmap male map (excl. complimentary phase) =\", len(male_het_loci_kept)/2\n",
    "    print \"loci in MSTmap female map (excl. complimentary phase) =\", len(female_het_loci_kept)/2\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "if len(sys.argv) == 1:\n",
    "    # sys.exit(help(VCF_MSTmap_converter))\n",
    "    print VCF_MSTmap_converter.__doc__\n",
    "elif len(sys.argv) < 4: ## If not enough args are supplied print error message\n",
    "    sys.exit(\"\\n##Error, not enough arguments, run script with no arguments to see help message\\n\")\n",
    "\n",
    "elif len(sys.argv) == 4:\n",
    "    specified_vcfpath = sys.argv[1]\n",
    "    sex_info_file = sys.argv[2]\n",
    "    sample_presence_threshold = float(sys.argv[3])\n",
    "\n",
    "    VCF_MSTmap_converter(specified_vcfpath, sex_info_file, sample_presence_threshold)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### For testing =======================================================\n",
    "\n",
    "specified_vcfpath = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/populations_better/batch_1_test.vcf\"\n",
    "sex_info_file = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/pop_codes_linkage.txt\"\n",
    "sample_presence_threshold = 0.75\n",
    "\n",
    "VCF_MSTmap_converter(specified_vcfpath, sex_info_file, sample_presence_threshold)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
