{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import shutil\n",
    "from __future__ import division\n",
    "from Bio.Application import _Option, AbstractCommandline, _Switch\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class PopulationsCommandLine(AbstractCommandline):\n",
    "    def __init__(self, cmd=\"populations\", **kwargs):\n",
    "        \"\"\"\n",
    "        # Any populations option and switch\n",
    "        >>> cline_args = {'b': 'fastq',\n",
    "        ...               'P': '../../dan_data/OBY01_1.fil.fq_1',\n",
    "        ...               'M': '../../dan_data/OBY01_1.fi/M_tests/M_1',\n",
    "        ...               's': 1,\n",
    "        ...               'B': 6,\n",
    "        ...               'W': 1,\n",
    "        ...               'e': 3,\n",
    "        ...               't': True,\n",
    "        ...               'v': True}\n",
    "        \n",
    "        # This creates a command line object\n",
    "        >>> populations_cline = PopulationsCommandLine(**cline_args)\n",
    "        \n",
    "        # This prints a string representation\n",
    "        >>> print populations_cline \n",
    "        populations -t fastq -f ../../dan_data/OBY01_1.fil.fq_1 -o ../../dan_data/OBY01_1.fi/M_tests/M_1 -i 1 -m 6 -M 1 -p 3 -r -d\n",
    "        \n",
    "        # This runs populations\n",
    "        >>> stdout, stderr = populations_cline()\n",
    "        \n",
    "        \"\"\"\n",
    "        \n",
    "        self.parameters = [\n",
    "            _Option([\"-b\", \"b\"],\n",
    "                    \"Batch ID to examine when exporting from the catalog\",\n",
    "                    is_required=True,\n",
    "                    equate=False),\n",
    "            _Option([\"-P\", \"P\"],\n",
    "                    \"path to the Stacks output files\",\n",
    "                    is_required=True,\n",
    "                    filename=True,\n",
    "                    equate=False),\n",
    "            _Option([\"-M\", \"M\"],\n",
    "                    \"path to the population map, a tab separated file describing which individuals belong in which population.\",\n",
    "                    is_required=True,\n",
    "                    filename=True,\n",
    "                    equate=False),\n",
    "            _Option([\"-s\", \"s\"],\n",
    "                    \"output a file to import results into an SQL database.\",\n",
    "                    is_required=False,\n",
    "                    equate=False),\n",
    "            _Option([\"-B\", \"B\"],\n",
    "                    \"specify a file containing Blacklisted markers to be excluded from the export.\",\n",
    "                    is_required=False,\n",
    "                    filename=True,\n",
    "                    equate=False),\n",
    "            _Option([\"-W\", \"W\"],\n",
    "                    \"specify a file containing Whitelisted markers to include in the export.\",\n",
    "                    is_required=False,\n",
    "                    filename=True,\n",
    "                    equate=False),\n",
    "            _Option([\"-e\", \"e\"],\n",
    "                    \"restriction enzyme, required if generating 'genomic' output.\",\n",
    "                    is_required=False,\n",
    "                    equate=False),\n",
    "            _Option([\"-t\", \"t\"],\n",
    "                    \"number of threads to run in parallel sections of code.\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,int),\n",
    "                    equate=False),  \n",
    "            _Switch([\"-v\", \"v\"],\n",
    "                    \"print program version.\"),\n",
    "            _Switch([\"-h\", \"h\"],\n",
    "                    \"display this help messsage\"),\n",
    "\n",
    "            \n",
    "            # Merging and Phasing:\n",
    "            _Switch([\"--merge_sites\", \"-merge_sites\", \"merge_sites\"],\n",
    "                    \"merge loci that were produced from the same restriction enzyme cutsite (requires reference-aligned data).\"), \n",
    "            _Switch([\"--merge_prune_lim\", \"-merge_prune_lim\", \"merge_prune_lim\"],\n",
    "                    \"when merging adjacent loci, if at least X% samples posses both loci prune the remaining samples out of the analysis.\"),\n",
    "            \n",
    "            # Data Filtering:\n",
    "            _Option([\"-r\", \"r\"],\n",
    "                    \"minimum percentage of individuals in a population required to process a locus for that population.\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,float),\n",
    "                    equate=False),\n",
    "            _Option([\"-p\", \"p\"],\n",
    "                    \"minimum number of populations a locus must be present in to process a locus.\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,int),\n",
    "                    equate=False),\n",
    "            _Option([\"-m\", \"m\"],\n",
    "                    \"specify a minimum stack depth required for individuals at a locus.\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,int),\n",
    "                    equate=False),\n",
    "            _Option([\"-f\", \"f\"],\n",
    "                    \"specify a correction to be applied to Fst values: 'p_value', 'bonferroni_win', or 'bonferroni_gen'.\",\n",
    "                    is_required=False,\n",
    "                    equate=False),\n",
    "            _Option([\"--min_maf\", \"-min_maf\", \"min_maf\"],\n",
    "                    \"specify a minimum minor allele frequency required to process a nucleotide site at a locus (0 < min_maf < 0.5).\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,float),\n",
    "                    equate=False),\n",
    "            _Option([\"--max_obs_het\", \"-max_obs_het\", \"max_obs_het\"],\n",
    "                    \"specify a maximum observed heterozygosity required to process a nucleotide site at a locus\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,float),\n",
    "                    equate=False),\n",
    "            _Option([\"--p_value_cutoff\", \"-p_value_cutoff\", \"p_value_cutoff\"],\n",
    "                    \"required p-value to keep an Fst measurement (0.05 by default). Also used as base for Bonferroni correction\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,float),\n",
    "                    equate=False),\n",
    "            _Option([\"--lnl_lim\", \"-lnl_lim\", \"lnl_lim\"],\n",
    "                    \"filter loci with log likelihood values below this threshold\",\n",
    "                    is_required=False,\n",
    "                    equate=False),\n",
    "            _Switch([\"--write_single_snp\", \"-write_single_snp\", \"write_single_snp\"],\n",
    "                    \"restrict data analysis to only the first SNP per locus\"),\n",
    "            _Switch([\"--write_random_snp\", \"-write_random_snp\", \"write_random_snp\"],\n",
    "                    \"restrict data analysis to one random SNP per locus\"),\n",
    "            \n",
    "            \n",
    "            ## Add missing options here\n",
    "            \n",
    "                    \n",
    "            ## Fstats\n",
    "            _Switch([\"--fstats\", \"-fstats\", \"fstats\"],\n",
    "                    \"enable SNP and haplotype-based F statistics.\"), \n",
    "            \n",
    "                    \n",
    "            # Bootstrap Resampling: \n",
    "            _Switch([\"--bootstrap\", \"-bootstrap\", \"bootstrap\"],\n",
    "                    \"turn on boostrap resampling for all smoothed statistics.\"),\n",
    "            _Switch([\"--bootstrap_pifis\", \"-bootstrap_pifis\", \"bootstrap_pifis\"],\n",
    "                    \"turn on boostrap resampling for smoothed SNP-based Pi and Fis calculations..\"),\n",
    "            _Switch([\"--bootstrap_fst\", \"-bootstrap_fst\", \"bootstrap_fst\"],\n",
    "                    \"turn on boostrap resampling for smoothed Fst calculations based on pairwise population comparison of SNPs..\"),\n",
    "            _Switch([\"--bootstrap_div\", \"-bootstrap_div\", \"bootstrap_div\"],\n",
    "                    \"turn on boostrap resampling for smoothed haplotype diveristy and gene diversity calculations based on haplotypes.\"),\n",
    "            _Switch([\"--bootstrap_phist\", \"-bootstrap_phist\", \"bootstrap_phist\"],\n",
    "                    \"turn on boostrap resampling for smoothed Phi_st calculations based on haplotypes.\"),\n",
    "            _Option([\"--bootstrap_reps\", \"-bootstrap_reps\",\"bootstrap_reps\"],\n",
    "                    \"number of bootstrap resamplings to calculate (default 100)\",\n",
    "                    is_required=False,\n",
    "                    checker_function=lambda value : isinstance(value,int),\n",
    "                    equate=False),\n",
    "            _Option([\"--bootstrap_wl\", \"-bootstrap_wl\",\"bootstrap_wl\"],\n",
    "                    \"only bootstrap loci contained in this whitelist.\",\n",
    "                    is_required=False,\n",
    "                    filename=True,\n",
    "                    equate=False),\n",
    "            \n",
    "            # File ouput options:\n",
    "            _Switch([\"--ordered_export\", \"-ordered_export\", \"ordered_export\"],\n",
    "                    \"if data is reference aligned, exports will be ordered; only a single representative of each overlapping site..\"),\n",
    "            _Switch([\"--genomic\", \"-genomic\", \"genomic\"],\n",
    "                    \"output each nucleotide position (fixed or polymorphic) in all population members to a file.\"),\n",
    "            _Switch([\"--fasta\", \"-fasta\", \"fasta\"],\n",
    "                    \"output full sequence for each unique haplotype, from each sample locus in FASTA format, regardless of plausibility.\"),\n",
    "            _Switch([\"--fasta_strict\", \"-fasta_strict\", \"fasta_strict\"],\n",
    "                    \"output full sequence for each haplotype, from each sample locus in FASTA format, only for biologically plausible loci.\"),\n",
    "            _Switch([\"--vcf\", \"-vcf\", \"vcf\"],\n",
    "                    \"output SNPs in Variant Call Format (VCF).\"),\n",
    "            _Switch([\"--vcf_haplotypes\", \"-vcf_haplotypes\", \"vcf_haplotypes\"],\n",
    "                    \"output haplotypes in Variant Call Format (VCF).\"),\n",
    "            _Switch([\"--genepop\", \"-genepop\", \"genepop\"],\n",
    "                    \"output results in GenePop format.\"),\n",
    "            _Switch([\"--structure\", \"-structure\", \"structure\"],\n",
    "                    \"output results in Structure format.\"),\n",
    "            _Switch([\"--phase\", \"-phase\", \"phase\"],\n",
    "                    \"output genotypes in PHASE format.\"),\n",
    "            _Switch([\"--fastphase\", \"-fastphase\", \"fastphase\"],\n",
    "                    \"output genotypes in fastPHASE format.\"),\n",
    "            _Switch([\"--beagle\", \"-beagle\", \"beagle\"],\n",
    "                    \"output genotypes in Beagle format.\"),\n",
    "            _Switch([\"--beagle_phased\", \"-beagle_phased\", \"beagle_phased\"],\n",
    "                    \"output haplotypes in Beagle format.\"),\n",
    "            _Switch([\"--plink\", \"-plink\", \"plink\"],\n",
    "                    \"output genotypes in PLINK format.\"),\n",
    "            _Switch([\"--hzar\", \"-hzar\", \"hzar\"],\n",
    "                    \"output genotypes in Hybrid Zone Analysis using R (HZAR) format.\"),\n",
    "            _Switch([\"--phylip\", \"-phylip\", \"phylip\"],\n",
    "                    \"output nucleotides that are fixed-within, and variant among populations in Phylip format for phylogenetic tree construction.\"),\n",
    "            _Switch([\"--phylip_var\", \"-phylip_var\", \"phylip_var\"],\n",
    "                    \"include variable sites in the phylip output encoded using IUPAC notation.\"),\n",
    "            _Switch([\"--phylip_var_all\", \"-phylip_var_all\", \"phylip_var_all\"],\n",
    "                    \"include all sequence as well as variable sites in the phylip output encoded using IUPAC notation.\"),\n",
    "            _Switch([\"--treemix\", \"-treemix\", \"treemix\"],\n",
    "                    \"output SNPs in a format useable for the TreeMix program (Pickrell and Pritchard).\"),\n",
    "            \n",
    "            # Debugging:\n",
    "            _Switch([\"--verbose\", \"-verbose\", \"verbose\"],\n",
    "                    \"turn on additional logging.\"),\n",
    "            _Switch([\"--log_fst_comp\", \"-log_fst_comp\", \"log_fst_comp\"],\n",
    "                    \"log components of Fst/Phi_st calculations to a file.\")]\n",
    "        AbstractCommandline.__init__(self, cmd, **kwargs) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def make_and_run_populations_cline(Setup_dict, param, value):\n",
    "    \n",
    "    \"\"\"\n",
    "    Makes and runs the command line using the PopulationsCommandLine class\n",
    "    \"\"\"\n",
    "    batch_params = {}\n",
    "    \n",
    "    populations_params = ['b','P','M','s','B','W','e','t','v','h','merge_sites','merge_prune_lim','r','p','m','f','min_maf','max_obs_het','p_value_cutoff','lnl_lim',\n",
    "              'write_single_snp','write_random_snp','fstats','k','window_size','bootstrap','bootstrap_pifis','bootstrap_fst','bootstrap_div','bootstrap_phist',\n",
    "              'bootstrap_reps','bootstrap_wl','ordered_export','genomic','fasta','fasta_strict','vcf','vcf_haplotypes','enepop','structure','phase',\n",
    "              'fastphase','beagle','beagle_phased','plink','hzar','phylip','phylip_var','phylip_var_all','treemix','verbose','log_fst_comp','-b','-P','-M','-s','-B','-W','-e','-t','-v','-h','--merge_sites','--merge_prune_lim','-r','-p','-m','-f','--min_maf','--max_obs_het','--p_value_cutoff','--lnl_lim',\n",
    "              '--write_single_snp','--write_random_snp','--fstats','-k','--window_size','--bootstrap','--bootstrap_pifis','--bootstrap_fst','--bootstrap_div','--bootstrap_phist',\n",
    "              '--bootstrap_reps','--bootstrap_wl','--ordered_export','--genomic','--fasta','--fasta_strict','--vcf','--vcf_haplotypes','--genepop','--structure','--phase',\n",
    "              '--fastphase','--beagle','--beagle_phased','--plink','--hzar','--phylip','--phylip_var','--phylip_var_all','--treemix','--verbose','--log_fst_comp']\n",
    "    \n",
    "    \n",
    "     \n",
    "    args = {}\n",
    "    \n",
    "    ## Pass the values of the Test parameters\n",
    "    \n",
    "    if param == \"p\":\n",
    "    \n",
    "        args[\"-p\"] = value\n",
    "        args[\"-r\"] = Setup_dict[\"r_default\"]\n",
    "        args[\"-m\"] = Setup_dict[\"m_default\"]\n",
    "    \n",
    "    elif param == \"r\":\n",
    "    \n",
    "        args[\"-p\"] = Setup_dict[\"p_default\"]\n",
    "        args[\"-r\"] = value\n",
    "        args[\"-m\"] = Setup_dict[\"m_default\"]\n",
    "    \n",
    "    elif param == \"m\":\n",
    "    \n",
    "        args[\"-p\"] = Setup_dict[\"p_default\"]\n",
    "        args[\"-r\"] = Setup_dict[\"r_default\"]\n",
    "        args[\"-m\"] = value\n",
    "            \n",
    "    \n",
    "    for parameter in Setup_dict:\n",
    "        if parameter in populations_params:\n",
    "            args[parameter] = Setup_dict[parameter]\n",
    "     \n",
    "    args[\"--vcf\"] = True ## overrides the input dictionary as the vcf is a necassary output for getting plotting data\n",
    "    \n",
    "    ## Make command line ---------------------------------------------------------------------------\n",
    "    \n",
    "\n",
    "    cline = PopulationsCommandLine(**args)\n",
    "        \n",
    "    if Setup_dict['run_cline_switch'] == 1:\n",
    "        stderr, stdout = cline()        \n",
    "\n",
    "        \n",
    "    return cline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def sample_Avcov_from_vcf(vcf_path):\n",
    "    \n",
    "    \"\"\"\n",
    "    Finds the average coverage of each sample in a vcf file.\n",
    "    Just requires the vcf path as input. Not, it does not incorporate information\n",
    "    about missing loci, the average is only calculated from those loci called is each sample. \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    import vcf\n",
    "    import numpy as np\n",
    "    \n",
    "    sample_cov = {}\n",
    "    \n",
    "    for record in vcf.Reader(open(vcf_path, 'r')):\n",
    "        samples_called = 0\n",
    "        for sample in record.samples:\n",
    "            if sample.called  == True:\n",
    "                samples_called += 1\n",
    "                \n",
    "                if sample.sample not in sample_cov.keys():\n",
    "                    sample_cov[sample.sample] = []\n",
    "                else:\n",
    "                    sample_cov[sample.sample].append(sample['DP'])\n",
    "    \n",
    "    sample_avgs = {}\n",
    "\n",
    "    for sample in sample_cov.keys():\n",
    "        sample_avgs[sample] = np.round(np.mean(sample_cov[sample]), 2)\n",
    "    \n",
    "    return sample_avgs\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def Missing_data_finder(vcf_path):\n",
    "    \n",
    "    import vcf\n",
    "    import numpy as np\n",
    "    \n",
    "    Missing_data_dict = {}\n",
    "    locus_counter = 0\n",
    "    \n",
    "    for record in vcf.Reader(open(vcf_path, 'r')):\n",
    "        locus_counter += 1\n",
    "        for sample in record.samples:\n",
    "            if sample.sample not in Missing_data_dict:\n",
    "                Missing_data_dict[sample.sample] = 0\n",
    "            if sample.called  == False:\n",
    "                Missing_data_dict[sample.sample] += 1\n",
    "    \n",
    "    for sample in Missing_data_dict:\n",
    "        Missing_data_dict[sample] = np.round(float(Missing_data_dict[sample]/locus_counter), 2)\n",
    "                \n",
    "    return Missing_data_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def heterozygostity_finder(vcf_path):\n",
    "    import numpy as np\n",
    "    import vcf\n",
    "    \n",
    "    sample_het_loci = {}\n",
    "    sample_called_loci = {}\n",
    "    sample_heterozygosities = {}\n",
    "\n",
    "    for record in vcf.Reader(open(vcf_path, 'r')):\n",
    "        for sample in record.samples:\n",
    "            if sample.sample not in sample_het_loci:\n",
    "                sample_het_loci[sample.sample] = 0\n",
    "            if sample.sample not in sample_called_loci:\n",
    "                sample_called_loci[sample.sample] = 0\n",
    "            if sample.called:\n",
    "                sample_called_loci[sample.sample] += 1\n",
    "            if sample.is_het:\n",
    "                sample_het_loci[sample.sample] += 1\n",
    "\n",
    "    for sample in sample_het_loci:\n",
    "        sample_heterozygosities[sample] = np.round(sample_het_loci[sample]/sample_called_loci[sample],2)\n",
    "    \n",
    "    return sample_heterozygosities, sample_called_loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#vcf_path = \"/home/djeffrie/Data/Caspers_data/populations/r_075_goodsamples/batch_1c.vcf.altered\"\n",
    "\n",
    "def Per_locus_stats(vcf_path, sample_list):\n",
    "    \n",
    "    import numpy as np\n",
    "    import vcf\n",
    "    \n",
    "    locus_dict = {}\n",
    "    \n",
    "    print vcf_path\n",
    "    \n",
    "    for record in vcf.Reader(open(vcf_path, 'r')):\n",
    "\n",
    "        LocID = \"%s_%s\" % (record.ID, record.POS)\n",
    "        locus_dict[LocID] = {}\n",
    "        REF = record.REF        \n",
    "        \n",
    "        locus_coverages = []\n",
    "        het_samples = 0\n",
    "        sample_count = 0\n",
    "        REF_counts = 0\n",
    "        \n",
    "        for sample in record.samples:\n",
    "            if sample.sample in sample_list:\n",
    "                if sample.called:\n",
    "                    #print LocID, \"is called, sample ID should be > 0\"\n",
    "                    sample_count += 1\n",
    "                    locus_coverages.append(sample['DP'])\n",
    "                    REF_counts += sample.gt_bases.count(REF)\n",
    "                if sample.is_het:\n",
    "                    het_samples += 1\n",
    "\n",
    "        locus_dict[LocID][\"N_samples_called\"] = sample_count\n",
    "        \n",
    "        if sample_count > 0:\n",
    "            locus_dict[LocID][\"Major_freq\"] = REF_counts/(sample_count*2)\n",
    "            locus_dict[LocID][\"Average_coverage\"] = np.round(np.mean(locus_coverages),2)\n",
    "            locus_dict[LocID][\"Loc_heterozygosity\"] = het_samples/sample_count\n",
    "        else:\n",
    "            locus_dict[LocID][\"Major_freq\"] = 0 ## this may cause problems downstream, look out for this! Need to deal with it cos this will be common in p_tests\n",
    "            locus_dict[LocID][\"Average_coverage\"] = 0\n",
    "            locus_dict[LocID][\"Loc_heterozygosity\"] = 0\n",
    "        \n",
    "    \n",
    "    \n",
    "    return locus_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def Nsnp_Avcov_plotter(Sample_data, group_info):\n",
    "\n",
    "    import math\n",
    "    import numpy as np\n",
    "    from matplotlib import pyplot as plt\n",
    "    from matplotlib import cm\n",
    "    ## Get data for groups so this can be plotted separately\n",
    "\n",
    "    Mean_N_SNPs_per_sample = {}\n",
    "    Mean_coverage_per_sample = {}\n",
    "    \n",
    "    fig = plt.figure(figsize = (7*(len(Sample_data)),10))\n",
    "    \n",
    "    param_plot_counter = 0\n",
    "        \n",
    "    N_params = 0\n",
    "    \n",
    "    for param in Sample_data:\n",
    "        if len(Sample_data[param]) > 0:\n",
    "            N_params += 1\n",
    "    \n",
    "    for param in Sample_data:\n",
    "        \n",
    "        if len(Sample_data[param]) > 0:\n",
    "    \n",
    "            Mean_N_SNPs_per_sample[param] = {}\n",
    "            Mean_coverage_per_sample[param] = {}\n",
    "            param_plot_counter += 1\n",
    "            \n",
    "            for value in Sample_data[param]:\n",
    "                \n",
    "                if len(Sample_data[param][value]) > 0:\n",
    "                    \n",
    "                    \n",
    "                    Mean_N_SNPs_per_sample[param][value] = {}\n",
    "                    N_SNPs_per_sample_group = {}\n",
    "                    Mean_coverage_per_sample[param][value] = {}\n",
    "                    Coverage_per_sample_group = {}\n",
    "\n",
    "                    if len(Sample_data[param][value]) >= 1:\n",
    "\n",
    "                        #print param, value, Sample_data[param][value].keys()\n",
    "\n",
    "                        for sample in Sample_data[param][value]:\n",
    "                            for group in group_info:\n",
    "                                if sample in group_info[group]:\n",
    "                                    sample_group = group\n",
    "                            #print Sample_data[param][value][sample]\n",
    "                            if sample_group not in N_SNPs_per_sample_group:\n",
    "                                N_SNPs_per_sample_group[sample_group] = []\n",
    "                                Coverage_per_sample_group[sample_group] = []\n",
    "                            N_SNPs_per_sample_group[sample_group].append(Sample_data[param][value][sample][\"N_snps\"])\n",
    "                            Coverage_per_sample_group[sample_group].append(Sample_data[param][value][sample][\"Average_coverage\"])\n",
    "                    for group in N_SNPs_per_sample_group:\n",
    "                        Mean_N_SNPs_per_sample[param][value][group] = np.round(np.mean(N_SNPs_per_sample_group[group]))\n",
    "                        Mean_coverage_per_sample[param][value][group] = np.round(np.mean(Coverage_per_sample_group[group]))\n",
    "\n",
    "            \n",
    "            #Now plot the data for all groups on each graph separately.\n",
    "\n",
    "            blues = cm.get_cmap(\"Blues\")\n",
    "\n",
    "            ax1 = fig.add_subplot(1,N_params,param_plot_counter)\n",
    "\n",
    "\n",
    "            group_counter = 1\n",
    "            for group in group_info:\n",
    "                group_color = blues(group_counter/len(group_info))\n",
    "                group_counter += 1\n",
    "\n",
    "                for param in Mean_N_SNPs_per_sample:\n",
    "                    group_snp_vals = []\n",
    "                    group_cov_vals = []\n",
    "\n",
    "                    for val in Mean_N_SNPs_per_sample[param]:\n",
    "                        if Mean_N_SNPs_per_sample[param][val] > 1:\n",
    "\n",
    "                            group_snp_vals.append(Mean_N_SNPs_per_sample[param][val][group])\n",
    "                            group_cov_vals.append(Mean_coverage_per_sample[param][val][group])\n",
    "\n",
    "                ax1.scatter(range(1,len(Mean_N_SNPs_per_sample[param].keys())+1,1), group_snp_vals, color = group_color)\n",
    "                ax1.plot(range(1,len(Mean_N_SNPs_per_sample[param].keys())+1,1), group_snp_vals, color = group_color, linewidth = 2, label = \"%s Nsnps\" % group)\n",
    "                ax1.set_ylabel(\"Mean number of SNPs per sample\")\n",
    "                #ax1.legend()\n",
    "\n",
    "                ax2 = ax1.twinx()\n",
    "                ax2.scatter(range(1,len(Mean_coverage_per_sample[param].keys())+1,1), group_cov_vals, color = group_color, marker = \"D\")\n",
    "                ax2.plot(range(1,len(Mean_coverage_per_sample[param].keys())+1,1), group_cov_vals, color = group_color, linewidth = 2, linestyle = \"dashed\", label = \"%s Cov\" % group)\n",
    "                ax2.set_ylabel(\"Mean coverage per sample\")\n",
    "                \n",
    "                if group_counter != len(group_info):\n",
    "                    ax2.axes.get_yaxis().set_visible(False)\n",
    "                \n",
    "                plt.xlabel(\"%s value\" % param)\n",
    "                plt.xticks(range(1,len(Mean_N_SNPs_per_sample[param].keys())+1,1), sorted(Mean_N_SNPs_per_sample[param].keys()))\n",
    "                plt.title(\"Parameter = %s\" % param)\n",
    "\n",
    "            ax1.legend(loc = \"upper left\")\n",
    "            ax2.legend(loc = \"upper right\")\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "### Next, plotting the amount of missin data for each sample in each group for each param value for each param\n",
    "\n",
    "def Missing_data_plotter(Sample_data, group_info):\n",
    "\n",
    "    import numpy as np\n",
    "    from matplotlib import cm\n",
    "    from matplotlib import pyplot as plt\n",
    "\n",
    "    colours = cm.get_cmap(\"Set1\")\n",
    "    \n",
    "    N_params = 0\n",
    "    for param in Sample_data:\n",
    "        if len(Sample_data[param]) > 0:\n",
    "            N_params += 1\n",
    "    \n",
    "    param_counter = 1\n",
    "    for param in Sample_data:\n",
    "        if len(Sample_data[param]) > 0:\n",
    "            \n",
    "            Missing_per_sample = {}\n",
    "            Missing_all_samples = []\n",
    "            Missing_per_group = {}\n",
    "            Mean_missing_per_sample = {}\n",
    "            Mean_missing_per_group = {}\n",
    "            Sample_labs = []\n",
    "            \n",
    "            for value in Sample_data[param]:\n",
    "                if len(Sample_data[param][value]) > 0:\n",
    "                    for group in group_info:\n",
    "                        Missing_per_group[group] = []\n",
    "                        for sample in Sample_data[param][value]:\n",
    "                            if sample in group_info[group]:\n",
    "                                if sample not in Missing_per_sample:\n",
    "                                    Missing_per_sample[sample] = []\n",
    "                                    Sample_labs.append(sample)\n",
    "                                Missing_per_sample[sample].append(Sample_data[param][value][sample][\"Missing_data\"])\n",
    "                                Missing_per_group[group].append(Missing_per_sample[sample])\n",
    "\n",
    "            for sample in Missing_per_sample:\n",
    "                Missing_all_samples.append(Missing_per_sample[sample])\n",
    "\n",
    "            fig = plt.figure(figsize=(0.3*(len(Sample_labs)), 10*N_params))\n",
    "            \n",
    "\n",
    "            fig.add_subplot(N_params,1,param_counter)\n",
    "\n",
    "            plt.boxplot(Missing_all_samples)\n",
    "            plt.xticks(range(1, len(Sample_labs)+1, 1), Sample_labs, rotation = \"vertical\")\n",
    "            plt.ylabel(\"Proportion of loci missing\")\n",
    "            plt.title(\"Parameter = %s\" % param)\n",
    "            plt.grid(b=True, which='both', color='0.65',linestyle='-', axis = \"x\")\n",
    "            plt.hlines(np.mean([np.mean(i) for i in Missing_all_samples]),1,len(Sample_labs), color =\"grey\", linestyle = \"dashed\", label = \"All\")\n",
    "\n",
    "            group_counter = 1\n",
    "            for group in Missing_per_group:\n",
    "                mycolour = colours(group_counter/len(Missing_per_group))\n",
    "                plt.hlines(np.mean([np.mean(i) for i in Missing_per_group[group]]),1,len(Sample_labs), color =mycolour, label = group, linewidth = 2)\n",
    "\n",
    "                group_counter += 1\n",
    "\n",
    "            plt.legend(loc='upper left', title = \"Means\")\n",
    "\n",
    "            param_counter += 1\n",
    "            plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def Incremental_Pop(Analysis_parameters):\n",
    "    \n",
    "    import vcf\n",
    "    import sys\n",
    "    import os\n",
    "    \n",
    "    ## Inside Incremental_Populations function.\n",
    "\n",
    "    ## Making the necassary directories ---------------------------------------------------------------------------------\n",
    "\n",
    "    p_default = Analysis_parameters[\"p_default\"]\n",
    "    r_default = 0.75\n",
    "    m_default = 3\n",
    "\n",
    "    outdirs = []\n",
    "\n",
    "    for param in  Analysis_parameters['parameters']: \n",
    "        \n",
    "        if len(Analysis_parameters[\"%s_vals\" % param]) > 0:\n",
    "\n",
    "            for value in Analysis_parameters[\"%s_vals\" % param]:\n",
    "\n",
    "                indirname = Analysis_parameters[\"P\"]\n",
    "\n",
    "                if isinstance(value, float):\n",
    "                    dirvalue = str(value).replace(\".\", \"\")\n",
    "                    outdirname = '%s/%s_tests/%s_%s'%(Analysis_parameters[\"Populations_test_dir_path\"],param,param,str(dirvalue))\n",
    "                    outdirs.append(outdirname)\n",
    "\n",
    "                else:\n",
    "                    outdirname = '%s/%s_tests/%s_%s'%(Analysis_parameters[\"Populations_test_dir_path\"],param,param,str(value))\n",
    "                    outdirs.append(outdirname)\n",
    "\n",
    "\n",
    "                if not os.path.exists(outdirname):\n",
    "                    try:\n",
    "                        os.makedirs(outdirname)\n",
    "                    except:\n",
    "                        os.mkdir(outdirname)\n",
    "\n",
    "                ## Make and run the command line\n",
    "\n",
    "                if Analysis_parameters[\"run_cline_switch\"] == 1:\n",
    "                    print \"## Running command lines ##\"\n",
    "\n",
    "                pop_cline = make_and_run_populations_cline(Analysis_parameters,param, value)\n",
    "\n",
    "                print \"%s    ## FINISHED \" % pop_cline\n",
    "\n",
    "                ## Move files \n",
    "\n",
    "                if Analysis_parameters[\"run_cline_switch\"] == 1:\n",
    "                    for fil in os.listdir(Analysis_parameters1[\"P\"]):\n",
    "\n",
    "                        pop_outfile = \"%s/%s\" % (Analysis_parameters1[\"P\"],fil)\n",
    "                        pop_testdir_outfile = \"%s/%s\" % (outdirname, fil)\n",
    "\n",
    "                        if \"batch_%s\" % Analysis_parameters1[\"b\"] in fil and \"catalog\" not in fil:\n",
    "                            os.rename(pop_outfile, pop_testdir_outfile)\n",
    "                        #if \"populations.log\" in fil:\n",
    "                            #os.rename(pop_outfile, pop_testdir_outfile)\n",
    "                    print \"Files moved to %s\" % outdirname\n",
    "            \n",
    "    \n",
    "    \n",
    "    ## Next get the data for plotting ## --------------------------------------------------------------------------------\n",
    "    \n",
    "    print \"##### Getting and plotting the data ###### \"\n",
    "    \n",
    "    ## Make two dictionaries, one for sample level data, and one for locus level data. \n",
    "\n",
    "    Sample_level_data={}\n",
    "    Locus_level_data={}\n",
    "    group_dict = {}\n",
    "    Samples = []\n",
    "\n",
    "    Popmap = Analysis_parameters['M'] ## Get the popmap file used in Stacks for the population information\n",
    "\n",
    "    for line in open(Popmap, 'r').readlines():\n",
    "        sample_name = line.split()[0]\n",
    "        pop_name = line.split()[1]\n",
    "\n",
    "        if pop_name not in group_dict:\n",
    "            group_dict[pop_name] = []\n",
    "        group_dict[pop_name].append(sample_name)\n",
    "        \n",
    "        Samples.append(sample_name)\n",
    "\n",
    "    ## Get only data for samples in the popmap. If other samples present in the vcf they won't be examined.\n",
    "    \n",
    "    ## And get only data for the parameters specified to the function\n",
    "\n",
    "    test_path = Analysis_parameters[\"Populations_test_dir_path\"]\n",
    "\n",
    "    ## find the vcf's for the parameters and values specified\n",
    "    \n",
    "    vcf_files_found = 0\n",
    "    for param in Analysis_parameters['parameters']:\n",
    "        \n",
    "        Sample_level_data[param] = {}\n",
    "        Locus_level_data[param] = {}\n",
    "        \n",
    "        for value in Analysis_parameters[\"%s_vals\" % param]:\n",
    "            \n",
    "            if isinstance(value, float):\n",
    "                value = str(value).replace(\".\", \"\")\n",
    "            \n",
    "            Sample_level_data[param][value] = {}\n",
    "            Locus_level_data[param][value] = {}\n",
    "                        \n",
    "            \n",
    "            for outdirectory in outdirs:\n",
    "            \n",
    "                if param in outdirectory.rpartition(\"/\")[-1].split(\"_\")[0] and str(value) in outdirectory.rpartition(\"/\")[-1].split(\"_\")[1]:\n",
    "                    \n",
    "                    for fil in os.listdir(outdirectory):\n",
    "                        if fil.endswith(\"vcf\") or fil.endswith(\"vcf.altered\"):\n",
    "                            if \"haplotypes\" not in fil: ## what about haplotypes files?\n",
    "                                vcf_files_found += 1\n",
    "                                vcffile = \"%s/%s\" % (outdirectory, fil)\n",
    "\n",
    "                                alteredvcfpath = \"%s.altered\" % vcffile\n",
    "                                alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "                                for line in open(vcffile, 'r').readlines():\n",
    "                                    if \"Allele Depth\" not in line:\n",
    "                                        alteredvcf.write(line)\n",
    "                                    elif \"Allele Depth\" in line:\n",
    "                                        line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "                                        alteredvcf.write(line)\n",
    "                                alteredvcf.close()\n",
    "\n",
    "\n",
    "                                ## Now use all the functions you have to get the data needed from the vcfs! Can get sumstats files with this loop too if needed. \n",
    "\n",
    "                                ### SAMPLE LEVEL STATISTICS ### -----------------------------------------------------------------------------------------------\n",
    "\n",
    "\n",
    "                                ### AVERAGE COVERAGES ###\n",
    "                                sample_Av_coverages = sample_Avcov_from_vcf(alteredvcfpath)\n",
    "\n",
    "                                for sample in sample_Av_coverages:\n",
    "                                    if sample in Samples:\n",
    "                                        if sample not in Sample_level_data[param][value]:\n",
    "                                            Sample_level_data[param][value][sample] = {}\n",
    "                                            for group in group_dict:\n",
    "                                                if sample in group_dict[group]:\n",
    "                                                    Sample_level_data[param][value][sample][\"Pop\"] = group\n",
    "\n",
    "                                        Sample_level_data[param][value][sample][\"Average_coverage\"] = sample_Av_coverages[sample]\n",
    "\n",
    "                                ### AVERAGE HETEROZYGOSITY and Number of SNP loci ###\n",
    "                                Per_sample_heterozygosities, N_snps_per_sample = heterozygostity_finder(alteredvcfpath)\n",
    "\n",
    "                                for sample in Per_sample_heterozygosities:\n",
    "                                    if sample in Samples:\n",
    "                                        Sample_level_data[param][value][sample][\"Heterozygosity\"] = Per_sample_heterozygosities[sample]\n",
    "\n",
    "\n",
    "                                ### PERCENTAGE MISSING DATA ###\n",
    "                                Per_sample_missing_data = Missing_data_finder(alteredvcfpath)\n",
    "\n",
    "                                for sample in Per_sample_missing_data:\n",
    "                                    #print sample\n",
    "                                    if sample in Samples:\n",
    "                                        Sample_level_data[param][value][sample][\"Missing_data\"] = Per_sample_missing_data[sample]\n",
    "\n",
    "                                for sample in N_snps_per_sample:\n",
    "                                    #print sample\n",
    "                                    if sample in Samples:\n",
    "                                        Sample_level_data[param][value][sample][\"N_snps\"] = N_snps_per_sample[sample]\n",
    "\n",
    "\n",
    "                                ### lOCUS LEVEL STATISTICS ### ------------------------------------------------------------------------------------------------\n",
    "\n",
    "                                ## For all samples\n",
    "\n",
    "                                Locus_level_data[param][value][\"All_samples\"] = Per_locus_stats(alteredvcfpath, Sample_level_data[param][value].keys())\n",
    "\n",
    "                                ## For sample groups\n",
    "\n",
    "                                for group in group_dict:\n",
    "                                    Locus_level_data[param][value][group] = Per_locus_stats(alteredvcfpath, group_dict[group])\n",
    "\n",
    "    #print Sample_level_data\n",
    "    \n",
    "    if vcf_files_found == 0:\n",
    "        sys.exit(\"No vcf files found in test directories\")\n",
    "    else:\n",
    "        ### Plotting ### --------------------------------------------------------------------------------------------------------------------------------------\n",
    "\n",
    "        ### Sample and group level data ###\n",
    "\n",
    "        ## Number of snps and mean coverage accross parameter values -----------------\n",
    "\n",
    "        Nsnp_Avcov_plotter(Sample_level_data, group_dict)\n",
    "\n",
    "        ## Missing data for all parameter tests averaged across values -------------------\n",
    "\n",
    "        Missing_data_plotter(Sample_level_data, group_dict)\n",
    "\n",
    "        return Sample_level_data, Locus_level_data, group_dict\n",
    "    \n",
    "\n",
    "        ### Locus level data ###\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## params for the Incremental populations analyses\n",
    "\n",
    "Analysis_parameters1 = {}\n",
    "\n",
    "## Test parameters\n",
    "Analysis_parameters1[\"parameters\"] = [\"r\",\"m\"]\n",
    "Analysis_parameters1[\"p_vals\"] = []\n",
    "Analysis_parameters1[\"p_default\"] = 2 ## The value of p when other params are being tested. Set as 2 for males and females!\n",
    "Analysis_parameters1[\"r_vals\"] = [0.75, 0.85, 0.9, 0.95]\n",
    "Analysis_parameters1[\"r_default\"] = 0.75 ## The value of r when other params are being tested. \n",
    "Analysis_parameters1[\"m_vals\"] = [2,3,4,5,6]\n",
    "Analysis_parameters1[\"m_default\"] = 3 ## The value of m when other params are being tested. \n",
    "Analysis_parameters1[\"Populations_test_dir_path\"] = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/Populations_tests_outs\"\n",
    "Analysis_parameters1[\"run_cline_switch\"] = 0 ## run command line or not. \n",
    "\n",
    "## Mandatory Populations options\n",
    "Analysis_parameters1[\"b\"] = 1 ## batch ID\n",
    "Analysis_parameters1[\"P\"] = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests\"## Path to stacks outputs\n",
    "Analysis_parameters1[\"M\"] = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt\" ## All samples\n",
    "\n",
    "## Outputs parameters\n",
    "Analysis_parameters1[\"vcf\"] = True\n",
    "Analysis_parameters1[\"vcf_haplotypes\"] = True\n",
    "\n",
    "## Optional parameters\n",
    "Analysis_parameters1[\"t\"] = 7 ## number of threads\n",
    "Analysis_parameters1[\"min_maf\"] = 0.05 ## number of threads\n",
    "\n",
    "#Analysis_parameters1[\"B\"] = None ## Locus blacklist\n",
    "Analysis_parameters1[\"write_single_snp\"] = True ## switch to write first snp of locus or not.\n",
    "Analysis_parameters1[\"fstats\"] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 3 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.85 -p 2 -m 3 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.9 -p 2 -m 3 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.95 -p 2 -m 3 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 2 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 3 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 4 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 5 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "populations -b 1 -P /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests -M /home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/pop_codes.txt -t 7 -r 0.75 -p 2 -m 6 --min_maf 0.05 --write_single_snp --vcf --vcf_haplotypes    ## FINISHED \n",
      "##### Getting and plotting the data ###### \n",
      "/home/djeffrie/Data/RADseq/Lib_Rjapor/Incremental_tests/Populations_tests/Populations_tests_outs/r_tests/r_075\n"
     ]
    },
    {
     "ename": "ZeroDivisionError",
     "evalue": "division by zero",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-56-c6ae242b65d3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mSample_data\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mLocus_data\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgroup_info\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mIncremental_Pop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mAnalysis_parameters1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-52-9d2d4ef9dc87>\u001b[0m in \u001b[0;36mIncremental_Pop\u001b[1;34m(Analysis_parameters)\u001b[0m\n\u001b[0;32m    151\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    152\u001b[0m                                 \u001b[1;31m### AVERAGE HETEROZYGOSITY and Number of SNP loci ###\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 153\u001b[1;33m                                 \u001b[0mPer_sample_heterozygosities\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mN_snps_per_sample\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mheterozygostity_finder\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malteredvcfpath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    154\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    155\u001b[0m                                 \u001b[1;32mfor\u001b[0m \u001b[0msample\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mPer_sample_heterozygosities\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-34-3d763021ec3b>\u001b[0m in \u001b[0;36mheterozygostity_finder\u001b[1;34m(vcf_path)\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     20\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0msample\u001b[0m \u001b[1;32min\u001b[0m \u001b[0msample_het_loci\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 21\u001b[1;33m         \u001b[0msample_heterozygosities\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0msample\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mround\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msample_het_loci\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0msample\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m/\u001b[0m\u001b[0msample_called_loci\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0msample\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     22\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     23\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0msample_heterozygosities\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msample_called_loci\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mZeroDivisionError\u001b[0m: division by zero"
     ]
    }
   ],
   "source": [
    "Sample_data, Locus_data, group_info = Incremental_Pop(Analysis_parameters1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
