{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def missing_data_finder(input_file, output_file_path, blacklisted_tags_cutoff=100):\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    This function will count the number of tags that droupout in all samples\n",
    "    in the specified VCF. It will output the data and simple bar plots of both \n",
    "    the number of dropouts per sample, and the number of dropouts per tag. It will also \n",
    "    output a file called blacklist.txt containing locus IDs which can be used to remove \n",
    "    these loci from subsequent runs of the stacks populations module.\n",
    "    \n",
    "        input_file              -  full path to vcf\n",
    "        output_file_path        -  path to output data, plots and blacklist to\n",
    "        blacklisted_tags_cutoff -  number of 'worst' tags to blacklist\n",
    "        \n",
    "    '''\n",
    "    \n",
    "    \n",
    "    import sys\n",
    "    import vcf \n",
    "    import numpy as np\n",
    "    import matplotlib.pyplot as plt \n",
    "    from collections import Counter\n",
    "    from operator import itemgetter\n",
    "\n",
    "\n",
    "    ## Alter the allele depth header line - which doesn't seem to be compatible between Stacks and PyVCF. \n",
    "    \n",
    "    myvcf = open(input_file, 'r').readlines()\n",
    "    alteredvcfpath = \"%s%s\" %(input_file, \".altered\")\n",
    "    alteredvcf = open(alteredvcfpath, 'w')\n",
    "    \n",
    "    for line in myvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "    \n",
    "    ## Now open altered file and continue\n",
    "    \n",
    "    alteredvcf = open(alteredvcfpath, 'r')\n",
    "    \n",
    "    vcf_reader = vcf.Reader(alteredvcf) \n",
    "    record = next(vcf_reader)\n",
    "    \n",
    "    samples = vcf_reader.samples ## get all the samples in the file\n",
    "\n",
    "    ## Count the missing data --------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "    missing_data_count = {}\n",
    "    missing_data_tags = {}\n",
    "    \n",
    "    for record in vcf.Reader(open(alteredvcfpath, 'r')): \n",
    "        for sample in samples: ## for each sample\n",
    "            if sample not in missing_data_count:\n",
    "                missing_data_count[sample] = 0\n",
    "            \n",
    "            if record.ID not in missing_data_tags:\n",
    "                missing_data_tags[record.ID] = 0\n",
    "            \n",
    "            if record.genotype(sample)['GT'] == None: ## If the current samples genotype for this record is missing ('./.'), count it with no_genotype_counter\n",
    "                missing_data_count[sample] += 1\n",
    "                missing_data_tags[record.ID] += 1\n",
    "\n",
    "    \n",
    "    ######### TAG DROPOUTS PER SAMPLE ######## -------------------------------------------------------------------------------------------\n",
    "    \n",
    "    sorted_tags_per_sample = sorted(missing_data_count.items(), key=itemgetter(1), reverse=True) ## write file from this sorted tags list\n",
    "    \n",
    "    ## Print tag dropouts per sample to file\n",
    "    f = open(output_file_path+'/Tag_dropouts_per_sample.txt', 'w')\n",
    "    f.write('sample\\t#_missing_loci\\n')\n",
    "    for sample in sorted_tags_per_sample:\n",
    "        f.write(\"%s\\t%s\\n\" % (sorted_tags_per_sample[0], sorted_tags_per_sample[1]))\n",
    "    f.close()\n",
    "    \n",
    "\n",
    "    \n",
    "    ######### TAG DROPOUTS PER SAMPLE ######## -------------------------------------------------------------------------------------------\n",
    "    \n",
    "    x = []\n",
    "    y = []\n",
    "    xlabs = []\n",
    "    counter = 1\n",
    "\n",
    "    plt.figure(figsize=(70,50))\n",
    "\n",
    "    for i, j in sorted_tags_per_sample:\n",
    "        xlabs.append(i)        \n",
    "        x.append(counter)\n",
    "        y.append(int(j))\n",
    "        counter += 1\n",
    "\n",
    "    ## Plot the per-tag droupouts\n",
    "    plt.bar(x,y, edgecolor = \"none\",align=\"center\")\n",
    "    plt.yticks(range(0,max(y),100), fontsize = 30)\n",
    "    plt.xticks(x, xlabs, fontsize = 30, rotation = \"vertical\")\n",
    "    plt.ylabel(\"Number of dropouts\", fontsize = 30)\n",
    "    plt.xlabel(\"Sample\", fontsize = 30)\n",
    "    plt.axhline(np.mean(y), linewidth=4, color='r')\n",
    "    plt.title(\"Tag dropouts per sample\", fontsize = 40)\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"%s/Tag_dropouts_per_sample.pdf\" % output_file_path)\n",
    "    plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    \n",
    "    ######### TAG DROPOUTS PER TAG ######## -------------------------------------------------------------------------------------------\n",
    "    \n",
    "    sorted_tags = sorted(missing_data_tags.items(), key=itemgetter(1), reverse=True)\n",
    "    \n",
    "    x = []\n",
    "    y = []\n",
    "    xlabs = []\n",
    "    blacklisted = []\n",
    "\n",
    "    cut_off = blacklisted_tags_cutoff\n",
    "    counter = 1\n",
    "    plt.figure(figsize=(70,50))\n",
    "\n",
    "    for i, j in sorted_tags:\n",
    "        xlabs.append(i)\n",
    "        x.append(counter)\n",
    "        y.append(int(j))\n",
    "        if counter <= blacklisted_tags_cutoff:\n",
    "            blacklisted.append(j) ## write this to file\n",
    "        counter += 1\n",
    "\n",
    "    xarray = np.asarray(x)\n",
    "    yarray = np.asarray(y)\n",
    "\n",
    "    mask1 = xarray < cut_off\n",
    "    mask2 = xarray >= cut_off\n",
    "\n",
    "    ## write per-tag droupout data to file\n",
    "\n",
    "    tag_drop_outs = open(\"%s/Tag_dropouts_per_tag.txt\" % output_file_path, 'w')\n",
    "    tag_drop_outs.write(\"TagID\\tNumber_of_dropouts\\n\")\n",
    "\n",
    "    for tag in sorted_tags:\n",
    "        tag_drop_outs.write(\"%s\\t%s\\n\" % (tag[0], tag[1]))\n",
    "    tag_drop_outs.close()\n",
    "    \n",
    "    ## Write the blacklist file\n",
    "    \n",
    "    blacklist = open(output_file_path+'/blacklist.txt', 'w')\n",
    "    for tag_ID in set(blacklisted):\n",
    "        blacklist.write(\"%s\\n\" % (tag_ID))\n",
    "    blacklist.close()\n",
    "\n",
    "    ## Plot the per-tag droupouts\n",
    "    plt.bar(xarray[mask1],yarray[mask1], edgecolor = \"none\")\n",
    "    plt.bar(xarray[mask2],yarray[mask2], color = \"grey\", edgecolor = \"none\")\n",
    "    plt.yticks(range(0,max(y),50), fontsize = 30)\n",
    "    plt.xticks(range(0,max(x),100), fontsize = 30, rotation = \"vertical\")\n",
    "    plt.ylabel(\"Number of dropouts\", fontsize = 30)\n",
    "    plt.xlabel(\"Tag\", fontsize = 30)\n",
    "    plt.axhline(np.mean(yarray), linewidth=4, color='r')\n",
    "    plt.title(\"Tag dropouts per tag\", fontsize = 40)\n",
    "\n",
    "    #plt.tight_layout()\n",
    "    \n",
    "    plt.savefig(\"%s/Tag_dropouts_per_tag.pdf\" % output_file_path)\n",
    "    #plt.show()\n",
    "    plt.close()\n",
    "    \n",
    "    print 'Outputs written to:'\n",
    "    print \"%s/Tag_dropouts_per_sample.txt\" % output_file_path\n",
    "    print \"%s/Tag_dropouts_per_tag.txt\" % output_file_path\n",
    "    print \"%s/Tag_dropouts_per_sample.pdf\" % output_file_path\n",
    "    print \"%s/Tag_dropouts_per_tag.pdf\" % output_file_path\n",
    "    print \"%s/blacklist.txt\" % output_file_path\n",
    "    \n",
    "    return missing_data_count\n",
    "\n",
    "\n",
    "def missing_data_per_pop(missing_data_per_sample_dict, pop_file_path, outpath):\n",
    "    \n",
    "    '''\n",
    "    This function takes the outputs from the missing_data_finder function\n",
    "    along with the pop_map file used in stacks and creates an additional \n",
    "    dataset and barplot for the tag dropouts per population.\n",
    "    \n",
    "        missing_data_per_sample_dict     - Outputs from missing_data_finder\n",
    "        pop_file_path                    - Full path to the pop map text file used in stacks (popualtions)\n",
    "        \n",
    "    '''\n",
    "    \n",
    "    pop_map = open(pop_file_path, 'r').readlines()\n",
    "    \n",
    "    pop_missing_dict = {}\n",
    "    \n",
    "    for line in pop_map:\n",
    "        sample_name = line.strip().split()[0]\n",
    "        pop_name = line.strip().split()[1]\n",
    "        \n",
    "        if sample_name in missing_data_per_sample_dict:\n",
    "            \n",
    "            if pop_name not in pop_missing_dict:\n",
    "                pop_missing_dict[pop_name] = [int(missing_data_per_sample_dict[sample_name])]\n",
    "            else:\n",
    "                pop_missing_dict[pop_name].append(int(missing_data_per_sample_dict[sample_name]))\n",
    "                \n",
    "    x = pop_missing_dict.keys()\n",
    "    y = [sum (i) for i in pop_missing_dict.values()]\n",
    "    z = zip(x,y)            \n",
    "    ## write to file\n",
    "    \n",
    "    pop_dropouts = open(\"%s/Tag_dropouts_per_population.txt\" % outpath, 'w')\n",
    "    \n",
    "    for pop in z:\n",
    "        pop_dropouts.write(\"%s\\t%s\\n\" % (pop[0], pop[1]))\n",
    "    pop_dropouts.close()\n",
    "    \n",
    "    ## Plot population tag dropouts\n",
    "    plt.figure(figsize=(70,50))\n",
    "    plt.bar(range(len(x)), y, align='center', width = 0.6)\n",
    "    plt.xticks(range(len(x)), x, fontsize = 30)\n",
    "    plt.yticks(range(0,max(y),(np.round(max(y)/10))),range(0,max(y),(np.round(max(y)/10))) , fontsize = 30)\n",
    "    plt.ylabel(\"Number of tag dropouts\", fontsize = 30)\n",
    "    plt.title(\"Tag dropouts per population\", fontsize = 40)\n",
    "    plt.savefig(\"%s/Tag_dropouts_per_population.pdf\" % outpath)\n",
    "    #plt.show()\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "def Cstacks_script_maker(input_dir, output_dir, pop_file_path, nval, pval):\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    Makes Cstacks scripts and saves them to .sh file for running.\n",
    "    \n",
    "    input_dir        -  full path to the ustacks output files\n",
    "    output_dir       -  path to write the catalog to\n",
    "    pop_file_path    -  full path to the pop_map file\n",
    "    nval             -  value for n (mismatch)\n",
    "    pval             -  value for p (threads)\n",
    "    \n",
    "    '''\n",
    "    \n",
    "    import os\n",
    "    import re\n",
    "    import sys\n",
    "    \n",
    "    samples = []\n",
    "    pop_list = []\n",
    "    samples_ready = []\n",
    "    \n",
    "    for root, dirs, files in os.walk(input_dir):\n",
    "        for fil in files:\n",
    "            #print fil\n",
    "            if 'tags.tsv' in fil:\n",
    "                if fil.endswith('tsv'):\n",
    "                    samples.append(fil.rpartition('.')[0].rpartition('.')[0])\n",
    "                elif fil.endswith('tsv.gz'):\n",
    "                        samples.append(fil.rpartition('.')[0].rpartition('.')[0].rpartition('.')[0]) \n",
    "    samples = sorted(set(samples))\n",
    "    \n",
    "    for sample in samples:\n",
    "        samples_ready.append(' -s '+input_dir+sample)\n",
    "    samples = ''.join(samples_ready)\n",
    "    \n",
    "    Ccommand = 'cstacks -b 1 -n%s -p %s %s -o %s'  % (nval,pval,samples,output_dir)\n",
    "    \n",
    "    f = open(\"%s/Cstacks_scripts.sh\" % output_dir,'w')\n",
    "    f.write('#!/bin/bash\\n\\n')\n",
    "    \n",
    "    \n",
    "    for command in Ccommand:\n",
    "        f.write(command)\n",
    "    f.close()\n",
    "    \n",
    "    return Ccommand\n",
    "\n",
    "\n",
    "def Sstacks_script_maker(input_dir,threads, batch_id = 1):\n",
    "    \n",
    "    '''\n",
    "    Makes Sstacks scripts\n",
    "    input_dir\n",
    "    threads\n",
    "    '''\n",
    "    \n",
    "    import os\n",
    "    \n",
    "    sstacks_commands = []\n",
    "    \n",
    "    ## Make script\n",
    "    sstacks_script = open(\"%s/%s\" % (input_dir, \"Sstacks_scripts.sh\"), 'w')\n",
    "    sstacks_script.write(\"#!/bin/bash/\\n\")\n",
    "\n",
    "    ## First pass to get catalog file\n",
    "    for root, dirs, files in os.walk(input_dir):\n",
    "        for fil in files:\n",
    "            if \"tags.tsv\" in fil and \"catalog\" in fil:\n",
    "                catalog = \"%s/%s\" % (root, fil)\n",
    "    \n",
    "    ## Second pass to make script\n",
    "    for root, dirs, files in os.walk(input_dir):\n",
    "        for fil in files:\n",
    "            if \"tags.tsv\" in fil and \"catalog\" not in fil:\n",
    "                infile = \"%s/%s\" % (root, fil.partition(\".\")[0])\n",
    "\n",
    "                command = \"sstacks -b %s -c %s -s %s -o %s -p %s\" % (batch_id, catalog, infile, input_dir, threads)\n",
    "                script_command = \"sstacks -b %s -c %s -s %s -o %s -p %s;\\n\" % (batch_id, catalog, infile, input_dir, threads)\n",
    "                sstacks_script.write(script_command)\n",
    "                sstacks_commands.append(command)\n",
    "                \n",
    "                \n",
    "    sstacks_script.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
