{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq > 0.4.\n",
    "Written by D. Jeffries 09/2015 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Workflow:\n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "       i.e. female - male > 0.4 and femREF_freq > 0.95 for Xlinked or the reverse for Z linked.\n",
    "    4. Outputs all male and female frequencies to separate files called \"yourinput.vcf.female.freqs\", \"yourinput.vcf.male.freqs\" (where yourinput = the name and path of your vcf file)\n",
    "    5. Outputs all female-male frequency ratios to a file called \"yourinput.vcf.fem-male_freqs.tsv\"\n",
    "    6. Outputs all putative X or Z linked markers to separate files if any are identified.\n",
    "    7. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    8. Everything is recorded to a log file, with a summary at the end of this file. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from __future__ import division\n",
    "import vcf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import sys\n",
    "import time\n",
    "import gzip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## ARGS to be passed from command line eventually\n",
    "\n",
    "myvcfpath = \"/home/djeffrie/Data/Guillaumes_test_data/Rchen/Rchen_dp7.recode.vcf\"\n",
    "alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "popmappath = \"/home/djeffrie/Data/Guillaumes_test_data/Rchen/Rche_sex_info.txt\"\n",
    "catalog_tags_file = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/batch_1.catalog.tags.tsv\"\n",
    "X_or_Z_freq_threshold = 0.4\n",
    "sample_presence_cutoff = 0.75\n",
    "coverage_threshold = 7\n",
    "maf_threshold = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## ARGS to be passed from command line eventually\n",
    "\n",
    "myvcfpath = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/populations_1pop_kept_samples_outs/batch_1.vcf\"\n",
    "alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "popmappath = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/populations_1pop_kept_samples_outs/pop_codes.txt\"\n",
    "catalog_tags_file = \"/home/djeffrie/Data/Pnig_RAD/Stacks_outs/batch_1.catalog.tags.tsv\"\n",
    "X_or_Z_freq_threshold = 0.4\n",
    "sample_presence_cutoff = 0.75\n",
    "coverage_threshold = 7\n",
    "maf_threshold = 0.05"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "myvcf = open(myvcfpath, 'r').readlines()\n",
    "alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "for line in myvcf:\n",
    "    if \"Allele Depth\" not in line:\n",
    "        alteredvcf.write(line)\n",
    "    elif \"Allele Depth\" in line:\n",
    "        line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "        alteredvcf.write(line)\n",
    "alteredvcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now calculate allele frequencies for males and females at each SNP\n",
    "####Requires:\n",
    "    1. pyvcf module installed (can use pip, remember to add to python path. This is on the cluster!\n",
    "    2. altered vcf file from above\n",
    "    3. pop_map.txt file. Same format as used for stacks. Sample names must be the same. And males and females must be denoted by M or F (case sensitive) respectively. Must be the same file as used in populations to creat the VCF. If there are additional samples in this file the allele frequencies will be wrong!\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 64476\n",
      "Number of samples = 39\n",
      "Number of loci with too few samples = 45653\n",
      "Number of loci with low MAF = 14122\n",
      "Number of loci with enough data = 4701\n",
      "Number of putative X linked markers = 1\n",
      "Number of putative Z linked markers = 1\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEOCAYAAACTqoDjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAG4RJREFUeJzt3X+UJGV97/H3h18hKAiEXH4tskRB3FyvYCJq0DCKIHgM\nohgV1KAgxyPHoDcaBbmysyEhSIwhvzQJCm5AMYi5iESFBXYQjPwKLIgrsNywuazJLpIgkaBXkO/9\no56Bonm6p2amq6uq+/M6p89UV9fUfGu6n/5WPU/VtxQRmJmZ9dqs6QDMzKydnCDMzCzLCcLMzLKc\nIMzMLMsJwszMspwgzMwsq5YEIWkPSaslfVfSHZJOSvN3lLRK0t2SrpC0fel3TpG0TtKdkg6tIy4z\nM6tOdVwHIWkXYJeIWCPpmcA/AUcC7wIeiIizJH0E2CEiTpa0DPgC8GJgd+BKYJ+IeHzowZmZWSW1\nHEFExMaIWJOmHwa+R/HFfwSwMi22kiJpALweuDAiHo2I9cA9wAF1xGZmZtXUPgYhaSmwP3ADsHNE\nbEovbQJ2TtO7ARtKv7aBIqGYmVlDak0QqXvpy8D7I+JH5dei6Nsa1L/lGiBmZg3aoq4VS9qSIjmc\nHxGXpNmbJO0SERsl7Qrcn+Z/H9ij9OtL0rzedTppmJktQERovr9T11lMAj4LrI2Is0svXQocm6aP\nBS4pzX+rpK0k7QXsDdyYW3dEjO3joIOWNx5DXY/ly8d3256yfRSHvk3H4/dvfo9xbnsRC9+vrusI\n4kDg7cDtkm5N804BzgQuknQ8sB54M0BErJV0EbAWeAw4MRazVR11zTVNR2A2mdz28mpJEBFxHf2P\nTl7d53fOAM6oIx4zM5s/X0ndKlNNB1CbqamppkOolbev66aaDqCVarlQri6SxrrnSYIx3rzJoDQO\n6DeyU8a97Uki2jJIbWZm3Vfbaa42f8uXNx2BzZf01J2ycT7CHWdue3nuYjJbBElMT08DMD097QRh\nreQuJjMzGyonCDMzy3KCMDOzLCcIMzPLcoJokTTWaV0mPXkthHWG216eE0SLrFjRdATWBElPedjo\nue3lOUGYtcD09PQTp8uatYUThJmZZTlBmJlZlhOEmZlluRZTi7gezBjoKbXhWk3d4LaX5yOIFvEY\n5XjyAHT7+e3J8xGE2RD5NFUbJz6CMBuicmVXHzVY1zlBmJlZlhOEmZllOUG0iHskum/59DSDzlNy\nSY12ctvLc4JoEdeDGX/lsQknivZw28tzgjBriAeyre2cIMzMLMsJwszMspwgzMwsywmiRVwPpvtW\nTE/jYefucdvLc4JoEY9XmjXDbS/PCcLMzLKcIMzMLMsJwszMspwgzMwsywmiRTxQ1n1z1WKydnLb\ny3OCaBHXgzFrhttenu8oZ9ZSvp+1Nc1HEGYt5oJ+1iQnCDMzy3KCMDOzLCeIFnE9mOaV7/hWHgPo\nN7+XazF1k9tenhNEi7iruR369ft7PGB8+W3Nc4IwM7MsJwgzM8tygjAzsywnCDMzy3KCaBEPlHWf\nazF1k9tenhNEi7gejFkz3PbyaksQks6VtEnSd0rzpiVtkHRrehxeeu0USesk3Snp0LriMjOzauo8\ngjgPOKxnXgCfjIj90+PrAJKWAW8BlqXf+ZQkH92YmTWoti/hiLgWeDDzUu5C09cDF0bEoxGxHrgH\nOKCu2MzMbG5N7KX/tqTbJH1W0vZp3m7AhtIyG4DdRx+amZnNGvX9ID4N/F6aPh34Y+D4PstmTwYp\nlzqYmppiampqeNE1zPVgum9FKscx3XQgNi/j1vZmZmaYmZlZ9HpGmiAi4v7ZaUmfAb6ann4f2KO0\n6JI072nGuRbOGG+aWauNW9vr3XlescDTtEbaxSRp19LTNwCzZzhdCrxV0laS9gL2Bm4cZWxmZvZU\ntR1BSLoQOAjYSdJ9wHJgStJ+FN1H9wLvAYiItZIuAtYCjwEnhu+vaGbWqNoSREQcnZl97oDlzwDO\nqCseMzObH19rYGZmWU4QLTJuA2WTyLWYusltL88JokVcD8asGW57eU4QZmaWNeoL5cxaR8pVfxn+\n75h1jY8gzCguwJzPRZjzXd6si5wgzMwsywmiRcatHswkWjE9nS1XbO3mtpfnBNEi7rEwa4bbXp4T\nhJmZZTlBmJlZlhOEmZllOUGYmVmWE0SLeKCs+4ZRi0mSL8QbMbe9PCeIFnE9GANfhNcEt708l9ow\nG8B78jbJfARhNoD35m2SOUGYmVmWE4SZmWU5QbSI68F0n2sxdZPbXt6cCULS4enncyV9StKv1x/W\nZHJXt1kz3PbyqhxBfCj9PBW4APhkfeGYjcbstQY+S8msvyoJ4pmS9gQei4h/BB6uOSazkfAZSmaD\nVUkQfwicDnxC0tbA9fWGZGZVlI+CfCRkdeh7oZykvYEPA0uAu4EfRsRPgJNHFJuZzWH2CMhHQlaH\nQUcQ5wLnAEcClwGfGklEE8xtvPuGUYvJRs9tL29QgngsIm6MiP8XEauAHUYV1KRyPRizZrjt5Q2q\nxfQiSdeWnr9g9nlEvKLesMwWp7dPPsL79Wbz1TdBRMSzRhmI2bC5f95scQYNUp+emR0AEXFabRGZ\nmVkrDOpiuir9nD02fx7wAeDfa43IzMxaoe8gdUTMRMQMsCXwfuBA4B0ef6iP68F0n2sxdZPbXt6g\nLqbjgWOAbwEnRsTGkUU1odxVXp/ZQWsPVluO217eoNNcPwZsC7wS+JKka2cfownNbHg8UG02f4PO\nYlo6wjjMzKxl+h5BSHq3pO3T9MskXS3pSklTI4vOzMwaM6iL6biI+GGa/hvgOOAoisJ9ZmY25gYl\niJ8BSHoO8GBErI+Ih4DHRxLZBHI3ef3qrnzqWkzd5LaXNyhB3Czpb4EvAJ8GkOR6TDVyPZj6+R4Q\nluO2lzdokPp/SvrvwCMR8c+ll46rPywz61X1yMd1qGxYBt4wKCLuKCeHiHgwIv5P/WGZWa9ybam5\njoJ8pGTDUOWOcmZmNoEGneb6d+mnz1oyM5tAg44glkj6Q+Adkk4vPX5vVMFNGteD6T7XYuomt728\nQdVcDwVeCLwCuHI04Uw2dxmbNcNtL29QNdf/ioh/jIiXAz8Gdgd+HBHXjCw6MzNrzJyD1JL+FDiB\nonDfCZL+vPaozMyscVXOYtovIk6IiL+OiBMoup3mJOlcSZskfac0b0dJqyTdLemK2VpP6bVTJK2T\ndKekQ+e/KWZmNkxVEsSPJL1N0vMlvQP4UcV1nwcc1jPvZGBVROxDcce6kwEkLQPeAixLv/MpST4F\n18ysQVW+hN8G7AacBOxMcROhOUXEtcCDPbOPAFam6ZXAkWn69cCFEfFoRKwH7gEOqPJ3xokHyrrP\ntZi6yW0vb84EEREPRcQfRcR7I+ITqWDfQu0cEZvS9CaKhANFAtpQWm4DxaD4RHE9GLNmuO3lDTrN\ntVYREZIG7WxlXyuXD5iammJqamq4gZmZddzMzAwzMzOLXs/ABKGi6tdrIuIbi/5LhU2SdomIjZJ2\nBe5P878P7FFabkma9zSuL2P91FnG26xLeneeVyzwEGmuYn0BvHdBa867FDg2TR8LXFKa/1ZJW0na\nC9gbuHGIf9cmhIvUmQ1PlS4mSfo6cAvpZkER8bEKv3QhcBCwk6T7gNOAM4GLJB0PrAfenNa3VtJF\nwFrgMeDEcI1iM7NGVUkQn6TPeMAgEXF0n5de3Wf5M4Az5vt3xonrwXTfinQEM910IDYvbnt5VU5z\nXQO8DHgt8C1gu1ojmmDuGTFrhtteXpUEcQFwG/BrEfEY8IF6QzIzszaokiB+PiK+TjE2YNYISU95\nmFn9qiSIuyWdDOwg6XeA79Yck1mWz1AyG60qCeJE4A7g88A9EXFSvSGZmVkbVEkQWwDPALYCtpG0\nZb0hTS7vHHefazF1k9teXpUE8SVgT+Am4NnpudXA9WBsMTw+s3Bue3lVroPYOiLOStNXSLq8zoDM\nbGFmx2c8TmPD0jdBpKudAX4i6a8prod4IfCfowjMzMyaNegI4mcUV1DP1ksK4HoWcFW1mZl1T98E\nERGfm52WtAOwA+AOTjOzCTHnILWkc4AvAh8D/ld6WA1cD6b7VkxPey+qg9z28qoMUu8TEQfVHon5\nVDuzhrjt5VVJEJ+XdBJPXkEdEXF1jTGZmVkLVEkQR1Dc3W2H0jwnCDOzMVfpntQR8Z66AzEzs3ap\nkiB+IulPKLqYgqKL6dx6wzIzs6ZVKbVxGcVFco9SlPz+Wa0RTTAPlHWfazF1k9teXpUEsRqY6XlY\nDVwPxqwZbnt5VbqYptNPAb8MPAgcWldAZmbWDnMmiIh41+y0ilKRruZqreHqpWb1mTNBSDqYJ+sv\n7QYsrTMgs/lyFVOzelTpYnp5+hnAQ8Ab6gvHzMzaou8gtaTnSHoOcEF6fJ7ijCbfUa4mrgfTfa7F\n1E1ue3mDjiByRfkOAJ43x+/ZArmHxKwZbnt5g8p9vwtA0mbAUcAJwFXAkaMJzczMmjTojnLbUSSF\n3wAuBX4zIh4aVWBmZtasQV1FG4D7gK9QFOr7UDqlMCLitBHEZmZmDRqUII5IP105wGwM9F4zEuGm\nbYP1PYspImbS45rexygDnCQeKOu+ttdimp6e9vUiGf6X5FWpxWQj4nowZs1w28tzgjAzsyxfz2Ct\n5TpLZs1ygrBW61dnycmjurn+V+XXPXBtZe5isk7yYGt1Vf5X/l9ajhNEi7geTPe5FlM3ue3lOUG0\niHfizJrhtpfnBGFmZllOEGZmluUEYWZmWU4QZmaW5QTRIh4o676212KyPLe9PCeIFnE9GLNmuO3l\nOUGYmVmWS21Yq7iEhll7+AjCWsdlNMzawQnCzMyyGkkQktZLul3SrZJuTPN2lLRK0t2SrpC0fROx\nNcn1YLrPtZi6yW0vr6kjiACmImL/iDggzTsZWBUR+wBXpecTxb0qZs1w28trsoupd0frCGBlml4J\nHDnacMxM0hMPsyaPIK6UdLOkE9K8nSNiU5reBOzcTGhmk8snCFhZU6e5HhgR/ybpF4FVku4svxgR\nISl7QWr5wzs1NcXU1FSdcZqZdc7MzAwzMzOLXk8jCSIi/i39/IGk/w0cAGyStEtEbJS0K3B/7ne9\nd2NmNljvzvOKBV4qPvIuJknbSNo2TT8DOBT4DnApcGxa7FjgklHH1jTnvu5zLaZuctvLa2IMYmfg\nWklrgBuAyyLiCuBM4BBJdwOvSs8niuvB2Ch5MPpJbnt5I+9iioh7gf0y8/8DePWo4zGbVLPdte62\ntX58JbWZmWW5WJ81prd7I8K9921Sfn/83kwmH0FYo3zefXv5vTEniBZxPZjum6RaTOWrrrs+2O22\nl+cE0SLeWbOuGZejjDHYhFo4QZiZWZYThJmZZfksJhuprvdVm00SH0HYyI1Lv7XZuHOCaBF/Z3af\nazF1k9tenhNEi7gejLXVuJzO2o/bXp4ThJnNyd2Ck8kJwszMspwgrDXGtfvCrKucIKw13IVh1i5O\nEC3iejDdN0m1mMaJ216eE0SLeAfarBlue3lOEGZmluUEYWZmWU4QZmaW5QRhZmZZThAt4oGy7nMt\npm5y28tzgmiRcaoH03s7ynGu42P9deX9H6e2N0xOEFabcv0e1/KZXH7vu8sJwszMsnxHORuaNnch\n2PD4fZ4cThA2UO+XQcTgIdhyl5KNp37vcfmzMtfnxLrBXUwt0tZ6MO5Drm6SazF1+XPS1rbXNCeI\nFulo2zLrPLe9PCcIMzPL8hiELYoHLM3Gl48gbNG63PdsZv05Qdi8tP2KWDMbHieIFunCTriPFAZz\nLaZu8sc6zwmiRVwPxiZNW2o1ue3leZDazEaqNxH44sr28hGEmY2cT2zoBicIMxsZn+DQLU4QZjYy\nPmroFieIFnE9mO6b5FpMXea2l+dB6hbxzpVNuqYqwrrt5TlB2IK4L9nq4DOa2sUJwp6mype/G7LZ\n+HOCsCwnAFsMH2GOByeIMVLl7m9uuDYK3sEYDz6LqUXmaktVyhJUuQCpvEy/aVsY12Iavt7PfR2l\nOfyxz2tVgpB0mKQ7Ja2T9JGm4xm1KvVg5vMl3oYaN2YLVf7s9tuRGdZn3LWY8lqTICRtDvwFcBiw\nDDha0vObjWrUZoa6tlxDasq9997b2N8eBW/f8M33aHhxn/GZBf7eeGtNggAOAO6JiPUR8SjwReD1\nDcc0YjO1rbnp7qP169c39rdHwdvXvMV9vmeGFMV4aVOC2B24r/R8Q5rXmDVr1rBs2TL23Xdf9t13\nXz74wQ82GY6ZVTBozGIxRxltKU0+Sm06i6l1Y3uPPPII69atY9ttt+WnP/3pUA6zh3mm0aR8SM3m\nY64TMKanpzNtZzmS5rx6ez5nZ1Vp68P8vTqoyT9eJumlwHREHJaenwI8HhEfLy3TjmDNzDomIua9\nR9mmBLEFcBdwMPCvwI3A0RHxvUYDMzObUK3pYoqIxyS9D7gc2Bz4rJODmVlzWnMEYWZm7dKms5ie\nRtKOklZJulvSFZK277Pc9pIulvQ9SWvTeEarVd22tOzmkm6V9NVRxrgYVbZP0h6SVkv6rqQ7JJ3U\nRKzzUeViTkl/ll6/TdL+o45xMebaPklvS9t1u6RvSfofTcS5EFUvxJX0YkmPSXrjKONbrIqfzan0\nXXKHpJk5VxoRrX0AZwEfTtMfAc7ss9xK4Lg0vQXwrKZjH9a2pdd/B/g8cGnTcQ9z+4BdgP3S9DMp\nxqCe33TsA7Zpc+AeYCmwJbCmN17gtcDX0vRLgOubjnvI2/ey2fZFcVFrJ7avyraVlrsauAw4qum4\nh/zebQ98F1iSnu8013pbfQQBHEHx5U/6eWTvApKeBbwiIs6FYiwjIh4aXYgLNue2AUhaQvGl8xno\n1M3K5ty+iNgYEWvS9MPA94DdRhbh/FW5mPOJ7Y6IG4DtJe082jAXbM7ti4hvl9rXDcCSEce4UFUv\nxP1t4GLgB6MMbgiqbN8xwJcjYgNARDww10rbniB2johNaXoTkGtoewE/kHSepFsknSNpm9GFuGBV\ntg3gT4DfBR4fSVTDU3X7AJC0FNif4kunrapczJlbpitfovO9WPV44Gu1RjQ8c26bpN0pvlQ/nWZ1\naYC2ynu3N7Bj6ta9WdI75lpp42cxSVpF0dXQ69Tyk4iIPtdBbAG8CHhfRNwk6WzgZOC0oQc7T4vd\nNkmvA+6PiFslTdUT5cIN4b2bXc8zKfba3p+OJNqq6hdG75FeV75oKscp6ZXAccCB9YUzVFW27Wzg\n5PR5Fd06Yq+yfVtSfFceDGwDfFvS9RGxrt8vNJ4gIuKQfq9J2iRpl4jYKGlX4P7MYhuADRFxU3p+\nMUWCaNwQtu3XgCMkvRbYGthO0t9GxG/VFPK8DGH7kLQl8GXggoi4pKZQh+X7wB6l53tQfP4GLbMk\nzeuCKttHGpg+BzgsIh4cUWyLVWXbfgX4YrqSeSfgcEmPRsSlowlxUaps333AAxHxY+DHkr4JvBDo\nmyDa3sV0KXBsmj4WeNoXSERsBO6TtE+a9WqKgZi2q7JtH42IPSJiL+CtwNVtSQ4VzLl9aS/ts8Da\niDh7hLEt1M3A3pKWStoKeAvFdpZdCvwWPFEd4Ielrra2m3P7JD0b+Hvg7RFxTwMxLtSc2xYRvxQR\ne6X2djHw3o4kB6j22fwK8PJ0VuQ2FCdRrB241qZH3+cYmd8RuBK4G7gC2D7N3w34h9JyLwRuAm6j\n+PB24SymSttWWv4gunUW05zbB7ycYmxlDXBrehzWdOxzbNfhFGdb3QOckua9B3hPaZm/SK/fBryo\n6ZiHuX0UJ0v8e+n9urHpmIf53pWWPQ94Y9MxD3v7gA9R7EB/BzhprnX6QjkzM8tqexeTmZk1xAnC\nzMyynCDMzCzLCcLMzLKcIMzMLMsJwszMspwgbGhSKeF/SbVeVkvabkjrXSrp/Hksf5ekqyVdL+nI\nNO+dqRTybGwv7pn3D2m5d0u6SdLhw4g9E9uMpD8rPb9M0nkDlr92Huv+XNrm6yT9fpq3NF3VPrvd\n75a0Z8+8fSS9KNUyO2VxW2jjpPFSGzZWAlgZEU3Xwbo/Il4laWvgGxRXcQdwVqSqvwCSlvXOA34T\nODAiflpaTjG8C4aCVLxP0rbAdvQpQ7LAdR8TEf+s4h4c26b5V0TEE4XZUmHE3nkfpbi46vLSvM0i\nomtFIm2IfARhw/aUAmeSTkt7qVelPdelkq6V9KW0x/pGSZenedtI2lLSlZKuUXETqM161ve69Nq3\nJL1mjli2BR7tF1vvPElvoiibfLmk56i4Mc75wIclvSRtx3WS3lmK5WZJ50q6rvq/iJskHUC6d8Rs\nDJLOTkcY35RUrquDpOem/9OMpFMz6ywtqs2Bn6NiBWBJzwVOAD4u6aj0Nz4OrJS0k6SvpCOyv0zL\n75WOVL6S3os957Ht1iVNXx7ux/g8gClgPbAaOBd4AfBX6bXnA38F7AnckuYdDfx9mj4FeEOa3jr9\nPJ2ittaewPkUX6SrKXZstgKu7BPHncAM8EPgkDTvnWn+6vTYo2feH6TlVgObpekHgJ9P09+guKmR\ngFUUlTGvo6iKuQRYV/F/tBrYFziT4r4RLwDOS6/N/q2Dgd9P09emn38H7J6mvzA73bPu84Drgf8L\nnJHmLQU2lrb7den/OTtvtuzJcuBVpRhfkqb/GHhpmj4TeCnwlxQ3DtocuAN4dtOfPT/qebiLyYYp\ngPMj4mMAkt4MTElanV7/1/Rzbel5eXoHSc8AzpG0G8U9JO7myWqTO1EkmqvS81/sE8cPImJK0sEU\ne8aryHcxPW1ej7uiqHwJRb2v2Vu+/kL6249HxCPAI5KecoMZSW+nuF/C1yPirJ71rgN+GfgR8J+l\n+R+R9CqK5NNbRO15wAUqKo0+i6KmVa5K7DHAw8AVkma7+lbF07uYnjJv9qXS9D+ln/sCZ6b/1TOA\nG4FfAm6NiJ9Jup38kZmNAScIq9NdFH3dJwFI2oJib3u2P188tY69gNdQfDEfkwZay11MD1AUGXtN\nRDye1tdXRFyVurj+W2n9vQZ9uZW7aG4F3hQRj0jaIiIek7SZiqqYO9KTrCLiAuCCAeu+mNLYg6Qd\ngYMi4tclHULxRV92J/CBKMqnb0b/+v+KiPslfQ34DeCWATEMMrvtd1GUYr8lxbk5xVHdfpJuojgC\nckG3MeUEYcP2xJdFRNwmaWM6ggjgQorKrlFaNnp+9wbgVEm/CjxEcQSRVhch6ZPAVWmPdi3wvkEx\nAJ+jKDd+P/C7ac8e4A8yyw5az3Lgqyp24f8DeBPwceCbFF/CGwes52nrjYiVAKn/PoAHgYclXQXc\nnonrVOBcST9HMa5yFPBfA2L+DPCnKbZDSkdxl1Ekp2xcmXlnAH+j4ta+jwPvBj5B0c21iWK7fQQx\nplzN1WwIJK2OiFc2HceoqThFdzoi/qXpWGz4fARhnSbpDIoB01mfjoiLGghlpHtaaXznvaVZ346I\nj44yhhLvZY4pH0GYmVmWr4MwM7MsJwgzM8tygjAzsywnCDMzy3KCMDOzLCcIMzPL+v/v4U5MAjJf\nmAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fd619b791d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "\n",
    "## Open file for all frequency and locus info for all kept genotypes\n",
    "all_frequencies = open(\"%s%s\" %(myvcfpath, \".all_frequencies.tsv\"), 'w')\n",
    "all_frequencies.write(\"#Locus_ID\\tPOS\\tN_MaleChroms\\tN_FemChroms\\tMalefreq_REF\\tMalefreq_ALT\\tFemalefreq_REF\\tFemalefreq_ALT\\tFemREF-MaleREF\\tSex_linked\\n\")\n",
    "\n",
    "## Open files to output X or Z linked loci to and add headers\n",
    "\n",
    "Putative_Xlinked_makers = []\n",
    "\n",
    "Putative_Zlinked_makers = []\n",
    "\n",
    "## Make a log file for all auxillary info\n",
    "freq_ratios_log = open(\"%s%s\" %(myvcfpath, \".freq_ratios.log\"), 'w')\n",
    "freq_ratios_log.write(\"Script run on %s\\n \" % (time.strftime(\"%c\")))\n",
    "\n",
    "## Make list to keep frequency information in for histogram\n",
    "freq_ratio_data = []\n",
    "\n",
    "## get male/female info for samples for working out frequencies below38      3460    50      26      0.940   0.060   0.846   0.154   -0.094  NotSexLinked\n",
    "\n",
    "pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "sample_dict = {}\n",
    "sample_dict[\"females\"] = []\n",
    "sample_dict[\"males\"] = []\n",
    "fem_samples = 0\n",
    "male_samples = 0\n",
    "sample_counter = 0\n",
    "low_maf_counter = 0\n",
    "kept_loci = 0\n",
    "\n",
    "for sample in pop_map:\n",
    "    name = sample.split()[0]\n",
    "    sex = sample.strip().split()[1]\n",
    "    #sample_counter += 1\n",
    "    if sex == \"F\":\n",
    "        sample_dict[\"females\"].append(name)\n",
    "        fem_samples += 1\n",
    "    elif sex == \"M\":\n",
    "        sample_dict[\"males\"].append(name)\n",
    "        male_samples += 1\n",
    "        \n",
    "locus_dict = {}\n",
    "\n",
    "\n",
    "## Find number of loci in input file\n",
    "locus_counter = 0\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    locus_counter += 1\n",
    "print \"Number of loci = %s\" % (locus_counter)\n",
    "    \n",
    "## Write some general stats and input options to the log file\n",
    "\n",
    "freq_ratios_log.write(\"\\n## User specified options:\\n\")\n",
    "freq_ratios_log.write(\"Input vcf: %s\\n\" % (myvcfpath))\n",
    "freq_ratios_log.write(\"pop_map file used: %s\\n\" % (popmappath))\n",
    "freq_ratios_log.write(\"Catalog file used: %s\\n\" % (catalog_tags_file))\n",
    "freq_ratios_log.write(\"Female-male threshold: %s\\n\" % (X_or_Z_freq_threshold))\n",
    "freq_ratios_log.write(\"Min percentage samples present: %s\\n\" % (sample_presence_cutoff))\n",
    "freq_ratios_log.write(\"Min coverage per genotype: %s\\n\" % (coverage_threshold))\n",
    "freq_ratios_log.write(\"Min maf per locus: %s\\n\" % (maf_threshold))\n",
    "freq_ratios_log.write(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "freq_ratios_log.write(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "freq_ratios_log.write(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "                      \n",
    "  \n",
    "                      \n",
    "## add options specified here!\n",
    "\n",
    "## Some quick counters\n",
    "\n",
    "numb_putative_Xlinked = 0\n",
    "numb_putative_Zlinked = 0\n",
    "low_data_loci = 0\n",
    "\n",
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    femREF_count = 0    ## set the counters for the reference and alternative allele (encoded as 0 in the vcf)\n",
    "    femALT_count = 0\n",
    "    malREF_count = 0\n",
    "    malALT_count = 0   \n",
    "    fem_none_count = 0\n",
    "    male_none_count = 0\n",
    "    low_cov_samples = 0\n",
    "    n_genotypes = 0\n",
    "    male_genotypes = 0\n",
    "    fem_genotypes = 0\n",
    "    number_of_samples = len(record.samples)\n",
    "    \n",
    "    if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "        low_maf_counter += 1\n",
    "        freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "        freq_ratios_log.write(\"Minor allele frequence of locus is lower than specified cutoff (%s)\\n\" % (record.aaf[0]))\n",
    "        pass  \n",
    "    \n",
    "    elif record.aaf >= maf_threshold:\n",
    "    \n",
    "        for sample in record.samples: \n",
    "            \n",
    "            name = sample.sample\n",
    "            \n",
    "            ### For each sample, if the coverage is too low, remove the genotype for that individual.\n",
    "            \n",
    "            \n",
    "            if sample['DP'] < coverage_threshold:\n",
    "                genotype = None\n",
    "                low_cov_samples += 1\n",
    "                #freq_ratios_log.write(\"Sample %s thrown out due to low coverage (%s)\\n\" % (name, sample['DP']))\n",
    "            elif sample['DP'] >= coverage_threshold:\n",
    "                genotype = sample['GT']\n",
    "            \n",
    "            ## Now calculate the female and male frequencies separately\n",
    "            \n",
    "            if name in sample_dict[\"females\"]: \n",
    "                #print \"Locus=\", record.ID, \"Female=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                #print \"FEMALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    femREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    femALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                #print \"\\tN_REF=\", femREF_count, \"N_ALT=\", femALT_count, \"Nnone=\", fem_none_count\n",
    "            elif name in sample_dict[\"males\"]:\n",
    "                #print \"Locus=\", record.ID, \"Male=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                #print \"MALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    malREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    malALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "            #else:\n",
    "                #print \"\\n##SAMPLE NAME NOT IN POP_MAP.TXT: Sample = %s\" % (name)\n",
    "                #print \"\\tN_REF=\", malREF_count, \"N_ALT=\", malALT_count, \"Nnone=\", male_none_count\n",
    "                \n",
    "        ## Filter loci that have too many missing samples, including samples thrown out due to low coverage!\n",
    "        samples_at_locus = n_genotypes\n",
    "        chromosomes_at_locus = n_genotypes*2\n",
    "        percent_samples_present = n_genotypes/number_of_samples\n",
    "        #print \"Locus\", record.ID\n",
    "        #print samples_at_locus\n",
    "        #print percent_samples_present\n",
    "        \n",
    "        if percent_samples_present >= sample_presence_cutoff:\n",
    "            kept_loci += 1\n",
    "            ## Calculate frequencies\n",
    "            \n",
    "            femREF_freq = femREF_count/(fem_genotypes*2)\n",
    "            femALT_freq = femALT_count/(fem_genotypes*2)\n",
    "            \n",
    "            maleREF_freq = malREF_count/(male_genotypes*2)\n",
    "            maleALT_freq = malALT_count/(male_genotypes*2)\n",
    "                                \n",
    "                        \n",
    "            ## Output female stats\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s\\n\\n\" %(record.ID))   \n",
    "            freq_ratios_log.write(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "            freq_ratios_log.write(\"Female reference count = %s\\n\" % (femREF_count))\n",
    "            freq_ratios_log.write(\"Female alternative count = %s\\n\" % (femALT_count))\n",
    "            freq_ratios_log.write(\"Female reference frequency = %.3f\\n\" % (femREF_freq))\n",
    "            freq_ratios_log.write(\"Female alternative frequency = %.3f\\n\" % (femALT_freq))\n",
    "                                  \n",
    "            ## check fem freqs\n",
    "            if not (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (femREF_freq + femALT_freq))\n",
    "        \n",
    "            ## Output male stats\n",
    "            freq_ratios_log.write(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "            freq_ratios_log.write(\"Male reference count = %s\\n\" % (malREF_count))\n",
    "            freq_ratios_log.write(\"Male alternative count = %s\\n\" % (malALT_count))\n",
    "            freq_ratios_log.write(\"Male reference frequency = %.3f\\n\" % (maleREF_freq))\n",
    "            freq_ratios_log.write(\"Male alternative frequency = %.3f\\n\" % (maleALT_freq))\n",
    "                                  \n",
    "            ## check male freqs\n",
    "            if not (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (maleREF_freq + maleALT_freq))\n",
    "        \n",
    "            locus_dict[record.ID] = {}\n",
    "            locus_dict[record.ID][\"female_freqs\"] = [\"%.3f\" % (femREF_freq), \"%.3f\" % (femALT_freq)]\n",
    "            locus_dict[record.ID][\"male_freqs\"] = [\"%.3f\" % (maleREF_freq), \"%.3f\" % (maleALT_freq)]\n",
    "            \n",
    "\n",
    "            ####### ==============================================================================================\n",
    "            ### So now I have the allele frequencies for males and females, I can subtract them and see if the distribution fits the expectition for X linked or Z linked!\n",
    "            \n",
    "            freq_ratio = femREF_freq - maleREF_freq\n",
    "            freq_ratio_data.append(freq_ratio)\n",
    "            \n",
    "\n",
    "            ## Write files for X or Z linked loci\n",
    "            \n",
    "            if freq_ratio > X_or_Z_freq_threshold and femREF_freq >= 0.95:  ## for X linked\n",
    "                linked_status = \"Xlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Xlinked += 1\n",
    "            \n",
    "            elif freq_ratio < -X_or_Z_freq_threshold and maleREF_freq >= 0.95:  ## for Z linked\n",
    "                linked_status = \"Zlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Zlinked += 1\n",
    "            else:\n",
    "                freq_ratios_log.write(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                linked_status =\"NotSexLinked\"\n",
    "        \n",
    "            ## Write the main info file for male and female frequencies, ratios etc.\n",
    "            \n",
    "            all_frequencies.write(\"%s\\t%s\\t%s\\t%s\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%s\\n\" % (record.ID, record.POS, male_genotypes*2, fem_genotypes*2, maleREF_freq, maleALT_freq, femREF_freq, femALT_freq, freq_ratio, linked_status))\n",
    "            \n",
    "        \n",
    "        elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            freq_ratios_log.write(\"Number of samples at locus is lower than sample presence cutoff (%s)\\n\" % (samples_at_locus))\n",
    "            low_data_loci += 1\n",
    "    \n",
    "    \n",
    "## Print some quick summary stats - also written at the end of the log file\n",
    "print \"Number of samples =\", number_of_samples\n",
    "print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "print \"Number of putative X linked markers = %s\" % (numb_putative_Xlinked)\n",
    "print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "\n",
    "\n",
    "freq_ratios_log.write(\"\\nSUMMARY....\\n\\n\")\n",
    "freq_ratios_log.write(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "freq_ratios_log.write(\"Number of loci with enough data = %s\\n\" % (locus_counter - low_data_loci))\n",
    "freq_ratios_log.write(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "freq_ratios_log.write(\"Number of putative X linked markers = %s\\n\" % (numb_putative_Xlinked))\n",
    "freq_ratios_log.write(\"Number of putative Z linked markers = %s\\n\" % (numb_putative_Zlinked))\n",
    "\n",
    "\n",
    "\n",
    "## plot histogram of frequency ratios\n",
    "\n",
    "plt.hist(freq_ratio_data, bins = 100, color = '0.5')\n",
    "plt.xlabel('Female_REFfreq - Male_REFfreq', fontsize=8)\n",
    "plt.ylabel(\"Number of SNPs\", fontsize=8)\n",
    "plt.axvline(x = np.mean(freq_ratio_data), color='r', linestyle='dashed', linewidth = 2)\n",
    "plt.axvline(x = X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "plt.axvline(x = -X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "plt.savefig(\"%s%s\" %(myvcfpath, \".fem-male_freqs.pdf\"), format = 'pdf')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "## Write fasta files of putative X or Z linked loci if there are any\n",
    "\n",
    "if catalog_tags_file.endswith(\"gz\"):\n",
    "    catalog = gzip.open(catalog_tags_file, 'r').readlines()\n",
    "else:\n",
    "    catalog = open(catalog_tags_file, 'r').readlines()\n",
    "\n",
    "if numb_putative_Xlinked > 0:\n",
    "    Putative_Xlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Xlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Xlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Xlinked_makers_file.write(\">LocusID_%s\\n\" % (locus))\n",
    "                Putative_Xlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Xlinked_makers_file.close()\n",
    "\n",
    "if numb_putative_Zlinked > 0:\n",
    "    Putative_Zlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Zlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Zlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Zlinked_makers_file.write(\">LocusID_%s\\n\" % (locus))\n",
    "                Putative_Zlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Zlinked_makers_file.close()\n",
    "\n",
    "\n",
    "## close all unclosed files \n",
    "freq_ratios_log.close()\n",
    "all_frequencies.close()\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
