{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "import MISC_RAD_tools as MISC\n",
    "import shutil\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Nan_chunks=\"/home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei/R_latastei_Rtemp_to_Nan_8kchunks.fa\"\n",
    "Xen_db = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Xenopus_genome/Xenoupus_blastdb\"\n",
    "Xenblastouts=\"/home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei/R_latastei_Rtemp_to_Nan_8kchunks_to_Xen_blastn_test.xml\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class NcbiblastxCommandline in module Bio.Blast.Applications:\n",
      "\n",
      "class NcbiblastxCommandline(_NcbiblastMain2SeqCommandline)\n",
      " |  Wrapper for the NCBI BLAST+ program blastx (nucleotide query, protein database).\n",
      " |  \n",
      " |  With the release of BLAST+ (BLAST rewritten in C++ instead of C), the NCBI\n",
      " |  replaced the old blastall tool with separate tools for each of the searches.\n",
      " |  This wrapper therefore replaces BlastallCommandline with option -p blastx.\n",
      " |  \n",
      " |  >>> from Bio.Blast.Applications import NcbiblastxCommandline\n",
      " |  >>> cline = NcbiblastxCommandline(query=\"m_cold.fasta\", db=\"nr\", evalue=0.001)\n",
      " |  >>> cline\n",
      " |  NcbiblastxCommandline(cmd='blastx', query='m_cold.fasta', db='nr', evalue=0.001)\n",
      " |  >>> print(cline)\n",
      " |  blastx -query m_cold.fasta -db nr -evalue 0.001\n",
      " |  \n",
      " |  You would typically run the command line with cline() or via the Python\n",
      " |  subprocess module, as described in the Biopython tutorial.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      NcbiblastxCommandline\n",
      " |      _NcbiblastMain2SeqCommandline\n",
      " |      _Ncbiblast2SeqCommandline\n",
      " |      _NcbiblastCommandline\n",
      " |      _NcbibaseblastCommandline\n",
      " |      Bio.Application.AbstractCommandline\n",
      " |      __builtin__.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, cmd='blastx', **kwargs)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  best_hit_overhang\n",
      " |      Best Hit algorithm overhang value (float, recommended value: 0.1)\n",
      " |      \n",
      " |                          Float between 0.0 and 0.5 inclusive.\n",
      " |      \n",
      " |                          Incompatible with: culling_limit.\n",
      " |      \n",
      " |      This controls the addition of the -best_hit_overhang parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  best_hit_score_edge\n",
      " |      Best Hit algorithm score edge value (float, recommended value: 0.1)\n",
      " |      \n",
      " |                          Float between 0.0 and 0.5 inclusive.\n",
      " |      \n",
      " |                          Incompatible with: culling_limit.\n",
      " |      \n",
      " |      This controls the addition of the -best_hit_score_edge parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  comp_based_stats\n",
      " |      Use composition-based statistics for blastp, blastx, or tblastn:\n",
      " |      \n",
      " |                          D or d: default (equivalent to 2 )\n",
      " |                          0 or F or f: no composition-based statistics\n",
      " |                          1: Composition-based statistics as in NAR 29:2994-3005, 2001\n",
      " |                          2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties\n",
      " |                          3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally\n",
      " |      \n",
      " |                          For programs other than tblastn, must either be absent or be D, F or 0\n",
      " |                          Default = `2'\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the -comp_based_stats parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  culling_limit\n",
      " |      Hit culling limit (integer).\n",
      " |      \n",
      " |                          If the query range of a hit is enveloped by that of at\n",
      " |                          least this many higher-scoring hits, delete the hit.\n",
      " |      \n",
      " |                          Incompatible with: best_hit_overhang, best_hit_score_edge.\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the -culling_limit parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db\n",
      " |      The database to BLAST against.\n",
      " |      \n",
      " |      This controls the addition of the -db parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db_hard_mask\n",
      " |      Filtering algorithm for hard masking (integer).\n",
      " |      \n",
      " |                          Filtering algorithm ID to apply to the BLAST database as hard masking.\n",
      " |      \n",
      " |                          Incompatible with: db_soft_mask, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -db_hard_mask parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  db_soft_mask\n",
      " |      Filtering algorithm for soft masking (integer).\n",
      " |      \n",
      " |                          Filtering algorithm ID to apply to the BLAST database as soft masking.\n",
      " |      \n",
      " |                          Incompatible with: db_hard_mask, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -db_soft_mask parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  dbsize\n",
      " |      Effective length of the database (integer).\n",
      " |      \n",
      " |      This controls the addition of the -dbsize parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  entrez_query\n",
      " |      Restrict search with the given Entrez query (requires remote).\n",
      " |      \n",
      " |      This controls the addition of the -entrez_query parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  evalue\n",
      " |      Expectation value cutoff.\n",
      " |      \n",
      " |      This controls the addition of the -evalue parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  export_search_strategy\n",
      " |      File name to record the search strategy used.\n",
      " |      \n",
      " |                          Incompatible with: import_search_strategy\n",
      " |      \n",
      " |      This controls the addition of the -export_search_strategy parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  frame_shift_penalty\n",
      " |      Frame shift penalty (integer, at least 1, default ignored) (OBSOLETE).\n",
      " |      \n",
      " |                          This was removed in BLAST 2.2.27+\n",
      " |      \n",
      " |      This controls the addition of the -frame_shift_penalty parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapextend\n",
      " |      Cost to extend a gap (integer).\n",
      " |      \n",
      " |      This controls the addition of the -gapextend parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gapopen\n",
      " |      Cost to open a gap (integer).\n",
      " |      \n",
      " |      This controls the addition of the -gapopen parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  gilist\n",
      " |      Restrict search of database to list of GI's.\n",
      " |      \n",
      " |                          Incompatible with: negative_gilist, seqidlist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -gilist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  h\n",
      " |      Print USAGE and DESCRIPTION;  ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -h switch, treat this property as a boolean.\n",
      " |  \n",
      " |  help\n",
      " |      Print USAGE, DESCRIPTION and ARGUMENTS description; ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -help switch, treat this property as a boolean.\n",
      " |  \n",
      " |  html\n",
      " |      Produce HTML output? See also the outfmt option.\n",
      " |      \n",
      " |      This property controls the addition of the -html switch, treat this property as a boolean.\n",
      " |  \n",
      " |  import_search_strategy\n",
      " |      Search strategy to use.\n",
      " |      \n",
      " |                          Incompatible with: export_search_strategy\n",
      " |      \n",
      " |      This controls the addition of the -import_search_strategy parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  lcase_masking\n",
      " |      Use lower case filtering in query and subject sequence(s)?\n",
      " |      \n",
      " |      This property controls the addition of the -lcase_masking switch, treat this property as a boolean.\n",
      " |  \n",
      " |  line_length\n",
      " |      Line length for formatting alignments (integer, at least 1, default 60).\n",
      " |      \n",
      " |                          Not applicable for outfmt > 4.\n",
      " |      \n",
      " |                          Added in BLAST+ 2.2.30.\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the -line_length parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  matrix\n",
      " |      Scoring matrix name (default BLOSUM62).\n",
      " |      \n",
      " |      This controls the addition of the -matrix parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_hsps\n",
      " |      Set maximum number of HSPs per subject sequence to save (default 0 means no limit).\n",
      " |      \n",
      " |      This controls the addition of the -max_hsps parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_hsps_per_subject\n",
      " |      Override maximum number of HSPs per subject to save for ungapped searches (integer).\n",
      " |      \n",
      " |      This controls the addition of the -max_hsps_per_subject parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_intron_length\n",
      " |      Maximum intron length (integer).\n",
      " |      \n",
      " |                          Length of the largest intron allowed in a translated nucleotide\n",
      " |                          sequence when linking multiple distinct alignments (a negative\n",
      " |                          value disables linking). Default zero.\n",
      " |      \n",
      " |      This controls the addition of the -max_intron_length parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  max_target_seqs\n",
      " |      Maximum number of aligned sequences to keep (integer, at least one).\n",
      " |      \n",
      " |      This controls the addition of the -max_target_seqs parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  negative_gilist\n",
      " |      Restrict search of database to everything except the listed GIs.\n",
      " |      \n",
      " |                          Incompatible with: gilist, seqidlist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -negative_gilist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  num_alignments\n",
      " |      Number of database sequences to show num_alignments for.\n",
      " |      \n",
      " |                          Integer argument (at least zero). Default is 200.\n",
      " |                          See also num_alignments.\n",
      " |      \n",
      " |      This controls the addition of the -num_alignments parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  num_descriptions\n",
      " |      Number of database sequences to show one-line descriptions for.\n",
      " |      \n",
      " |                          Integer argument (at least zero). Default is 500.\n",
      " |                          See also num_alignments.\n",
      " |      \n",
      " |      This controls the addition of the -num_descriptions parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  num_threads\n",
      " |      Number of threads to use in the BLAST search (integer, at least one).\n",
      " |      \n",
      " |                          Default is one.\n",
      " |                          Incompatible with: remote\n",
      " |      \n",
      " |      This controls the addition of the -num_threads parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  out\n",
      " |      Output file for alignment.\n",
      " |      \n",
      " |      This controls the addition of the -out parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  outfmt\n",
      " |      Alignment view.  Integer 0-11.  Use 5 for XML output (differs from classic BLAST which used 7 for XML).\n",
      " |      \n",
      " |      This controls the addition of the -outfmt parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  parse_deflines\n",
      " |      Should the query and subject defline(s) be parsed?\n",
      " |      \n",
      " |      This property controls the addition of the -parse_deflines switch, treat this property as a boolean.\n",
      " |  \n",
      " |  qcov_hsp_perc\n",
      " |      Percent query coverage per hsp (float, 0 to 100).\n",
      " |      \n",
      " |                          Added in BLAST+ 2.2.30.\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the -qcov_hsp_perc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  query\n",
      " |      The sequence to search with.\n",
      " |      \n",
      " |      This controls the addition of the -query parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  query_gencode\n",
      " |      Genetic code to use to translate query (integer, default 1).\n",
      " |      \n",
      " |      This controls the addition of the -query_gencode parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  query_loc\n",
      " |      Location on the query sequence (Format: start-stop)\n",
      " |      \n",
      " |      This controls the addition of the -query_loc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  remote\n",
      " |      Execute search remotely?\n",
      " |      \n",
      " |                          Incompatible with: gilist, negative_gilist, subject_loc, num_threads, ...\n",
      " |      \n",
      " |      This property controls the addition of the -remote switch, treat this property as a boolean.\n",
      " |  \n",
      " |  searchsp\n",
      " |      Effective length of the search space (integer).\n",
      " |      \n",
      " |      This controls the addition of the -searchsp parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seg\n",
      " |      Filter query sequence with SEG (string).\n",
      " |      \n",
      " |                          Format: \"yes\", \"window locut hicut\", or \"no\" to disable.\n",
      " |                          Default is \"12 2.2 2.5\n",
      " |      \n",
      " |      This controls the addition of the -seg parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  seqidlist\n",
      " |      Restrict search of database to list of SeqID's.\n",
      " |      \n",
      " |                          Incompatible with: gilist, negative_gilist, remote, subject, subject_loc\n",
      " |      \n",
      " |      This controls the addition of the -seqidlist parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  show_gis\n",
      " |      Show NCBI GIs in deflines?\n",
      " |      \n",
      " |      This property controls the addition of the -show_gis switch, treat this property as a boolean.\n",
      " |  \n",
      " |  soft_masking\n",
      " |      Apply filtering locations as soft masks (Boolean, Default = true)\n",
      " |      \n",
      " |      This controls the addition of the -soft_masking parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  strand\n",
      " |      Query strand(s) to search against database/subject.\n",
      " |      \n",
      " |                          Values allowed are \"both\" (default), \"minus\", \"plus\".\n",
      " |      \n",
      " |      This controls the addition of the -strand parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  subject\n",
      " |      Subject sequence(s) to search.\n",
      " |      \n",
      " |                          Incompatible with: db, gilist, negative_gilist.\n",
      " |                          See also subject_loc.\n",
      " |      \n",
      " |      This controls the addition of the -subject parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  subject_loc\n",
      " |      Location on the subject sequence (Format: start-stop).\n",
      " |      \n",
      " |                          Incompatible with: db, gilist, seqidlist, negative_gilist,\n",
      " |                          db_soft_mask, db_hard_mask, remote.\n",
      " |      \n",
      " |                          See also subject.\n",
      " |      \n",
      " |      This controls the addition of the -subject_loc parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  sum_statistics\n",
      " |      Use sum statistics.\n",
      " |      \n",
      " |      This property controls the addition of the -sum_statistics switch, treat this property as a boolean.\n",
      " |  \n",
      " |  sum_stats\n",
      " |      Use sum statistics (boolean).\n",
      " |      \n",
      " |                          Added in BLAST+ 2.2.30.\n",
      " |                          \n",
      " |      \n",
      " |      This controls the addition of the -sum_stats parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  task\n",
      " |      Task to execute (string, blastx (default) or blastx-fast).\n",
      " |      \n",
      " |      This controls the addition of the -task parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  threshold\n",
      " |      Minimum score for words to be added to the BLAST lookup table (float).\n",
      " |      \n",
      " |      This controls the addition of the -threshold parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ungapped\n",
      " |      Perform ungapped alignment only?\n",
      " |      \n",
      " |      This property controls the addition of the -ungapped switch, treat this property as a boolean.\n",
      " |  \n",
      " |  use_sw_tback\n",
      " |      Compute locally optimal Smith-Waterman alignments?\n",
      " |      \n",
      " |      This property controls the addition of the -use_sw_tback switch, treat this property as a boolean.\n",
      " |  \n",
      " |  version\n",
      " |      Print version number;  ignore other arguments.\n",
      " |      \n",
      " |      This property controls the addition of the -version switch, treat this property as a boolean.\n",
      " |  \n",
      " |  window_size\n",
      " |      Multiple hits window size, use 0 to specify 1-hit algorithm (integer).\n",
      " |      \n",
      " |      This controls the addition of the -window_size parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  word_size\n",
      " |      Word size for wordfinder algorithm.\n",
      " |      \n",
      " |                          Integer. Minimum 2.\n",
      " |      \n",
      " |      This controls the addition of the -word_size parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_gap\n",
      " |      X-dropoff value (in bits) for preliminary gapped extensions (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_gap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_gap_final\n",
      " |      X-dropoff value (in bits) for final gapped alignment (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_gap_final parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  xdrop_ungap\n",
      " |      X-dropoff value (in bits) for ungapped extensions (float).\n",
      " |      \n",
      " |      This controls the addition of the -xdrop_ungap parameter and its associated value.  Set this property to the argument value required.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __call__(self, stdin=None, stdout=True, stderr=True, cwd=None, env=None)\n",
      " |      Executes the command, waits for it to finish, and returns output.\n",
      " |      \n",
      " |      Runs the command line tool and waits for it to finish. If it returns\n",
      " |      a non-zero error level, an exception is raised. Otherwise two strings\n",
      " |      are returned containing stdout and stderr.\n",
      " |      \n",
      " |      The optional stdin argument should be a string of data which will be\n",
      " |      passed to the tool as standard input.\n",
      " |      \n",
      " |      The optional stdout and stderr argument may be filenames (string),\n",
      " |      but otherwise are treated as a booleans, and control if the output\n",
      " |      should be captured as strings (True, default), or ignored by sending\n",
      " |      it to /dev/null to avoid wasting memory (False). If sent to a file\n",
      " |      or ignored, then empty string(s) are returned.\n",
      " |      \n",
      " |      The optional cwd argument is a string giving the working directory\n",
      " |      to run the command from. See Python's subprocess module documentation\n",
      " |      for more details.\n",
      " |      \n",
      " |      The optional env argument is a dictionary setting the environment\n",
      " |      variables to be used in the new process. By default the current\n",
      " |      process' environment variables are used. See Python's subprocess\n",
      " |      module documentation for more details.\n",
      " |      \n",
      " |      Default example usage::\n",
      " |      \n",
      " |          from Bio.Emboss.Applications import WaterCommandline\n",
      " |          water_cmd = WaterCommandline(gapopen=10, gapextend=0.5,\n",
      " |                                       stdout=True, auto=True,\n",
      " |                                       asequence=\"a.fasta\", bsequence=\"b.fasta\")\n",
      " |          print(\"About to run: %s\" % water_cmd)\n",
      " |          std_output, err_output = water_cmd()\n",
      " |      \n",
      " |      This functionality is similar to subprocess.check_output() added in\n",
      " |      Python 2.7. In general if you require more control over running the\n",
      " |      command, use subprocess directly.\n",
      " |      \n",
      " |      As of Biopython 1.56, when the program called returns a non-zero error\n",
      " |      level, a custom ApplicationError exception is raised. This includes\n",
      " |      any stdout and stderr strings captured as attributes of the exception\n",
      " |      object, since they may be useful for diagnosing what went wrong.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the command line object for debugging.\n",
      " |      \n",
      " |      e.g.\n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> cline\n",
      " |      WaterCommandline(cmd='water', outfile='temp_water.txt', asequence='asis:ACCCGGGCGCGGT', bsequence='asis:ACCCGAGCGCGGT', gapopen=10, gapextend=0.5)\n",
      " |  \n",
      " |  __setattr__(self, name, value)\n",
      " |      Set attribute name to value (PRIVATE).\n",
      " |      \n",
      " |      This code implements a workaround for a user interface issue.\n",
      " |      Without this __setattr__ attribute-based assignment of parameters\n",
      " |      will silently accept invalid parameters, leading to known instances\n",
      " |      of the user assuming that parameters for the application are set,\n",
      " |      when they are not.\n",
      " |      \n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5, stdout=True)\n",
      " |      >>> cline.asequence = \"a.fasta\"\n",
      " |      >>> cline.bsequence = \"b.fasta\"\n",
      " |      >>> cline.csequence = \"c.fasta\"\n",
      " |      Traceback (most recent call last):\n",
      " |      ...\n",
      " |      ValueError: Option name csequence was not found.\n",
      " |      >>> print(cline)\n",
      " |      water -stdout -asequence=a.fasta -bsequence=b.fasta -gapopen=10 -gapextend=0.5\n",
      " |      \n",
      " |      This workaround uses a whitelist of object attributes, and sets the\n",
      " |      object attribute list as normal, for these.  Other attributes are\n",
      " |      assumed to be parameters, and passed to the self.set_parameter method\n",
      " |      for validation and assignment.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Make the commandline string with the currently set options.\n",
      " |      \n",
      " |      e.g.\n",
      " |      >>> from Bio.Emboss.Applications import WaterCommandline\n",
      " |      >>> cline = WaterCommandline(gapopen=10, gapextend=0.5)\n",
      " |      >>> cline.asequence = \"asis:ACCCGGGCGCGGT\"\n",
      " |      >>> cline.bsequence = \"asis:ACCCGAGCGCGGT\"\n",
      " |      >>> cline.outfile = \"temp_water.txt\"\n",
      " |      >>> print(cline)\n",
      " |      water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5\n",
      " |      >>> str(cline)\n",
      " |      'water -outfile=temp_water.txt -asequence=asis:ACCCGGGCGCGGT -bsequence=asis:ACCCGAGCGCGGT -gapopen=10 -gapextend=0.5'\n",
      " |  \n",
      " |  set_parameter(self, name, value=None)\n",
      " |      Set a commandline option for a program (OBSOLETE).\n",
      " |      \n",
      " |      Every parameter is available via a property and as a named\n",
      " |      keyword when creating the instance. Using either of these is\n",
      " |      preferred to this legacy set_parameter method which is now\n",
      " |      OBSOLETE, and likely to be DEPRECATED and later REMOVED in\n",
      " |      future releases.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from Bio.Application.AbstractCommandline:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(NcbiblastxCommandline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "blastn_cline = NcbiblastnCommandline(query=Nan_chunks, db=Xen_db, outfmt=5, out=Xenblastouts)\n",
    "\n",
    "stdout, stderr = blastn_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "##  Mapping tags to R. temporaria  ##\n",
      "\n",
      "Number of multi-alingments kept: 7\n",
      "Number of unique alingments kept: 7\n",
      "Getting subject scaffold segments from /home/djeffrie/Data/Genomes/Rtemp/V2/Rtemp_gapfilled_iter2_flted2.fa . . . \n",
      "14 sequence scaffold segments are in /home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei//blast_8000_chunks.fa\n",
      "\n",
      "##  Mapping Rtemp chunks Nanorana  ##\n",
      "\n",
      "Number of multi-alingments kept: 6\n",
      "Number of unique alingments kept: 1\n",
      "Getting subject scaffold segments from /home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Nanorana_blast_db/Nanorana_parkeri.genome.v2.fasta . . . \n",
      "7 sequence scaffold segments are in /home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei//blast_8000_chunks.fa\n",
      "\n",
      "##  Mapping Nanorana chunks Xenopus  ##\n",
      "\n",
      "Number of multi-alingments kept: 2\n",
      "Number of unique alingments kept: 1\n",
      "Getting subject scaffold segments from /home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Xenopus_genome/Xtropicalis.v9.repeatMasked.fa . . . \n",
      "3 sequence scaffold segments are in /home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei//blast_8000_chunks.fa\n",
      "\n",
      "##  Final Xenopus hits  ##\n",
      "\n",
      "scaffold294 Chr02\n",
      "scaffold1218 Chr02\n",
      "scaffold525 scaffold_269\n"
     ]
    }
   ],
   "source": [
    "Rtemp_fasta = \"/home/djeffrie/Data/Genomes/Rtemp/V2/Rtemp_gapfilled_iter2_flted2.fa\"\n",
    "Rtemp_db = \"/home/djeffrie/Data/Genomes/Rtemp/V2/Rtemp_gapfilled_iter2_flted2-blastn-db\"\n",
    "\n",
    "Nano_fasta = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Nanorana_blast_db/Nanorana_parkeri.genome.v2.fasta\"\n",
    "Nanorana_db = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Nanorana_blast_db/Nanorana_blast_db\"\n",
    "\n",
    "Xen_db = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Xenopus_genome/Xenoupus_blastdb\"\n",
    "Xen_fasta = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Genome_mapping/Xenopus_genome/Xtropicalis.v9.repeatMasked.fa\"\n",
    "\n",
    "\n",
    "best_hit_crit = 1e-5\n",
    "Eval_threshold = 1e-20\n",
    "Window = 8000\n",
    "\n",
    "wd = \"/home/djeffrie/Data/Mapping_SL_tags_to_genomes/Sex_linked_tags/Manual_mapping/R_latastei/\"\n",
    "\n",
    "### Tags to Rtemp\n",
    "\n",
    "print \"\\n##  Mapping tags to R. temporaria  ##\\n\"\n",
    "\n",
    "R_lat_tags =\"%s/%s\" % (wd, \"R_latastei.fa\")\n",
    "R_lat_Rtemp_blastnouts = \"%s/%s\" % (wd, \"R_lat_tags_Rtemp_blastnouts.xml\")\n",
    "\n",
    "blastn_cline = NcbiblastnCommandline(query=R_lat_tags, db=Rtemp_db, outfmt=5, out=R_lat_Rtemp_blastnouts)\n",
    "stdout, stderr = blastn_cline()\n",
    "\n",
    "Rtemp_records = MISC.BlastParseExtra(R_lat_Rtemp_blastnouts,Rtemp_fasta , best_hit_crit, Eval_threshold,Window)\n",
    "\n",
    "Rtemp_blast_chunks = \"%s/%s\" % (wd,\"R_lat_Rtemp_8k_chunks.fa\")\n",
    "\n",
    "shutil.move(\"%s/%s\" % (wd,\"blast_8000_chunks.fa\"), Rtemp_blast_chunks)\n",
    "\n",
    "\n",
    "### Rtemp chunks to Nanorana\n",
    "\n",
    "print \"\\n##  Mapping Rtemp chunks Nanorana  ##\\n\"\n",
    "\n",
    "R_lat_Rtemp_Nano_blastnouts = \"%s/%s\" % (wd, \"R_lat_Rtemp_Nano_blastnouts.xml\")\n",
    "\n",
    "blastn_cline = NcbiblastnCommandline(query=Rtemp_blast_chunks, db=Nanorana_db, outfmt=5, out=R_lat_Rtemp_Nano_blastnouts)\n",
    "stdout, stderr = blastn_cline()\n",
    "\n",
    "Nano_records = MISC.BlastParseExtra(R_lat_Rtemp_Nano_blastnouts, Nano_fasta , best_hit_crit, Eval_threshold,Window)\n",
    "\n",
    "Rtemp_Nano_blast_chunks = \"%s/%s\" % (wd,\"R_lat_Rtemp_Nano_8k_chunks.fa\")\n",
    "\n",
    "shutil.move(\"%s/%s\" % (wd,\"blast_8000_chunks.fa\"), Rtemp_Nano_blast_chunks)\n",
    "\n",
    "\n",
    "\n",
    "### Nanorana chunks to Xenopus\n",
    "\n",
    "print \"\\n##  Mapping Nanorana chunks Xenopus  ##\\n\"\n",
    "\n",
    "R_lat_Rtemp_Nano_Xen_blastnouts = \"%s/%s\" % (wd, \"R_lat_tags_Rtemp_Nano_Xen_blastnouts.xml\")\n",
    "\n",
    "blastn_cline = NcbiblastnCommandline(query=Rtemp_Nano_blast_chunks, db=Xen_db, outfmt=5, out=R_lat_Rtemp_Nano_Xen_blastnouts)\n",
    "stdout, stderr = blastn_cline()\n",
    "\n",
    "Xen_records = MISC.BlastParseExtra(R_lat_Rtemp_Nano_Xen_blastnouts, Xen_fasta , best_hit_crit, Eval_threshold,Window)\n",
    "\n",
    "shutil.move(\"%s/%s\" % (wd,\"blast_8000_chunks.fa\"), \"%s/%s\" % (wd,\"R_lat_Rtemp_Nano_Xen_8kb_chunks.fa\"))\n",
    "\n",
    "\n",
    "\n",
    "### Get the Xenopus summary\n",
    "\n",
    "print \"\\n##  Final Xenopus hits  ##\\n\"\n",
    "\n",
    "from collections import Counter \n",
    "Xen_hits = []\n",
    "for i in Xen_records:\n",
    "    Xen_hits.append(Xen_records[i][\"Ref_hit_id\"])\n",
    "    print i, Xen_records[i][\"Ref_hit_id\"]\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
