{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def SL_markers_from_sexed_fam(vcf_path, sex_info_file, catalog_path, offspring_presence_threshold, mendelian_threshold, het_sex_thresh, hom_sex_thresh):\n",
    "    \n",
    "    \"\"\"\n",
    "    Usage: MST_input_maker  <vcf_path>  <sex_info_file>  <offspring_presence_threshold>  <mendelian_threshold>\n",
    "    \n",
    "    ### Informative locus filters:\n",
    "    -Loci are used only if the number of offspring present is equal to or above the percentage specified by the user\n",
    "    -Loci are used only if they are heterozygous in one sex and homozygous in another.\n",
    "    \n",
    "    ###Null allele filters:\n",
    "\n",
    "    Null allele in Parents:\n",
    "    -If an offspring is found to have an allele that isn't present in the parent, this locus is discarded\n",
    "\n",
    "    Null allele in offspring:\n",
    "    -If only 1 offspring is homozygous for the parental minor allele then that sample is assumed to possess a null allele and the data for that individual is coded as missing. **Note, the sample presence/absence filter works after this step, so some loci may be pushed over the \"missing data\" limit by this null allele filter. \n",
    "    -If two or more samples are found to be homozygous for the minor parental allele, it is more likely that this is due to the homozygous parent having a null allele. In this case, the locus is again discarded.\n",
    "    -Also, loci with non-mendelian properties indicative of allele dropout or similar are also filtered\n",
    "    \n",
    "    \"\"\"\n",
    "    #from __future__ import division\n",
    "    import sys\n",
    "    import gzip\n",
    "    import vcf\n",
    "    from collections import Counter\n",
    "    import pprint as pp\n",
    "    from pydoc import help\n",
    "    \n",
    "\n",
    "    \n",
    "    ## Alter the header in the vcf file to be compatible with PyVCF ==========================\n",
    "    \n",
    "    alteredvcfpath = \"%s%s\" % (vcf_path, \".altered\")\n",
    "\n",
    "    oldvcf = open(vcf_path, 'r').readlines()\n",
    "    alteredvcf = open(alteredvcfpath, 'w')    \n",
    "    \n",
    "    for line in oldvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "    \n",
    "    altered_vcf = open(alteredvcfpath, 'r')\n",
    "    \n",
    "    ## ========================================================================================\n",
    "    \n",
    "    \n",
    "    #####\n",
    "    #####\n",
    "    #####\n",
    "        \n",
    "    \n",
    "    ## Read in VCF, make counters and lists, open output files ================================\n",
    "    \n",
    "    myvcf = vcf.Reader(altered_vcf)\n",
    "        \n",
    "    Working_dir = \"%s/\" % (vcf_path.rpartition(\"/\")[0])\n",
    "    \n",
    "    loci_used_for_male_map = 0\n",
    "    loci_used_for_female_map = 0\n",
    "    \n",
    "    Loc_w_null_alleles = 0\n",
    "    loc_w_excess_hom = 0\n",
    "    \n",
    "    offspring = []\n",
    "    male_offspring = []\n",
    "    female_offspring = []\n",
    "    \n",
    "    #ZW_tags = open(\"%sPutative_ZW_linked_tags.txt\" % (MST_path), 'w')\n",
    "    #XY_tags = open(\"%sPutative_XY_linked_tags.txt\" % (MST_path), 'w')\n",
    "    log_file = open(\"%sSL_markers_from_sexed_fam.log\" % (Working_dir), 'w')\n",
    "    \n",
    "    ZW_tags = []\n",
    "    XY_tags = []\n",
    "    \n",
    "    ## =========================================================================================================\n",
    "\n",
    "       \n",
    "    ## Get info for Parents and offspring ======================================================================\n",
    "    \n",
    "    sex_file = open(sex_info_file, 'r').readlines()\n",
    "        \n",
    "    for line in sex_file: ## Note, only samples in the sex_info file are used\n",
    "        \n",
    "        sample_name = line.split()[0]\n",
    "        sex = line.split()[1]                    \n",
    "        \n",
    "        if sex == \"F\" or sex == \"f\":\n",
    "            Mother = sample_name\n",
    "            log_file.write(\"Mother = %s\" % (sample_name))\n",
    "        \n",
    "        elif sex == \"M\" or sex == \"m\":\n",
    "            Father = sample_name\n",
    "            log_file.write(\"\\nFather = %s\\n\" % (sample_name))        \n",
    "    \n",
    "        elif sex == \"O\" or sex == \"o\" or sex == \"MO\" or sex == \"FO\" or sex == \"mo\" or sex == \"fo\":\n",
    "            offspring.append(sample_name)\n",
    "            if sex == \"MO\" or  sex == \"mo\":\n",
    "                male_offspring.append(sample_name)\n",
    "            elif sex == \"FO\" or  sex == \"fo\":\n",
    "                female_offspring.append(sample_name)\n",
    "            \n",
    "    offspring = sorted(offspring) ## sort the list of offspring - used to retain correct order throughout script\n",
    "    male_offspring = sorted(male_offspring)\n",
    "    female_offspring = sorted(female_offspring)\n",
    "    \n",
    "    \n",
    "    \n",
    "    ## ===========================================================================================================\n",
    "\n",
    "    \n",
    "    ## Looping over loci to get info =============================================================================\n",
    "    ## First check loci for signs of wrong genotype calls, assign them to a list of \"Good loci\".\n",
    "    ## And then check these good loci for XY or ZW signal    \n",
    "    \n",
    "    Good_loci = []\n",
    "    \n",
    "    Male_offs_GTs = {}\n",
    "    Female_offs_GTs = {}\n",
    "    N_uncalled_male_offs = 0\n",
    "    N_uncalled_female_offs = 0\n",
    "    \n",
    "    \n",
    "    N_ZW_testers = 0\n",
    "    N_ZW_testers_with_enough_het_females = 0\n",
    "    N_ZW_testers_with_enough_het_females_and_hom_males = 0\n",
    "            \n",
    "            \n",
    "    N_XY_testers = 0\n",
    "    N_XY_testers_with_enough_het_males = 0\n",
    "    N_XY_testers_with_enough_het_males_and_hom_females = 0\n",
    "\n",
    "    for record in myvcf:\n",
    "        \n",
    "        ## Make all variable defaults, lists and dicts.\n",
    "        \n",
    "        Mat_ref = None\n",
    "        Mat_alt = None\n",
    "        Pat_ref = None\n",
    "        Pat_alt = None\n",
    "        mat_het = None\n",
    "        pat_het = None\n",
    "        Mat_genotype = None\n",
    "        Pat_genotype = None\n",
    "        Mat_genotype_bases = None\n",
    "        Pat_genotype_bases = None\n",
    "        Non_par_allele = False\n",
    "        N_offspring = 0\n",
    "        offspr_w_allele_dropout = 0\n",
    "        N_offsp_hom_for_minor_par_allele = 0\n",
    "        \n",
    "        Sample_headers = []\n",
    "        sample_names = []\n",
    "        Parental_alleles = []\n",
    "        offspring_gt_types = []\n",
    "        \n",
    "        offspring_gt_types_dict = {}\n",
    "        offspring_gt_bases_dict = {}\n",
    "        allele_dropout_scores = {}\n",
    "        \n",
    "        ## lists for constructing the lines of the MST output file\n",
    "        malemap_offspring_haploid_genotypes = []\n",
    "        malemap_offspring_haploid_genotypes_compli = []\n",
    "        femalemap_offspring_haploid_genotypes = []\n",
    "        femalemap_offspring_haploid_genotypes_compli = []\n",
    "    \n",
    "    \n",
    "        Loc_Id = \"%s_%s\" % (record.ID, record.POS)\n",
    "        \n",
    "        log_file.write(\"\\n\\nLocus: %s -------------\\n\" % (Loc_Id))\n",
    "        \n",
    "        for sample in record.samples:\n",
    "            \n",
    "            name = sample.sample\n",
    "            genotype = sample['GT']\n",
    "            allele_dropout_scores[record] = {}\n",
    "            \n",
    "            sample_names.append(name)\n",
    "            \n",
    "            \n",
    "            ## Get Mother and Father information\n",
    "            \n",
    "            if name == Mother:\n",
    "                \n",
    "                if not sample.called == False:\n",
    "                    Mat_ref = sample.gt_bases.split(\"/\")[0]\n",
    "                    Mat_alt = sample.gt_bases.split(\"/\")[1]\n",
    "                    Mat_genotype = genotype\n",
    "                    Mat_genotype_bases = sample.gt_bases\n",
    "                    \n",
    "                    Parental_alleles.append(Mat_ref)\n",
    "                    Parental_alleles.append(Mat_alt)\n",
    "                    \n",
    "                    if sample.gt_type == 0 or sample.gt_type == 2:\n",
    "                        mat_het = False\n",
    "                        log_file.write(\"\\nMother is homozygous (%s)\" % (sample.gt_bases))\n",
    "                    elif sample.gt_type == 1:\n",
    "                        mat_het = True   \n",
    "                        log_file.write(\"\\nMother is heterozygous (%s)\" % (sample.gt_bases))\n",
    "                elif sample.called == False:\n",
    "                    log_file.write(\"\\nLocus not called in mother\")\n",
    "                    \n",
    "                \n",
    "                    \n",
    "            elif name == Father:\n",
    "                \n",
    "                if not sample.called == False:\n",
    "                    Pat_ref = sample.gt_bases.split(\"/\")[0]\n",
    "                    Pat_alt = sample.gt_bases.split(\"/\")[1]\n",
    "                    Pat_genotype = genotype\n",
    "                    Pat_genotype_bases = sample.gt_bases\n",
    "                    \n",
    "                    Parental_alleles.append(Pat_ref)\n",
    "                    Parental_alleles.append(Pat_alt)\n",
    "                    \n",
    "                    if sample.gt_type == 0 or sample.gt_type == 2:\n",
    "                        pat_het = False\n",
    "                        log_file.write(\"\\nFather is homozygous (%s)\" % (sample.gt_bases))\n",
    "                    elif sample.gt_type == 1:\n",
    "                        pat_het = True\n",
    "                        log_file.write(\"\\nFather is heterozygous (%s)\" % (sample.gt_bases))\n",
    "                elif sample.called == False:\n",
    "                    log_file.write(\"\\nLocus not called in father\")\n",
    "            \n",
    "        \n",
    "        ## Get parental allele information and filter for unacceptable loci\n",
    "        \n",
    "        Par_allele_counts = Counter(Parental_alleles)\n",
    "        \n",
    "        \n",
    "        if None in Par_allele_counts or Mat_genotype == None or Pat_genotype == None: ## If there is a missing call in the parents, discard locus\n",
    "            log_file.write(\"\\nMissing data in parents, locus discarded\")\n",
    "                        \n",
    "        elif len(Par_allele_counts) > 2: ## If there are more than two alleles in the parents, discard locus\n",
    "            log_file.write(\"\\nparents contain more than 2 alleles, locus not used\")\n",
    "            \n",
    "        elif pat_het == True and mat_het == True: ## If both parents are heterozygous, discard locus\n",
    "            log_file.write(\"\\nBoth parents heterozygous, locus not used\")\n",
    "            \n",
    "        elif pat_het == False and mat_het == False: ## If both parents are homozygous, discard locus\n",
    "            log_file.write(\"\\nBoth parents homozygous, locus not used\")\n",
    "            \n",
    "\n",
    "            \n",
    "        elif len(Par_allele_counts) == 2 and not 2 in Par_allele_counts.values(): ## if both parents are called and both do not have the same genotype\n",
    "            \n",
    "            for allele in Par_allele_counts:\n",
    "                if Par_allele_counts[allele] == 3:\n",
    "                    Par_major_allele = allele\n",
    "                elif Par_allele_counts[allele] == 1:\n",
    "                    Par_minor_allele = allele\n",
    "                    \n",
    "        \n",
    "            ## Get offspring information -----------------------------------------------------------------------\n",
    "                 \n",
    "            \n",
    "            ## First - look for signs of allele dropout - using all offspring for this\n",
    "            \n",
    "            for sample in record.samples:            \n",
    "                name = sample.sample\n",
    "                genotype = sample['GT']            \n",
    "                \n",
    "                if name in offspring:  ## Only samples included in the pop_map file are used\n",
    "                    N_offspring += 1\n",
    "\n",
    "\n",
    "                    ### Filter sample genotypes for signs of allele droupout \n",
    "\n",
    "                    if genotype == None: ## If sample not called, record the lack of genotype\n",
    "                        \n",
    "                        offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                        offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "                        \n",
    "                        \n",
    "                    elif not genotype == None: ## If sample has been called\n",
    "\n",
    "                        off_ref = sample.gt_bases.split(\"/\")[0] ## get the samples ref allele\n",
    "                        off_alt = sample.gt_bases.split(\"/\")[1] ## get the samples alt allele\n",
    "                        \n",
    "                        if sample.gt_bases == \"%s/%s\" % (Par_minor_allele,Par_minor_allele): ## If sample homozygous for the parental minor allele, remove sample info, record the event for tallying later.\n",
    "                            \n",
    "                            log_file.write(\"\\n%s is homozygous for minor parental allele (%s), likely allele dropout in offspring (if rare) or parent (if common)\" % (name, sample.gt_bases))\n",
    "                            offspr_w_allele_dropout += 1\n",
    "\n",
    "                            offspring_gt_types.append(None) # Count as missing data\n",
    "                            offspring_gt_bases_dict[name] = (None, None)\n",
    "                            N_offsp_hom_for_minor_par_allele += 1\n",
    "\n",
    "                        elif off_ref not in Parental_alleles or off_alt not in Parental_alleles: ## If sample has an allele that is not in parents, remove locus\n",
    "                            \n",
    "                            offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                            offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "\n",
    "                            log_file.write(\"\\n%s contains non-parental allele in genotype (%s), likely allele dropout in a Parent\" % (name, sample.gt_bases))\n",
    "                            Non_par_allele = True\n",
    "\n",
    "\n",
    "                        if off_ref in Parental_alleles and off_alt in Parental_alleles: ## And if the sample's alleles are both in the Parental genotypes then the locus is good\n",
    "                            \n",
    "                            offspring_gt_types.append(sample.gt_type) # record hom or het\n",
    "                            offspring_gt_bases_dict[name] = (sample.gt_type, sample.gt_bases)\n",
    "\n",
    "                            ## Record male and female genotypes for later use if the locus is ok.\n",
    "\n",
    "                            if name in male_offspring:\n",
    "                                if Loc_Id not in Male_offs_GTs:\n",
    "                                    Male_offs_GTs[Loc_Id] = {}\n",
    "                                Male_offs_GTs[Loc_Id][name] = (sample.gt_type, sample.gt_bases)\n",
    "\n",
    "                                if sample.gt_type == None:\n",
    "                                    N_uncalled_male_offs += 1\n",
    "\n",
    "                            elif name in female_offspring:\n",
    "                                if Loc_Id not in Female_offs_GTs:\n",
    "                                    Female_offs_GTs[Loc_Id] = {}\n",
    "                                Female_offs_GTs[Loc_Id][name] = (sample.gt_type, sample.gt_bases)\n",
    "\n",
    "                                if sample.gt_type == None:\n",
    "                                    N_uncalled_female_offs += 1\n",
    "\n",
    "            \n",
    "            counted = Counter(offspring_gt_types) ## Count the numbers of each genotype at this locus for looking at offspring presence threshold\n",
    "\n",
    "            \n",
    "\n",
    "            ## Caluculate the percentage of offspring missing at the locus\n",
    "\n",
    "            if None in counted:\n",
    "                perc_offspring_missing = counted[None]/N_offspring\n",
    "            else:\n",
    "                perc_offspring_missing = 0\n",
    "\n",
    "            ## calculate the percentages of homozygotes/heterozygotes for mendelian frequency filters below\n",
    "\n",
    "            perc_hom_0 = counted[0]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_hom_2 = counted[2]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_het = counted[1]/(counted[0]+counted[2]+counted[1])\n",
    "            perc_off_w_all_dropout = offspr_w_allele_dropout/N_offspring                  \n",
    "\n",
    "            ## Filter loci with non-mendelian genotype proportions and write data to repsective map files\n",
    "\n",
    "            if perc_offspring_missing > 1-offspring_presence_threshold:\n",
    "                log_file.write(\"\\nToo many missing genotypes, locus not used\")\n",
    "                #print \"Too many missing genotypes, locus not used\"\n",
    "\n",
    "            elif Non_par_allele == True:\n",
    "                log_file.write(\"\\nNon parental allele found, locus not used\")\n",
    "                #print \"Non parental allele found, locus not used\"\n",
    "\n",
    "            elif N_offsp_hom_for_minor_par_allele > 1: ## if there is more than one offspring homozygous for the minor parental allele, then it is likely to be allele dropout in the parent, locus is discarded\n",
    "                log_file.write(\"\\nMore than one offspring homozygous for minor allele, locus not used\")\n",
    "                #print \"More than one offspring homozygous for minor allele, locus not used\"\n",
    "            elif perc_hom_0 > mendelian_threshold or perc_hom_2 > mendelian_threshold:\n",
    "                log_file.write(\"\\nHomozygosity excess (>%s)\" % (mendelian_threshold))\n",
    "                log_file.write(\"\\nLocus discarded, Perc_hom_0= %s, Perc_hom_2= %s\" % (perc_hom_0, perc_hom_2))\n",
    "                loc_w_excess_hom += 1\n",
    "                #print \"Homozygosity excess\"\n",
    "            else:\n",
    "                #print \"Locus passed null allele filters\"\n",
    "                Good_loci.append(Loc_Id)\n",
    "\n",
    "\n",
    "            ## Now look through offspring again, to test for XY, look only at loci where Father is heterozygous and find those that \n",
    "            ## are heterozygous in all (or more than a threshold number) of males\n",
    "\n",
    "            ##============================================================================================================================= \n",
    "            ## Now look specifically at loci that are het in only one parent. =============================================================\n",
    "            ## ============================================================================================================================\n",
    "\n",
    "            ## Get total number of CALLED females\n",
    "\n",
    "            N_called_fem_offspring = len(female_offspring) - N_uncalled_female_offs\n",
    "            N_called_male_offspring = len(male_offspring) - N_uncalled_male_offs\n",
    "            \n",
    "            \n",
    "\n",
    "            \n",
    "\n",
    "            if Loc_Id in Good_loci:            \n",
    "\n",
    "                if pat_het == False and mat_het == True: ## If mother is het and father is hom, use to test for ZW\n",
    "                    log_file.write(\"\\nMother is heterozygous, Father is homozygous, locus used to test for ZW\")\n",
    "                    \n",
    "                    N_ZW_testers += 1\n",
    "                    \n",
    "                    ## Now look through the male and female offspring - look for ZW patterns\n",
    "\n",
    "                    N_het_females = 0\n",
    "                    N_hom_males = 0\n",
    "\n",
    "                    ## Find out the percentage of females called that were heterozygous\n",
    "                    \n",
    "                    for female in Female_offs_GTs[Loc_Id]:\n",
    "                        if Female_offs_GTs[Loc_Id][female][0] == 1 and Mat_alt in Female_offs_GTs[Loc_Id][female][1].split(\"/\"): ## If female is heterozygous count it. \n",
    "                            N_het_females += 1\n",
    "                    \n",
    "                    ## And then find out the percentage of males called that were homozygous\n",
    "                    \n",
    "                    for male in Male_offs_GTs[Loc_Id]:\n",
    "                        if Male_offs_GTs[Loc_Id][male][0] in [0,2]: ## If male is homozygous count it. \n",
    "                            N_hom_males += 1\n",
    "                        \n",
    "                    \n",
    "\n",
    "                    if N_het_females > 1:\n",
    "                        perc_het_females = N_het_females/N_called_fem_offspring\n",
    "                        log_file.write(\"\\nN called female offspring = %s\" % N_called_fem_offspring)\n",
    "                        log_file.write(\"\\nN heterozygous females = %s\" % N_het_females)\n",
    "                        log_file.write(\"\\nPercent heterozygous female offspring = %s\" % perc_het_females)\n",
    "                        \n",
    "                        N_ZW_testers_with_enough_het_females += 1\n",
    "                        \n",
    "                        if perc_het_females >= het_sex_thresh:\n",
    "                            log_file.write(\"\\nEnough heterozygous females!\")\n",
    "\n",
    "                            if N_hom_males > 1:\n",
    "                                perc_hom_males = N_hom_males/N_called_male_offspring\n",
    "                                log_file.write(\"\\nPercent homozygous males = %s\" % perc_hom_males)\n",
    "                                \n",
    "                                if perc_hom_males >= hom_sex_thresh:\n",
    "                                    ZW_tags.append(Loc_Id)\n",
    "                                    log_file.write(\"\\nEnough homozygous males!\")\n",
    "                                    log_file.write(\"\\n##Locus supports ZW system!\")\n",
    "                                    N_ZW_testers_with_enough_het_females_and_hom_males += 1\n",
    "                                else:\n",
    "                                    log_file.write(\"\\nNot enough homozygous males\")\n",
    "                            else:\n",
    "                                log_file.write(\"\\nNo homozygous male offspring\")\n",
    "                        else:\n",
    "                            log_file.write(\"\\nNot enough heterozygous female offspring\")\n",
    "                    else:\n",
    "                        log_file.write(\"\\nNo heterozygous female offspring\")\n",
    "\n",
    "                elif pat_het == True and mat_het == False: ## If father is het and mother is hom, use to test for XY\n",
    "                    log_file.write(\"\\nFather is heterozygous, Mother is homozygous, locus used to test for XY\")\n",
    "                    N_XY_testers += 1\n",
    "                    \n",
    "                    \n",
    "                    N_het_males = 0\n",
    "                    N_hom_females = 0\n",
    "                    \n",
    "                    ## Find out the percentage of males called that were heterozygous\n",
    "                    \n",
    "                    for male in Male_offs_GTs[Loc_Id]:\n",
    "                        if Male_offs_GTs[Loc_Id][male][0] == 1 and Pat_alt in Male_offs_GTs[Loc_Id][male][1].split(\"/\"): ## If male is heterozygous count it. \n",
    "                            N_het_males += 1\n",
    "                    \n",
    "                    ## Find out the percentage of females called that were homozygous\n",
    "                    \n",
    "                    for female in Female_offs_GTs[Loc_Id]:\n",
    "                        if Female_offs_GTs[Loc_Id][female][0] in [0,2]: ## If female is homozygous count it. \n",
    "                            N_hom_females += 1\n",
    "                            \n",
    "                    \n",
    "\n",
    "                    if N_het_males > 1:\n",
    "                        perc_het_males = N_het_males/N_called_male_offspring\n",
    "                        \n",
    "                        log_file.write(\"\\nN called male offspring = %s\" % N_called_male_offspring)\n",
    "                        log_file.write(\"\\nN heterozygous male offspring = %s\" % N_het_males)\n",
    "                        log_file.write(\"\\nPercent heterozygous males = %s\" % perc_het_males)\n",
    "                        \n",
    "                        if perc_het_males >= het_sex_thresh:\n",
    "                            log_file.write(\"\\nEnough heterozygous males!\")\n",
    "                            N_XY_testers_with_enough_het_males += 1\n",
    "                            \n",
    "                            if N_hom_females > 1:\n",
    "                                perc_hom_females = N_hom_females/N_called_fem_offspring\n",
    "                                log_file.write(\"\\nPercent homozygous females = %s\" % perc_hom_females)\n",
    "                                \n",
    "                                if perc_hom_females >= hom_sex_thresh:\n",
    "                                    XY_tags.append(Loc_Id)\n",
    "                                    log_file.write(\"\\nEnough homozygous females!\")\n",
    "                                    log_file.write(\"\\n##Locus supports XY system!\")\n",
    "                                    N_XY_testers_with_enough_het_males_and_hom_females += 1\n",
    "                                else:\n",
    "                                    log_file.write(\"\\nNot enough homozygous females\")\n",
    "\n",
    "                            else:\n",
    "                                log_file.write(\"\\nNo homozygous female offspring\")        \n",
    "                        else:\n",
    "                            log_file.write(\"\\nNot enough heterozygous male offspring\")           \n",
    "                    else:\n",
    "                        log_file.write(\"\\nNo heterozygous male offspring\")\n",
    "\n",
    "\n",
    "    print \"Number of good loci = %s\" % len(Good_loci)\n",
    "\n",
    "    print \"\\nN loci suitable for XY testing:\", N_XY_testers\n",
    "    print \"N XY test loci with enough heterozygous males:\", N_XY_testers_with_enough_het_males\n",
    "    print \"N loci that fit the specified XY criteria:\", N_XY_testers_with_enough_het_males_and_hom_females\n",
    "    \n",
    "    print \"\\nN loci suitable for ZW testing:\", N_ZW_testers\n",
    "    print \"N ZW test loci with enough heterozygous females:\", N_ZW_testers_with_enough_het_females\n",
    "    print \"N loci that fit the specified ZW criteria:\", N_ZW_testers_with_enough_het_females_and_hom_males\n",
    "    \n",
    "\n",
    "    \n",
    "    ## Finally, make fasta files and vcf files of the sex linked markers\n",
    "    \n",
    "    XYset = set(XY_tags)\n",
    "    ZWset = set(ZW_tags)\n",
    "    \n",
    "    if catalog_path.endswith(\"gz\"):\n",
    "        catalog = gzip.open(catalog_path, 'r').readlines()\n",
    "    else:\n",
    "        catalog = open(catalog_path, 'r').readlines()\n",
    "\n",
    "        \n",
    "    ## Outdir same as vcf dir:\n",
    "\n",
    "    outdir = vcf_path.rpartition(\"/\")[0]\n",
    "\n",
    "    if len(XYset) > 0:\n",
    "        Putative_XYlinked_makers_file = open(\"%s/%s\" %(outdir, \"Putative_XYlinked_makers.fa\"), 'w')\n",
    "\n",
    "        for locus in XYset:\n",
    "            for tag in catalog:\n",
    "                if locus.split(\"_\")[0] == tag.split()[2]:\n",
    "                    Putative_XYlinked_makers_file.write(\">X_linkedLocusID_%s\\n\" % locus)\n",
    "                    Putative_XYlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_XYlinked_makers_file.close()\n",
    "\n",
    "    if len(ZWset) > 0:\n",
    "        Putative_ZWlinked_makers_file = open(\"%s/%s\" %(outdir, \"Putative_ZWlinked_makers.fa\"), 'w')\n",
    "\n",
    "        for locus in ZWset:\n",
    "            for tag in catalog:\n",
    "                if locus.split(\"_\")[0] == tag.split()[2]:\n",
    "                    Putative_ZWlinked_makers_file.write(\">Z_linked|LocusID_%s\\n\" % locus)\n",
    "                    Putative_ZWlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_ZWlinked_makers_file.close()\n",
    "\n",
    "    return XY_tags, ZW_tags\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of good loci = 2250\n",
      "\n",
      "N loci suitable for XY testing: 1349\n",
      "N XY test loci with enough heterozygous males: 219\n",
      "N loci that fit the specified XY criteria: 137\n",
      "\n",
      "N loci suitable for ZW testing: 901\n",
      "N ZW test loci with enough heterozygous females: 896\n",
      "N loci that fit the specified ZW criteria: 0\n"
     ]
    }
   ],
   "source": [
    "## Analyses parameters\n",
    "\n",
    "vcf_path = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Final_SNP_data/batch_1_strict.vcf.altered\"\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Final_SNP_data/popmap.txt\"\n",
    "catalog_file = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Final_SNP_data/batch_1.catalog.tags.tsv.gz\"\n",
    "offspring_presence_threshold = 0.9   # 90% of offspring must have data for a given locus\n",
    "medelian_cut_off = 0.75   # 75% mendelian segregation frequency threshold\n",
    "min_thresh_for_heterozygous_heterogametic_sex = 0.75  # 75% of the heterogametic sex must be heterozygous\n",
    "min_thresh_for_homozygous_homogametic_sex = 0.9  # 90% of the homogametic sex must be homozygous\n",
    "\n",
    "\n",
    "XYs, ZWs = SL_markers_from_sexed_fam(vcf_path,\n",
    "                                     popmap_path, \n",
    "                                     catalog_file, \n",
    "                                     offspring_presence_threshold, \n",
    "                                     medelian_cut_off, \n",
    "                                     min_thresh_for_heterozygous_heterogametic_sex, \n",
    "                                     min_thresh_for_homozygous_homogametic_sex)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Something wrong with the rpy2 thing, so the plotting isn't working at the moment. Fix next week. In the mean time . . . it looks from the PCA like there is some misassignment. Also one of the females stands out a bit, which might mean it was from a different family. Re-run the analyses without these two samples. . . . (Still have 8 females and 7 males in the offspring)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of good loci = 2250\n",
      "\n",
      "N loci suitable for XY testing: 1349\n",
      "N XY test loci with enough heterozygous males: 276\n",
      "N loci that fit the specified XY criteria: 144\n",
      "\n",
      "N loci suitable for ZW testing: 901\n",
      "N ZW test loci with enough heterozygous females: 886\n",
      "N loci that fit the specified ZW criteria: 2\n"
     ]
    }
   ],
   "source": [
    "## Analyses parameters\n",
    "\n",
    "vcf_path = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Final_SNP_data/batch_1_strict.vcf.altered\"\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/STOECK/Bviridis/Final_SNP_data/pop_map_edited.txt\"\n",
    "catalog_file = \"\"\n",
    "offspring_presence_threshold = 0.9   # 90% of offspring must have data for a given locus\n",
    "medelian_cut_off = 0.75   # 75% mendelian segregation frequency threshold\n",
    "min_thresh_for_heterozygous_heterogametic_sex = 0.75  # 75% of the heterogametic sex must be heterozygous\n",
    "min_thresh_for_homozygous_homogametic_sex = 0.9  # 90% of the homogametic sex must be homozygous\n",
    "\n",
    "\n",
    "XYs, ZWs = MST_input_maker(vcf_path,\n",
    "                           popmap_path, \n",
    "                           offspring_presence_threshold, \n",
    "                           medelian_cut_off, \n",
    "                           min_thresh_for_heterozygous_heterogametic_sex, \n",
    "                           min_thresh_for_homozygous_homogametic_sex)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
