{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "def BlastParseExtra(infile, genome_fa, best_hit_criteria, Eval_thresh, get_frags_switch = 1, window_size = 2000, verb = 0):\n",
    "    \"\"\"\n",
    "    Usage: BlastParseExtra.py  <blast_xml_output> <genome_fasta> <best_hit_criteria> <Eval_thresh> <get_frags_switch> <window_size> <verb>\n",
    "    \n",
    "    <blast_xml_output>  -  absolute path to blast xml output\n",
    "    <genome_fasta>      -  absolute path to genome fasta file\n",
    "    <best_hit_criteria> -  factor by which the best e-value must be lower than the second best (recommended: 1e-5)\n",
    "    <Eval_thresh>       -  e-value threshold for unique alignments\n",
    "    <get_frags_switch>  -  switch (1 = on, 0 = off) for getting fragments around the mapping\n",
    "    <window_size>       -  size of the window (in bp) around the mapping coordinates used to extract the subject scaffold segment \n",
    "    <verb>              -  make it chatty\n",
    "\n",
    "    This script first filters the mappings in the <blast_xml_output> for uniq hits with evalues better than\n",
    "    <Eval_thresh> or for multi hits where the best hit is <best_hit_criteria> orders of magnitude better than \n",
    "    the second.\n",
    "    \n",
    "    It will then retrieve a segment of the scaffold from <genome_fasta> which is + and - the <window_size> around\n",
    "    the mapping coordinates for each query. If the ends of the scaffold are not within this window, then the \n",
    "    length of the segment will be (length of mapped query sequence + 2 x <window_size>). However if an end of a\n",
    "    scaffold is within this window, the segment will be trimmed to this length.\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "\n",
    "    import sys\n",
    "    from Bio.Blast import NCBIXML\n",
    "    from Bio import SeqIO\n",
    "    import gzip\n",
    "\n",
    "    if infile.endswith(\"gz\"):\n",
    "        handle = gzip.open(infile, 'r')\n",
    "    else:\n",
    "        handle = open(infile, 'r')\n",
    "\n",
    "    blast = NCBIXML.parse(handle)\n",
    "\n",
    "    good_blast_outs = {}\n",
    "    multi_counter = 0\n",
    "    unique_counter = 0\n",
    "    \n",
    "    ## From Alan's script: Returns blast hits only when the best e-value is 5 orders of magnitude better than the second best.\n",
    "\n",
    "    for record in blast :\n",
    "        if len(record.alignments)==1:\n",
    "            if record.alignments[0].hsps[0].expect <= Eval_thresh:\n",
    "                unique_counter += 1\n",
    "                good_blast_outs[record.query] = {}\n",
    "                good_blast_outs[record.query][\"Ref_hit_id\"] = str(record.alignments[0].hit_def)\n",
    "                good_blast_outs[record.query][\"Evalue\"] = float(record.alignments[0].hsps[0].expect)\n",
    "                good_blast_outs[record.query][\"Hit_start_coord\"] = int(record.alignments[0].hsps[0].sbjct_start)\n",
    "                good_blast_outs[record.query][\"Hit_end_coord\"] = int(record.alignments[0].hsps[0].sbjct_end)\n",
    "                good_blast_outs[record.query][\"aln_len\"] = int(record.alignments[0].hsps[0].align_length)\n",
    "                good_blast_outs[record.query][\"Hit_identity\"] = int(record.alignments[0].hsps[0].identities)\n",
    "                #print \"Uniq\\t%s\\t%s\\t%s\\t%s\\t%s\" % (record.query, good_blast_outs[record.query][\"Ref_hit_id\"], good_blast_outs[record.query][\"Evalue\"], good_blast_outs[record.query][\"Hit_start_coord\"], good_blast_outs[record.query][\"Hit_end_coord\"])\n",
    "\n",
    "\n",
    "        elif len(record.alignments)>1:\n",
    "            if all([record.alignments[0].hsps[0].expect <= Eval_thresh, record.alignments[0].hsps[0].expect < best_hit_criteria * record.alignments[1].hsps[0].expect]):\n",
    "                multi_counter += 1\n",
    "                good_blast_outs[record.query] = {}\n",
    "                good_blast_outs[record.query][\"Ref_hit_id\"] = str(record.alignments[0].hit_def)\n",
    "                good_blast_outs[record.query][\"Evalue\"] = float(record.alignments[0].hsps[0].expect)\n",
    "                good_blast_outs[record.query][\"Hit_start_coord\"] = int(record.alignments[0].hsps[0].sbjct_start)\n",
    "                good_blast_outs[record.query][\"Hit_end_coord\"] = int(record.alignments[0].hsps[0].sbjct_end)\n",
    "                good_blast_outs[record.query][\"aln_len\"] = int(record.alignments[0].hsps[0].align_length)\n",
    "                good_blast_outs[record.query][\"Hit_identity\"] = int(record.alignments[0].hsps[0].identities)\n",
    "\n",
    "    if verb == \"1\":\n",
    "\n",
    "        print \"Number of multi-alingments kept:\", multi_counter\n",
    "        print \"Number of unique alingments kept:\", unique_counter\n",
    "        print \"Total = %s\", % str(multi_counter + unique_counter)\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "    if get_frags_switch == \"1\":\n",
    "        \n",
    "        Rtemp_summary_out = open(\"%s/filtered_blast_%s_summary.out\" % (infile.rpartition(\"/\")[0], window_size), 'w')\n",
    "        Rtemp_summary_out.write(\"query\\tsubject\\thit_STRT\\thit_END\\tEvalue\\tAln_len\\tAln_identities\\n\")\n",
    "        \n",
    "        #Rtemp_summary_out = open(\"%s/blast_%s_summary.out\" % (infile.rpartition(\"/\")[0], window_size), 'w')\n",
    "        #Rtemp_summary_out.write(\"query\\tsubject\\n\")\n",
    "\n",
    "        print \"\\nGetting subject scaffold segments from %s . . . \" % (genome_fa)\n",
    "\n",
    "        Rtemp_chunks = open(\"%s/filtered_blast_%s_chunks.fa\" % (infile.rpartition(\"/\")[0], window_size), 'w')\n",
    "\n",
    "        segment_counter = 0\n",
    "\n",
    "        Rtemp = SeqIO.parse(genome_fa, \"fasta\")\n",
    "\n",
    "        recorded_scaffolds = []\n",
    "\n",
    "        for scaffold in Rtemp:\n",
    "            \n",
    "            \n",
    "            for query in good_blast_outs:\n",
    "            \n",
    "                           \n",
    "                if scaffold.id == good_blast_outs[query][\"Ref_hit_id\"]: ## If the scaffold has a hit\n",
    "            \n",
    "\n",
    "                    Rtemp_summary_out.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (query, good_blast_outs[query][\"Ref_hit_id\"], good_blast_outs[query][\"Hit_start_coord\"], good_blast_outs[query][\"Hit_end_coord\"], good_blast_outs[query][\"Evalue\"], good_blast_outs[query][\"aln_len\"], good_blast_outs[query][\"Hit_identity\"]))\n",
    "                    \n",
    "\n",
    "\n",
    "                    if good_blast_outs[query][\"Hit_start_coord\"] - window_size <= 0 and good_blast_outs[query][\"Hit_end_coord\"] + window_size >= len(scaffold.seq): # if the beginning and if the end of the scaffold is within the rang of the window\n",
    "\n",
    "                        SeqIO.write(scaffold, Rtemp_chunks, 'fasta') ## just print whole scaffold\n",
    "                        segment_counter += 1\n",
    "\n",
    "\n",
    "                    elif good_blast_outs[query][\"Hit_start_coord\"] - window_size <= 0 and good_blast_outs[query][\"Hit_end_coord\"] + window_size < len(scaffold.seq): ## or if the begninning is in range of the window but the end isn't\n",
    "\n",
    "                        SeqIO.write(scaffold[:good_blast_outs[query][\"Hit_end_coord\"]+ window_size], Rtemp_chunks, 'fasta') ## print from beginning to upper end of window\n",
    "                        segment_counter += 1\n",
    "\n",
    "\n",
    "                    elif good_blast_outs[query][\"Hit_start_coord\"] - window_size > 0 and good_blast_outs[query][\"Hit_end_coord\"] + window_size >= len(scaffold.seq): ## or the end of the scaffold is in range but the beginning isnt\n",
    "\n",
    "                        SeqIO.write(scaffold[good_blast_outs[query][\"Hit_start_coord\"]- window_size:], Rtemp_chunks, 'fasta') ## print from lower window limit to the end of the scaffold\n",
    "                        segment_counter += 1\n",
    "\n",
    "                    elif good_blast_outs[query][\"Hit_end_coord\"] + window_size < len(scaffold.seq) and good_blast_outs[query][\"Hit_start_coord\"] - window_size > 0: ## or if neither end of the scaffold is in range of the window\n",
    "                        SeqIO.write(scaffold[good_blast_outs[query][\"Hit_start_coord\"]- window_size:good_blast_outs[query][\"Hit_end_coord\"] + window_size], Rtemp_chunks, 'fasta') ## print from lower window limit to the end of the scaffold\n",
    "                        segment_counter += 1\n",
    "        \n",
    "        print \"%s sequence scaffold segments are in %s/filtered_blast_%s_chunks.fa\" % (segment_counter, infile.rpartition(\"/\")[0], window_size)\n",
    "        print \"\\nOutputting filtered hits summary to %s/filtered_blast_%s_summary.out\" % (infile.rpartition(\"/\")[0], window_size)\n",
    "        \n",
    "        Rtemp_chunks.close()\n",
    "    \n",
    "    \n",
    "    ## If fragments not requested, output summary anyway\n",
    "    \n",
    "    elif get_frags_switch == \"0\":\n",
    "        \n",
    "        Rtemp_summary_out = open(\"%s/filtered_blast_summary.out\" % (infile.rpartition(\"/\")[0]), 'w')\n",
    "        Rtemp_summary_out.write(\"query\\tsubject\\thit_STRT\\thit_END\\tEvalue\\tAln_len\\tAln_identities\\n\")\n",
    "        \n",
    "        print \"\\nNot getting genome fragments\"\n",
    "        print \"\\nOutputting filtered hits summary to %s/filtered_blast_summary.out\\n\" % (infile.rpartition(\"/\")[0])\n",
    "        \n",
    "        for query in good_blast_outs:\n",
    "            Rtemp_summary_out.write(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (query, good_blast_outs[query][\"Ref_hit_id\"], good_blast_outs[query][\"Hit_start_coord\"], good_blast_outs[query][\"Hit_end_coord\"], good_blast_outs[query][\"Evalue\"], good_blast_outs[query][\"aln_len\"], good_blast_outs[query][\"Hit_identity\"]))\n",
    "        \n",
    "    \n",
    "    Rtemp_summary_out.close()\n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of multi-alingments kept: 20\n",
      "Number of unique alingments kept: 7\n",
      "\n",
      "Not outputting genome fragments\n",
      "\n",
      "Outputting filtered hits summary to /home/djeffrie/Data/RADseq/Barret/filtered_blast_summary.out\n"
     ]
    }
   ],
   "source": [
    "input_file = \"/home/djeffrie/Data/RADseq/Barret/ammarnas_small.xml\"\n",
    "genome_fasta = \"/home/djeffrie/Data/Genomes/Rtemp_hybrid/Genome/Rtemp_hyb_final_assemb.fasta\"\n",
    "best_hit_eval_diff = 1e-5\n",
    "eval_threshold = 1e-20\n",
    "get_frags = \"0\"\n",
    "window = 2000\n",
    "verbose = \"1\"\n",
    "\n",
    "\n",
    "good_blasts = BlastParseExtra(input_file, genome_fasta, best_hit_eval_diff, eval_threshold, get_frags, window, verbose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "### Cline compatibility\n",
    "\n",
    "if len(sys.argv) < 8:\n",
    "    sys.exit(\"\\nNot enough arguments\\n%s\" % BlastParseExtra.__doc__)\n",
    "else:\n",
    "    print \"\\n  Running . . . \\n\"\n",
    "    input_file = sys.argv[1]\n",
    "    genome_fasta = sys.argv[2]\n",
    "    best_hit_eval_diff = float(sys.argv[3])\n",
    "    eval_threshold = float(sys.argv[4])\n",
    "    get_frags = sys.argv[5]\n",
    "    window = int(sys.argv[6])\n",
    "    verbose = sys.argv[7]\n",
    "\n",
    "BlastParseExtra(input_file, genome_fasta, best_hit_eval_diff, eval_threshold, get_frags, window, verbose)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
