{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import vcf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq > 0.4.\n",
    "Written by D. Jeffries 09/2015 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Workflow:\n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci identified as X or Z linked are labelled as such in this file.\n",
    "    6. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    7. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    8. All suplus information is recorded to a log file, with a summary at the end of this file. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "SystemExit",
     "evalue": "\n##Error, not enough arguments\n\n##USAGE (args in this order):\nSL_snp_finder.py <path/to/.vcf> <path/to/pop_map.txt> <path_to_catalog.tags.tsv_file> <female-male_cutoff> <sample_presence_cutoff> <coverage_per_sample_cutoff> <maf_cutoff>\n\n##All paths should be absolute, not relative\n##pop_map.txt must have the same number of samples as the pop_map.txt file used in populations\n\n",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[1;31mSystemExit\u001b[0m\u001b[1;31m:\u001b[0m \n##Error, not enough arguments\n\n##USAGE (args in this order):\nSL_snp_finder.py <path/to/.vcf> <path/to/pop_map.txt> <path_to_catalog.tags.tsv_file> <female-male_cutoff> <sample_presence_cutoff> <coverage_per_sample_cutoff> <maf_cutoff>\n\n##All paths should be absolute, not relative\n##pop_map.txt must have the same number of samples as the pop_map.txt file used in populations\n\n\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "To exit: use 'exit', 'quit', or Ctrl-D.\n"
     ]
    }
   ],
   "source": [
    "## Get arguments from the command line\n",
    "\n",
    "Usage_message = \"\\n##USAGE (args in this order):\\nSL_snp_finder.py <path/to/.vcf> <path/to/pop_map.txt> <path_to_catalog.tags.tsv_file> <female-male_cutoff> <sample_presence_cutoff> <coverage_per_sample_cutoff> <maf_cutoff>\\n\\n##All paths should be absolute, not relative\\n##pop_map.txt must have the same number of samples as the pop_map.txt file used in populations\\n\\n\"\n",
    "\n",
    "if len(sys.argv) < 8: ## If not enough args are supplied print error message\n",
    "    sys.exit(\"\\n##Error, not enough arguments\\n\"+Usage_message)\n",
    "\n",
    "elif len(sys.argv) == 8:\n",
    "    myvcfpath = sys.argv[1]\n",
    "    alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "    popmappath = sys.argv[2]\n",
    "    catalog_tags_file = sys.argv[3]\n",
    "    X_or_Z_freq_threshold = float(sys.argv[4])\n",
    "    sample_presence_cutoff = float(sys.argv[5])\n",
    "    coverage_threshold = int(sys.argv[6])\n",
    "    maf_threshold = float(sys.argv[7])\n",
    "    \n",
    "else:\n",
    "    sys.exit(\"Unknown Error\\n\"+Usage_message)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'myvcfpath' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-5-f463b64d2515>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mmyvcf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmyvcfpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[0malteredvcf\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malteredvcfpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'w'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mmyvcf\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[1;34m\"Allele Depth\"\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mline\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'myvcfpath' is not defined"
     ]
    }
   ],
   "source": [
    "myvcf = open(myvcfpath, 'r').readlines()\n",
    "alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "for line in myvcf:\n",
    "    if \"Allele Depth\" not in line:\n",
    "        alteredvcf.write(line)\n",
    "    elif \"Allele Depth\" in line:\n",
    "        line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "        alteredvcf.write(line)\n",
    "alteredvcf.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now calculate allele frequencies for males and females at each SNP\n",
    "####Requires:\n",
    "    1. pyvcf module installed (can use pip, remember to add to python path. This is on the cluster!\n",
    "    2. altered vcf file from above\n",
    "    3. pop_map.txt file. Same format as used for stacks. Sample names must be the same. And males and females must be denoted by M or F (case sensitive) respectively. Must be the same file as used in populations to creat the VCF. If there are additional samples in this file the allele frequencies will be wrong!\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'alteredvcfpath' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-6-6d8a914496c1>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mvcf_reader\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvcf\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0malteredvcfpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'r'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m## load in altered vcf file\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      2\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;31m## Open file for all frequency and locus info for all kept genotypes\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      5\u001b[0m \u001b[0mall_frequencies\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"%s%s\"\u001b[0m \u001b[1;33m%\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmyvcfpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\".all_frequencies.tsv\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'w'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'alteredvcfpath' is not defined"
     ]
    }
   ],
   "source": [
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "\n",
    "## Open file for all frequency and locus info for all kept genotypes\n",
    "all_frequencies = open(\"%s%s\" %(myvcfpath, \".all_frequencies.tsv\"), 'w')\n",
    "all_frequencies.write(\"#Locus_ID\\tPOS\\tN_MaleChroms\\tN_FemChroms\\tMalefreq_REF\\tMalefreq_ALT\\tFemalefreq_REF\\tFemalefreq_ALT\\tFemREF-MaleREF\\tSex_linked\\n\")\n",
    "\n",
    "## Open files to output X or Z linked loci to and add headers\n",
    "\n",
    "Putative_Xlinked_makers = []\n",
    "\n",
    "Putative_Zlinked_makers = []\n",
    "\n",
    "## Make a log file for all auxillary info\n",
    "freq_ratios_log = open(\"%s%s\" %(myvcfpath, \".freq_ratios.log\"), 'w')\n",
    "\n",
    "## Make list to keep frequency information in for histogram\n",
    "freq_ratio_data = []\n",
    "\n",
    "## get male/female info for samples for working out frequencies below38      3460    50      26      0.940   0.060   0.846   0.154   -0.094  NotSexLinked\n",
    "\n",
    "pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "sample_dict = {}\n",
    "sample_dict[\"females\"] = []\n",
    "sample_dict[\"males\"] = []\n",
    "fem_samples = 0\n",
    "male_samples = 0\n",
    "sample_counter = 0\n",
    "low_maf_counter = 0\n",
    "kept_loci = 0\n",
    "\n",
    "for sample in pop_map:\n",
    "    name = sample.split()[0]\n",
    "    sex = sample.strip().split()[1]\n",
    "    sample_counter += 1\n",
    "    if sex == \"F\":\n",
    "        sample_dict[\"females\"].append(name)\n",
    "        fem_samples += 1\n",
    "    elif sex == \"M\":\n",
    "        sample_dict[\"males\"].append(name)\n",
    "        male_samples += 1\n",
    "        \n",
    "locus_dict = {}\n",
    "number_of_samples = sample_counter\n",
    "print \"Number of samples =\", number_of_samples\n",
    "\n",
    "## Find number of loci in input file\n",
    "locus_counter = 0\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    locus_counter += 1\n",
    "print \"Number of loci = %s\" % (locus_counter)\n",
    "    \n",
    "## Write some general stats to the log file\n",
    "freq_ratios_log.write(\"Input file %s\\n\" % (myvcfpath))\n",
    "freq_ratios_log.write(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "freq_ratios_log.write(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "freq_ratios_log.write(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "freq_ratios_log.write(\"\\n## User specified options:\\n\")\n",
    "## add options specified here!\n",
    "\n",
    "## Some quick counters\n",
    "\n",
    "numb_putative_Xlinked = 0\n",
    "numb_putative_Zlinked = 0\n",
    "low_data_loci = 0\n",
    "\n",
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    femREF_count = 0    ## set the counters for the reference and alternative allele (encoded as 0 in the vcf)\n",
    "    femALT_count = 0\n",
    "    malREF_count = 0\n",
    "    malALT_count = 0   \n",
    "    fem_none_count = 0\n",
    "    male_none_count = 0\n",
    "    low_cov_samples = 0\n",
    "    n_genotypes = 0\n",
    "    male_genotypes = 0\n",
    "    fem_genotypes = 0\n",
    "    \n",
    "    if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "        low_maf_counter += 1\n",
    "        freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "        freq_ratios_log.write(\"Minor allele frequence of locus is lower than specified cutoff (%s)\\n\" % (record.aaf[0]))\n",
    "        pass  \n",
    "    \n",
    "    elif record.aaf >= maf_threshold:\n",
    "    \n",
    "        for sample in record.samples: \n",
    "            \n",
    "            name = sample.sample\n",
    "            \n",
    "            ### For each sample, if the coverage is too low, remove the genotype for that individual.\n",
    "            \n",
    "            \n",
    "            if sample['DP'] < coverage_threshold:\n",
    "                genotype = None\n",
    "                low_cov_samples += 1\n",
    "                freq_ratios_log.write(\"Sample %s thrown out due to low coverage (%s)\\n\" % (name, sample['DP']))\n",
    "            elif sample['DP'] >= coverage_threshold:\n",
    "                genotype = sample['GT']\n",
    "            \n",
    "            ## Now calculate the female and male frequencies separately\n",
    "            \n",
    "            if name in sample_dict[\"females\"]: \n",
    "                #print \"Locus=\", record.ID, \"Female=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    femREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    femALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                #print \"\\tN_REF=\", femREF_count, \"N_ALT=\", femALT_count, \"Nnone=\", fem_none_count\n",
    "            elif name in sample_dict[\"males\"]:\n",
    "                #print \"Locus=\", record.ID, \"Male=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    malREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    malALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "            #else:\n",
    "                #print \"\\n##SAMPLE NAME NOT IN POP_MAP.TXT: Sample = %s\" % (name)\n",
    "                #print \"\\tN_REF=\", malREF_count, \"N_ALT=\", malALT_count, \"Nnone=\", male_none_count\n",
    "                \n",
    "        ## Filter loci that have too many missing samples, including samples thrown out due to low coverage!\n",
    "        samples_at_locus = n_genotypes\n",
    "        chromosomes_at_locus = n_genotypes*2\n",
    "        percent_samples_present = n_genotypes/sample_counter\n",
    "        \n",
    "        if percent_samples_present >= sample_presence_cutoff:\n",
    "            kept_loci += 1\n",
    "            ## Calculate frequencies\n",
    "            \n",
    "            femREF_freq = femREF_count/(fem_genotypes*2)\n",
    "            femALT_freq = femALT_count/(fem_genotypes*2)\n",
    "            \n",
    "            maleREF_freq = malREF_count/(male_genotypes*2)\n",
    "            maleALT_freq = malALT_count/(male_genotypes*2)\n",
    "                                \n",
    "                        \n",
    "            ## Output female stats\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s\\n\\n\" %(record.ID))   \n",
    "            freq_ratios_log.write(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "            freq_ratios_log.write(\"Female reference count = %s\\n\" % (femREF_count))\n",
    "            freq_ratios_log.write(\"Female alternative count = %s\\n\" % (femALT_count))\n",
    "            freq_ratios_log.write(\"Female reference frequency = %.3f\\n\" % (femREF_freq))\n",
    "            freq_ratios_log.write(\"Female alternative frequency = %.3f\\n\" % (femALT_freq))\n",
    "                                  \n",
    "            ## check fem freqs\n",
    "            if not (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (femREF_freq + femALT_freq))\n",
    "        \n",
    "            ## Output male stats\n",
    "            freq_ratios_log.write(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "            freq_ratios_log.write(\"Male reference count = %s\\n\" % (malREF_count))\n",
    "            freq_ratios_log.write(\"Male alternative count = %s\\n\" % (malALT_count))\n",
    "            freq_ratios_log.write(\"Male reference frequency = %.3f\\n\" % (maleREF_freq))\n",
    "            freq_ratios_log.write(\"Male alternative frequency = %.3f\\n\" % (maleALT_freq))\n",
    "                                  \n",
    "            ## check male freqs\n",
    "            if not (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (maleREF_freq + maleALT_freq))\n",
    "        \n",
    "            locus_dict[record.ID] = {}\n",
    "            locus_dict[record.ID][\"female_freqs\"] = [\"%.3f\" % (femREF_freq), \"%.3f\" % (femALT_freq)]\n",
    "            locus_dict[record.ID][\"male_freqs\"] = [\"%.3f\" % (maleREF_freq), \"%.3f\" % (maleALT_freq)]\n",
    "            \n",
    "\n",
    "            ####### ==============================================================================================\n",
    "            ### So now I have the allele frequencies for males and females, I can subtract them and see if the distribution fits the expectition for X linked or Z linked!\n",
    "            \n",
    "            freq_ratio = femREF_freq - maleREF_freq\n",
    "            freq_ratio_data.append(freq_ratio)\n",
    "            \n",
    "\n",
    "            ## Write files for X or Z linked loci\n",
    "            \n",
    "            if freq_ratio > X_or_Z_freq_threshold:  ## for X linked\n",
    "                linked_status = \"Xlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Xlinked += 1\n",
    "            \n",
    "            elif freq_ratio < -X_or_Z_freq_threshold:  ## for Z linked\n",
    "                linked_status = \"Zlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Zlinked += 1\n",
    "            else:\n",
    "                freq_ratios_log.write(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                linked_status =\"NotSexLinked\"\n",
    "        \n",
    "            ## Write the main info file for male and female frequencies, ratios etc.\n",
    "            \n",
    "            all_frequencies.write(\"%s\\t%s\\t%s\\t%s\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%s\\n\" % (record.ID, record.POS, male_genotypes*2, fem_genotypes*2, maleREF_freq, maleALT_freq, femREF_freq, femALT_freq, freq_ratio, linked_status))\n",
    "            \n",
    "        \n",
    "        elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            freq_ratios_log.write(\"Number of samples at locus is lower than sample presence cutoff (%s)\\n\" % (samples_at_locus))\n",
    "            low_data_loci += 1\n",
    "    \n",
    "    \n",
    "## Print some quick summary stats - also written at the end of the log file\n",
    "print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "print \"Number of putative X linked markers = %s\" % (numb_putative_Xlinked)\n",
    "print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "\n",
    "\n",
    "freq_ratios_log.write(\"\\nSUMMARY....\\n\\n\")\n",
    "freq_ratios_log.write(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "freq_ratios_log.write(\"Number of loci with enough data = %s\\n\" % (locus_counter - low_data_loci))\n",
    "freq_ratios_log.write(\"Number of putative X linked markers = %s\\n\" % (numb_putative_Xlinked))\n",
    "freq_ratios_log.write(\"Number of putative Z linked markers = %s\\n\" % (numb_putative_Zlinked))\n",
    "freq_ratios_log.write(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "\n",
    "\n",
    "## plot histogram of frequency ratios\n",
    "\n",
    "plt.hist(freq_ratio_data, bins = 100)\n",
    "plt.axvline(x = np.mean(freq_ratio_data), color='r', linestyle='dashed', linewidth = 2)\n",
    "plt.savefig(\"%s%s\" %(myvcfpath, \".fem-male_freqs.pdf\"), format = 'pdf')\n",
    "plt.show()\n",
    "\n",
    "\n",
    "## Write fasta files of putative X or Z linked loci if there are any\n",
    "\n",
    "catalog = open(catalog_tags_file, 'r').readlines()\n",
    "\n",
    "if numb_putative_Xlinked > 0:\n",
    "    Putative_Xlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Xlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Xlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Xlinked_makers_file.write(\">LocusID_%s\\n\" % (locus))\n",
    "                Putative_Xlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Xlinked_makers_file.close()\n",
    "\n",
    "if numb_putative_Zlinked > 0:\n",
    "    Putative_Zlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Zlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Zlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Zlinked_makers_file.write(\">LocusID_%s\\n\" % (locus))\n",
    "                Putative_Zlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Zlinked_makers_file.close()\n",
    "\n",
    "\n",
    "## close all unclosed files \n",
    "freq_ratios_log.close()\n",
    "all_frequencies.close()\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IT WORKS!!\n",
    "### Issues to fix:\n",
    "    1. put in coverage cutoff\n",
    "    2. put in maf cut off"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
