{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import vcf\n",
    "import matplotlib\n",
    "#matplotlib.use('Agg') ## this allows the drawing of plots in matplotlib on the cluster, which doesn't use the X-server backend. This has something to do with display (but I don't know what)\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import sys\n",
    "import time\n",
    "import gzip\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "####Workflow:\n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci identified as X or Z linked are labelled as such in this file.\n",
    "    5. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    6. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    7. All suplus information is recorded to a log file, with a summary at the end of this file. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NO\n"
     ]
    }
   ],
   "source": [
    "femREF_freq = 0.95\n",
    "male_heterozygosity = 0.61\n",
    "lower_thresh = 0.4\n",
    "upper_thresh = 0.6\n",
    "\n",
    "if all([femREF_freq >= 0.95, male_heterozygosity >= lower_thresh, male_heterozygosity <= upper_thresh]):\n",
    "    print \"YES\"\n",
    "else:\n",
    "    print \"NO\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def SL_snp_finder2(myvcfpath, popmappath, catalog_tags_file, homogamtic_homozygosity_threshold, heterogametic_heterozygosity_threshold, sample_presence_cutoff, coverage_threshold, maf_threshold):\n",
    "    \n",
    "   \n",
    "    \"\"\"\n",
    "    This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "    Arguments:\n",
    "    \n",
    "    myvcfpath                            - path to vcf file (note this will be altered to make header compatible with Pyvcf. \n",
    "                                           New vcf will have same name with \".altered\" appended to the end)\n",
    "    \n",
    "    popmappath                           - path to population map file containing sex information. Same format as Stacks pop map file.\n",
    "    \n",
    "    catalog_tags_file                    - The catalog tags file used to create the vcf\n",
    "    \n",
    "    homogamtic_homozygosity_threshold    - The lower threshold for the proportion of homozygotes in the homogametic sex at a locus\n",
    "    \n",
    "    homogamtic_homozygosity_threshold    - The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus\n",
    "        \n",
    "    sample_presence_cutoff               - a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "    \n",
    "    coverage_threshold                   - a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this \n",
    "                                           threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will\n",
    "                                           then remove the locus.\n",
    "    \n",
    "    maf_threshold                        - minor allele frequency cutoff for a locus across all samples. \n",
    "    \n",
    "    \n",
    "    Workflow: ### NOTE*** different to SL_snp_finder()\n",
    "    \n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for male-s and females separately\n",
    "    3. Looks for loci that are heterozygous in all or most of the homogametic sex and homozygous in all or most\n",
    "       of the heterogametic sex\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \n",
    "       \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci \n",
    "       identified as X or Z linked are labelled as such in this file.\n",
    "    5. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    6. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    7. All suplus information is recorded to a log file, with a summary at the end of this file.\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    import vcf\n",
    "    import matplotlib\n",
    "    #matplotlib.use('Agg') ## this allows the drawing of plots in matplotlib on the cluster, which doesn't use the X-server backend. This has something to do with display (but I don't know what)\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy as np\n",
    "    import os.path\n",
    "    import sys\n",
    "    import time\n",
    "    import gzip\n",
    "    \n",
    "    \n",
    "    ## set the window around the freq threshold. The window automatically tightens and relaxes around 0.5 or -0.5 \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    # First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue.\n",
    "\n",
    "    myvcf = open(myvcfpath, 'r').readlines()\n",
    "    \n",
    "    alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "    alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "    for line in myvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "\n",
    "\n",
    "    # ### Now calculate allele frequencies for males and females at each SNP\n",
    "    # ####Requires:\n",
    "    #     1. pyvcf module installed (can use pip, remember to add to python path. This is on the cluster!\n",
    "    #     2. altered vcf file from above\n",
    "    #     3. pop_map.txt file. Same format as used for stacks. Sample names must be the same. And males and females must be denoted by M or F (case sensitive) respectively. Must be the same file as used in populations to creat the VCF. If there are additional samples in this file the allele frequencies will be wrong!\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "\n",
    "    ## Open file for all frequency and locus info for all kept genotypes\n",
    "    all_frequencies = open(\"%s%s\" %(myvcfpath, \".all_frequencies.tsv\"), 'w')\n",
    "    all_frequencies.write(\"#Locus_ID\\tPOS\\tN_MaleChroms\\tN_FemChroms\\tMalefreq_REF\\tMalefreq_ALT\\tFemalefreq_REF\\tFemalefreq_ALT\\tFemREF-MaleREF\\tSex_linked\\n\")\n",
    "\n",
    "    ## Open files to output X or Z linked loci to and add headers\n",
    "\n",
    "    Putative_Xlinked_makers = []\n",
    "\n",
    "    Putative_Zlinked_makers = []\n",
    "\n",
    "    ## Make a log file for all auxillary info\n",
    "    freq_ratios_log = open(\"%s%s\" %(myvcfpath, \".freq_ratios.log\"), 'w')\n",
    "    freq_ratios_log.write(\"Script run on %s\\n \" % (time.strftime(\"%c\")))\n",
    "\n",
    "    ## Make list to keep frequency information in for histogram\n",
    "    freq_ratio_data = []\n",
    "\n",
    "    ## get male/female info for samples for working out frequencies below38      3460    50      26      0.940   0.060   0.846   0.154   -0.094  NotSexLinked\n",
    "\n",
    "    pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "    sample_dict = {}\n",
    "    sample_dict[\"females\"] = []\n",
    "    sample_dict[\"males\"] = []\n",
    "    fem_samples = 0\n",
    "    male_samples = 0\n",
    "    sample_counter = 0\n",
    "    low_maf_counter = 0\n",
    "    kept_loci = 0\n",
    "    sample_missing_dict = {}\n",
    "    sample_cov_dict = {}\n",
    "    sample_cov_kept_dict = {}\n",
    "\n",
    "    for sample in pop_map:\n",
    "        name = sample.split()[0]\n",
    "        sex = sample.strip().split()[1]\n",
    "        sample_missing_dict[name] = 0\n",
    "\n",
    "        #sample_counter += 1\n",
    "        if sex == \"F\" or sex == \"f\" or sex == \"Female\" or sex == \"female\" or sex == \"Fem\" or sex == \"fem\":\n",
    "            sample_dict[\"females\"].append(name)\n",
    "            fem_samples += 1\n",
    "        elif sex == \"M\" or sex == \"m\" or sex == \"Male\" or sex == \"male\" or sex == \"Mal\" or sex == \"mal\":\n",
    "            sample_dict[\"males\"].append(name)\n",
    "            male_samples += 1\n",
    "\n",
    "    #print sample_dict        \n",
    "    locus_dict = {}\n",
    "\n",
    "\n",
    "    ## Find number of loci in input file\n",
    "    locus_counter = 0\n",
    "\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        locus_counter += 1\n",
    "    print \"Number of loci = %s\" % (locus_counter)\n",
    "\n",
    "    ## Write some general stats and input options to the log file\n",
    "\n",
    "    freq_ratios_log.write(\"\\n## User specified options:\\n\")\n",
    "    freq_ratios_log.write(\"Input vcf: %s\\n\" % (myvcfpath))\n",
    "    freq_ratios_log.write(\"pop_map file used: %s\\n\" % (popmappath))\n",
    "    freq_ratios_log.write(\"Catalog file used: %s\\n\" % (catalog_tags_file))\n",
    "    #freq_ratios_log.write(\"Specified female-male threshold: %s\\n\" % (X_or_Z_freq_threshold))\n",
    "    #freq_ratios_log.write(\"Female-male thresh range: (+/-) %s-%s\\n\" % (lower_thresh, upper_thresh))\n",
    "    freq_ratios_log.write(\"Min percentage samples present: %s\\n\" % (sample_presence_cutoff))\n",
    "    freq_ratios_log.write(\"Min coverage per genotype: %s\\n\" % (coverage_threshold))\n",
    "    freq_ratios_log.write(\"Min maf per locus: %s\\n\" % (maf_threshold))\n",
    "    freq_ratios_log.write(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "    freq_ratios_log.write(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "    freq_ratios_log.write(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "\n",
    "\n",
    "    ## Some quick counters\n",
    "\n",
    "    numb_putative_Xlinked = 0\n",
    "    numb_putative_Zlinked = 0\n",
    "    low_data_loci = 0\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "    info_rec = vcf_reader.next()\n",
    "\n",
    "    for sample in info_rec:\n",
    "        name = sample.sample\n",
    "        print name\n",
    "        sample_cov_dict[name] = []\n",
    "        sample_cov_kept_dict[name] = []\n",
    "\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        femREF_count = 0    ## set the counters for the reference and alternative allele (encoded as 0 in the vcf)\n",
    "        femALT_count = 0\n",
    "        malREF_count = 0\n",
    "        malALT_count = 0\n",
    "        het_males_count = 0 ## NEW\n",
    "        het_females_count = 0 ## NEW\n",
    "        hom_males_count = 0 ## NEW\n",
    "        hom_females_count = 0 ## NEW    \n",
    "        fem_none_count = 0\n",
    "        male_none_count = 0\n",
    "        low_cov_samples = 0\n",
    "        n_genotypes = 0\n",
    "        male_genotypes = 0\n",
    "        fem_genotypes = 0\n",
    "        number_of_samples = len(record.samples) ## this will change at each locus no???\n",
    "\n",
    "\n",
    "        for sample in record.samples:\n",
    "\n",
    "            name = sample.sample  \n",
    "            sample_cov_dict[name].append(sample['DP'])\n",
    "\n",
    "\n",
    "        if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "            low_maf_counter += 1\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            freq_ratios_log.write(\"Minor allele frequence of locus is lower than specified cutoff (%s)\\n\" % (record.aaf[0]))\n",
    "\n",
    "\n",
    "        elif record.aaf >= maf_threshold:\n",
    "\n",
    "            for sample in record.samples: \n",
    "\n",
    "                name = sample.sample\n",
    "\n",
    "                ### For each sample, if the coverage is too low, remove the genotype for that individual.\n",
    "\n",
    "\n",
    "                if sample['DP'] < coverage_threshold:\n",
    "                    genotype = None\n",
    "                    low_cov_samples += 1\n",
    "                    #freq_ratios_log.write(\"Sample %s thrown out due to low coverage (%s)\\n\" % (name, sample['DP']))\n",
    "\n",
    "                elif sample['DP'] >= coverage_threshold:\n",
    "                    genotype = sample['GT']\n",
    "                    sample_cov_kept_dict[name].append(sample['DP'])\n",
    "                ## Now calculate the female and male frequencies separately\n",
    "\n",
    "                if name in sample_dict[\"females\"]: \n",
    "                    #print \"Locus=\", record.ID, \"Female=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                    #print \"FEMALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                    if genotype == None: ## if no genotype exists\n",
    "                        sample_missing_dict[name] += 1\n",
    "                        \n",
    "                    elif genotype == \"0/0\":\n",
    "                        femREF_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                        hom_females_count += 1 ## new\n",
    "                    elif genotype == \"0/1\":\n",
    "                        het_females_count += 1 ## new\n",
    "                        femREF_count += 1\n",
    "                        femALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    elif genotype == \"1/0\":\n",
    "                        het_females_count += 1 ## new\n",
    "                        femREF_count += 1\n",
    "                        femALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    elif genotype == \"1/1\":\n",
    "                        femALT_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                        hom_females_count += 1 ## new\n",
    "                    #print \"\\tN_REF=\", femREF_count, \"N_ALT=\", femALT_count, \"Nnone=\", fem_none_count\n",
    "                elif name in sample_dict[\"males\"]:\n",
    "                    #print \"Locus=\", record.ID, \"Male=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                    #print \"MALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                    if genotype == None: ## if no genotype exists\n",
    "                        sample_missing_dict[name] += 1\n",
    "                        \n",
    "                    elif genotype == \"0/0\":\n",
    "                        malREF_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                        hom_males_count += 1 ## new\n",
    "                    elif genotype == \"0/1\":\n",
    "                        het_males_count += 1 ## new\n",
    "                        malREF_count += 1\n",
    "                        malALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                    elif genotype == \"1/0\":\n",
    "                        het_males_count += 1\n",
    "                        malREF_count += 1\n",
    "                        malALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1 ## new\n",
    "                    elif genotype == \"1/1\":\n",
    "                        malALT_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                        hom_males_count += 1 ## new\n",
    "                #else:\n",
    "                    #print \"\\n##SAMPLE NAME NOT IN POP_MAP.TXT: Sample = %s\" % (name)\n",
    "                    #print \"\\tN_REF=\", malREF_count, \"N_ALT=\", malALT_count, \"Nnone=\", male_none_count\n",
    "\n",
    "            ## Filter loci that have too many missing samples, including samples thrown out due to low coverage!\n",
    "            samples_at_locus = n_genotypes\n",
    "            chromosomes_at_locus = n_genotypes*2\n",
    "            percent_samples_present = n_genotypes/number_of_samples\n",
    "\n",
    "\n",
    "            if percent_samples_present >= sample_presence_cutoff:\n",
    "                \n",
    "                ## Calculate frequencies per sex\n",
    "\n",
    "                femREF_freq = femREF_count/(fem_genotypes*2)\n",
    "                femALT_freq = femALT_count/(fem_genotypes*2)\n",
    "\n",
    "                maleREF_freq = malREF_count/(male_genotypes*2)\n",
    "                maleALT_freq = malALT_count/(male_genotypes*2)\n",
    "                \n",
    "                male_homozygosity = hom_males_count/male_genotypes\n",
    "                female_homozygosity = hom_females_count/fem_genotypes\n",
    "                \n",
    "                male_heterozygosity = het_males_count/male_genotypes\n",
    "                female_heterozygosity = het_females_count/fem_genotypes\n",
    "                \n",
    "                \n",
    "                ## check freqs add up\n",
    "                \n",
    "                if not all([(femREF_freq) + (femALT_freq) == 1, (maleREF_freq) + (maleALT_freq) == 1, male_homozygosity + male_heterozygosity == 1, female_homozygosity + female_heterozygosity == 1]):\n",
    "                    freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "                else:\n",
    "                    freq_ratios_log.write(\"Summed frequencies are OK!\")\n",
    "                    kept_loci += 1\n",
    "\n",
    "\n",
    "                    ## Output female stats\n",
    "                    freq_ratios_log.write(\"\\n#LOCUS_ID: %s_%s\\n\\n\" %(record.ID, record.POS))   \n",
    "                    freq_ratios_log.write(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "                    freq_ratios_log.write(\"Female reference count = %s\\n\" % (femREF_count))\n",
    "                    freq_ratios_log.write(\"Female alternative count = %s\\n\" % (femALT_count))\n",
    "                    freq_ratios_log.write(\"Female reference frequency = %.3f\\n\" % (femREF_freq))\n",
    "                    freq_ratios_log.write(\"Female alternative frequency = %.3f\\n\" % (femALT_freq))\n",
    "                    freq_ratios_log.write(\"Female homozygosity = %.3f\\n\" % (female_homozygosity))\n",
    "                    freq_ratios_log.write(\"Female heterozygosity = %.3f\\n\" % (female_heterozygosity))\n",
    "\n",
    "\n",
    "\n",
    "                    ## Output male stats\n",
    "                    freq_ratios_log.write(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "                    freq_ratios_log.write(\"Male reference count = %s\\n\" % (malREF_count))\n",
    "                    freq_ratios_log.write(\"Male alternative count = %s\\n\" % (malALT_count))\n",
    "                    freq_ratios_log.write(\"Male reference frequency = %.3f\\n\" % (maleREF_freq))\n",
    "                    freq_ratios_log.write(\"Male alternative frequency = %.3f\\n\" % (maleALT_freq))\n",
    "                    freq_ratios_log.write(\"Male homozygosity = %.3f\\n\" % (male_homozygosity))\n",
    "                    freq_ratios_log.write(\"Male heterozygosity = %.3f\\n\" % (male_heterozygosity))\n",
    "\n",
    "\n",
    "                    locus_dict[record.ID] = {}\n",
    "                    locus_dict[record.ID][\"female_freqs\"] = [\"%.3f\" % (femREF_freq), \"%.3f\" % (femALT_freq)]\n",
    "                    locus_dict[record.ID][\"male_freqs\"] = [\"%.3f\" % (maleREF_freq), \"%.3f\" % (maleALT_freq)]\n",
    "\n",
    "\n",
    "                    ####### ==============================================================================================\n",
    "                    ### So now I have the allele frequencies and heterozygosities for males and females - this is the new bit. \n",
    "                    ### Not using freq ratio any more . . . . \n",
    "\n",
    "\n",
    "                    freq_ratio = femREF_freq - maleREF_freq\n",
    "                    #print freq_ratio\n",
    "                    freq_ratio_data.append(freq_ratio)\n",
    "\n",
    "\n",
    "                    ## Write files for X or Z linked loci\n",
    "\n",
    "                    if all([female_homozygosity >= homogamtic_homozygosity_threshold, male_heterozygosity >= heterogametic_heterozygosity_threshold]):\n",
    "                        linked_status = \"Xlinked\"\n",
    "                        freq_ratios_log.write(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                        Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                        numb_putative_Xlinked += 1\n",
    "\n",
    "                    elif all([male_homozygosity >= homogamtic_homozygosity_threshold, female_heterozygosity >= heterogametic_heterozygosity_threshold]):  ## for Z linked\n",
    "                        linked_status = \"Zlinked\"\n",
    "                        freq_ratios_log.write(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                        Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                        numb_putative_Zlinked += 1\n",
    "                    else:\n",
    "                        freq_ratios_log.write(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                        linked_status =\"NotSexLinked\"\n",
    "\n",
    "                    ## Write the main info file for male and female frequencies, ratios etc.\n",
    "\n",
    "                    all_frequencies.write(\"%s\\t%s\\t%s\\t%s\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%s\\n\" % (record.ID, record.POS, male_genotypes*2, fem_genotypes*2, maleREF_freq, maleALT_freq, femREF_freq, femALT_freq, freq_ratio, linked_status))\n",
    "\n",
    "\n",
    "            elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "                freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "                freq_ratios_log.write(\"Number of samples at locus is lower than sample presence cutoff (%s)\\n\" % (samples_at_locus))\n",
    "                low_data_loci += 1\n",
    "\n",
    "    ## Look at coverage...\n",
    "\n",
    "\n",
    "\n",
    "    ## Print some quick summary stats - also written at the end of the log file\n",
    "    print \"Number of samples =\", number_of_samples\n",
    "    print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "    print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "    print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "    print \"Number of putative X linked snps = %s\" % (numb_putative_Xlinked)\n",
    "    print \"Number of putative X linked tags = %s\" % (len(set(Putative_Xlinked_makers)))\n",
    "    print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "    print \"Number of putative Z linked tags = %s\" % (len(set(Putative_Zlinked_makers)))\n",
    "\n",
    "    freq_ratios_log.write(\"\\nSUMMARY....\\n\\n\")\n",
    "    freq_ratios_log.write(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "    freq_ratios_log.write(\"Number of loci with enough data = %s\\n\" % (kept_loci))\n",
    "    freq_ratios_log.write(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "    freq_ratios_log.write(\"Number of putative X linked snps = %s\\n\" % (numb_putative_Xlinked))\n",
    "    freq_ratios_log.write(\"Number of putative X linked tags = %s\\n\" % (len(set(Putative_Xlinked_makers))))\n",
    "    freq_ratios_log.write(\"Number of putative Z linked snps = %s\\n\" % (numb_putative_Zlinked))\n",
    "    freq_ratios_log.write(\"Number of putative Z linked tags = %s\\n\" % (len(set(Putative_Zlinked_makers))))\n",
    "\n",
    "\n",
    "\n",
    "    ## plot histogram of missing data per sample\n",
    "\n",
    "    plt.bar(range(len(sample_missing_dict)), sample_missing_dict.values(), align='center', color = '0.5')\n",
    "    plt.xticks(range(len(sample_missing_dict)), sample_missing_dict.keys(), rotation = 90, fontsize=8)\n",
    "    plt.ylabel(\"Number of loci with missing data\", fontsize=8)\n",
    "    plt.savefig(\"%s%s\" %(myvcfpath, \".missing_data_by_sample.pdf\"), format = 'pdf')\n",
    "    plt.close()\n",
    "    #plt.show()\n",
    "\n",
    "    ## Plot coverage per sample\n",
    "\n",
    "    n_groups = len(sample_cov_dict)\n",
    "\n",
    "    Xlabs = []\n",
    "    all_means = []\n",
    "    all_std = []\n",
    "\n",
    "    kept_means = []\n",
    "    kept_std = []\n",
    "\n",
    "    for sample in sorted(sample_cov_dict.keys()):\n",
    "        Xlabs.append(sample)\n",
    "        all_means.append(np.mean(sample_cov_dict[sample]))\n",
    "        all_std.append(np.std(sample_cov_dict[sample]))\n",
    "    for sample in sample_cov_kept_dict.keys():\n",
    "        kept_means.append(np.mean(sample_cov_kept_dict[sample]))\n",
    "        kept_std.append(np.std(sample_cov_kept_dict[sample]))\n",
    "\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    index = np.arange(n_groups)\n",
    "    bar_width = 0.35\n",
    "\n",
    "    opacity = 0.4\n",
    "    error_config = {'ecolor': '0.3'}\n",
    "\n",
    "    rects1 = plt.bar(index, all_means, bar_width,\n",
    "                     alpha=opacity,\n",
    "                     color='0.5',\n",
    "                     #yerr=all_std,\n",
    "                     error_kw=error_config,\n",
    "                     label='All')\n",
    "\n",
    "    rects2 = plt.bar(index + bar_width, kept_means, bar_width,\n",
    "                     alpha=opacity,\n",
    "                     color='0',\n",
    "                     #yerr=kept_std,\n",
    "                     error_kw=error_config,\n",
    "                     label='Kept')\n",
    "\n",
    "    plt.xlabel('Sample')\n",
    "    plt.ylabel('Mean coverage')\n",
    "    plt.title('Average coverage per sample')\n",
    "    plt.xticks(index + bar_width, (Xlabs), rotation = 90, size = 8)\n",
    "    plt.legend()\n",
    "\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"%s%s\" %(myvcfpath, \".coverage_by_sample.pdf\"), format = 'pdf')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ## Write fasta files of putative X or Z linked loci if there are any\n",
    "\n",
    "    Putative_Xlinked_makers = set(Putative_Xlinked_makers)\n",
    "    Putative_Zlinked_makers = set(Putative_Zlinked_makers)\n",
    "\n",
    "    if catalog_tags_file.endswith(\"gz\"):\n",
    "        catalog = gzip.open(catalog_tags_file, 'r').readlines()\n",
    "    else:\n",
    "        catalog = open(catalog_tags_file, 'r').readlines()\n",
    "\n",
    "    if numb_putative_Xlinked > 0:\n",
    "        Putative_Xlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Xlinked_makers.fa\"), 'w')\n",
    "\n",
    "        for locus in Putative_Xlinked_makers:\n",
    "            for tag in catalog:\n",
    "                if locus == tag.split()[2]:\n",
    "                    Putative_Xlinked_makers_file.write(\">X_linkedLocusID_%s\\n\" % (locus))\n",
    "                    Putative_Xlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_Xlinked_makers_file.close()\n",
    "\n",
    "    if numb_putative_Zlinked > 0:\n",
    "        Putative_Zlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Zlinked_makers.fa\"), 'w')\n",
    "\n",
    "        for locus in Putative_Zlinked_makers:\n",
    "            for tag in catalog:\n",
    "                if locus == tag.split()[2]:\n",
    "                    Putative_Zlinked_makers_file.write(\">Z_linked|LocusID_%s\\n\" % (locus))\n",
    "                    Putative_Zlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_Zlinked_makers_file.close()\n",
    "\n",
    "\n",
    "    ## close all unclosed files \n",
    "    freq_ratios_log.close()\n",
    "    all_frequencies.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    print \"\\n***DONE!***\\n\"\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/Populations_combined_pops/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/pop_codes.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.7\n",
    "hom_thresh = 0.8\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 59892\n",
      "Rjap_h_F_1\n",
      "Rjap_h_F_2\n",
      "Rjap_h_F_3\n",
      "Rjap_h_F_4\n",
      "Rjap_h_F_5\n",
      "Rjap_t_F_1\n",
      "Rjap_t_F_2\n",
      "Rjap_t_F_3\n",
      "Rjap_t_F_4\n",
      "Rjap_t_F_5\n",
      "Rjap_h_M_1\n",
      "Rjap_h_M_2\n",
      "Rjap_h_M_3\n",
      "Rjap_h_M_4\n",
      "Rjap_h_M_5\n",
      "Rjap_h_M_6\n",
      "Rjap_t_M_1\n",
      "Rjap_t_M_2\n",
      "Rjap_t_M_3\n",
      "Rjap_t_M_4\n",
      "Rjap_t_M_5\n",
      "Rjap_t_M_6\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-129-08f372727736>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mSL_snp_finder2\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvcfpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpopmap_path\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcatalog_tags\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mhom_thresh\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mhet_thresh\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0msample_presence_cut_off\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mcoverage_thresh\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmaf_thresh\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;31m## NEW FUNCTION\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-127-dd40d60cede7>\u001b[0m in \u001b[0;36mSL_snp_finder2\u001b[1;34m(myvcfpath, popmappath, catalog_tags_file, homogamtic_homozygosity_threshold, heterogametic_heterozygosity_threshold, sample_presence_cutoff, coverage_threshold, maf_threshold)\u001b[0m\n\u001b[0;32m    205\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    206\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 207\u001b[1;33m         \u001b[1;32melif\u001b[0m \u001b[0mrecord\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0maaf\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[0mmaf_threshold\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    208\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    209\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0msample\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrecord\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msamples\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/vcf/model.pyc\u001b[0m in \u001b[0;36maaf\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    227\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_type\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    228\u001b[0m                 \u001b[0mallele_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_alleles\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 229\u001b[1;33m                 \u001b[0mallele_counts\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_alleles\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    230\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mallele_counts\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m/\u001b[0m\u001b[0mnum_chroms\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mALT\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    231\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/vcf/model.pyc\u001b[0m in \u001b[0;36mgt_alleles\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     55\u001b[0m         \u001b[1;34m'''The numbers of the alleles called at a given sample'''\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     56\u001b[0m         \u001b[1;31m# grab the numeric alleles of the gt string; tokenize by phasing\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 57\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_nums\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_phase_char\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     58\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     59\u001b[0m     \u001b[1;33m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/vcf/model.pyc\u001b[0m in \u001b[0;36mgt_phase_char\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     49\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     50\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mgt_phase_char\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 51\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[1;34m\"/\"\u001b[0m \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mphased\u001b[0m \u001b[1;32melse\u001b[0m \u001b[1;34m\"|\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     52\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     53\u001b[0m     \u001b[1;33m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m/usr/local/lib/python2.7/dist-packages/vcf/model.pyc\u001b[0m in \u001b[0;36mphased\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     98\u001b[0m            \u001b[0mthe\u001b[0m \u001b[0mgenotype\u001b[0m \u001b[1;32mis\u001b[0m \u001b[0mphased\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mthis\u001b[0m \u001b[0msample\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     99\u001b[0m         '''\n\u001b[1;32m--> 100\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_nums\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mNone\u001b[0m \u001b[1;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgt_nums\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"|\"\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m>=\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    101\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    102\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__getitem__\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "SL_snp_finder2(vcfpath, popmap_path, catalog_tags,hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh) ## NEW FUNCTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Arvalis tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/sex_info.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.7\n",
    "hom_thresh = 0.9\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 200608\n",
      "Rarv_008\n",
      "Rarv_037\n",
      "Rarv_040\n",
      "Rarv_041\n",
      "Rarv_047\n",
      "Rarv_053\n",
      "Rarv_055\n",
      "Rarv_058\n",
      "Rarv_063\n",
      "Rarv_070\n",
      "Rarv_076\n",
      "Rarv_081\n",
      "Rarv_082\n",
      "Rarv_083\n",
      "Rarv_085\n",
      "Rarv_086\n",
      "Rarv_087\n",
      "Rarv_088\n",
      "Rarv_089\n",
      "Rarv_091\n",
      "Rarv_092\n",
      "Rarv_094\n",
      "Rarv_095\n",
      "Rarv_097\n",
      "Rarv_100\n",
      "Rarv_101\n",
      "Rarv_102\n",
      "Rarv_105\n",
      "Rarv_106\n",
      "Rarv_108\n",
      "Rarv_014\n",
      "Rarv_016\n",
      "Rarv_023\n",
      "Rarv_033\n",
      "Rarv_034\n",
      "Rarv_036\n",
      "Rarv_039\n",
      "Rarv_049\n",
      "Rarv_057\n",
      "Rarv_066\n",
      "Rarv_067\n",
      "Rarv_084\n",
      "Rarv_098\n",
      "Rarv_107\n",
      "Rarv_110\n",
      "Rarv_115\n",
      "Rarv_118\n",
      "Rarv_122\n",
      "Rarv_125\n",
      "Rarv_128\n",
      "Rarv_134\n",
      "Number of samples = 51\n",
      "Number of loci with too few samples = 89027\n",
      "Number of loci with low MAF = 76806\n",
      "Number of loci with enough data = 34774\n",
      "Number of putative X linked snps = 72\n",
      "Number of putative X linked tags = 69\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dalmatina tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/Rdal_SL_script_tests/vcf/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/Rdal_SL_script_tests/vcf/pop_codes.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.7\n",
    "hom_thresh = 0.9\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 9272\n",
      "RdalM01\n",
      "RdalM03\n",
      "RdalM04\n",
      "RdalM05\n",
      "RdalM07\n",
      "RdalM08\n",
      "RdalM13\n",
      "RdalM15\n",
      "RdalM16\n",
      "RdalM17\n",
      "RdalM19\n",
      "RdalM20\n",
      "RdalM22\n",
      "RdalM23\n",
      "RdalM24\n",
      "RdalM25\n",
      "RdalM27\n",
      "RdalM30\n",
      "RdalM32\n",
      "RdalM02\n",
      "RdalM06\n",
      "RdalM09\n",
      "RdalM10\n",
      "RdalM11\n",
      "RdalM12\n",
      "RdalM14\n",
      "RdalM18\n",
      "RdalM21\n",
      "RdalM26\n",
      "RdalM31\n",
      "RdalM28\n",
      "RdalM29\n",
      "Number of samples = 32\n",
      "Number of loci with too few samples = 1779\n",
      "Number of loci with low MAF = 2901\n",
      "Number of loci with enough data = 4591\n",
      "Number of putative X linked snps = 150\n",
      "Number of putative X linked tags = 131\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 9272\n",
      "RdalM01\n",
      "RdalM03\n",
      "RdalM04\n",
      "RdalM05\n",
      "RdalM07\n",
      "RdalM08\n",
      "RdalM13\n",
      "RdalM15\n",
      "RdalM16\n",
      "RdalM17\n",
      "RdalM19\n",
      "RdalM20\n",
      "RdalM22\n",
      "RdalM23\n",
      "RdalM24\n",
      "RdalM25\n",
      "RdalM27\n",
      "RdalM30\n",
      "RdalM32\n",
      "RdalM02\n",
      "RdalM06\n",
      "RdalM09\n",
      "RdalM10\n",
      "RdalM11\n",
      "RdalM12\n",
      "RdalM14\n",
      "RdalM18\n",
      "RdalM21\n",
      "RdalM26\n",
      "RdalM31\n",
      "RdalM28\n",
      "RdalM29\n",
      "Number of samples = 32\n",
      "Number of loci with too few samples = 1779\n",
      "Number of loci with low MAF = 2901\n",
      "Number of loci with enough data = 4591\n",
      "Number of putative X linked snps = 150\n",
      "Number of putative X linked tags = 131\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder2(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rjap  h pop only ------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/Populations_h_pop_only/Populations_h_pop_only/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/Populations_h_pop_only/Populations_h_pop_only/pop_codes_h_pop_only.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.8\n",
    "hom_thresh = 1\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 39479\n",
      "Rjap_h_F_1\n",
      "Rjap_h_F_2\n",
      "Rjap_h_F_3\n",
      "Rjap_h_F_4\n",
      "Rjap_h_F_5\n",
      "Rjap_h_M_1\n",
      "Rjap_h_M_2\n",
      "Rjap_h_M_3\n",
      "Rjap_h_M_4\n",
      "Rjap_h_M_5\n",
      "Rjap_h_M_6\n",
      "Number of samples = 11\n",
      "Number of loci with too few samples = 1657\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 37821\n",
      "Number of putative X linked snps = 61\n",
      "Number of putative X linked tags = 48\n",
      "Number of putative Z linked markers = 4\n",
      "Number of putative Z linked tags = 3\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rjap  t pop only ------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/Populations_t_pop_only/Populations_t_pop_only/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/Populations_t_pop_only/pop_codes_t_pop_only.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rjap/Stacks_outs/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.8\n",
    "hom_thresh = 1\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 46222\n",
      "Rjap_t_F_1\n",
      "Rjap_t_F_2\n",
      "Rjap_t_F_3\n",
      "Rjap_t_F_4\n",
      "Rjap_t_F_5\n",
      "Rjap_t_M_1\n",
      "Rjap_t_M_2\n",
      "Rjap_t_M_3\n",
      "Rjap_t_M_4\n",
      "Rjap_t_M_5\n",
      "Rjap_t_M_6\n",
      "Number of samples = 11\n",
      "Number of loci with too few samples = 2419\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 43802\n",
      "Number of putative X linked snps = 731\n",
      "Number of putative X linked tags = 662\n",
      "Number of putative Z linked markers = 76\n",
      "Number of putative Z linked tags = 69\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder2(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    This is potentially very interesting . . . but I need to check this thoroughly. First check to see if the male that falls out in these tests is always the same one because I do not find any loci if I require them to be hetrozygous in ALL males, I have to let one be hom."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rorn combinedG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/Populations_pops_combined/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/pop_codes_combined.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 0.8\n",
    "hom_thresh = 0.9\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 66337\n",
      "Ror_h_F_1\n",
      "Ror_h_F_2\n",
      "Ror_h_F_3\n",
      "Ror_h_F_4\n",
      "Ror_h_F_5\n",
      "Ror_h_F_6\n",
      "Ror_o_F_1\n",
      "Ror_o_F_2\n",
      "Ror_o_F_3\n",
      "Ror_o_F_4\n",
      "Ror_o_F_5\n",
      "Ror_o_F_6\n",
      "Ror_h_M_1\n",
      "Ror_h_M_2\n",
      "Ror_h_M_3\n",
      "Ror_h_M_4\n",
      "Ror_h_M_5\n",
      "Ror_h_M_6\n",
      "Ror_o_M_1\n",
      "Ror_o_M_2\n",
      "Ror_o_M_3\n",
      "Ror_o_M_4\n",
      "Ror_o_M_5\n",
      "Ror_o_M_6\n",
      "Number of samples = 24\n",
      "Number of loci with too few samples = 4137\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 62199\n",
      "Number of putative X linked snps = 6\n",
      "Number of putative X linked tags = 6\n",
      "Number of putative Z linked markers = 8\n",
      "Number of putative Z linked tags = 7\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rorn h pop only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/Populations_h_pop_only/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/pop_codes_h_pop_only.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 1\n",
    "hom_thresh = 1\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 36871\n",
      "Ror_h_F_1\n",
      "Ror_h_F_2\n",
      "Ror_h_F_3\n",
      "Ror_h_F_4\n",
      "Ror_h_F_5\n",
      "Ror_h_F_6\n",
      "Ror_h_M_1\n",
      "Ror_h_M_2\n",
      "Ror_h_M_3\n",
      "Ror_h_M_4\n",
      "Ror_h_M_5\n",
      "Ror_h_M_6\n",
      "Number of samples = 12\n",
      "Number of loci with too few samples = 962\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 35908\n",
      "Number of putative X linked snps = 70\n",
      "Number of putative X linked tags = 59\n",
      "Number of putative Z linked markers = 80\n",
      "Number of putative Z linked tags = 66\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder2(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rorn o Pop only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcfpath = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/Populations_o_pop_only/batch_1.vcf\"\n",
    "\n",
    "popmap_path = \"/home/djeffrie/Data/RADseq/Lib_Rjapor/Rorn/Stacks_outs/pop_codes_o_pop_only.txt\"\n",
    "catalog_tags = \"/home/djeffrie/Data/Rarvalis_SL_script_tests/stacks/batch_1.catalog.tags.tsv.gz\"\n",
    "het_thresh = 1\n",
    "hom_thresh = 1\n",
    "sample_presence_cut_off = 0.75\n",
    "coverage_thresh = 6\n",
    "maf_thresh = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci = 57369\n",
      "Ror_o_F_1\n",
      "Ror_o_F_2\n",
      "Ror_o_F_3\n",
      "Ror_o_F_4\n",
      "Ror_o_F_5\n",
      "Ror_o_F_6\n",
      "Ror_o_M_1\n",
      "Ror_o_M_2\n",
      "Ror_o_M_3\n",
      "Ror_o_M_4\n",
      "Ror_o_M_5\n",
      "Ror_o_M_6\n",
      "Number of samples = 12\n",
      "Number of loci with too few samples = 1551\n",
      "Number of loci with low MAF = 0\n",
      "Number of loci with enough data = 55817\n",
      "Number of putative X linked snps = 13\n",
      "Number of putative X linked tags = 12\n",
      "Number of putative Z linked markers = 33\n",
      "Number of putative Z linked tags = 27\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SL_snp_finder(vcfpath, popmap_path, catalog_tags, hom_thresh, het_thresh, sample_presence_cut_off, coverage_thresh, maf_thresh)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
