{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import vcf\n",
    "import matplotlib\n",
    "#matplotlib.use('Agg') ## this allows the drawing of plots in matplotlib on the cluster, which doesn't use the X-server backend. This has something to do with display (but I don't know what)\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os.path\n",
    "import sys\n",
    "import time\n",
    "import gzip\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "####Workflow:\n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci identified as X or Z linked are labelled as such in this file.\n",
    "    6. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    7. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    8. All suplus information is recorded to a log file, with a summary at the end of this file. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## Get arguments from the command line\n",
    "\n",
    "Usage_message = \"\\n##USAGE (args in this order):\\nSL_snp_finder.py <path/to/.vcf> <path/to/pop_map.txt> <path_to_catalog.tags.tsv_file> <female-male_cutoff> <sample_presence_cutoff> <coverage_per_sample_cutoff> <maf_cutoff>\\n\\n##All paths should be absolute, not relative\\n\\n\"\n",
    "\n",
    "if len(sys.argv) == 1:\n",
    "    sys.exit(Usage_message)\n",
    "\n",
    "elif len(sys.argv) < 8: ## If not enough args are supplied print error message\n",
    "    sys.exit(\"\\n##Error, not enough arguments\\n\"+Usage_message)\n",
    "\n",
    "elif len(sys.argv) == 8:\n",
    "    myvcfpath = sys.argv[1]\n",
    "    alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "    popmappath = sys.argv[2]\n",
    "    catalog_tags_file = sys.argv[3]\n",
    "    X_or_Z_freq_threshold = float(sys.argv[4])\n",
    "    sample_presence_cutoff = float(sys.argv[5])\n",
    "    coverage_threshold = int(sys.argv[6])\n",
    "    maf_threshold = float(sys.argv[7])\n",
    "    \n",
    "else:\n",
    "    sys.exit(\"Unknown Error\\n\"+Usage_message)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "myvcfpath = \"/media/djeffrie/OS/Users/djeffrie/Data/Crucian_carp/Crucian_sex_chroms/populations_outs/batch_1.vcf\"\n",
    "alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "popmappath = \"/media/djeffrie/OS/Users/djeffrie/Data/Crucian_carp/Crucian_sex_chroms/sex_info.txt\"\n",
    "catalog_tags_file = \"/media/djeffrie/OS/Users/djeffrie/Data/Crucian_carp/Crucian_sex_chroms/batch_1.catalog.tags.tsv.gz\"\n",
    "X_or_Z_freq_threshold = 0.4\n",
    "sample_presence_cutoff = 0.75\n",
    "coverage_threshold = 6\n",
    "maf_threshold = 0.05"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SWE10_1_1.fil\n",
      "SWE10_2_1.fil\n",
      "SWE10_3_1.fil\n",
      "SWE10_4_1.fil\n",
      "SWE10_5_1.fil\n",
      "SWE10_6_1.fil\n",
      "SWE10_7_1.fil\n",
      "SWE10_8_1.fil\n",
      "SWE10_9_1.fil\n",
      "SWE14_10_1.fil\n",
      "SWE14_11_1.fil\n",
      "SWE14_13_1.fil\n",
      "SWE14_16_1.fil\n",
      "SWE14_2_1.fil\n",
      "SWE14_3_1.fil\n",
      "SWE14_4_1.fil\n",
      "SWE14_6_1.fil\n",
      "SWE14_8_1.fil\n",
      "SWE8_10_RD-P1-166_1.fil\n",
      "SWE8_1_RD-P1-128_1.fil\n",
      "SWE8_2_RD-P1-167_1.fil\n",
      "SWE8_3_RD-P1-169_1.fil\n",
      "SWE8_4_RD-P1-193_1.fil\n",
      "SWE8_5_RD-P1-195_1.fil\n",
      "SWE8_6_RD-P1-101_1.fil\n",
      "SWE8_7_RD-P1-104_1.fil\n",
      "SWE8_8_RD-P1-129_1.fil\n",
      "SWE8_9_RD-P1-130_1.fil\n",
      "SWE9_10_RD-P1-149_1.fil\n",
      "SWE9_1_RD-P1-147_1.fil\n",
      "SWE9_2_RD-P1-153_1.fil\n",
      "SWE9_3_RD-P1-154_1.fil\n",
      "SWE9_4_RD-P1-182_1.fil\n",
      "SWE9_5_RD-P1-185_1.fil\n",
      "SWE9_6_RD-P1-113_1.fil\n",
      "SWE9_7_RD-P1-115_1.fil\n",
      "SWE9_8_RD-P1-141_1.fil\n",
      "SWE9_9_RD-P1-142_1.fil\n",
      "Number of loci = 9108\n",
      "SWE10_1_1.fil\n",
      "SWE10_3_1.fil\n",
      "SWE10_4_1.fil\n",
      "SWE10_7_1.fil\n",
      "SWE10_9_1.fil\n",
      "SWE14_16_1.fil\n",
      "SWE14_2_1.fil\n",
      "SWE14_3_1.fil\n",
      "SWE14_4_1.fil\n",
      "SWE14_8_1.fil\n",
      "SWE8_10_RD-P1-166_1.fil\n",
      "SWE8_1_RD-P1-128_1.fil\n",
      "SWE8_6_RD-P1-101_1.fil\n",
      "SWE8_8_RD-P1-129_1.fil\n",
      "SWE9_10_RD-P1-149_1.fil\n",
      "SWE9_1_RD-P1-147_1.fil\n",
      "SWE9_2_RD-P1-153_1.fil\n",
      "SWE9_4_RD-P1-182_1.fil\n",
      "SWE9_5_RD-P1-185_1.fil\n",
      "SWE9_7_RD-P1-115_1.fil\n",
      "SWE9_8_RD-P1-141_1.fil\n",
      "SWE9_9_RD-P1-142_1.fil\n",
      "SWE10_2_1.fil\n",
      "SWE10_5_1.fil\n",
      "SWE10_6_1.fil\n",
      "SWE10_8_1.fil\n",
      "SWE14_10_1.fil\n",
      "SWE14_11_1.fil\n",
      "SWE14_13_1.fil\n",
      "SWE14_6_1.fil\n",
      "SWE8_2_RD-P1-167_1.fil\n",
      "SWE8_3_RD-P1-169_1.fil\n",
      "SWE8_4_RD-P1-193_1.fil\n",
      "SWE8_5_RD-P1-195_1.fil\n",
      "SWE8_7_RD-P1-104_1.fil\n",
      "SWE8_9_RD-P1-130_1.fil\n",
      "SWE9_3_RD-P1-154_1.fil\n",
      "SWE9_6_RD-P1-113_1.fil\n",
      "Number of samples = 38\n",
      "Number of loci with too few samples = 1152\n",
      "Number of loci with low MAF = 2011\n",
      "Number of loci with enough data = 5944\n",
      "Number of putative X linked snps = 0\n",
      "Number of putative X linked tags = 0\n",
      "Number of putative Z linked markers = 0\n",
      "Number of putative Z linked tags = 0\n",
      "\n",
      "***DONE!***\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "## set the window around the freq threshold. The window automatically tightens and relaxes around 0.5 or -0.5 \n",
    "\n",
    "lower_thresh = X_or_Z_freq_threshold\n",
    "upper_thresh = 0.5 + (0.5 - X_or_Z_freq_threshold)\n",
    "\n",
    "\n",
    "# First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue.\n",
    "\n",
    "myvcf = open(myvcfpath, 'r').readlines()\n",
    "alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "for line in myvcf:\n",
    "    if \"Allele Depth\" not in line:\n",
    "        alteredvcf.write(line)\n",
    "    elif \"Allele Depth\" in line:\n",
    "        line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "        alteredvcf.write(line)\n",
    "alteredvcf.close()\n",
    "\n",
    "\n",
    "# ### Now calculate allele frequencies for males and females at each SNP\n",
    "# ####Requires:\n",
    "#     1. pyvcf module installed (can use pip, remember to add to python path. This is on the cluster!\n",
    "#     2. altered vcf file from above\n",
    "#     3. pop_map.txt file. Same format as used for stacks. Sample names must be the same. And males and females must be denoted by M or F (case sensitive) respectively. Must be the same file as used in populations to creat the VCF. If there are additional samples in this file the allele frequencies will be wrong!\n",
    "\n",
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "\n",
    "## Open file for all frequency and locus info for all kept genotypes\n",
    "all_frequencies = open(\"%s%s\" %(myvcfpath, \".all_frequencies.tsv\"), 'w')\n",
    "all_frequencies.write(\"#Locus_ID\\tPOS\\tN_MaleChroms\\tN_FemChroms\\tMalefreq_REF\\tMalefreq_ALT\\tFemalefreq_REF\\tFemalefreq_ALT\\tFemREF-MaleREF\\tSex_linked\\n\")\n",
    "\n",
    "## Open files to output X or Z linked loci to and add headers\n",
    "\n",
    "Putative_Xlinked_makers = []\n",
    "\n",
    "Putative_Zlinked_makers = []\n",
    "\n",
    "## Make a log file for all auxillary info\n",
    "freq_ratios_log = open(\"%s%s\" %(myvcfpath, \".freq_ratios.log\"), 'w')\n",
    "freq_ratios_log.write(\"Script run on %s\\n \" % (time.strftime(\"%c\")))\n",
    "\n",
    "## Make list to keep frequency information in for histogram\n",
    "freq_ratio_data = []\n",
    "\n",
    "## get male/female info for samples for working out frequencies below38      3460    50      26      0.940   0.060   0.846   0.154   -0.094  NotSexLinked\n",
    "\n",
    "pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "sample_dict = {}\n",
    "sample_dict[\"females\"] = []\n",
    "sample_dict[\"males\"] = []\n",
    "fem_samples = 0\n",
    "male_samples = 0\n",
    "sample_counter = 0\n",
    "low_maf_counter = 0\n",
    "kept_loci = 0\n",
    "sample_missing_dict = {}\n",
    "sample_cov_dict = {}\n",
    "sample_cov_kept_dict = {}\n",
    "\n",
    "for sample in pop_map:\n",
    "    name = sample.split()[0]\n",
    "    print name\n",
    "    sex = sample.strip().split()[1]\n",
    "    sample_missing_dict[name] = 0\n",
    "\n",
    "    #sample_counter += 1\n",
    "    if sex == \"F\":\n",
    "        sample_dict[\"females\"].append(name)\n",
    "        fem_samples += 1\n",
    "    elif sex == \"M\":\n",
    "        sample_dict[\"males\"].append(name)\n",
    "        male_samples += 1\n",
    "\n",
    "#print sample_dict        \n",
    "locus_dict = {}\n",
    "\n",
    "\n",
    "## Find number of loci in input file\n",
    "locus_counter = 0\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    locus_counter += 1\n",
    "print \"Number of loci = %s\" % (locus_counter)\n",
    "    \n",
    "## Write some general stats and input options to the log file\n",
    "\n",
    "freq_ratios_log.write(\"\\n## User specified options:\\n\")\n",
    "freq_ratios_log.write(\"Input vcf: %s\\n\" % (myvcfpath))\n",
    "freq_ratios_log.write(\"pop_map file used: %s\\n\" % (popmappath))\n",
    "freq_ratios_log.write(\"Catalog file used: %s\\n\" % (catalog_tags_file))\n",
    "freq_ratios_log.write(\"Specified female-male threshold: %s\\n\" % (X_or_Z_freq_threshold))\n",
    "freq_ratios_log.write(\"Female-male thresh range: (+/-) %s-%s\\n\" % (lower_thresh, upper_thresh))\n",
    "freq_ratios_log.write(\"Min percentage samples present: %s\\n\" % (sample_presence_cutoff))\n",
    "freq_ratios_log.write(\"Min coverage per genotype: %s\\n\" % (coverage_threshold))\n",
    "freq_ratios_log.write(\"Min maf per locus: %s\\n\" % (maf_threshold))\n",
    "freq_ratios_log.write(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "freq_ratios_log.write(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "freq_ratios_log.write(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "                      \n",
    "\n",
    "## Some quick counters\n",
    "\n",
    "numb_putative_Xlinked = 0\n",
    "numb_putative_Zlinked = 0\n",
    "low_data_loci = 0\n",
    "\n",
    "vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "info_rec = vcf_reader.next()\n",
    "\n",
    "for sample in info_rec:\n",
    "    name = sample.sample\n",
    "    print name\n",
    "    sample_cov_dict[name] = []\n",
    "    sample_cov_kept_dict[name] = []\n",
    "\n",
    "\n",
    "for record in vcf_reader:\n",
    "    femREF_count = 0    ## set the counters for the reference and alternative allele (encoded as 0 in the vcf)\n",
    "    femALT_count = 0\n",
    "    malREF_count = 0\n",
    "    malALT_count = 0   \n",
    "    fem_none_count = 0\n",
    "    male_none_count = 0\n",
    "    low_cov_samples = 0\n",
    "    n_genotypes = 0\n",
    "    male_genotypes = 0\n",
    "    fem_genotypes = 0\n",
    "    number_of_samples = len(record.samples)\n",
    "    \n",
    "    \n",
    "    for sample in record.samples:\n",
    "        \n",
    "        name = sample.sample  \n",
    "        sample_cov_dict[name].append(sample['DP'])\n",
    "    \n",
    "    \n",
    "    if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "        low_maf_counter += 1\n",
    "        freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "        freq_ratios_log.write(\"Minor allele frequence of locus is lower than specified cutoff (%s)\\n\" % (record.aaf[0]))\n",
    "        pass  \n",
    "    \n",
    "    elif record.aaf >= maf_threshold:\n",
    "    \n",
    "        for sample in record.samples: \n",
    "            \n",
    "            name = sample.sample\n",
    "            \n",
    "            ### For each sample, if the coverage is too low, remove the genotype for that individual.\n",
    "            \n",
    "            \n",
    "            if sample['DP'] < coverage_threshold:\n",
    "                genotype = None\n",
    "                low_cov_samples += 1\n",
    "                #freq_ratios_log.write(\"Sample %s thrown out due to low coverage (%s)\\n\" % (name, sample['DP']))\n",
    "                \n",
    "            elif sample['DP'] >= coverage_threshold:\n",
    "                genotype = sample['GT']\n",
    "                sample_cov_kept_dict[name].append(sample['DP'])\n",
    "            ## Now calculate the female and male frequencies separately\n",
    "            \n",
    "            if name in sample_dict[\"females\"]: \n",
    "                #print \"Locus=\", record.ID, \"Female=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                #print \"FEMALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    sample_missing_dict[name] += 1\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    femREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    femREF_count += 1\n",
    "                    femALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    femALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    fem_genotypes +=1\n",
    "                #print \"\\tN_REF=\", femREF_count, \"N_ALT=\", femALT_count, \"Nnone=\", fem_none_count\n",
    "            elif name in sample_dict[\"males\"]:\n",
    "                #print \"Locus=\", record.ID, \"Male=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                #print \"MALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                if genotype == None: ## if no genotype exists\n",
    "                    sample_missing_dict[name] += 1\n",
    "                    pass\n",
    "                elif genotype == \"0/0\":\n",
    "                    malREF_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"0/1\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/0\":\n",
    "                    malREF_count += 1\n",
    "                    malALT_count += 1\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "                elif genotype == \"1/1\":\n",
    "                    malALT_count += 2\n",
    "                    n_genotypes +=1\n",
    "                    male_genotypes +=1\n",
    "            #else:\n",
    "                #print \"\\n##SAMPLE NAME NOT IN POP_MAP.TXT: Sample = %s\" % (name)\n",
    "                #print \"\\tN_REF=\", malREF_count, \"N_ALT=\", malALT_count, \"Nnone=\", male_none_count\n",
    "                \n",
    "        ## Filter loci that have too many missing samples, including samples thrown out due to low coverage!\n",
    "        samples_at_locus = n_genotypes\n",
    "        chromosomes_at_locus = n_genotypes*2\n",
    "        percent_samples_present = n_genotypes/number_of_samples\n",
    "        #print \"Locus\", record.ID\n",
    "        #print samples_at_locus\n",
    "        #print percent_samples_present\n",
    "        \n",
    "        if percent_samples_present >= sample_presence_cutoff:\n",
    "            kept_loci += 1\n",
    "            ## Calculate frequencies\n",
    "            \n",
    "            femREF_freq = femREF_count/(fem_genotypes*2)\n",
    "            femALT_freq = femALT_count/(fem_genotypes*2)\n",
    "            \n",
    "            maleREF_freq = malREF_count/(male_genotypes*2)\n",
    "            maleALT_freq = malALT_count/(male_genotypes*2)\n",
    "                                \n",
    "                        \n",
    "            ## Output female stats\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s\\n\\n\" %(record.ID))   \n",
    "            freq_ratios_log.write(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "            freq_ratios_log.write(\"Female reference count = %s\\n\" % (femREF_count))\n",
    "            freq_ratios_log.write(\"Female alternative count = %s\\n\" % (femALT_count))\n",
    "            freq_ratios_log.write(\"Female reference frequency = %.3f\\n\" % (femREF_freq))\n",
    "            freq_ratios_log.write(\"Female alternative frequency = %.3f\\n\" % (femALT_freq))\n",
    "                                  \n",
    "            ## check fem freqs\n",
    "            if not (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (femREF_freq) + (femALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (femREF_freq + femALT_freq))\n",
    "        \n",
    "            ## Output male stats\n",
    "            freq_ratios_log.write(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "            freq_ratios_log.write(\"Male reference count = %s\\n\" % (malREF_count))\n",
    "            freq_ratios_log.write(\"Male alternative count = %s\\n\" % (malALT_count))\n",
    "            freq_ratios_log.write(\"Male reference frequency = %.3f\\n\" % (maleREF_freq))\n",
    "            freq_ratios_log.write(\"Male alternative frequency = %.3f\\n\" % (maleALT_freq))\n",
    "                                  \n",
    "            ## check male freqs\n",
    "            if not (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "            elif (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                freq_ratios_log.write(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (maleREF_freq + maleALT_freq))\n",
    "        \n",
    "            locus_dict[record.ID] = {}\n",
    "            locus_dict[record.ID][\"female_freqs\"] = [\"%.3f\" % (femREF_freq), \"%.3f\" % (femALT_freq)]\n",
    "            locus_dict[record.ID][\"male_freqs\"] = [\"%.3f\" % (maleREF_freq), \"%.3f\" % (maleALT_freq)]\n",
    "            \n",
    "\n",
    "            ####### ==============================================================================================\n",
    "            ### So now I have the allele frequencies for males and females, I can subtract them and see if the distribution fits the expectition for X linked or Z linked!\n",
    "            \n",
    "            freq_ratio = femREF_freq - maleREF_freq\n",
    "            #print freq_ratio\n",
    "            freq_ratio_data.append(freq_ratio)\n",
    "            \n",
    "\n",
    "            ## Write files for X or Z linked loci\n",
    "            \n",
    "            if freq_ratio >= lower_thresh and freq_ratio <= upper_thresh and femREF_freq >= 0.95:  ## for X linked\n",
    "                linked_status = \"Xlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Xlinked += 1\n",
    "            \n",
    "            elif freq_ratio >= -upper_thresh and freq_ratio <= -lower_thresh and maleREF_freq >= 0.95:  ## for Z linked\n",
    "                linked_status = \"Zlinked\"\n",
    "                freq_ratios_log.write(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                numb_putative_Zlinked += 1\n",
    "            else:\n",
    "                freq_ratios_log.write(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                linked_status =\"NotSexLinked\"\n",
    "        \n",
    "            ## Write the main info file for male and female frequencies, ratios etc.\n",
    "            \n",
    "            all_frequencies.write(\"%s\\t%s\\t%s\\t%s\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%s\\n\" % (record.ID, record.POS, male_genotypes*2, fem_genotypes*2, maleREF_freq, maleALT_freq, femREF_freq, femALT_freq, freq_ratio, linked_status))\n",
    "            \n",
    "        \n",
    "        elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "            freq_ratios_log.write(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            freq_ratios_log.write(\"Number of samples at locus is lower than sample presence cutoff (%s)\\n\" % (samples_at_locus))\n",
    "            low_data_loci += 1\n",
    "    \n",
    "## Look at coverage...\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "## Print some quick summary stats - also written at the end of the log file\n",
    "print \"Number of samples =\", number_of_samples\n",
    "print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "print \"Number of putative X linked snps = %s\" % (numb_putative_Xlinked)\n",
    "print \"Number of putative X linked tags = %s\" % (len(set(Putative_Xlinked_makers)))\n",
    "print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "print \"Number of putative Z linked tags = %s\" % (len(set(Putative_Zlinked_makers)))\n",
    "\n",
    "freq_ratios_log.write(\"\\nSUMMARY....\\n\\n\")\n",
    "freq_ratios_log.write(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "freq_ratios_log.write(\"Number of loci with enough data = %s\\n\" % (kept_loci))\n",
    "freq_ratios_log.write(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "freq_ratios_log.write(\"Number of putative X linked snps = %s\\n\" % (numb_putative_Xlinked))\n",
    "freq_ratios_log.write(\"Number of putative X linked tags = %s\\n\" % (len(set(Putative_Xlinked_makers))))\n",
    "freq_ratios_log.write(\"Number of putative Z linked snps = %s\\n\" % (numb_putative_Zlinked))\n",
    "freq_ratios_log.write(\"Number of putative Z linked tags = %s\\n\" % (len(set(Putative_Zlinked_makers))))\n",
    "\n",
    "\n",
    "## plot histogram of frequency ratios\n",
    "\n",
    "plt.hist(freq_ratio_data, bins = 100, color = '0.5')\n",
    "plt.xlabel('Female_REFfreq - Male_REFfreq', fontsize=8)\n",
    "plt.ylabel(\"Number of SNPs\", fontsize=8)\n",
    "plt.axvline(x = np.mean(freq_ratio_data), color='r', linestyle='dashed', linewidth = 2)\n",
    "plt.axvline(x = X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "plt.axvline(x = -X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "plt.savefig(\"%s%s\" %(myvcfpath, \".fem-male_freqs.pdf\"), format = 'pdf')\n",
    "plt.close()\n",
    "#plt.show()\n",
    "\n",
    "\n",
    "## plot histogram of missing data per sample\n",
    "\n",
    "plt.bar(range(len(sample_missing_dict)), sample_missing_dict.values(), align='center', color = '0.5')\n",
    "plt.xticks(range(len(sample_missing_dict)), sample_missing_dict.keys(), rotation = 90, fontsize=8)\n",
    "plt.ylabel(\"Number of loci with missing data\", fontsize=8)\n",
    "plt.savefig(\"%s%s\" %(myvcfpath, \".missing_data_by_sample.pdf\"), format = 'pdf')\n",
    "plt.close()\n",
    "#plt.show()\n",
    "\n",
    "## Plot coverage per sample\n",
    "\n",
    "n_groups = len(sample_cov_dict)\n",
    "\n",
    "Xlabs = []\n",
    "all_means = []\n",
    "all_std = []\n",
    "\n",
    "kept_means = []\n",
    "kept_std = []\n",
    "\n",
    "for sample in sorted(sample_cov_dict.keys()):\n",
    "    Xlabs.append(sample)\n",
    "    all_means.append(np.mean(sample_cov_dict[sample]))\n",
    "    all_std.append(np.std(sample_cov_dict[sample]))\n",
    "for sample in sample_cov_kept_dict.keys():\n",
    "    kept_means.append(np.mean(sample_cov_kept_dict[sample]))\n",
    "    kept_std.append(np.std(sample_cov_kept_dict[sample]))\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "index = np.arange(n_groups)\n",
    "bar_width = 0.35\n",
    "\n",
    "opacity = 0.4\n",
    "error_config = {'ecolor': '0.3'}\n",
    "\n",
    "rects1 = plt.bar(index, all_means, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color='0.5',\n",
    "                 #yerr=all_std,\n",
    "                 error_kw=error_config,\n",
    "                 label='All')\n",
    "\n",
    "rects2 = plt.bar(index + bar_width, kept_means, bar_width,\n",
    "                 alpha=opacity,\n",
    "                 color='0',\n",
    "                 #yerr=kept_std,\n",
    "                 error_kw=error_config,\n",
    "                 label='Kept')\n",
    "\n",
    "plt.xlabel('Sample')\n",
    "plt.ylabel('Mean coverage')\n",
    "plt.title('Average coverage per sample')\n",
    "plt.xticks(index + bar_width, (Xlabs), rotation = 90, size = 8)\n",
    "plt.legend()\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"%s%s\" %(myvcfpath, \".coverage_by_sample.pdf\"), format = 'pdf')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Write fasta files of putative X or Z linked loci if there are any\n",
    "\n",
    "Putative_Xlinked_makers = set(Putative_Xlinked_makers)\n",
    "Putative_Zlinked_makers = set(Putative_Zlinked_makers)\n",
    "\n",
    "if catalog_tags_file.endswith(\"gz\"):\n",
    "    catalog = gzip.open(catalog_tags_file, 'r').readlines()\n",
    "else:\n",
    "    catalog = open(catalog_tags_file, 'r').readlines()\n",
    "\n",
    "if numb_putative_Xlinked > 0:\n",
    "    Putative_Xlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Xlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Xlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Xlinked_makers_file.write(\">X_linkedLocusID_%s\\n\" % (locus))\n",
    "                Putative_Xlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Xlinked_makers_file.close()\n",
    "\n",
    "if numb_putative_Zlinked > 0:\n",
    "    Putative_Zlinked_makers_file = open(\"%s%s\" %(myvcfpath, \".Putative_Zlinked_makers.fa\"), 'w')\n",
    "    \n",
    "    for locus in Putative_Zlinked_makers:\n",
    "        for tag in catalog:\n",
    "            if locus == tag.split()[2]:\n",
    "                Putative_Zlinked_makers_file.write(\">Z_linked|LocusID_%s\\n\" % (locus))\n",
    "                Putative_Zlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "    Putative_Zlinked_makers_file.close()\n",
    "\n",
    "\n",
    "## close all unclosed files \n",
    "freq_ratios_log.close()\n",
    "all_frequencies.close()\n",
    "\n",
    "            \n",
    "           \n",
    "\n",
    "print \"\\n***DONE!***\\n\"\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[35.980815853272901, 32.452935596078845, 31.53683988616001, 28.29598397807526, 13.059344365974491, 16.847053863181195, 17.997154000210816, 0.38589648993359332, 33.993780963423632, 31.564562032254663, 19.745968166965319, 22.12511858332455, 28.709497206703912, 19.644777063349846, 14.860756825129124, 22.606303362496046, 36.792347422789078, 24.166543691367135, 30.206598503214927, 9.1648571729735426, 39.040054811847789, 32.570464846632234, 37.747022240961314, 39.462316854643198, 22.799936755560239, 28.239801834088752, 40.818593865289344, 34.209866132602507, 34.402867081269108, 27.799831348160641, 31.577421735005796, 44.87066512069147, 19.789501422999894, 26.295351533677664, 36.824075050068515, 34.528512701591652]\n",
      "[33.214922048997771, 30.349659090909093, 22.877109638704837, 29.671615720524017, 25.284648318042812, 23.193898134140191, 37.261904761904759, 33.075583270616065, 46.075449231435378, 27.320365458296493, 27.296422300263387, 38.519367333763718, 33.709716157205243, 35.630705394190869, 29.193297825122059, 32.522707518470227, 39.863709764237271, 21.888308387461169, 28.172250065945661, 34.585685159351861, 29.575246132208157, 20.252793088038135, 19.09853012437409, 13.070652173913043, 22.510667026735078, 32.934567085261072, 35.904443225452553, 23.126123672023972, 37.932551319648091, 35.370543987790256, 30.921814254859612, 30.885217391304348, 37.495863328677338, 40.227727028479315, 28.719133848296821, 41.954707985697262]\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
