{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####This script defines the Super_SLM_finder function, which combines the three approaches for finding sex linked markers described in Brelsford & Lavanchy et al (2016)\n",
    "\n",
    "The three appraches are:\n",
    "    1. SNP Frequency difference between sexes\n",
    "    2. SNP Heterozygosity differences between sexes\n",
    "    3. RADtag presence or absence between sexes\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "% matplotlib inline\n",
    "import sys\n",
    "import os\n",
    "import gzip\n",
    "import linecache\n",
    "import pprint as pp\n",
    "import math\n",
    "from matplotlib import pyplot as plt\n",
    "from __future__ import division"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def SL_snp_freq(myvcfpath, popmappath, catalog_tags_file, X_or_Z_freq_threshold = 0.4, sample_presence_cutoff = 0.75, coverage_threshold = 3, maf_threshold = 0.05):    \n",
    "   \n",
    "    \"\"\"\n",
    "    This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "    Arguments:\n",
    "    \n",
    "    myvcfpath                     - path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name \n",
    "                                    with \".altered\" appended to the end)\n",
    "    \n",
    "    popmappath                    - path to population map file containing sex information. Same format as Stacks pop map file.\n",
    "    \n",
    "    catalog_tags_file             - The catalog tags file used to create the vcf\n",
    "    \n",
    "    X_or_Z_freq_threshold         - The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold\n",
    "                                    of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be\n",
    "                                    the same distance above 0.5 as the lower threshold is below 0.5)\n",
    "    \n",
    "    sample_presence_cutoff        - a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "    \n",
    "    coverage_threshold            - a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this \n",
    "                                    threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will\n",
    "                                    then remove the locus.\n",
    "    \n",
    "    maf_threshold                 - minor allele frequency cutoff for a locus across all samples. \n",
    "    \n",
    "    \n",
    "    Workflow:\n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \n",
    "       \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci \n",
    "       identified as X or Z linked are labelled as such in this file.\n",
    "    5. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    6. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    7. All suplus information is recorded to a log file, with a summary at the end of this file.\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    import vcf\n",
    "    import matplotlib\n",
    "    #matplotlib.use('Agg') ## this allows the drawing of plots in matplotlib on the cluster, which doesn't use the X-server backend. This has something to do with display (but I don't know what)\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy as np\n",
    "    import os.path\n",
    "    import sys\n",
    "    import time\n",
    "    import gzip\n",
    "    \n",
    "    \n",
    "    print \"\\n##### Using SNP frequency approach #### \\n\"\n",
    "    \n",
    "    ## set the window around the freq threshold. The window automatically tightens and relaxes around 0.5 or -0.5 \n",
    "\n",
    "    lower_thresh = X_or_Z_freq_threshold\n",
    "    upper_thresh = 1 - X_or_Z_freq_threshold\n",
    "\n",
    "\n",
    "    # First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue.\n",
    "\n",
    "    myvcf = open(myvcfpath, 'r').readlines()\n",
    "    \n",
    "    alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "    alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "    for line in myvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "\n",
    "\n",
    "    # ### Now calculate allele frequencies for males and females at each SNP\n",
    "    # ####Requires:\n",
    "    #     1. pyvcf module installed (can use pip, remember to add to python path. This is on the cluster!\n",
    "    #     2. altered vcf file from above\n",
    "    #     3. pop_map.txt file. Same format as used for stacks. Sample names must be the same. And males and females must be denoted by M or F (case sensitive) respectively. Must be the same file as used in populations to creat the VCF. If there are additional samples in this file the allele frequencies will be wrong!\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "\n",
    "    ## Open file for all frequency and locus info for all kept genotypes\n",
    "    all_frequencies = open(\"%s%s\" %(myvcfpath, \".all_frequencies.tsv\"), 'w')\n",
    "    all_frequencies.write(\"#Locus_ID\\tPOS\\tN_MaleChroms\\tN_FemChroms\\tMalefreq_REF\\tMalefreq_ALT\\tFemalefreq_REF\\tFemalefreq_ALT\\tFemREF-MaleREF\\tSex_linked\\n\")\n",
    "\n",
    "    ## Open files to output X or Z linked loci to and add headers\n",
    "\n",
    "    Putative_Xlinked_makers = []\n",
    "\n",
    "    Putative_Zlinked_makers = []\n",
    "\n",
    "    ## Make a log file for all auxillary info\n",
    "    freq_ratios_log = []\n",
    "    #open(\"%s%s\" %(myvcfpath, \".freq_ratios.log\"), 'w')\n",
    "    freq_ratios_log.append(\"Script run on %s\\n \" % (time.strftime(\"%c\")))\n",
    "\n",
    "    ## Make list to keep frequency information in for histogram\n",
    "    freq_ratio_data = []\n",
    "\n",
    "    ## get male/female info for samples for working out frequencies below38      3460    50      26      0.940   0.060   0.846   0.154   -0.094  NotSexLinked\n",
    "\n",
    "    pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "    sample_dict = {}\n",
    "    sample_dict[\"females\"] = []\n",
    "    sample_dict[\"males\"] = []\n",
    "    fem_samples = 0\n",
    "    male_samples = 0\n",
    "    sample_counter = 0\n",
    "    low_maf_counter = 0\n",
    "    kept_loci = 0\n",
    "    sample_missing_dict = {}\n",
    "    sample_cov_dict = {}\n",
    "    sample_cov_kept_dict = {}\n",
    "\n",
    "    for sample in pop_map:\n",
    "        name = sample.split()[0]\n",
    "        sex = sample.strip().split()[1]\n",
    "        sample_missing_dict[name] = 0\n",
    "\n",
    "        #sample_counter += 1\n",
    "        if sex == \"F\" or sex == \"f\" or sex == \"Female\" or sex == \"female\" or sex == \"Fem\" or sex == \"fem\":\n",
    "            sample_dict[\"females\"].append(name)\n",
    "            fem_samples += 1\n",
    "        elif sex == \"M\" or sex == \"m\" or sex == \"Male\" or sex == \"male\" or sex == \"Mal\" or sex == \"mal\":\n",
    "            sample_dict[\"males\"].append(name)\n",
    "            male_samples += 1\n",
    "\n",
    "    #print sample_dict        \n",
    "    locus_dict = {}\n",
    "\n",
    "\n",
    "    ## Find number of loci in input file\n",
    "    locus_counter = 0\n",
    "\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        locus_counter += 1\n",
    "    print \"Number of loci = %s\" % (locus_counter)\n",
    "\n",
    "    ## Write some general stats and input options to the log file\n",
    "\n",
    "    freq_ratios_log.append(\"\\n## User specified options:\\n\")\n",
    "    freq_ratios_log.append(\"Input vcf: %s\\n\" % (myvcfpath))\n",
    "    freq_ratios_log.append(\"pop_map file used: %s\\n\" % (popmappath))\n",
    "    freq_ratios_log.append(\"Catalog file used: %s\\n\" % (catalog_tags_file))\n",
    "    freq_ratios_log.append(\"Specified female-male threshold: %s\\n\" % (X_or_Z_freq_threshold))\n",
    "    freq_ratios_log.append(\"Female-male thresh range: (+/-) %s-%s\\n\" % (lower_thresh, upper_thresh))\n",
    "    freq_ratios_log.append(\"Min percentage samples present: %s\\n\" % (sample_presence_cutoff))\n",
    "    freq_ratios_log.append(\"Min coverage per genotype: %s\\n\" % (coverage_threshold))\n",
    "    freq_ratios_log.append(\"Min maf per locus: %s\\n\" % (maf_threshold))\n",
    "    freq_ratios_log.append(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "    freq_ratios_log.append(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "    freq_ratios_log.append(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "\n",
    "\n",
    "    ## Some quick counters\n",
    "\n",
    "    numb_putative_Xlinked = 0\n",
    "    numb_putative_Zlinked = 0\n",
    "    low_data_loci = 0\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "    info_rec = vcf_reader.next()\n",
    "\n",
    "    for sample in info_rec:\n",
    "        name = sample.sample\n",
    "        sample_cov_dict[name] = []\n",
    "        sample_cov_kept_dict[name] = []\n",
    "\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        femREF_count = 0    ## set the counters for the reference and alternative allele (encoded as 0 in the vcf)\n",
    "        femALT_count = 0\n",
    "        malREF_count = 0\n",
    "        malALT_count = 0   \n",
    "        fem_none_count = 0\n",
    "        male_none_count = 0\n",
    "        low_cov_samples = 0\n",
    "        n_genotypes = 0\n",
    "        male_genotypes = 0\n",
    "        fem_genotypes = 0\n",
    "        number_of_samples = len(record.samples)\n",
    "\n",
    "\n",
    "        for sample in record.samples:\n",
    "\n",
    "            name = sample.sample  \n",
    "            sample_cov_dict[name].append(sample['DP'])\n",
    "\n",
    "\n",
    "        if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "            low_maf_counter += 1\n",
    "            freq_ratios_log.append(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            freq_ratios_log.append(\"Minor allele frequence of locus is lower than specified cutoff (%s)\\n\" % (record.aaf[0]))\n",
    "            pass  \n",
    "\n",
    "        elif record.aaf >= maf_threshold:\n",
    "\n",
    "            for sample in record.samples: \n",
    "\n",
    "                name = sample.sample\n",
    "\n",
    "                ### For each sample, if the coverage is too low, remove the genotype for that individual.\n",
    "\n",
    "\n",
    "                if sample['DP'] < coverage_threshold:\n",
    "                    genotype = None\n",
    "                    low_cov_samples += 1\n",
    "                    \n",
    "\n",
    "                elif sample['DP'] >= coverage_threshold:\n",
    "                    genotype = sample['GT']\n",
    "                    sample_cov_kept_dict[name].append(sample['DP'])\n",
    "                ## Now calculate the female and male frequencies separately\n",
    "\n",
    "                if name in sample_dict[\"females\"]: \n",
    "\n",
    "                    if genotype == None: ## if no genotype exists\n",
    "                        sample_missing_dict[name] += 1\n",
    "                        pass\n",
    "                    elif genotype == \"0/0\":\n",
    "                        femREF_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    elif genotype == \"0/1\":\n",
    "                        femREF_count += 1\n",
    "                        femALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    elif genotype == \"1/0\":\n",
    "                        femREF_count += 1\n",
    "                        femALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    elif genotype == \"1/1\":\n",
    "                        femALT_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        fem_genotypes +=1\n",
    "                    #print \"\\tN_REF=\", femREF_count, \"N_ALT=\", femALT_count, \"Nnone=\", fem_none_count\n",
    "                elif name in sample_dict[\"males\"]:\n",
    "                    #print \"Locus=\", record.ID, \"Male=\", sample.sample, \"depth=\", sample['DP'], \"Orig_GT=\", sample['GT'], \"assignedGT=\", genotype                    \n",
    "                    #print \"MALE Locus=\", record.ID, \"n_genotypes\", n_genotypes\n",
    "                    if genotype == None: ## if no genotype exists\n",
    "                        sample_missing_dict[name] += 1\n",
    "                        pass\n",
    "                    elif genotype == \"0/0\":\n",
    "                        malREF_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                    elif genotype == \"0/1\":\n",
    "                        malREF_count += 1\n",
    "                        malALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                    elif genotype == \"1/0\":\n",
    "                        malREF_count += 1\n",
    "                        malALT_count += 1\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                    elif genotype == \"1/1\":\n",
    "                        malALT_count += 2\n",
    "                        n_genotypes +=1\n",
    "                        male_genotypes +=1\n",
    "                #else:\n",
    "                    #print \"\\n##SAMPLE NAME NOT IN POP_MAP.TXT: Sample = %s\" % (name)\n",
    "                    #print \"\\tN_REF=\", malREF_count, \"N_ALT=\", malALT_count, \"Nnone=\", male_none_count\n",
    "\n",
    "            ## Filter loci that have too many missing samples, including samples thrown out due to low coverage!\n",
    "            samples_at_locus = n_genotypes\n",
    "            chromosomes_at_locus = n_genotypes*2\n",
    "            percent_samples_present = n_genotypes/number_of_samples\n",
    "            #print \"Locus\", record.ID\n",
    "            #print samples_at_locus\n",
    "            #print percent_samples_present\n",
    "\n",
    "            if percent_samples_present >= sample_presence_cutoff:\n",
    "                kept_loci += 1\n",
    "                ## Calculate frequencies\n",
    "\n",
    "                femREF_freq = femREF_count/(fem_genotypes*2)\n",
    "                femALT_freq = femALT_count/(fem_genotypes*2)\n",
    "\n",
    "                maleREF_freq = malREF_count/(male_genotypes*2)\n",
    "                maleALT_freq = malALT_count/(male_genotypes*2)\n",
    "\n",
    "\n",
    "                ## Output female stats\n",
    "                freq_ratios_log.append(\"\\n#LOCUS_ID: %s\\n\\n\" %(record.ID))   \n",
    "                freq_ratios_log.append(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "                freq_ratios_log.append(\"Female reference count = %s\\n\" % (femREF_count))\n",
    "                freq_ratios_log.append(\"Female alternative count = %s\\n\" % (femALT_count))\n",
    "                freq_ratios_log.append(\"Female reference frequency = %.3f\\n\" % (femREF_freq))\n",
    "                freq_ratios_log.append(\"Female alternative frequency = %.3f\\n\" % (femALT_freq))\n",
    "\n",
    "                ## check fem freqs\n",
    "                if not (femREF_freq) + (femALT_freq) == 1:\n",
    "                    freq_ratios_log.append(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "                elif (femREF_freq) + (femALT_freq) == 1:\n",
    "                    freq_ratios_log.append(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (femREF_freq + femALT_freq))\n",
    "\n",
    "                ## Output male stats\n",
    "                freq_ratios_log.append(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "                freq_ratios_log.append(\"Male reference count = %s\\n\" % (malREF_count))\n",
    "                freq_ratios_log.append(\"Male alternative count = %s\\n\" % (malALT_count))\n",
    "                freq_ratios_log.append(\"Male reference frequency = %.3f\\n\" % (maleREF_freq))\n",
    "                freq_ratios_log.append(\"Male alternative frequency = %.3f\\n\" % (maleALT_freq))\n",
    "\n",
    "                ## check male freqs\n",
    "                if not (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                    freq_ratios_log.append(\"\\n******ERROR, summed frequencies do not add up to 1******\\n\")\n",
    "                elif (maleREF_freq) + (maleALT_freq) == 1:\n",
    "                    freq_ratios_log.append(\"Summed female ref and alt frequencies OK! (= %.3f)\\n\" % (maleREF_freq + maleALT_freq))\n",
    "\n",
    "                locus_dict[record.ID] = {}\n",
    "                locus_dict[record.ID][\"female_freqs\"] = [\"%.3f\" % (femREF_freq), \"%.3f\" % (femALT_freq)]\n",
    "                locus_dict[record.ID][\"male_freqs\"] = [\"%.3f\" % (maleREF_freq), \"%.3f\" % (maleALT_freq)]\n",
    "\n",
    "\n",
    "                ####### ==============================================================================================\n",
    "                ### So now I have the allele frequencies for males and females, I can subtract them and see if the distribution fits the expectition for X linked or Z linked!\n",
    "\n",
    "                freq_ratio = femREF_freq - maleREF_freq\n",
    "                #print freq_ratio\n",
    "                freq_ratio_data.append(freq_ratio)\n",
    "\n",
    "\n",
    "                ## Write files for X or Z linked loci\n",
    "\n",
    "                if freq_ratio >= lower_thresh and freq_ratio <= upper_thresh and femREF_freq >= 0.95:  ## for X linked\n",
    "                    linked_status = \"Xlinked\"\n",
    "                    freq_ratios_log.append(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                    Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                    numb_putative_Xlinked += 1\n",
    "\n",
    "                elif freq_ratio >= -upper_thresh and freq_ratio <= -lower_thresh and maleREF_freq >= 0.95:  ## for Z linked\n",
    "                    linked_status = \"Zlinked\"\n",
    "                    freq_ratios_log.append(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                    Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                    numb_putative_Zlinked += 1\n",
    "                else:\n",
    "                    freq_ratios_log.append(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                    linked_status =\"NotSexLinked\"\n",
    "\n",
    "                ## Write the main info file for male and female frequencies, ratios etc.\n",
    "\n",
    "                all_frequencies.write(\"%s\\t%s\\t%s\\t%s\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%.3f\\t%s\\n\" % (record.ID, record.POS, male_genotypes*2, fem_genotypes*2, maleREF_freq, maleALT_freq, femREF_freq, femALT_freq, freq_ratio, linked_status))\n",
    "\n",
    "\n",
    "            elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "                freq_ratios_log.append(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "                freq_ratios_log.append(\"Number of samples at locus is lower than sample presence cutoff (%s)\\n\" % (samples_at_locus))\n",
    "                low_data_loci += 1\n",
    "\n",
    "\n",
    "    ## Print some quick summary stats - also written at the end of the log file\n",
    "    print \"Number of samples =\", number_of_samples\n",
    "    print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "    print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "    print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "    print \"Number of putative X linked snps = %s\" % (numb_putative_Xlinked)\n",
    "    print \"Number of putative X linked tags = %s\" % (len(set(Putative_Xlinked_makers)))\n",
    "    print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "    print \"Number of putative Z linked tags = %s\" % (len(set(Putative_Zlinked_makers)))\n",
    "\n",
    "    freq_ratios_log.append(\"\\nSUMMARY....\\n\\n\")\n",
    "    freq_ratios_log.append(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "    freq_ratios_log.append(\"Number of loci with enough data = %s\\n\" % (kept_loci))\n",
    "    freq_ratios_log.append(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "    freq_ratios_log.append(\"Number of putative X linked snps = %s\\n\" % (numb_putative_Xlinked))\n",
    "    freq_ratios_log.append(\"Number of putative X linked tags = %s\\n\" % (len(set(Putative_Xlinked_makers))))\n",
    "    freq_ratios_log.append(\"Number of putative Z linked snps = %s\\n\" % (numb_putative_Zlinked))\n",
    "    freq_ratios_log.append(\"Number of putative Z linked tags = %s\\n\" % (len(set(Putative_Zlinked_makers))))\n",
    "\n",
    "\n",
    "    ## plot histogram of frequency ratios\n",
    "\n",
    "    plt.hist(freq_ratio_data, bins = 100, color = '0.5')\n",
    "    plt.xlabel('Female_REFfreq - Male_REFfreq', fontsize=8)\n",
    "    plt.ylabel(\"Number of SNPs\", fontsize=8)\n",
    "    plt.axvline(x = np.mean(freq_ratio_data), color='r', linestyle='dashed', linewidth = 2)\n",
    "    plt.axvline(x = X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "    plt.axvline(x = -X_or_Z_freq_threshold, color='b', linestyle='dashed', linewidth = 1)\n",
    "    plt.savefig(\"%s%s\" %(myvcfpath, \".fem-male_freqs.pdf\"), format = 'pdf')\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "\n",
    "    all_frequencies.close()\n",
    "    \n",
    "\n",
    "    print \"\\n***DONE!***\\n\"\n",
    "\n",
    "    return set(Putative_Xlinked_makers), set(Putative_Zlinked_makers), freq_ratios_log\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "def SL_snp_het(myvcfpath, popmappath, catalog_tags_file, homogamtic_homozygosity_threshold = 0.9, heterogametic_heterozygosity_threshold = 0.5, sample_presence_cutoff = 0.75, coverage_threshold = 3, maf_threshold = 0.05):\n",
    "    \n",
    "   \n",
    "    \"\"\"\n",
    "    This script identifies sex linked markers from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "    Arguments:\n",
    "    \n",
    "    myvcfpath                            - path to vcf file (note this will be altered to make header compatible with Pyvcf. \n",
    "                                           New vcf will have same name with \".altered\" appended to the end)\n",
    "    \n",
    "    popmappath                           - path to population map file containing sex information. Same format as Stacks pop map file.\n",
    "    \n",
    "    catalog_tags_file                    - The catalog tags file used to create the vcf\n",
    "    \n",
    "    homogamtic_homozygosity_threshold    - The lower threshold for the proportion of homozygotes in the homogametic sex at a locus\n",
    "    \n",
    "    heterogamtic_heterozygosity_threshold- The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus\n",
    "        \n",
    "    sample_presence_cutoff               - a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "    \n",
    "    coverage_threshold                   - a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this \n",
    "                                           threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will\n",
    "                                           then remove the locus.\n",
    "    \n",
    "    maf_threshold                        - minor allele frequency cutoff for a locus across all samples. \n",
    "    \n",
    "    \n",
    "    Workflow: ### NOTE*** different to SL_snp_finder()\n",
    "    \n",
    "    1. Filters loci that are present in the user-specified number of samples\n",
    "    2. Looks for loci that are heterozygous in all or most of the homogametic sex and homozygous in all or most of the heterogametic sex\n",
    "    3. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    4. All suplus information is recorded to a log file, with a summary at the end of this file.\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    import vcf\n",
    "    import matplotlib\n",
    "    #matplotlib.use('Agg') ## this allows the drawing of plots in matplotlib on the cluster, which doesn't use the X-server backend. This has something to do with display (but I don't know what)\n",
    "    import matplotlib.pyplot as plt\n",
    "    import numpy as np\n",
    "    import os.path\n",
    "    import sys\n",
    "    import time\n",
    "    import gzip\n",
    "    \n",
    "    \n",
    "    \n",
    "    print \"\\n##### Using SNP heterozygosity approach #####\\n \"\n",
    "\n",
    "    # First thing to do is alter the metadata in the vcf outputted by stacks 1.30. I am not sure if it is stacks or pyvcf that is wrong, but stacks encodes the data in the allele depth field as an interger, while pyvcf expects a float. Changing the metadata line in the vcf header to contain \"Number=.\" instead of \"Number=1\" fixes the issue.  \n",
    "    \n",
    "    myvcf = open(myvcfpath, 'r').readlines()\n",
    "    \n",
    "    alteredvcfpath = \"%s%s\" %(myvcfpath, \".altered\")\n",
    "    alteredvcf = open(alteredvcfpath, 'w')\n",
    "\n",
    "    for line in myvcf:\n",
    "        if \"Allele Depth\" not in line:\n",
    "            alteredvcf.write(line)\n",
    "        elif \"Allele Depth\" in line:\n",
    "            line = '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allele Depth\">\\n'\n",
    "            alteredvcf.write(line)\n",
    "    alteredvcf.close()\n",
    "\n",
    "\n",
    "    # ### Now calculate heterozygosity for males and females at each SNP\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r')) ## load in altered vcf file\n",
    "\n",
    "    Putative_Xlinked_makers = []\n",
    "    Putative_Zlinked_makers = []\n",
    "\n",
    "    ## Make a log file for all auxillary info\n",
    "    het_approach_log = []\n",
    "    #open(\"%s%s\" %(myvcfpath, \".het_approach.log\"), 'w')\n",
    "    het_approach_log.append(\"Script run on %s\\n \" % (time.strftime(\"%c\")))\n",
    "\n",
    "    ## get male/female info \n",
    "    \n",
    "    pop_map = open(popmappath, 'r').readlines()\n",
    "\n",
    "    sample_dict = {}\n",
    "    sample_dict[\"females\"] = []\n",
    "    sample_dict[\"males\"] = []\n",
    "    fem_samples = 0\n",
    "    male_samples = 0\n",
    "    sample_counter = 0\n",
    "    low_maf_counter = 0\n",
    "    kept_loci = 0\n",
    "    sample_missing_dict = {}\n",
    "    sample_cov_dict = {}\n",
    "    sample_cov_kept_dict = {}\n",
    "\n",
    "    for sample in pop_map:\n",
    "        name = sample.split()[0]\n",
    "        sex = sample.strip().split()[1]\n",
    "        sample_missing_dict[name] = 0\n",
    "\n",
    "        #sample_counter += 1\n",
    "        if sex == \"F\" or sex == \"f\" or sex == \"Female\" or sex == \"female\" or sex == \"Fem\" or sex == \"fem\":\n",
    "            sample_dict[\"females\"].append(name)\n",
    "            fem_samples += 1\n",
    "        elif sex == \"M\" or sex == \"m\" or sex == \"Male\" or sex == \"male\" or sex == \"Mal\" or sex == \"mal\":\n",
    "            sample_dict[\"males\"].append(name)\n",
    "            male_samples += 1\n",
    "\n",
    "    #print sample_dict        \n",
    "    locus_dict = {}\n",
    "\n",
    "    ## Find number of loci in input file\n",
    "    locus_counter = 0\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        locus_counter += 1\n",
    "    print \"Number of loci = %s\" % (locus_counter)\n",
    "\n",
    "    ## Write some general stats and input options to the log file\n",
    "\n",
    "    het_approach_log.append(\"\\n## User specified options:\\n\")\n",
    "    het_approach_log.append(\"Input vcf: %s\\n\" % (myvcfpath))\n",
    "    het_approach_log.append(\"pop_map file used: %s\\n\" % (popmappath))\n",
    "    het_approach_log.append(\"Catalog file used: %s\\n\" % (catalog_tags_file))\n",
    "    het_approach_log.append(\"homogamtic_homozygosity_threshold: %s\\n\" % (homogamtic_homozygosity_threshold))\n",
    "    het_approach_log.append(\"heterogamtic_heterozygosity_threshold : %s\\n\" % (heterogametic_heterozygosity_threshold))\n",
    "    het_approach_log.append(\"Min percentage samples present: %s\\n\" % (sample_presence_cutoff))\n",
    "    het_approach_log.append(\"Min coverage per genotype: %s\\n\" % (coverage_threshold))\n",
    "    het_approach_log.append(\"Min maf per locus: %s\\n\" % (maf_threshold))\n",
    "    het_approach_log.append(\"Number of female samples = %s\\n\" % (fem_samples))\n",
    "    het_approach_log.append(\"Number of male samples = %s\\n\" % (male_samples))\n",
    "    het_approach_log.append(\"Number of loci = %s\\n\" % (locus_counter))\n",
    "\n",
    "\n",
    "    ## Some quick counters\n",
    "\n",
    "    numb_putative_Xlinked = 0\n",
    "    numb_putative_Zlinked = 0\n",
    "    low_data_loci = 0\n",
    "\n",
    "    vcf_reader = vcf.Reader(open(alteredvcfpath, 'r'))\n",
    "\n",
    "    info_rec = vcf_reader.next()\n",
    "\n",
    "    for sample in info_rec:\n",
    "        name = sample.sample\n",
    "        sample_cov_dict[name] = []\n",
    "        sample_cov_kept_dict[name] = []\n",
    "\n",
    "\n",
    "    for record in vcf_reader:\n",
    "        \n",
    "        number_of_samples = len(record.samples) ## Number of samples (not genotypes)\n",
    "        samples_present = record.num_called ## Number of samples called\n",
    "        percent_samples_present = samples_present/number_of_samples ## percentage of samples genotyped at this locus\n",
    "        \n",
    "        if percent_samples_present >= sample_presence_cutoff:\n",
    "                \n",
    "            if record.aaf[0] < maf_threshold: ## if locus has minor allele freq lower than specified threshold then skip it\n",
    "                low_maf_counter += 1\n",
    "                het_approach_log.append(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "                het_approach_log.append(\"MAF is lower than cutoff (%s), locus not used\\n\" % (record.aaf[0]))\n",
    "\n",
    "            elif record.aaf >= maf_threshold:\n",
    "\n",
    "                male_genotypes = 0\n",
    "                fem_genotypes = 0\n",
    "\n",
    "                het_males_count = 0 \n",
    "                het_females_count = 0 \n",
    "                hom_males_count = 0 \n",
    "                hom_females_count = 0 \n",
    "\n",
    "                for sample in record.samples:\n",
    "                    if sample.called:  ## If the sample is called, keep going\n",
    "                        name = sample.sample\n",
    "\n",
    "                        if sample['DP'] < coverage_threshold: ## If the sample has enough coverage keep going\n",
    "                            low_cov_samples += 1\n",
    "                        elif sample['DP'] >= coverage_threshold:\n",
    "                            sample_cov_kept_dict[name].append(sample['DP'])\n",
    "\n",
    "                            if name in sample_dict[\"females\"]: ## If the sample is female, adjust female counts accordingly\n",
    "                                fem_genotypes += 1\n",
    "\n",
    "                                if sample.is_het:\n",
    "                                    het_females_count += 1\n",
    "                                elif not sample.is_het:\n",
    "                                    hom_females_count += 1\n",
    "\n",
    "                            elif name in sample_dict['males']: ## or if the sample is male, adjust the male counts\n",
    "                                male_genotypes += 1\n",
    "\n",
    "                                if sample.is_het:\n",
    "                                    het_males_count += 1\n",
    "                                elif not sample.is_het:\n",
    "                                    hom_males_count += 1\n",
    "                    \n",
    "\n",
    "                ## Calculate frequencies per sex (these frequencies are based on the number of males or females called at each number, so accounts for missing data)\n",
    "\n",
    "                male_homozygosity = hom_males_count/male_genotypes\n",
    "                male_heterozygosity = het_males_count/male_genotypes\n",
    "\n",
    "                female_homozygosity = hom_females_count/fem_genotypes\n",
    "                female_heterozygosity = het_females_count/fem_genotypes\n",
    "\n",
    "\n",
    "                ## Output female stats\n",
    "                het_approach_log.append(\"\\n#LOCUS_ID: %s_%s\\n\\n\" %(record.ID, record.POS))   \n",
    "                het_approach_log.append(\"Number of female genotypes for this locus = %s\\n\" %(fem_genotypes))\n",
    "                het_approach_log.append(\"Female homozygosity = %.3f\\n\" % (female_homozygosity))\n",
    "                het_approach_log.append(\"Female heterozygosity = %.3f\\n\" % (female_heterozygosity))\n",
    "\n",
    "                ## Output male stats\n",
    "                het_approach_log.append(\"Number of male genotypes for this locus = %s\\n\" %(male_genotypes))\n",
    "                het_approach_log.append(\"Male homozygosity = %.3f\\n\" % (male_homozygosity))\n",
    "                het_approach_log.append(\"Male heterozygosity = %.3f\\n\" % (male_heterozygosity))\n",
    "\n",
    "\n",
    "                ## Find and write files for X or Z linked loci\n",
    "\n",
    "                if all([female_homozygosity >= homogamtic_homozygosity_threshold, male_heterozygosity >= heterogametic_heterozygosity_threshold]):\n",
    "                    linked_status = \"Xlinked\"\n",
    "                    het_approach_log.append(\"Locus %s DOES FIT X linked criteria <------------------------\\n\" % (record.ID))\n",
    "                    Putative_Xlinked_makers.append(\"%s\" % (record.ID))\n",
    "                    numb_putative_Xlinked += 1\n",
    "\n",
    "                elif all([male_homozygosity >= homogamtic_homozygosity_threshold, female_heterozygosity >= heterogametic_heterozygosity_threshold]):  ## for Z linked\n",
    "                    linked_status = \"Zlinked\"\n",
    "                    het_approach_log.append(\"Locus %s DOES FIT Z linked criteria <------------------------\\n\" % (record.ID))\n",
    "                    Putative_Zlinked_makers.append(\"%s\" % (record.ID))\n",
    "                    numb_putative_Zlinked += 1\n",
    "                else:\n",
    "                    het_approach_log.append(\"Locus %s does not fit X or Z linked criteria\\n\" % (record.ID))\n",
    "                    linked_status =\"NotSexLinked\"\n",
    "        \n",
    "\n",
    "        elif percent_samples_present < sample_presence_cutoff:  ## If not enough samples at a locus then log it but don't use for female-male calculations\n",
    "            het_approach_log.append(\"\\n#LOCUS_ID: %s, Locus_POS: %s\\n\\n\" %(record.ID, record.POS)) \n",
    "            het_approach_log.append(\"Number of samples at locus (%s) is lower than sample presence cutoff (%s)\\n\" % (percent_samples_present, samples_present))\n",
    "            low_data_loci += 1\n",
    "\n",
    "\n",
    "    ## Print some quick summary stats - also written at the end of the log file\n",
    "    print \"Number of samples =\", number_of_samples\n",
    "    print \"Number of loci with too few samples = %s\" % (low_data_loci)\n",
    "    print \"Number of loci with low MAF = %s\" % (low_maf_counter)\n",
    "    print \"Number of loci with enough data = %s\" % (kept_loci)\n",
    "    print \"Number of putative X linked snps = %s\" % (numb_putative_Xlinked)\n",
    "    print \"Number of putative X linked tags = %s\" % (len(set(Putative_Xlinked_makers)))\n",
    "    print \"Number of putative Z linked markers = %s\" % (numb_putative_Zlinked)\n",
    "    print \"Number of putative Z linked tags = %s\" % (len(set(Putative_Zlinked_makers)))\n",
    "\n",
    "    het_approach_log.append(\"\\nSUMMARY....\\n\\n\")\n",
    "    het_approach_log.append(\"Number of loci with too few samples = %s\\n\" % (low_data_loci))\n",
    "    het_approach_log.append(\"Number of loci with enough data = %s\\n\" % (kept_loci))\n",
    "    het_approach_log.append(\"Number of loci with low MAF = %s\\n\" % (low_maf_counter))\n",
    "    het_approach_log.append(\"Number of putative X linked snps = %s\\n\" % (numb_putative_Xlinked))\n",
    "    het_approach_log.append(\"Number of putative X linked tags = %s\\n\" % (len(set(Putative_Xlinked_makers))))\n",
    "    het_approach_log.append(\"Number of putative Z linked snps = %s\\n\" % (numb_putative_Zlinked))\n",
    "    het_approach_log.append(\"Number of putative Z linked tags = %s\\n\" % (len(set(Putative_Zlinked_makers))))\n",
    "\n",
    "\n",
    "    print \"\\n ### DONE! ### \\n\"\n",
    "    \n",
    "    return set(Putative_Xlinked_makers), set(Putative_Zlinked_makers), het_approach_log\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def SL_tag_finder(catalog_tags_file, popmappath, sex_presence_thresh = 0.5):\n",
    "    \n",
    "    ## Note popmappath here should point to a file that contains usual two popmap columns plus a\n",
    "    ## a column for the ID of each sample . . . \n",
    "    \n",
    "    print \"\\n##### Using Sex specific tag approach ##### \\n\"\n",
    "    \n",
    "    ## open log file and write params\n",
    "    log = []\n",
    "    #open(\"%s/%s\" % (catalog_tags_file.rpartition(\"/\")[0], \"Presence_absence.log\"), 'w')     \n",
    "    log.append(\"Input parameters:\\n\\nCatalog file: %s\\nPop_map_file: %s\\nMinimum percentage of M or F present: %1.2f\\n\\n\" % (catalog_tags_file, popmappath, sex_presence_thresh))\n",
    "\n",
    "    \n",
    "    ## First get sample ID and sex info for each sample\n",
    "    ## Include only samples in the pop_map file\n",
    "\n",
    "    sex_file = open(popmappath, 'r').readlines()\n",
    "\n",
    "    kept_sample_names = []\n",
    "    \n",
    "    sample_dict = {}\n",
    "    sex_dict = {}\n",
    "    sex_dict[\"Males\"] = []\n",
    "    sex_dict[\"Females\"] = []\n",
    "    male_count = 0\n",
    "    female_count = 0\n",
    "\n",
    "    ## Get sample IDs\n",
    "    \n",
    "    for line in sex_file:\n",
    "        sample_name = line.split()[0]\n",
    "        sex = line.split()[1]\n",
    "        ID = line.split()[2]\n",
    "        \n",
    "        kept_sample_names.append(sample_name)\n",
    "        sample_dict[sample_name] = []\n",
    "        sample_dict[sample_name].append(ID)\n",
    "        \n",
    "        if sex == \"M\" or sex == \"m\" or sex == \"Male\" or sex == \"male\":\n",
    "            sample_dict[sample_name].append(sex)\n",
    "            sex_dict[\"Males\"].append(ID)\n",
    "            male_count += 1\n",
    "        elif sex == \"F\" or sex == \"f\" or sex == \"Female\" or sex == \"female\":\n",
    "            sample_dict[sample_name].append(sex)\n",
    "            sex_dict[\"Females\"].append(ID)\n",
    "            female_count += 1\n",
    "\n",
    "    log.append(\"\\n\\n############################\\n\\nLOG:\\n\\n# 'BAD' tag status means that one or more samples had two or more tags at the same \\ncatalog locus, thus, this locus was not used in analyses\\n\\n\")\n",
    "    log.append(\"Tag_status\\tTag_ID\\tN_males\\tN_females\\tY_linked\\tW_linked\\n\") ## headers for log info\n",
    "    \n",
    "    \n",
    "    ## Now look through each tag in catalog file:\n",
    "        ## find tags which are present in sex_presence_thresh males and no females (Y linked)\n",
    "        ## find tags which are present in sex_presence_thresh females and no males (W linked)\n",
    "    \n",
    "    \n",
    "    if catalog_tags_file.endswith(\"gz\"):\n",
    "        cat_file = gzip.open(catalog_tags_file, 'r').readlines()\n",
    "    else:\n",
    "        cat_file = open(catalog_tags_file, 'r').readlines()\n",
    "    \n",
    "    n_males_required = math.ceil(male_count*sex_presence_thresh) ## math.ceil rounds up to nearest whole individual\n",
    "    n_females_required = math.ceil(female_count*sex_presence_thresh)\n",
    "\n",
    "    putative_Ylinked_tags = []\n",
    "    putative_Wlinked_tags = []\n",
    "\n",
    "\n",
    "    for tag in cat_file[1:]:\n",
    "\n",
    "        ## First get all the info I need\n",
    "\n",
    "        N_males_at_locus = 0 \n",
    "        N_females_at_locus = 0\n",
    "        duplicates_present = \"no\" ## Checking for replicate samples (Often a tag can be present in the same individual twice, need to remove these)\n",
    "        samples_at_locus = [] ## for checking for replicate samples\n",
    "\n",
    "        tag_ID =  tag.split()[2]\n",
    "        samples_field = tag.split()[7] ## the field in the catalog file that contains the samples present at that locus\n",
    "\n",
    "        Y_linked_str = \"NO\" ## defaults for log file\n",
    "        W_linked_str = \"NO\"\n",
    "        tag_status = \"BAD\"\n",
    "\n",
    "        for sample in samples_field.split(\",\"):\n",
    "\n",
    "            sample_ID = sample.split(\"_\")[0]\n",
    "\n",
    "            if sample_ID not in samples_at_locus: ## check that there are no replicates\n",
    "                if sample_ID in sex_dict[\"Males\"]:\n",
    "                    N_males_at_locus += 1\n",
    "                    samples_at_locus.append(sample_ID)\n",
    "                elif sample_ID in sex_dict[\"Females\"]:\n",
    "                    N_females_at_locus += 1\n",
    "                    samples_at_locus.append(sample_ID)\n",
    "\n",
    "            elif sample_ID in samples_at_locus:\n",
    "                duplicates_present = \"yes\"\n",
    "\n",
    "        if duplicates_present == \"no\": ## If there are no sample duplicates at the tag\n",
    "\n",
    "            tag_status = \"OK\"\n",
    "\n",
    "            ## Look for Y linked tags\n",
    "\n",
    "            if N_females_at_locus == 0 and N_males_at_locus > n_males_required:\n",
    "                putative_Ylinked_tags.append(tag_ID)\n",
    "                Y_linked_str = \"YES\"\n",
    "\n",
    "            ## look for W linked tags\n",
    "\n",
    "            elif N_males_at_locus == 0 and N_females_at_locus > n_females_required:\n",
    "                putative_Wlinked_tags.append(tag_ID)\n",
    "                W_linked_str = \"YES\"\n",
    "        \n",
    "\n",
    "        log.append(\"%s\\t%s\\t%s\\t%s\\t%s\\t%s\\n\" % (tag_status, tag_ID, N_males_at_locus, N_females_at_locus, Y_linked_str, W_linked_str))\n",
    "\n",
    "\n",
    "    ## Write log file summary\n",
    "\n",
    "    log.append(\"\\nSUMMARY:\\nNumber of males: %s\\n\" % (female_count))\n",
    "    log.append(\"Number of males: %s\\n\" % (male_count))\n",
    "    log.append(\"Number of Putative Y linked tags: %s\\n\" % (len(putative_Ylinked_tags)))\n",
    "    log.append(\"Number of Putative W linked tags: %s\\n\" % (len(putative_Wlinked_tags)))\n",
    "    \n",
    "    ## Print summary to STDOUT\n",
    "\n",
    "    print \"\\nSUMMARY:\\nNumber of males: %s\" % (female_count)\n",
    "    print \"Number of males: %s\" % (male_count)\n",
    "    print \"Number of Putative Y linked tags: %s\" % (len(putative_Ylinked_tags))\n",
    "    print \"Number of Putative W linked tags: %s\" % (len(putative_Wlinked_tags))\n",
    "    print \"\\n ### DONE! ###\\n\"\n",
    "    \n",
    "    \n",
    "    ## return list of sex linked tags\n",
    "    \n",
    "    return set(putative_Ylinked_tags), set(putative_Wlinked_tags), log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def Super_SLM_finder(Parameter_dict, execute_seq = \"111\"):\n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    Parameter_dict - contains all data and settings for all three approaches\n",
    "    execute_seq -    (default = \"111\") tells the wrapper which approach to use, e.g. 111 means run all, 101 means run approach 1 and 3 but not 2.\n",
    "    \n",
    "    \n",
    "    -----------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "    Approach 1. Identifies sex linked SNPs from a VCF file using the criteria of female-male allele freq \n",
    "\n",
    "    Workflow:\n",
    "    1. Filters loci by presence (number of samples), coverage and maf.\n",
    "    2. Calculates the allele frequencies for males and females separately\n",
    "    3. Subtracts male from female frequencies and filter loci that show signs of X or Z linkage\n",
    "    4. Outputs all male and female frequencies and female-male outputs to a single file called \n",
    "       \"yourinput.vcf.all_frequencies.tsv\" (where yourinput = the name and path of your vcf file). Loci \n",
    "       identified as X or Z linked are labelled as such in this file.\n",
    "    5. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    6. Outputs a histogram of the distribution of female-male frequencies called \"yourinput.vcf.fem-male_freqs.pdf\"\n",
    "    7. All suplus information is recorded to a log file, with a summary at the end of this file.\n",
    "    \n",
    "    -----------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "    Approach 2. Identifies sex linked SNPs from a VCF file using the heterozygosity difference between sexes \n",
    "    \n",
    "    Workflow:\n",
    "    1. Filters loci by presence (number of samples), coverage and maf.\n",
    "    3. Looks for loci that are heterozygous in all or most of the homogametic sex and homozygous in all or most\n",
    "       of the heterogametic sex\n",
    "    5. Outputs all putative X or Z linked markers to separate fasta files if any are identified.\n",
    "    6. All suplus information is recorded to a log file, with a summary at the end of this file.\n",
    "    \n",
    "    -----------------------------------------------------------------------------------------------------------------------------------------------------\n",
    "    \n",
    "    Approach 3. Identifies sex linked RAD tags from a catalog file using the presence and absence of tags in each sex \n",
    "    \n",
    "    Workflow:\n",
    "    # Using only samples specified in the Pop_map file:\n",
    "    1. Finds tags which are present in <sex_presence_threshold> males and no females (Y linked)\n",
    "    2. Find tags which are present in <sex_presence_threshold> females and no males (W linked)\n",
    "    3. Outputs all putative X or W linked tags to a fasta file\n",
    "    \n",
    "    ====================================================================================================================================================\n",
    "    \n",
    "    ## Example parameter dict\n",
    "    \n",
    "    Parameter_dict = {}\n",
    "\n",
    "    ##### Data ########################\n",
    "\n",
    "    Parameter_dict['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "    Parameter_dict['VCF'] =  \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/Guillaumes_vcf/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "    Parameter_dict['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/sex_info_ID.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "    ###### threshold parameters #######\n",
    "\n",
    "    # 1. Frequency approach\n",
    "    Parameter_dict['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "    Parameter_dict['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "    Parameter_dict['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "    Parameter_dict['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "    # 2. Heterozygosity approach\n",
    "    Parameter_dict['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "    Parameter_dict['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "    Parameter_dict['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "    Parameter_dict['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "    Parameter_dict['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "    # 3. Sex specific presence or absence approach\n",
    "    Parameter_dict['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n",
    "\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    from matplotlib_venn import venn2, venn3\n",
    "    import gzip\n",
    "    import time\n",
    " \n",
    "    now = time.strftime(\"%c\")\n",
    "    \n",
    "    Xlinked_freq = []\n",
    "    Zlinked_freq = []\n",
    "    freq_log = []\n",
    "    Xlinked_het = []\n",
    "    Zlinked_het = []\n",
    "    het_log = []\n",
    "    Ylinked_tags = []\n",
    "    Wlinked_tags = []\n",
    "    pres_abs_log = []\n",
    "    \n",
    "    \n",
    "    ## freq approach ## -------------------------------------------------------------------\n",
    "    \n",
    "    if execute_seq[0] == 1 or execute_seq[0] == \"1\":\n",
    "        Xlinked_freq, Zlinked_freq, freq_log = SL_snp_freq(Parameter_dict['VCF'], \\\n",
    "                                                 Parameter_dict['Pop_map'],\\\n",
    "                                                 Parameter_dict['Catalog'],\\\n",
    "                                                 Parameter_dict['X_or_Z_freq_threshold'],\\\n",
    "                                                 Parameter_dict['sample_presence_cutoff1'],\\\n",
    "                                                 Parameter_dict['coverage_threshold1'],\\\n",
    "                                                 Parameter_dict['maf_threshold1'])\n",
    "        \n",
    "    ## Heterozygosity approach ## ---------------------------------------------------------\n",
    "    \n",
    "    if execute_seq[1] == 1 or execute_seq[1] == \"1\":\n",
    "        Xlinked_het, Zlinked_het, het_log = SL_snp_het(Parameter_dict['VCF'],\\\n",
    "                   Parameter_dict['Pop_map'],\\\n",
    "                   Parameter_dict['Catalog'],\\\n",
    "                   Parameter_dict['homogamtic_homozygosity_threshold'],\\\n",
    "                   Parameter_dict['heterogamtic_heterozygosity_threshold'],\\\n",
    "                   Parameter_dict['sample_presence_cutoff2'],\\\n",
    "                   Parameter_dict['coverage_threshold2'],\\\n",
    "                   Parameter_dict['maf_threshold2'])\n",
    "    \n",
    "    \n",
    "    ## Presence absence approach ## --------------------------------------------------------\n",
    "    \n",
    "    if execute_seq[2] == 1 or execute_seq[2] == \"1\":\n",
    "        Ylinked_tags, Wlinked_tags, pres_abs_log = SL_tag_finder(Parameter_dict['Catalog'],\\\n",
    "                      Parameter_dict['Pop_map'],\\\n",
    "                      Parameter_dict['sex_presence_threshold'])\n",
    "\n",
    "    \n",
    "    ## Merge the tag lists and draw Venn diagrams for loci identified by each method. \n",
    "    \n",
    "    if execute_seq == \"111\":\n",
    "    \n",
    "        ## For XY tags\n",
    "\n",
    "        fig = plt.figure(figsize= (15,15))\n",
    "        fig.add_subplot(1,2,1)\n",
    "        a = venn3([Xlinked_freq,Xlinked_het, Ylinked_tags], [\"Xlinked_freq\",\"Xlinked_het\", \"Ylinked_tags\"])\n",
    "        plt.title(\"XY tags identified\")\n",
    "\n",
    "        XYset = Xlinked_freq.union(Xlinked_het,Ylinked_tags)        \n",
    "\n",
    "        ## For ZW tags\n",
    "\n",
    "        fig.add_subplot(1,2,2)\n",
    "        b = venn3([Zlinked_freq, Zlinked_het, Wlinked_tags], [\"Zlinked_freq\",\"Zlinked_het\", \"Wlinked_tags\"])\n",
    "        plt.title(\"ZW tags identified\")\n",
    "\n",
    "        ZWset = Zlinked_freq.union(Zlinked_het,Wlinked_tags)\n",
    "\n",
    "    elif execute_seq == \"101\":\n",
    "\n",
    "        ## For XY tags\n",
    "\n",
    "        fig = plt.figure(figsize= (15,15))\n",
    "        fig.add_subplot(1,2,1)\n",
    "        venn2([Xlinked_freq, Ylinked_tags], [\"Xlinked_freq\", \"Ylinked_tags\"])\n",
    "        plt.title(\"XY tags identified\")\n",
    "        XYset = Xlinked_freq.union(Ylinked_tags)\n",
    "\n",
    "        ## For ZW tags\n",
    "\n",
    "        fig.add_subplot(1,2,2)\n",
    "        venn2([Zlinked_freq, Wlinked_tags], [\"Zlinked_freq\", \"Wlinked_tags\"])\n",
    "        ZWset = Zlinked_freq.union(Wlinked_tags)\n",
    "\n",
    "    elif execute_seq == \"011\":\n",
    "\n",
    "        ## For XY tags\n",
    "\n",
    "        fig = plt.figure(figsize= (15,15))\n",
    "        fig.add_subplot(1,2,1)\n",
    "        venn2([Xlinked_het, Ylinked_tags], [\"Xlinked_freq\", \"Ylinked_tags\"])\n",
    "        plt.title(\"XY tags identified\")\n",
    "        XYset = Xlinked_het.union(Ylinked_tags)\n",
    "\n",
    "        ## For ZW tags\n",
    "\n",
    "        fig.add_subplot(1,2,2)\n",
    "        venn2([Zlinked_het, Wlinked_tags], [\"Zlinked_freq\", \"Wlinked_tags\"])\n",
    "        ZWset = Zlinked_het.union(Wlinked_tags)\n",
    "        plt.title(\"ZW tags identified\")\n",
    "\n",
    "    elif execute_seq == \"110\":\n",
    "\n",
    "        ## For XY tags\n",
    "\n",
    "        fig = plt.figure(figsize= (15,15))\n",
    "        fig.add_subplot(1,2,1)\n",
    "        venn2([Xlinked_freq,Xlinked_het], [\"Xlinked_freq\", \"Ylinked_tags\"])\n",
    "        plt.title(\"XY tags identified\")\n",
    "        XYset = Xlinked_freq.union(Xlinked_het)\n",
    "\n",
    "        ## For ZW tags\n",
    "\n",
    "        fig.add_subplot(1,2,2)\n",
    "        venn2([Zlinked_freq,Zlinked_het], [\"Zlinked_freq\", \"Wlinked_tags\"])\n",
    "        plt.title(\"ZW tags identified\")\n",
    "        ZWset = Zlinked_freq.union(Zlinked_het)\n",
    "\n",
    "\n",
    "    elif execute_seq == \"100\":\n",
    "\n",
    "        XYset = Xlinked_freq\n",
    "        ZWset = Zlinked_freq\n",
    "\n",
    "    elif execute_seq == \"010\":\n",
    "\n",
    "        XYset = Xlinked_het\n",
    "        ZWset = Zlinked_het\n",
    "\n",
    "    elif execute_seq == \"001\":\n",
    "\n",
    "        XYset = Ylinked_tags\n",
    "        ZWset = Wlinked_tags\n",
    "    \n",
    "    \n",
    "    ## Now just need to output the fasta and its done\n",
    "\n",
    "    \n",
    "    if Parameter_dict['Catalog'].endswith(\"gz\"):\n",
    "        catalog = gzip.open(Parameter_dict['Catalog'], 'r').readlines()\n",
    "    else:\n",
    "        catalog = open(Parameter_dict['Catalog'], 'r').readlines()\n",
    "        \n",
    "    ## Outdir same as catalog dir:\n",
    "    \n",
    "    outdir = Parameter_dict['Catalog'].rpartition(\"/\")[0]\n",
    "        \n",
    "    if len(XYset) > 0:\n",
    "        Putative_XYlinked_makers_file = open(\"%s/%s\" %(outdir, \"Putative_XYlinked_makers.fa\"), 'w')\n",
    "\n",
    "        ## record which method found which tag\n",
    "        for locus in XYset:\n",
    "            \n",
    "            if locus in Xlinked_het and locus in Xlinked_freq:\n",
    "                method = \"Xlinked_het_Xlinked_freq\"\n",
    "            \n",
    "            elif locus in Xlinked_freq:\n",
    "                method = \"Xlinked_freq\"\n",
    "            \n",
    "            elif locus in Xlinked_het:\n",
    "                method = \"Xlinked_het\"\n",
    "            \n",
    "            elif locus in Ylinked_tags:\n",
    "                method = \"Ylinked_tags\"\n",
    "\n",
    "            for tag in catalog:\n",
    "                if locus == tag.split()[2]:\n",
    "                    Putative_XYlinked_makers_file.write(\">X_linkedLocusID_%s_%s\\n\" % (locus, method))\n",
    "                    Putative_XYlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_XYlinked_makers_file.close()\n",
    "\n",
    "    if len(ZWset) > 0:\n",
    "        Putative_ZWlinked_makers_file = open(\"%s/%s\" %(outdir, \"Putative_ZWlinked_makers.fa\"), 'w')\n",
    "        \n",
    "        ## record which method found which tag\n",
    "        for locus in ZWset:\n",
    "\n",
    "            if locus in Zlinked_het and locus in Zlinked_freq:\n",
    "                method = \"Zlinked_het_Zlinked_freq\" \n",
    "            \n",
    "            elif locus in Zlinked_freq: \n",
    "                method = \"Zlinked_freq\"\n",
    "\n",
    "            elif locus in Zlinked_het:\n",
    "                method = \"Zlinked_het\"    \n",
    "           \n",
    "            elif locus in Wlinked_tags:\n",
    "                method = \"Wlinked_tags\"\n",
    "                \n",
    "            for tag in catalog:\n",
    "                if locus == tag.split()[2]:\n",
    "                    Putative_ZWlinked_makers_file.write(\">Z_linked|LocusID_%s_%s\\n\" % (locus, method))\n",
    "                    Putative_ZWlinked_makers_file.write(\"%s\\n\" % (tag.split()[8]))\n",
    "        Putative_ZWlinked_makers_file.close()\n",
    "    \n",
    "    \n",
    "    \n",
    "    ### Merge log files ###\n",
    "    \n",
    "    master_log = open(\"%s/%s\" %(outdir, \"Super_Sex_linked_marker.log\"), 'w')\n",
    "    master_log.write(\"### Analyses run on at \" + time.strftime(\"%c\")+\"\\n\\n\")\n",
    "    \n",
    "    master_log.write(\"\\n\\n##### Approach 1. SNP frequencies #####\\n\")\n",
    "    \n",
    "    if len(freq_log) > 0:\n",
    "        for line in freq_log:\n",
    "            master_log.write(line)\n",
    "    else:\n",
    "        master_log.write(\"\\n\\n## NOT USED ##\\n\")\n",
    "        \n",
    "        \n",
    "    master_log.write(\"\\n\\n##### Approach 2. SNP heterozygosities #####\\n\")\n",
    "    \n",
    "    if len(het_log) > 0:\n",
    "        for line in het_log:\n",
    "            master_log.write(line)\n",
    "    else:\n",
    "        master_log.write(\"\\n\\n## NOT USED ##\\n\")\n",
    "    \n",
    "    master_log.write(\"\\n\\n##### Approach 3. Sex specific tags #####\\n\")\n",
    "    \n",
    "    if len(pres_abs_log) > 0:    \n",
    "        for line in pres_abs_log:\n",
    "            master_log.write(line)\n",
    "    else:\n",
    "        master_log.write(\"\\n\\n## NOT USED ##\\n\")\n",
    "    \n",
    "    master_log.write(\"\\n\\n##### Final summary #####\\n\")\n",
    "    master_log.write(\"\\n ## After merging tags accross methods ## \\n\")\n",
    "    \n",
    "    master_log.write(\"Final number of XY tags = %s\\n\" % len(XYset))\n",
    "    master_log.write(\"Final number of ZW tags = %s\\n\" % len(ZWset))\n",
    "    master_log.write(\"Sex linked tags outputted to fastas 'Putative_XYlinked_makers.fa' and Putative_ZWlinked_makers.fa in the directory %s\" % outdir)\n",
    "    master_log.close()\n",
    "    \n",
    "    ## output summary\n",
    "    \n",
    "    print \"\\n ## After merging tags accross methods ## \\n\"\n",
    "    \n",
    "    print \"Final number of XY tags = %s\" % len(XYset)\n",
    "    print \"Final number of ZW tags = %s\" % len(ZWset)\n",
    "                                                       \n",
    "    print \"Sex linked tags outputted to fastas 'Putative_XYlinked_makers.fa' and Putative_ZWlinked_makers.fa\"\n",
    "    print \"in the directory %s\" % outdir\n",
    "    \n",
    "    return XYset, ZWset\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "Parameter_dict = {}\n",
    "\n",
    "##### Data ########################\n",
    "\n",
    "Parameter_dict['Catalog'] =  \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/batch_1.catalog.tags.tsv.gz\" ## Path to the catalog file - used by all approaches.\n",
    "Parameter_dict['VCF'] =  \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/Guillaumes_vcf/batch_1.vcf\" ## path to vcf file (note this will be altered to make header compatible with Pyvcf. New vcf will have same name with \".altered\" appended to the end). Used by Approach i) and ii)\n",
    "Parameter_dict['Pop_map'] = \"/home/djeffrie/Data/RADseq/Rarvalis/Final_populations_outs/sex_info_ID.txt\" ## path to population map file containing sex information. Same format as Stacks pop map file. Used by all approaches.\n",
    "\n",
    "###### threshold parameters #######\n",
    "\n",
    "# 1. Frequency approach\n",
    "Parameter_dict['X_or_Z_freq_threshold'] = 0.4  ## (Default = 0.4) The lower threshold for the freq caluclation to find sex linked snps, e.g. for an XY system, a threshold of 0.4 means that f(F) - f(M) can be >= 0.4 and <= 0.6 (the upper threshold is automatically calculated to be the same distance above 0.5 as the lower threshold is below 0.5) \n",
    "Parameter_dict['sample_presence_cutoff1'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict['coverage_threshold1'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci below this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict['maf_threshold1'] =  0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 2. Heterozygosity approach\n",
    "Parameter_dict['homogamtic_homozygosity_threshold'] = 0.9 ## (Default = 0.9) The minimum number of the homogametic sex which must not have the tag for that tag to be considered linked to the sex-limited chromosome\n",
    "Parameter_dict['heterogamtic_heterozygosity_threshold'] = 0.5 ## (Default = 0.5) The lower threshold for the proportion of heterozygotes in the heterogametic sex at a locus \n",
    "Parameter_dict['sample_presence_cutoff2'] = 0.75 ## (Default = 0.75) a locus must be called in at least this proportion of all samples (not within populations) to be considered\n",
    "Parameter_dict['coverage_threshold2'] = 3 ## (Default = 3) a locus must have at least this threshold in a sample to be considered for that sample. Note that loci bels this threshold will be removed from a sample, and this can push the locus below the sample presence cut-off, which will then remove the locus.\n",
    "Parameter_dict['maf_threshold2'] = 0.05 ## (Default = 0.05) minor allele frequency cutoff for a locus across all samples. \n",
    "\n",
    "# 3. Sex specific presence or absence approach\n",
    "Parameter_dict['sex_presence_threshold'] =  0.5 ## (Default = 0.5) The minimum percenatage of the heterogametic sex that a tag must be present in.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "XYset, ZWset = Super_SLM_finder(Parameter_dict, \"111\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
